// @bryntum/calendar/source/lib/Calendar/localization/Ar.js
import LocaleHelper from '../../Core/localization/LocaleHelper.js';
import '../../Scheduler/localization/Ar.js';

const locale = {

    localeName : 'Ar',
    localeDesc : 'اللغة العربية',
    localeCode : 'ar',
    localeRtl  : true,

    EventEdit : {
        Calendar  : 'التقويم',
        'All day' : 'طوال اليوم',
        day       : 'يوم',
        week      : 'أسبوع',
        month     : 'شهر',
        year      : 'سنة',
        decade    : 'عقد'
    },

    EventMenu : {
        duplicateEvent : 'حدث مكرر',
        copy           : 'نسخ'
    },

    Calendar : {
        toggleSidebar : 'تبديل رؤية الشريط الجانبي',
        Today         : 'اليوم',
        Tomorrow      : 'غدًا',
        next          : range => `${range} التالي`,
        previous      : range => `${range} السابق`,
        plusMore      : value => `+${value} أكثر`,
        moreEvents    : value => `${value} أحداث أخرى`,
        allDay        : 'طوال اليوم',
        endsOn        : d => `${d} النهايات`,
        weekOfYear    : ([y, w]) => `${y} ,${w} أسبوع`,
        loadFail      : 'فشل تحميل بيانات التقويم. يُرجى الاتصال بمسؤول النظام الخاص بك',
        gotoDate      : 'اذهب إلى التاريخ',
        timeRange     : 'النطاق الزمني'
    },

    CalendarDrag : {
        holdCtrlForRecurrence : 'اضغط باستمرار على CTRL من أجل حدث متكرر'
    },

    CalendarMixin : {
        eventCount : count => `حدث${(count && count > 1) ? 'الأ' : ''} ${count || 'لا'}`
    },

    EventTip : {
        'Edit event' : 'تعديل حدث',
        timeFormat   : 'LST'
    },

    ModeSelector : {
        includeWeekends : 'تضمين عطلات نهاية الأسبوع',
        weekends        : 'عطلات نهاية الأسبوع'
    },

    AgendaView : {
        Agenda : 'جدول أعمال'
    },

    MonthView : {
        Month     : 'شهر',
        monthUnit : 'شهر'
    },

    WeekView : {
        weekUnit : 'أسبوع'
    },

    YearView : {
        Year     : 'سنة',
        yearUnit : 'سنة',
        noEvents : 'لا توجد أحداث'
    },

    EventList : {
        List   : 'قائمة',
        Start  : 'بدء',
        Finish : 'إنهاء',
        days   : count => `${count === 1 ? 'يوم' : 'أيام'}${count > 1 ? ` ${count}` : ''}`
    },

    DayView : {
        Day                : 'يوم',
        dayUnit            : 'يوم',
        daysUnit           : 'أيام',
        expandAllDayRow    : 'توسيع قسم طوال اليوم',
        collapseAllDayRow  : 'تصغير قسم طوال اليوم',
        timeFormat         : 'LST',
        timeAxisTimeFormat : 'LST',
        noMatchingDates    : 'لا توجد تواريخ مطابقة'
    },

    DayAgendaView : {
        dayAgenda : 'جدول اليوم'
    },

    DayResourceView : {
        dayResourceView : 'موارد اليومية'
    },

    Sidebar : {
        selectDate      : 'حدد تاريخ التقويم',
        matchEvents     : 'أحداث المباراة',
        filterEvents    : 'تصفية الأحداث حسب اسم الحدث',
        filterResources : 'تصفية الأحداث حسب المورد المحدد'
    },

    WeekExpander : {
        expandTip   : 'انقر لتوسيع صف',
        collapseTip : 'انقر لتصغير صف'
    }
};

export default LocaleHelper.publishLocale(locale);
