// @bryntum/calendar/source/lib/Calendar/localization/Bg.js
import LocaleHelper from '../../Core/localization/LocaleHelper.js';
import '../../Scheduler/localization/Bg.js';

const locale = {

    localeName : 'Bg',
    localeDesc : 'Български',
    localeCode : 'bg',
    localeRtl  : false,

    EventEdit : {
        Calendar  : 'Календар',
        'All day' : 'Цял ден',
        day       : 'Ден',
        week      : 'Седмица',
        month     : 'Месец',
        year      : 'Година',
        decade    : 'Десетилетие'
    },

    EventMenu : {
        duplicateEvent : 'Дублиране на събитие ',
        copy           : 'копиране'
    },

    Calendar : {
        toggleSidebar : 'Превключване на видимостта на страничната лента',
        Today         : 'Днес',
        Tomorrow      : 'Утре',
        next          : range => `Следващ ${range}`,
        previous      : range => `Предишен ${range}`,
        plusMore      : value => `още +${value}`,
        moreEvents    : value => `Още ${value} събития`,
        allDay        : 'Цял ден',
        endsOn        : d => `Свършва ${d}`,
        weekOfYear    : ([y, w]) => `Седмица ${w}, ${y}`,
        loadFail      : 'Неуспешно зареждане на данните от календара. Моля, свържете се със системния администратор',
        gotoDate      : 'Отиди на дата',
        timeRange     : 'Времеви диапазон'
    },

    CalendarDrag : {
        holdCtrlForRecurrence : 'Задържане на CTRL за повтарящо се събитие'
    },

    CalendarMixin : {
        eventCount : count => `${count || 'Не'} събити${(count && count > 1) ? 'я' : 'е'}`
    },

    EventTip : {
        'Edit event' : 'Редактиране на събитие',
        timeFormat   : 'LST'
    },

    ModeSelector : {
        includeWeekends : 'включване на уикенди',
        weekends        : 'Уикенди'
    },

    AgendaView : {
        Agenda : 'Програма'
    },

    MonthView : {
        Month     : 'Месец',
        monthUnit : 'месец'
    },

    WeekView : {
        weekUnit : 'седмица'
    },

    YearView : {
        Year     : 'Година',
        yearUnit : 'година',
        noEvents : 'Няма събития'
    },

    EventList : {
        List   : 'Списък',
        Start  : 'Начало',
        Finish : 'Край',
        days   : count => `${count > 1 ? `${count} ` : ''}ден${count === 1 ? '' : 'а'}`
    },

    DayView : {
        Day                : 'Ден',
        dayUnit            : 'ден',
        daysUnit           : 'дни',
        expandAllDayRow    : 'Разгръщане на секцията за целия ден',
        collapseAllDayRow  : 'Свиване на секцията за целия ден',
        timeFormat         : 'LST',
        timeAxisTimeFormat : 'LST',
        noMatchingDates    : 'Няма съвпадащи дати'
    },

    DayAgendaView : {
        dayAgenda : 'Дневен ред'
    },

    DayResourceView : {
        dayResourceView : 'Дневни ресурси'
    },

    Sidebar : {
        selectDate      : 'Изберете дата от календара',
        matchEvents     : 'Събития на мача',
        filterEvents    : 'Филтриране на събития по име на събитието',
        filterResources : 'Филтриране на събития по избран ресурс'
    },

    WeekExpander : {
        expandTip   : 'Щракнете, за да разширите реда',
        collapseTip : 'Щракнете, за да свиете реда '
    }
};

export default LocaleHelper.publishLocale(locale);
