// @bryntum/calendar/source/lib/Calendar/localization/Ca.js
import LocaleHelper from '../../Core/localization/LocaleHelper.js';
import '../../Scheduler/localization/Ca.js';

const locale = {

    localeName : 'Ca',
    localeDesc : 'Català',
    localeCode : 'ca-ES',
    localeRtl  : false,

    EventEdit : {
        Calendar  : 'Calendari',
        'All day' : 'Tot el dia',
        day       : 'Dia',
        week      : 'Setmana',
        month     : 'Mes',
        year      : 'Any',
        decade    : 'Dècada'
    },

    EventMenu : {
        duplicateEvent : "Duplica l'esdeveniment",
        copy           : 'copia'
    },

    Calendar : {
        toggleSidebar : 'Canviar la visibilitat de la barra lateral',
        Today         : 'Avui',
        Tomorrow      : 'Demà',
        next          : range => `${range} següents`,
        previous      : range => `${range} anteriors`,
        plusMore      : value => `+${value} més`,
        moreEvents    : value => `${value} més esdeveniments`,
        allDay        : 'Tot el dia',
        endsOn        : d => `Acaba el dia ${d}`,
        weekOfYear    : ([y, w]) => `Setmana ${w}, ${y}`,
        loadFail      : "No s'han pogut carregar les dades del calendari. Poseu-vos en contacte amb l'administrador del sistema.",
        gotoDate      : 'Anar a la data',
        timeRange     : 'Interval de temps'
    },

    CalendarDrag : {
        holdCtrlForRecurrence : 'Manteniu premut CTRL per crear un esdeveniment recurrent'
    },

    CalendarMixin : {
        eventCount : count => `${count || 'No hi ha'} esdeveniment${count === 1 ? '' : 's'}`
    },

    EventTip : {
        'Edit event' : "Edita l'esdeveniment",
        timeFormat   : 'LST'
    },

    ModeSelector : {
        includeWeekends : 'Inclou caps de setmana',
        weekends        : 'Caps de setmana'
    },

    AgendaView : {
        Agenda : 'Agenda'
    },

    MonthView : {
        Month     : 'Mes',
        monthUnit : 'mes'
    },

    WeekView : {
        weekUnit : 'setmana'
    },

    YearView : {
        Year     : 'Any',
        yearUnit : 'any',
        noEvents : 'No hi ha cap esdeveniment'
    },

    EventList : {
        List   : 'Llista',
        Start  : 'Inici',
        Finish : 'Final',
        days   : count => `${count > 1 ? `${count} ` : ''}di${count === 1 ? 'a' : 'es'}`
    },

    DayView : {
        Day                : 'Dia',
        dayUnit            : 'dia',
        daysUnit           : 'dies',
        expandAllDayRow    : 'Expandeix la secció de tot el dia',
        collapseAllDayRow  : 'Redueix la secció de tot el dia',
        timeFormat         : 'LST',
        timeAxisTimeFormat : 'LST',
        noMatchingDates    : 'No hi ha dates coincidents'
    },

    DayAgendaView : {
        dayAgenda : 'Agenda del dia'
    },

    DayResourceView : {
        dayResourceView : 'Recursos del dia'
    },

    Sidebar : {
        selectDate      : 'Selecciona la data del calendari',
        matchEvents     : 'Esdeveniments del partit',
        filterEvents    : "Filtra esdeveniments pel nom de l'esdeveniment",
        filterResources : 'Filtra esdeveniments pel recurs seleccionat'
    },

    WeekExpander : {
        expandTip   : 'Feu clic per expandir la fila',
        collapseTip : 'Feu clic per reduir la fila'
    }
};

export default LocaleHelper.publishLocale(locale);
