// @bryntum/calendar/source/lib/Calendar/localization/Cs.js
import LocaleHelper from '../../Core/localization/LocaleHelper.js';
import '../../Scheduler/localization/Cs.js';

const locale = {

    localeName : 'Cs',
    localeDesc : 'Česky',
    localeCode : 'cs',
    localeRtl  : false,

    EventEdit : {
        Calendar  : 'Kalendář',
        'All day' : 'Celý den',
        day       : 'Den',
        week      : 'Týden',
        month     : 'Měsíc',
        year      : 'Rok',
        decade    : 'Desetiletí'
    },

    EventMenu : {
        duplicateEvent : 'Duplikovat událost',
        copy           : 'kopírovat'
    },

    Calendar : {
        toggleSidebar : 'Přepnout viditelnost postranního panelu',
        Today         : 'Dnes',
        Tomorrow      : 'Zítra',
        next          : range => `Další ${range}`,
        previous      : range => `Předchozí ${range}`,
        plusMore      : value => `+${value} další`,
        moreEvents    : value => `${value} další události`,
        allDay        : 'Celý den',
        endsOn        : d => `Končí ${d}`,
        weekOfYear    : ([y, w]) => `Týden ${w}, ${y}`,
        loadFail      : 'Načtení dat kalendáře se nezdařilo. Obraťte se na správce systému',
        gotoDate      : 'Přejít na datum',
        timeRange     : 'Časový rozsah'
    },

    CalendarDrag : {
        holdCtrlForRecurrence : 'Při opakující se události podržte CTRL'
    },

    CalendarMixin : {
        eventCount : count => {
            if (!count) {
                return 'Žádná událost';
            }
            else {
                const lastDigit = count[count.length - 1];
                const suffix = { 1 : '', 2 : 'i', 3 : 'i', 4 : 'i' }[lastDigit] || 'í';

                return count + ' událost' + suffix;
            }
        }
    },

    EventTip : {
        'Edit event' : 'Upravit událost',
        timeFormat   : 'LST'
    },

    ModeSelector : {
        includeWeekends : 'Zahrnout víkendy',
        weekends        : 'Víkendy'
    },

    AgendaView : {
        Agenda : 'Program'
    },

    MonthView : {
        Month     : 'Měsíc',
        monthUnit : 'měsíc'
    },

    WeekView : {
        weekUnit : 'týden'
    },

    YearView : {
        Year     : 'Rok',
        yearUnit : 'rok',
        noEvents : 'Žádné události'
    },

    EventList : {
        List   : 'Seznam',
        Start  : 'Začátek',
        Finish : 'Konec',
        days   : count => `${count > 1 ? `${count} ` : ''}den${count === 1 ? '' : 'ů'}`
    },

    DayView : {
        Day                : 'Den',
        dayUnit            : 'den',
        daysUnit           : 'dny',
        expandAllDayRow    : 'Rozšířit výběr na všechny dny',
        collapseAllDayRow  : 'Sloučit výběr na všechny dny',
        timeFormat         : 'LST',
        timeAxisTimeFormat : 'LST',
        noMatchingDates    : 'Žádné odpovídající data'
    },

    DayAgendaView : {
        dayAgenda : 'Denní program'
    },

    DayResourceView : {
        dayResourceView : 'Denní zdroje'
    },

    Sidebar : {
        selectDate      : 'Vyberte datum kalendáře',
        matchEvents     : 'Události zápasu',
        filterEvents    : 'Filtrovat události podle názvu události',
        filterResources : 'Filtrovat události podle vybraného zdroje'
    },

    WeekExpander : {
        expandTip   : 'Kliknutím rozšíříte řádek',
        collapseTip : 'Kliknutím sloučíte řádek'
    }
};

export default LocaleHelper.publishLocale(locale);
