// @bryntum/calendar/source/lib/Calendar/localization/Da.js
import LocaleHelper from '../../Core/localization/LocaleHelper.js';
import '../../Scheduler/localization/Da.js';

const locale = {

    localeName : 'Da',
    localeDesc : 'Dansk',
    localeCode : 'da',
    localeRtl  : false,

    EventEdit : {
        Calendar  : 'Kalender',
        'All day' : 'Hele dag',
        day       : 'dagen',
        week      : 'Uge',
        month     : 'Måned',
        year      : 'År',
        decade    : 'Årti'
    },

    EventMenu : {
        duplicateEvent : 'Dubleret begivenhed',
        copy           : 'kopi'
    },

    Calendar : {
        toggleSidebar : 'Skift sidebjælkevisning',
        Today         : 'I dag',
        Tomorrow      : 'I morgen',
        next          : range => `Næste ${range}`,
        previous      : range => `Tidligere ${range}`,
        plusMore      : value => `+${value} mere`,
        moreEvents    : value => `${value} flere begivenheder`,
        allDay        : 'Hele dagen',
        endsOn        : d => `Slutter ${d}`,
        weekOfYear    : ([y, w]) => `Uge ${w}, ${y}`,
        loadFail      : 'Indlæsning af kalenderdata mislykkedes. Kontakt venligst din systemadministrator',
        gotoDate      : 'Gå til dato',
        timeRange     : 'Tidsinterval'
    },

    CalendarDrag : {
        holdCtrlForRecurrence : 'Hold CTRL nede for en tilbagevendende begivenhed'
    },

    CalendarMixin : {
        eventCount : count => `${count || 'Ingen'} begivenhed${(count && count > 1) ? 'er' : ''}`
    },

    EventTip : {
        'Edit event' : 'Rediger begivenhed',
        timeFormat   : 'LST'
    },

    ModeSelector : {
        includeWeekends : 'Inkluder weekender',
        weekends        : 'Weekender'
    },

    AgendaView : {
        Agenda : 'Dagsorden'
    },

    MonthView : {
        Month     : 'måned',
        monthUnit : 'måned'
    },

    WeekView : {
        weekUnit : 'uge'
    },

    YearView : {
        Year     : 'år',
        yearUnit : 'år',
        noEvents : 'Ingen begivenheder'
    },

    EventList : {
        List   : 'Liste',
        Start  : 'Start',
        Finish : 'Afslut',
        days   : count => `${count > 1 ? `${count} ` : ''}dag${count === 1 ? '' : 'e'}`
    },

    DayView : {
        Day                : 'dag',
        dayUnit            : 'dag',
        daysUnit           : 'dage',
        expandAllDayRow    : 'Udvid heldagsafsnittet',
        collapseAllDayRow  : 'Skjul heldagsafsnittet',
        timeFormat         : 'LST',
        timeAxisTimeFormat : 'LST',
        noMatchingDates    : 'Ingen matchende datoer'
    },

    DayAgendaView : {
        dayAgenda : 'Dagsorden'
    },

    DayResourceView : {
        dayResourceView : 'Dagsressourcer'
    },

    Sidebar : {
        selectDate      : 'Vælg kalenderdato',
        matchEvents     : 'Kampbegivenheder',
        filterEvents    : 'Filtrer begivenheder efter begivenhedsnavn',
        filterResources : 'Filtrer begivenheder efter valgt ressource'
    },

    WeekExpander : {
        expandTip   : 'Klik for at udvide rækken',
        collapseTip : 'Klik for at skjule rækken'
    }
};

export default LocaleHelper.publishLocale(locale);
