// @bryntum/calendar/source/lib/Calendar/localization/De.js
import StringHelper from '../../Core/helper/StringHelper.js';
import LocaleHelper from '../../Core/localization/LocaleHelper.js';
import '../../Scheduler/localization/De.js';

const
    next = {
        tag       : 'Nächster',
        woche     : 'Nächste',
        monat     : 'Nächster',
        jahr      : 'Nächstes',
        jahrzehnt : 'Nächstes'
    },
    previous = {
        tag       : 'Vorheriger',
        woche     : 'Vorherige',
        monat     : 'Vorheriger',
        jahr      : 'Vorheriges',
        jahrzehnt : 'Vorheriges'
    };

const locale = {

    localeName : 'De',
    localeDesc : 'Deutsch',
    localeCode : 'de-DE',
    localeRtl  : false,

    EventEdit : {
        Calendar  : 'Kalendar',
        'All day' : 'Ganztägig',
        day       : 'Tag',
        week      : 'Woche',
        month     : 'Monat',
        year      : 'Jahr',
        decade    : 'Jahrzehnt'
    },

    EventMenu : {
        duplicateEvent : 'Ereignis duplizieren',
        copy           : 'kopieren'
    },

    Calendar : {
        toggleSidebar : 'Sichtbarkeit der Seitenleiste umschalten',
        Today         : 'Heute',
        Tomorrow      : 'Morgen',
        next          : range => `${next[range]} ${StringHelper.capitalize(range)}`,
        previous      : range => `${previous[range]} ${StringHelper.capitalize(range)}`,
        plusMore      : value => `+${value} mehr`,
        moreEvents    : value => `${value} weitere Ereignisse`,
        allDay        : 'Ganztägig',
        endsOn        : d => `Endet ${d}`,
        weekOfYear    : ([y, w]) => `Woche ${w}, ${y}`,
        loadFail      : 'Kalenderdaten konnten nicht geladen werden. Bitte wenden Sie sich an Ihren Systemadministrator',
        gotoDate      : 'Gehe zu Datum',
        timeRange     : 'Zeitraum'
    },

    CalendarDrag : {
        holdCtrlForRecurrence : 'CTRL für ein wiederkehrendes Ereignis halten'
    },

    CalendarMixin : {
        eventCount : count => `${count || 'Kein'} Ereignis${(count && count > 1) ? 'sen' : ''}`
    },

    EventTip : {
        'Edit event' : 'Ereignis bearbeiten',
        timeFormat   : 'LST'
    },

    ModeSelector : {
        includeWeekends : 'Wochenenden einbeziehen',
        weekends        : 'Wochenenden'
    },

    AgendaView : {
        Agenda : 'Agenda'
    },

    MonthView : {
        Month     : 'Monat',
        monthUnit : 'monat'
    },

    WeekView : {
        weekUnit : 'woche'
    },

    YearView : {
        Year     : 'Jahr',
        yearUnit : 'jahr',
        noEvents : 'Keine Veranstaltungen'
    },

    EventList : {
        List   : 'Liste',
        Start  : 'Start',
        Finish : 'Ende',
        days   : count => `${count > 1 ? `${count} ` : ''}Tag${count === 1 ? '' : 'e'}`
    },

    DayView : {
        Day                : 'Tag',
        dayUnit            : 'tag',
        daysUnit           : 'tage',
        expandAllDayRow    : 'Ganztagsbereich erweitern',
        collapseAllDayRow  : 'Ganztagsbereich einklappen',
        timeFormat         : 'LST',
        timeAxisTimeFormat : 'LST',
        noMatchingDates    : 'Keine passenden Daten'
    },

    DayAgendaView : {
        dayAgenda : 'Tagesordnung'
    },

    DayResourceView : {
        dayResourceView : 'Tagesressourcen'
    },

    Sidebar : {
        selectDate      : 'Kalenderdatum auswählen',
        matchEvents     : 'Spielereignisse',
        filterEvents    : 'Ereignisse nach Ereignisnamen filtern',
        filterResources : 'Ereignisse nach ausgewählter Ressource filtern'
    },

    WeekExpander : {
        expandTip   : 'Zum Erweitern der Zeile klicken',
        collapseTip : 'Zum Einklappen der Zeile klicken'
    }
};

export default LocaleHelper.publishLocale(locale);
