// @bryntum/calendar/source/lib/Calendar/localization/El.js
import LocaleHelper from '../../Core/localization/LocaleHelper.js';
import '../../Scheduler/localization/El.js';

const locale = {

    localeName : 'El',
    localeDesc : 'Ελληνικά',
    localeCode : 'el',
    localeRtl  : false,

    EventEdit : {
        Calendar  : 'Ημερολόγιο',
        'All day' : 'Ολοήμερο',
        day       : 'Ημέρα',
        week      : 'Εβδομάδα',
        month     : 'Μήνας',
        year      : 'Έτος',
        decade    : 'Δεκαετία'
    },

    EventMenu : {
        duplicateEvent : 'Αντιγραφή συμβάντος',
        copy           : 'αντιγραφή'
    },

    Calendar : {
        toggleSidebar : 'Εναλλαγή ορατότητας πλευρικής μπάρας',
        Today         : 'Σήμερα',
        Tomorrow      : 'Αύριο',
        next          : range => `Επόμενο/η ${range}`,
        previous      : range => `Προηγούμενο/η ${range}`,
        plusMore      : value => `+${value} παραπάνω`,
        moreEvents    : value => `${value} περισσότερες εκδηλώσεις`,
        allDay        : 'Ολοήμερο',
        endsOn        : d => `Τελειώνει στις ${d}`,
        weekOfYear    : ([y, w]) => `Εβδομάδα ${w}, ${y}`,
        loadFail      : 'Η φόρτωση δεδομένων ημερολογίου απέτυχε. Επικοινωνήστε με το διαχειριστή του συστήματός σας.',
        gotoDate      : 'Μετάβαση στην ημερομηνία',
        timeRange     : 'Εύρος χρόνου'
    },

    CalendarDrag : {
        holdCtrlForRecurrence : 'Για επαναλαμβανόμενο συμβάν, κρατήστε πατημένο το πλήκτρο CTRL'
    },

    CalendarMixin : {
        eventCount : count => `${count || 'Κανένα'} συμβάν${(count && count > 1) ? 'εις' : ''}`
    },

    EventTip : {
        'Edit event' : 'Επεξεργασία συμβάντος',
        timeFormat   : 'LST'
    },

    ModeSelector : {
        includeWeekends : 'Συμπερίληψη Σαββατοκύριακων',
        weekends        : 'Σαββατοκύριακα'
    },

    AgendaView : {
        Agenda : 'Πρόγραμμα'
    },

    MonthView : {
        Month     : 'Μήνας',
        monthUnit : 'μήνας'
    },

    WeekView : {
        weekUnit : 'εβδομάδα'
    },

    YearView : {
        Year     : 'Έτος',
        yearUnit : 'έτος',
        noEvents : 'Δεν υπάρχουν εκδηλώσεις'
    },

    EventList : {
        List   : 'Λίστα',
        Start  : 'Αρχή',
        Finish : 'Τέλος',
        days   : count => `${count > 1 ? `${count} ` : ''}ημέρα${count === 1 ? '' : 'ς'}`
    },

    DayView : {
        Day                : 'Ημέρα',
        dayUnit            : 'ημέρα',
        daysUnit           : 'ημέρες',
        expandAllDayRow    : 'Επεκτείνετε την ενότητα του ολοήμερου',
        collapseAllDayRow  : 'Καταργήστε την ενότητα του ολοήμερου',
        timeFormat         : 'LST',
        timeAxisTimeFormat : 'LST',
        noMatchingDates    : 'Δεν υπάρχουν αντίστοιχες ημερομηνίες'
    },

    DayAgendaView : {
        dayAgenda : 'Ημερήσια ατζέντα'
    },

    DayResourceView : {
        dayResourceView : 'Ημερήσιοι πόροι'
    },

    Sidebar : {
        selectDate      : 'Επιλέξτε ημερομηνία ημερολογίου',
        matchEvents     : 'Συμβάντα αγώνα',
        filterEvents    : 'Φιλτράρισμα συμβάντων ανά όνομα συμβάντος',
        filterResources : 'Φιλτράρισμα συμβάντων ανά επιλεγμένο πόρο'
    },

    WeekExpander : {
        expandTip   : 'Κάντε κλικ για να επεκτείνετε τη σειρά',
        collapseTip : 'Κάντε κλικ για να καταργήσετε τη σειρά'
    }
};

export default LocaleHelper.publishLocale(locale);
