// @bryntum/calendar/source/lib/Calendar/localization/Et.js
import LocaleHelper from '../../Core/localization/LocaleHelper.js';
import '../../Scheduler/localization/Et.js';

const locale = {

    localeName : 'Et',
    localeDesc : 'Eesti keel',
    localeCode : 'et',
    localeRtl  : false,

    EventEdit : {
        Calendar  : 'Kalender',
        'All day' : 'Kogu päev',
        day       : 'Päev',
        week      : 'Nädal',
        month     : 'Kuu',
        year      : 'Aasta',
        decade    : 'Kümnend'
    },

    EventMenu : {
        duplicateEvent : 'Dubleeri sündmus',
        copy           : 'kopeeri'
    },

    Calendar : {
        toggleSidebar : 'Külgriba nähtavuse lülitamine',
        Today         : 'Täna',
        Tomorrow      : 'Homme',
        next          : range => `Järgmine ${range}`,
        previous      : range => `Eelmine ${range}`,
        plusMore      : value => `+${value} veel`,
        moreEvents    : value => `${value} rohkem sündmusi`,
        allDay        : 'Kogu päev',
        endsOn        : d => `Lõpeb ${d}`,
        weekOfYear    : ([y, w]) => `Nädal ${w}, ${y}`,
        loadFail      : 'Kalendriandmete laadimine nurjus. Palun võtke ühendust oma süsteemiadministraatoriga',
        gotoDate      : 'Mine kuupäevale',
        timeRange     : 'Ajavahemik'
    },

    CalendarDrag : {
        holdCtrlForRecurrence : 'Korduva sündmuse korral hoidke all klahvi CTRL'
    },

    CalendarMixin : {
        eventCount : count => `${count || 'Ei'} sündmust${(count && count > 1) ? 'ed' : ''}`
    },

    EventTip : {
        'Edit event' : 'Redigeeri sündmust',
        timeFormat   : 'LST'
    },

    ModeSelector : {
        includeWeekends : 'Kaasa nädalavahetused',
        weekends        : 'Nädalavahetused'
    },

    AgendaView : {
        Agenda : 'Tegevuskava'
    },

    MonthView : {
        Month     : 'Kuu',
        monthUnit : 'kuu'
    },

    WeekView : {
        weekUnit : 'nädal'
    },

    YearView : {
        Year     : 'Aasta',
        yearUnit : 'aasta',
        noEvents : 'Sündmusi pole'
    },

    EventList : {
        List   : 'Loend',
        Start  : 'Algus',
        Finish : 'Lõpp',
        days   : count => `${count > 1 ? `${count} ` : ''}päev${count === 1 ? '' : 'a'}`
    },

    DayView : {
        Day                : 'Päev',
        dayUnit            : 'päev',
        daysUnit           : 'päeva',
        expandAllDayRow    : 'Laienda kogu päeva jaotist',
        collapseAllDayRow  : 'Koonda kogu päeva jaotis',
        timeFormat         : 'LST',
        timeAxisTimeFormat : 'LST',
        noMatchingDates    : 'Vastavaid kuupäevi pole'
    },

    DayAgendaView : {
        dayAgenda : 'Päevakava'
    },

    DayResourceView : {
        dayResourceView : 'Päevased ressursid'
    },

    Sidebar : {
        selectDate      : 'Vali kuupäeva kalendrist',
        matchEvents     : 'Matši sündmused',
        filterEvents    : 'Filtreeri sündmusi sündmuse nime järgi',
        filterResources : 'Filtreeri sündmusi valitud ressursi järgi'
    },

    WeekExpander : {
        expandTip   : 'Klõpsa rea laiendamiseks',
        collapseTip : 'Klõpsa rea koondamiseks'
    }
};

export default LocaleHelper.publishLocale(locale);
