// @bryntum/calendar/source/lib/Calendar/localization/Eu.js
import LocaleHelper from '../../Core/localization/LocaleHelper.js';
import '../../Scheduler/localization/Eu.js';

const locale = {

    localeName : 'Eu',
    localeDesc : 'Euskara',
    localeCode : 'eu-ES',
    localeRtl  : false,

    EventEdit : {
        Calendar  : 'Egutegia',
        'All day' : 'Egun osoa',
        day       : 'Eguna',
        week      : 'Astea',
        month     : 'Hilabetea',
        year      : 'Urtea',
        decade    : 'Hamarkada'
    },

    EventMenu : {
        duplicateEvent : 'Gertaera bikoiztua',
        copy           : 'kopiatu'
    },

    Calendar : {
        toggleSidebar : 'Alboko barraren ikusgarritasuna aldatu',
        Today         : 'Gaur',
        Tomorrow      : 'Bihar',
        next          : range => `Hurrengo ${range}`,
        previous      : range => `Aurreko ${range}`,
        plusMore      : value => `+${value} gehiago`,
        moreEvents    : value => `${value} gertaera gehiago`,
        allDay        : 'Egun osoa',
        endsOn        : d => `Bukatuta ${d}`,
        weekOfYear    : ([y, w]) => `Astea ${w}, ${y}`,
        loadFail      : 'Egutegiko datuak kargatzeak huts egin du. Jarri harremanetan sistema-administratzailearekin',
        gotoDate      : 'Joan datara',
        timeRange     : 'Denbora tartea'
    },

    CalendarDrag : {
        holdCtrlForRecurrence : 'Sakatu KTRL gertaera errepikari baterako'
    },

    CalendarMixin : {
        eventCount : count => `${count || 'Ez'} event${count === 1 ? '' : 's'}`
    },

    EventTip : {
        'Edit event' : 'Editatu gertaera',
        timeFormat   : 'LST'
    },

    ModeSelector : {
        includeWeekends : 'Asteburuak barne',
        weekends        : 'Asteburuak'
    },

    AgendaView : {
        Agenda : 'Agenda'
    },

    MonthView : {
        Month     : 'Hilabetea',
        monthUnit : 'hilabete'
    },

    WeekView : {
        weekUnit : 'aste'
    },

    YearView : {
        Year     : 'Urtea',
        yearUnit : 'urte',
        noEvents : 'Gertaerarik gabe'
    },

    EventList : {
        List   : 'Zerrenda',
        Start  : 'Hasiera',
        Finish : 'Amaiera',
        days   : count => `${count > 1 ? `${count} ` : ''}day${count === 1 ? '' : 's'}`
    },

    DayView : {
        Day                : 'Eguna',
        dayUnit            : 'egun',
        daysUnit           : 'egun',
        expandAllDayRow    : 'Zabaldu egun osoko sekzioa',
        collapseAllDayRow  : 'Itxi egun osoko sekzioa',
        timeFormat         : 'LST',
        timeAxisTimeFormat : 'LST',
        noMatchingDates    : 'Ez dago bat datorren datarik'
    },

    DayAgendaView : {
        dayAgenda : 'Eguneko agenda'
    },

    DayResourceView : {
        dayResourceView : 'Eguneko baliabideak'
    },

    Sidebar : {
        selectDate      : 'Hautatu egutegi-data',
        matchEvents     : 'Partidaren gertaerak',
        filterEvents    : 'Iragazi gertaerak gertaera izenaren arabera',
        filterResources : 'Iragazi gertaerak aukeratutako baliabidearen arabera'
    },

    WeekExpander : {
        expandTip   : 'Egin klik errenkada zabaltzeko',
        collapseTip : 'Egin klik errenkada ixteko'
    }
};

export default LocaleHelper.publishLocale(locale);
