// @bryntum/calendar/source/lib/Calendar/localization/Fi.js
import LocaleHelper from '../../Core/localization/LocaleHelper.js';
import '../../Scheduler/localization/Fi.js';

const locale = {

    localeName : 'Fi',
    localeDesc : 'Suomi',
    localeCode : 'fi',
    localeRtl  : false,

    EventEdit : {
        Calendar  : 'Kalenteri',
        'All day' : 'Koko päivä',
        day       : 'Päivä',
        week      : 'Viikko',
        month     : 'Kuukausi',
        year      : 'Vuosi',
        decade    : 'Vuosikymmen'
    },

    EventMenu : {
        duplicateEvent : 'Päällekkäinen tapahtuma',
        copy           : 'kopio'
    },

    Calendar : {
        toggleSidebar : 'Vaihda sivupalkin näkyvyys',
        Today         : 'Tänään',
        Tomorrow      : 'Huomenna',
        next          : range => `Seuraava ${range}`,
        previous      : range => `Edellinen ${range}`,
        plusMore      : value => `+${value} lisää`,
        moreEvents    : value => `${value} lisää tapahtumia`,
        allDay        : 'Koko päivä',
        endsOn        : d => `Loppuu ${d}`,
        weekOfYear    : ([y, w]) => `Viikko ${w}, ${y}`,
        loadFail      : 'Kalenteritietojen lataus epäonnistui. Ota yhteyttä järjestelmänvalvojaasi',
        gotoDate      : 'Siirry päivämäärään',
        timeRange     : 'Aikaväli'
    },

    CalendarDrag : {
        holdCtrlForRecurrence : 'Pidä CTRL pohjassa toistuvaa tapahtumaa varten'
    },

    CalendarMixin : {
        eventCount : count => `${count || 'Ei'} tapahtuma${(count && count > 1) ? 't' : 'a'}`
    },

    EventTip : {
        'Edit event' : 'Muokkaa tapahtumaa',
        timeFormat   : 'LST'
    },

    ModeSelector : {
        includeWeekends : 'Sisällytä viikonloput',
        weekends        : 'Viikonloput'
    },

    AgendaView : {
        Agenda : 'Asialista'
    },

    MonthView : {
        Month     : 'Kuukausi',
        monthUnit : 'kuukausi'
    },

    WeekView : {
        weekUnit : 'viikko'
    },

    YearView : {
        Year     : 'Vuosi',
        yearUnit : 'vuosi',
        noEvents : 'Ei tapahtumia'
    },

    EventList : {
        List   : 'Lista',
        Start  : 'Aloita',
        Finish : 'Lopeta',
        days   : count => `${count > 1 ? `${count} ` : ''}päivä${count === 1 ? '' : 'ä'}`
    },

    DayView : {
        Day                : 'Päivä',
        dayUnit            : 'päivä',
        daysUnit           : 'päivät',
        expandAllDayRow    : 'Laajenna kokopäivä osio',
        collapseAllDayRow  : 'Kokopäivän osion sulkeminen',
        timeFormat         : 'LST',
        timeAxisTimeFormat : 'LST',
        noMatchingDates    : 'Ei vastaavia päivämääriä'
    },

    DayAgendaView : {
        dayAgenda : 'Päiväohjelma'
    },

    DayResourceView : {
        dayResourceView : 'Päiväresurssit'
    },

    Sidebar : {
        selectDate      : 'Valitse kalenteripäivämäärä',
        matchEvents     : 'Ottelutapahtumat',
        filterEvents    : 'Suodata tapahtumia tapahtuman nimen mukaan',
        filterResources : 'Suodata tapahtumia valitun resurssin mukaan'
    },

    WeekExpander : {
        expandTip   : 'Napsauta laajentaaksesi riviä',
        collapseTip : 'Napsauta sulkeaksesi rivin'
    }
};

export default LocaleHelper.publishLocale(locale);
