// @bryntum/calendar/source/lib/Calendar/localization/FrFr.js
import LocaleHelper from '../../Core/localization/LocaleHelper.js';
import '../../Scheduler/localization/FrFr.js';

const locale = {

    localeName : 'FrFr',
    localeDesc : 'Français (France)',
    localeCode : 'fr-FR',
    localeRtl  : false,

    EventEdit : {
        Calendar  : 'Calendrier',
        'All day' : 'Jour entier',
        day       : 'Jour',
        week      : 'Semaine',
        month     : 'Mois',
        year      : 'Année',
        decade    : 'Décennie'
    },

    EventMenu : {
        duplicateEvent : "Dupliquer l'événement",
        copy           : 'copier'
    },

    Calendar : {
        toggleSidebar : 'Basculer la visibilité de la barre latérale',
        Today         : "Aujourd'hui",
        Tomorrow      : 'Demain',
        next          : range => `Suivant ${range}`,
        previous      : range => `Précédent ${range}`,
        plusMore      : value => `+${value} de plus`,
        moreEvents    : value => `${value} événements supplémentaires`,
        allDay        : 'Jour entier',
        endsOn        : d => `Prend fin le ${d}`,
        weekOfYear    : ([y, w]) => `Semaine ${w}, ${y}`,
        loadFail      : 'Échec du chargement des données de calendrier. Veuillez contacter votre administrateur système',
        gotoDate      : 'Aller à la date',
        timeRange     : 'Plage horaire'
    },

    CalendarDrag : {
        holdCtrlForRecurrence : 'Maintenir la touche CTRL enfoncée pour un événement récurrent'
    },

    CalendarMixin : {
        eventCount : count => `${count || 'Aucun'} événement${(count && count > 1) ? 's' : '' }`
    },

    EventTip : {
        'Edit event' : "Éditer l'événement",
        timeFormat   : 'LST'
    },

    ModeSelector : {
        includeWeekends : 'Inclure les week-ends',
        weekends        : 'Week-ends'
    },

    AgendaView : {
        Agenda : 'Agenda'
    },

    MonthView : {
        Month     : 'Mois',
        monthUnit : 'mois'
    },

    WeekView : {
        weekUnit : 'semaine'
    },

    YearView : {
        Year     : 'Année',
        yearUnit : 'année',
        noEvents : "Pas d'événements"
    },

    EventList : {
        List   : 'Liste',
        Start  : 'Début',
        Finish : 'Fin',
        days   : count => `${count > 1 ? `${count} ` : ''}jour${count === 1 ? '' : 's'}`
    },

    DayView : {
        Day                : 'Jour',
        dayUnit            : 'jour',
        daysUnit           : 'jours',
        expandAllDayRow    : 'Développer la section jour entier',
        collapseAllDayRow  : 'Réduire la section jour entier',
        timeFormat         : 'LST',
        timeAxisTimeFormat : 'LST',
        noMatchingDates    : ' Aucune date correspondante'
    },

    DayAgendaView : {
        dayAgenda : 'Agenda du jour'
    },

    DayResourceView : {
        dayResourceView : 'Ressources quotidiennes'
    },

    Sidebar : {
        selectDate      : 'Sélectionner la date du calendrier',
        matchEvents     : 'Événements de match',
        filterEvents    : "Filtrer les événements par nom de l'événement",
        filterResources : 'Filtrer les événements par ressource sélectionnée'
    },

    WeekExpander : {
        expandTip   : 'Cliquez pour développer la ligne',
        collapseTip : 'Cliquez pour réduire la ligne'
    }
};

export default LocaleHelper.publishLocale(locale);
