// @bryntum/calendar/source/lib/Calendar/localization/Gl.js
import LocaleHelper from '../../Core/localization/LocaleHelper.js';
import '../../Scheduler/localization/Gl.js';

const locale = {

    localeName : 'Gl',
    localeDesc : 'Galego',
    localeCode : 'gl-ES',
    localeRtl  : false,

    EventEdit : {
        Calendar  : 'Calendario',
        'All day' : 'Todo o día',
        day       : 'Día',
        week      : 'Semana',
        month     : 'Mes',
        year      : 'Ano',
        decade    : 'Década'
    },

    EventMenu : {
        duplicateEvent : 'Duplicar evento',
        copy           : 'copiar'
    },

    Calendar : {
        toggleSidebar : 'Alternar visibilidade da barra lateral',
        Today         : 'Hoxe',
        Tomorrow      : 'Mañá',
        next          : range => `${range} seguinte`,
        previous      : range => `${range} anterior`,
        plusMore      : value => `+${value} máis`,
        moreEvents    : value => `${value} eventos máis`,
        allDay        : 'Todo o día',
        endsOn        : d => `Finaliza o ${d}`,
        weekOfYear    : ([y, w]) => `Semana ${w}, ${y}`,
        loadFail      : 'Non se puideron cargar os datos do calendario. Póñase en conatacto co administrador do sistema',
        gotoDate      : 'Ir á data',
        timeRange     : 'Intervalo de tempo'
    },

    CalendarDrag : {
        holdCtrlForRecurrence : 'Manteña premido CTRL para un evento recorrente'
    },

    CalendarMixin : {
        eventCount : count => `${count || 'Non'} event${count === 1 ? '' : 's'}`
    },

    EventTip : {
        'Edit event' : 'Editar evento',
        timeFormat   : 'LST'
    },

    ModeSelector : {
        includeWeekends : 'Incluír fins de semana',
        weekends        : 'Fins de semana'
    },

    AgendaView : {
        Agenda : 'Axenda'
    },

    MonthView : {
        Month     : 'Mes',
        monthUnit : 'mes'
    },

    WeekView : {
        weekUnit : 'semana'
    },

    YearView : {
        Year     : 'Ano',
        yearUnit : 'ano',
        noEvents : 'Sen eventos'
    },

    EventList : {
        List   : 'Lista',
        Start  : 'Inicio',
        Finish : 'Fin',
        days   : count => `${count > 1 ? `${count} ` : ''}day${count === 1 ? '' : 's'}`
    },

    DayView : {
        Day                : 'Día',
        dayUnit            : 'día',
        daysUnit           : 'días',
        expandAllDayRow    : 'Expandir a sección todo o día',
        collapseAllDayRow  : 'Contraer a sección todo o día',
        timeFormat         : 'LST',
        timeAxisTimeFormat : 'LST',
        noMatchingDates    : 'Non hai datas coincidentes'
    },

    DayAgendaView : {
        dayAgenda : 'Axenda do día'
    },

    DayResourceView : {
        dayResourceView : 'Recursos de día'
    },

    Sidebar : {
        selectDate      : 'Selecciona a data do calendario',
        matchEvents     : 'Eventos do partido',
        filterEvents    : 'Filtrar eventos por nome do evento',
        filterResources : 'Filtrar eventos polo recurso seleccionado'
    },

    WeekExpander : {
        expandTip   : 'Prema para expandir a fila',
        collapseTip : 'Prema para contraer a fila'
    }
};

export default LocaleHelper.publishLocale(locale);
