// @bryntum/calendar/source/lib/Calendar/localization/He.js
import LocaleHelper from '../../Core/localization/LocaleHelper.js';
import '../../Scheduler/localization/He.js';

const locale = {

    localeName : 'He',
    localeDesc : 'עִברִית',
    localeCode : 'he',
    localeRtl  : true,

    EventEdit : {
        Calendar  : 'לוח שנה',
        'All day' : 'כל היום',
        day       : 'יום',
        week      : 'שבוע',
        month     : 'חודש',
        year      : 'שנה',
        decade    : 'עשור'
    },

    EventMenu : {
        duplicateEvent : 'הכפלת אירוע',
        copy           : 'העתק'
    },

    Calendar : {
        toggleSidebar : 'החלפת נראות של סרגל הצד',
        Today         : 'היום',
        Tomorrow      : 'מחר',
        next          : range => `${range} הבא`,
        previous      : range => `${range} קודם`,
        plusMore      : value => `+${value} עוד`,
        moreEvents    : value => `עוד ${value} אירועים`,
        allDay        : 'כל היום',
        endsOn        : d => `${d} מסתיים`,
        weekOfYear    : ([y, w]) => `${y}, ${w}, שבוע`,
        loadFail      : 'טעינת נתוני לוח השנה כשלה. אנא פנה למנהל המערכת שלך.',
        gotoDate      : 'עבור לתאריך',
        timeRange     : 'טווח זמן'
    },

    CalendarDrag : {
        holdCtrlForRecurrence : 'לחץ לחיצה ממושכת על מקש קונטרול לאירוע חוזר'
    },

    CalendarMixin : {
        eventCount : count => `${(count && count > 1) ? 'ים' : ''}אירוע ${count || 'אין'}`
    },

    EventTip : {
        'Edit event' : 'ערוך אירוע',
        timeFormat   : 'LST'
    },

    ModeSelector : {
        includeWeekends : 'כלול סופי שבוע',
        weekends        : 'סופי שבוע'
    },

    AgendaView : {
        Agenda : 'יומן'
    },

    MonthView : {
        Month     : 'חודש',
        monthUnit : 'חודש'
    },

    WeekView : {
        weekUnit : 'שבוע'
    },

    YearView : {
        Year     : 'שנה',
        yearUnit : 'שנה',
        noEvents : 'אין אירועים'
    },

    EventList : {
        List   : 'רשימה',
        Start  : 'התחל',
        Finish : 'סיים',
        days   : count => `${count === 1 ? 'יְוֹם' : 'ימים'}${count > 1 ? ` ${count}` : ''}`
    },

    DayView : {
        Day                : 'יום',
        dayUnit            : 'יום',
        daysUnit           : 'ימים',
        expandAllDayRow    : 'הרחב את מקטע כל הימים',
        collapseAllDayRow  : 'מזער את מקטע כל הימים',
        timeFormat         : 'LST',
        timeAxisTimeFormat : 'LST',
        noMatchingDates    : 'אין תאריכים תואמים'
    },

    DayAgendaView : {
        dayAgenda : 'לוח זמנים יומי'
    },

    DayResourceView : {
        dayResourceView : 'משאבי יום'
    },

    Sidebar : {
        selectDate      : 'בחר תאריך בלוח שנה',
        matchEvents     : 'אירועי משחק',
        filterEvents    : 'סנן אירועים לפי שם אירוע',
        filterResources : 'סנן אירועים לפי משאב נבחר'
    },

    WeekExpander : {
        expandTip   : 'הקלק להרחבת השורה',
        collapseTip : 'הקלק למזעור השורה'
    }
};

export default LocaleHelper.publishLocale(locale);
