// @bryntum/calendar/source/lib/Calendar/localization/Hi.js
import LocaleHelper from '../../Core/localization/LocaleHelper.js';
import '../../Scheduler/localization/Hi.js';

const locale = {

    localeName : 'Hi',
    localeDesc : 'हिन्दी',
    localeCode : 'hi',
    localeRtl  : false,

    EventEdit : {
        Calendar  : 'कैलेन्डर',
        'All day' : 'सारा दिन',
        day       : 'दिन',
        week      : 'सप्ताह',
        month     : 'महीना',
        year      : 'साल',
        decade    : 'दशक'
    },

    EventMenu : {
        duplicateEvent : 'डुप्लीकेट ईवेंट',
        copy           : 'कॉपी करें'
    },

    Calendar : {
        toggleSidebar : 'साइडबार दृश्यता को टॉगल करें',
        Today         : 'आज',
        Tomorrow      : 'कल',
        next          : range => `अगली ${range}`,
        previous      : range => `पिछली ${range}`,
        plusMore      : value => `+${value} और`,
        moreEvents    : value => `${value} अधिक इवेंट्स`,
        allDay        : 'सारा दिन',
        endsOn        : d => `समाप्ति तिथि ${d}`,
        weekOfYear    : ([y, w]) => `सप्ताह ${w}, ${y}`,
        loadFail      : 'कैलेंडर डेटा लोड विफल रहा। कृपया अपने सिस्टम प्रशासक से संपर्क करें',
        gotoDate      : 'तिथि पर जाएं',
        timeRange     : 'समय सीमा'
    },

    CalendarDrag : {
        holdCtrlForRecurrence : 'बार-बार होने वाले ईवेंट के लिए CTRL दबाएं'
    },

    CalendarMixin : {
        eventCount : count => `${count || 'शून्य'} ईवेंट`
    },

    EventTip : {
        'Edit event' : 'ईवेंट संपादित करें',
        timeFormat   : 'LST'
    },

    ModeSelector : {
        includeWeekends : 'सप्ताहांत शामिल करें',
        weekends        : 'सप्ताहांत'
    },

    AgendaView : {
        Agenda : 'एजेंडा'
    },

    MonthView : {
        Month     : 'महीना',
        monthUnit : 'महीना'
    },

    WeekView : {
        weekUnit : 'सप्ताह'
    },

    YearView : {
        Year     : 'साल',
        yearUnit : 'साल',
        noEvents : 'कोई घटनाएँ नहीं'
    },

    EventList : {
        List   : 'सूची',
        Start  : 'शुरू करें',
        Finish : 'समाप्त करें',
        days   : count => `${count > 1 ? `${count} ` : ''}दिन}`
    },

    DayView : {
        Day                : 'दिन',
        dayUnit            : 'दिन',
        daysUnit           : 'दिन',
        expandAllDayRow    : 'सारा-दिन सेक्शन फैलाएं',
        collapseAllDayRow  : 'सारा-दिन सेक्शन समेटें',
        timeFormat         : 'LST',
        timeAxisTimeFormat : 'LST',
        noMatchingDates    : 'कोई मेल खाने वाली तिथियाँ नहीं'
    },

    DayAgendaView : {
        dayAgenda : 'दिनचर्या'
    },

    DayResourceView : {
        dayResourceView : 'दैनिक संसाधन'
    },

    Sidebar : {
        selectDate      : 'कैलेंडर की तारीख का चयन करें',
        matchEvents     : 'मैच की घटनाएँ',
        filterEvents    : 'इवेंट नाम से इवेंट्स को फ़िल्टर करें',
        filterResources : 'चयनित संसाधन के अनुसार घटनाओं को फ़िल्टर करें'
    },

    WeekExpander : {
        expandTip   : 'पंक्ति फैलाने के लिए क्लिक करें',
        collapseTip : 'पंक्ति समेटने के लिए क्लिक करें'
    }
};

export default LocaleHelper.publishLocale(locale);
