// @bryntum/calendar/source/lib/Calendar/localization/Hu.js
import LocaleHelper from '../../Core/localization/LocaleHelper.js';
import '../../Scheduler/localization/Hu.js';

const locale = {

    localeName : 'Hu',
    localeDesc : 'Magyar',
    localeCode : 'hu',
    localeRtl  : false,

    EventEdit : {
        Calendar  : 'Naptár',
        'All day' : 'Egész nap',
        day       : 'Nap',
        week      : 'Hét',
        month     : 'Hónap',
        year      : 'Év',
        decade    : 'Évtized'
    },

    EventMenu : {
        duplicateEvent : 'Esemény duplikálása',
        copy           : 'másolás'
    },

    Calendar : {
        toggleSidebar : 'Oldalsáv láthatóságának váltása"',
        Today         : 'Másolás',
        Tomorrow      : 'Holnap',
        next          : range => `Következő ${range}`,
        previous      : range => `Előző ${range}`,
        plusMore      : value => `+${value} további`,
        moreEvents    : value => `${value} további események`,
        allDay        : 'Egész nap',
        endsOn        : d => `Vége ${d}`,
        weekOfYear    : ([y, w]) => `Hét ${w}, ${y}`,
        loadFail      : 'A naptáradatok betöltése sikertelen. Forduljon a rendszergazdához.',
        gotoDate      : 'Ugrás a dátumhoz',
        timeRange     : 'Időtartomány'
    },

    CalendarDrag : {
        holdCtrlForRecurrence : 'Ismétlődő eseményhez nyomja a CTRL-t'
    },

    CalendarMixin : {
        eventCount : count => `${count || 'Nincs'} esemény`
    },

    EventTip : {
        'Edit event' : 'Esemény szerkesztése',
        timeFormat   : 'LST'
    },

    ModeSelector : {
        includeWeekends : 'Hétvégék beszámítása',
        weekends        : 'Hétvégék'
    },

    AgendaView : {
        Agenda : 'Napirend'
    },

    MonthView : {
        Month     : 'Hónap',
        monthUnit : 'hónap'
    },

    WeekView : {
        weekUnit : 'hét'
    },

    YearView : {
        Year     : 'Év',
        yearUnit : 'év',
        noEvents : 'Nincsenek események'
    },

    EventList : {
        List   : 'Lista',
        Start  : 'Kezdés',
        Finish : 'Vége',
        days   : count => `${count > 1 ? `${count} ` : ''}nap${count === 1 ? '' : 'ok'}`
    },

    DayView : {
        Day                : 'Nap',
        dayUnit            : 'nap',
        daysUnit           : 'nap',
        expandAllDayRow    : 'Egész nap rész kibontása',
        collapseAllDayRow  : 'Egész nap rész összecsukása',
        timeFormat         : 'LST',
        timeAxisTimeFormat : 'LST',
        noMatchingDates    : 'Nincs egyező dátum'
    },

    DayAgendaView : {
        dayAgenda : 'Napi napirend'
    },

    DayResourceView : {
        dayResourceView : 'Napi erőforrások'
    },

    Sidebar : {
        selectDate      : 'Válasszon naptári dátumot',
        matchEvents     : 'Mérkőzés események',
        filterEvents    : 'Események szűrése eseménynév alapján',
        filterResources : 'Események szűrése kiválasztott erőforrás alapján'
    },

    WeekExpander : {
        expandTip   : 'Kattintson a sor kibontásához',
        collapseTip : 'Kattintson a sor összecsukásához'
    }
};

export default LocaleHelper.publishLocale(locale);
