// @bryntum/calendar/source/lib/Calendar/localization/Id.js
import LocaleHelper from '../../Core/localization/LocaleHelper.js';
import '../../Scheduler/localization/Id.js';

const locale = {

    localeName : 'Id',
    localeDesc : 'Bahasa Indonesia',
    localeCode : 'id',
    localeRtl  : false,

    EventEdit : {
        Calendar  : 'Kalender',
        'All day' : 'Sepanjang hari',
        day       : 'Hari',
        week      : 'Minggu',
        month     : 'Bulan',
        year      : 'Tahun',
        decade    : 'Dekade'
    },

    EventMenu : {
        duplicateEvent : 'Duplikasi acara',
        copy           : 'salin'
    },

    Calendar : {
        toggleSidebar : 'Alihkan visibilitas sidebar',
        Today         : 'Hari ini',
        Tomorrow      : 'Besok',
        next          : range => `${range} Berikutnya`,
        previous      : range => `${range} Sebelumnya`,
        plusMore      : value => `+${value} lagi`,
        moreEvents    : value => `${value} acara lainnya`,
        allDay        : 'Sepanjang hari',
        endsOn        : d => `Berakhir ${d}`,
        weekOfYear    : ([y, w]) => `Minggu ${w}, ${y}`,
        loadFail      : 'Pemuatan data kalender gagal. Harap hubungi administrator sistem Anda',
        gotoDate      : 'Pergi ke tanggal',
        timeRange     : 'Rentang waktu'
    },

    CalendarDrag : {
        holdCtrlForRecurrence : 'Tahan CTRL untuk acara rutin'
    },

    CalendarMixin : {
        eventCount : count => `${count || 'Tidak ada'} event`
    },

    EventTip : {
        'Edit event' : 'Edit acara',
        timeFormat   : 'LST'
    },

    ModeSelector : {
        includeWeekends : 'Termasuk akhir pekan',
        weekends        : 'Akhir pekan'
    },

    AgendaView : {
        Agenda : 'Agenda'
    },

    MonthView : {
        Month     : 'Bulan',
        monthUnit : 'bulan'
    },

    WeekView : {
        weekUnit : 'minggu'
    },

    YearView : {
        Year     : 'Tahun',
        yearUnit : 'tahun',
        noEvents : 'Tidak ada acara'
    },

    EventList : {
        List   : 'Daftar',
        Start  : 'Mulai',
        Finish : 'Selesai',
        days   : count => `${count > 1 ? `${count} ` : ''}hari${count === 1 ? '' : 's'}`
    },

    DayView : {
        Day                : 'Hari',
        dayUnit            : 'hari',
        daysUnit           : 'hari',
        expandAllDayRow    : 'Luaskan bagian sepanjang hari',
        collapseAllDayRow  : 'Ciutkan bagian sepanjang hari',
        timeFormat         : 'LST',
        timeAxisTimeFormat : 'LST',
        noMatchingDates    : 'Tidak ada tanggal yang cocok'
    },

    DayAgendaView : {
        dayAgenda : 'Agenda Harian'
    },

    DayResourceView : {
        dayResourceView : 'Sumber daya harian'
    },

    Sidebar : {
        selectDate      : 'Pilih tanggal kalender',
        matchEvents     : 'Acara pertandingan',
        filterEvents    : 'Saring acara berdasarkan nama acara',
        filterResources : 'Saring acara berdasarkan sumber daya yang dipilih'
    },

    WeekExpander : {
        expandTip   : 'Klik untuk memperluas baris',
        collapseTip : 'Klik untuk menciutkan baris'
    }
};

export default LocaleHelper.publishLocale(locale);
