// @bryntum/calendar/source/lib/Calendar/localization/Kk.js
import LocaleHelper from '../../Core/localization/LocaleHelper.js';
import '../../Scheduler/localization/Kk.js';

const locale = {

    localeName : 'Kk',
    localeDesc : 'Қазақ тілі (KZ)',
    localeCode : 'kk-KZ',
    localeRtl  : false,

    EventEdit : {
        Calendar  : 'Күнтізбе',
        'All day' : 'Бүкіл күн',
        day       : 'Күн',
        week      : 'Апта',
        month     : 'Ай',
        year      : 'Жыл',
        decade    : 'Он жылдық'
    },

    EventMenu : {
        duplicateEvent : 'Оқиға көшірмесі',
        copy           : 'көшірме('
    },

    Calendar : {
        toggleSidebar : 'Жанама тақтаның көрінуін ауыстыру',
        Today         : 'Бүгін',
        Tomorrow      : 'Ертең',
        next          : range => `Келесі ${range}`,
        previous      : range => `Алдыңғы ${range}`,
        plusMore      : value => `Тағы +${value}`,
        moreEvents    : value => `${value} тағы оқиға`,
        allDay        : 'Бүкіл күн',
        endsOn        : d => `${d} аяқталады`,
        weekOfYear    : ([y, w]) => `Апта ${w}, ${y}`,
        loadFail      : 'Күнтізбе деректерін жүктеу мүмкін болмады. Жүйе әкімшісіне хабарласыңыз',
        gotoDate      : 'Күнге өту',
        timeRange     : 'Уақыт аралығы'
    },

    CalendarDrag : {
        holdCtrlForRecurrence : 'Қайталанатын оқиға үшін CTRL пернесін басып ұстап тұрыңыз'
    },

    CalendarMixin : {
        eventCount : count => `${count || 'Жоқ'} event${count === 1 ? '' : 'с'}`
    },

    EventTip : {
        'Edit event' : 'Оқиғаны өңдеу',
        timeFormat   : 'LST'
    },

    ModeSelector : {
        includeWeekends : 'Демалыс күндерін қосу',
        weekends        : 'Демалыс күндері'
    },

    AgendaView : {
        Agenda : 'Күн тәртібі'
    },

    MonthView : {
        Month     : 'Ай',
        monthUnit : 'ай'
    },

    WeekView : {
        weekUnit : 'апта'
    },

    YearView : {
        Year     : 'Жыл',
        yearUnit : 'жыл',
        noEvents : 'Оқиғалар жоқ'
    },

    EventList : {
        List   : 'Тізім',
        Start  : 'Басталуы',
        Finish : 'Аяқталуы',
        days   : count => `${count > 1 ? `${count} ` : ''}day${count === 1 ? '' : 'с'}`
    },

    DayView : {
        Day                : 'Күн',
        dayUnit            : 'күн',
        daysUnit           : 'күндер',
        expandAllDayRow    : 'Бүкіл күн бөлімін жаю',
        collapseAllDayRow  : 'Бүкіл күн бөлімін жиыру',
        timeFormat         : 'LST',
        timeAxisTimeFormat : 'LST',
        noMatchingDates    : 'Сәйкес келетін күндер жоқ'
    },

    DayAgendaView : {
        dayAgenda : 'Күн тәртібі'
    },

    DayResourceView : {
        dayResourceView : 'Күн ресурстары'
    },

    Sidebar : {
        selectDate      : 'Күнтізбе күнін таңдаңыз',
        matchEvents     : 'Матч оқиғалары',
        filterEvents    : 'Іс-шара атауы бойынша оқиғаларды сүзу',
        filterResources : 'Таңдалған ресурстар бойынша оқиғаларды сүзу'
    },

    WeekExpander : {
        expandTip   : 'Қатарды жаю үшін басыңыз',
        collapseTip : 'Қатарды жиыру үшін басыңыз'
    }
};

export default LocaleHelper.publishLocale(locale);
