// @bryntum/calendar/source/lib/Calendar/localization/Ko.js
import LocaleHelper from '../../Core/localization/LocaleHelper.js';
import '../../Scheduler/localization/Ko.js';

const locale = {

    localeName : 'Ko',
    localeDesc : '한국어',
    localeCode : 'ko',
    localeRtl  : false,

    EventEdit : {
        Calendar  : '캘린더',
        'All day' : '하루 종일',
        day       : '일',
        week      : '주',
        month     : '월',
        year      : '년',
        decade    : '10년'
    },

    EventMenu : {
        duplicateEvent : '중복 일정',
        copy           : '복사하기'
    },

    Calendar : {
        toggleSidebar : '사이드바 표시 전환',
        Today         : '오늘',
        Tomorrow      : '내일',
        next          : range => `다음 ${range}`,
        previous      : range => `이전 ${range}`,
        plusMore      : value => `+${value} 추가`,
        moreEvents    : value => `${value}개의 추가 이벤트`,
        allDay        : '하루 종일',
        endsOn        : d => `${d}일에 종료 `,
        weekOfYear    : ([y, w]) => `${w}, ${y} 주 `,
        loadFail      : '캘린더 데이터 로드에 실패했습니다. 시스템 관리자에게 문의하십시오',
        gotoDate      : '날짜로 이동',
        timeRange     : '시간 범위'
    },

    CalendarDrag : {
        holdCtrlForRecurrence : '반복되는 일정에 Ctrl 키 누르기 유지'
    },

    CalendarMixin : {
        eventCount : count => `일정${count || ' 없음'}`
    },

    EventTip : {
        'Edit event' : '일정 편집',
        timeFormat   : 'LST'
    },

    ModeSelector : {
        includeWeekends : '주말 포함',
        weekends        : '주말'
    },

    AgendaView : {
        Agenda : '일정표'
    },

    MonthView : {
        Month     : '월',
        monthUnit : '달'
    },

    WeekView : {
        weekUnit : '주'
    },

    YearView : {
        Year     : '년',
        yearUnit : '년',
        noEvents : '이벤트가 없습니다'
    },

    EventList : {
        List   : '목록',
        Start  : '시작',
        Finish : '마감',
        days   : count => `${count > 1 ? `${count} ` : ''}${count === 1 ? '낮' : '날'}`
    },

    DayView : {
        Day                : '일',
        dayUnit            : '일',
        daysUnit           : '일',
        expandAllDayRow    : '하루 종일 섹션 펼치기',
        collapseAllDayRow  : '하루 종일 섹션 접기',
        timeFormat         : 'LST',
        timeAxisTimeFormat : 'LST',
        noMatchingDates    : '일치하는 날짜가 없습니다'
    },

    DayAgendaView : {
        dayAgenda : '하루 일정'
    },

    DayResourceView : {
        dayResourceView : '일일 자원'
    },

    Sidebar : {
        selectDate      : '달력 날짜 선택',
        matchEvents     : '경기 이벤트',
        filterEvents    : '이벤트 이름으로 이벤트 필터링',
        filterResources : '선택한 리소스로 이벤트 필터링'
    },

    WeekExpander : {
        expandTip   : '클릭하여 행 펼치기',
        collapseTip : '클릭하여 행 접기'
    }
};

export default LocaleHelper.publishLocale(locale);
