// @bryntum/calendar/source/lib/Calendar/localization/Lt.js
import LocaleHelper from '../../Core/localization/LocaleHelper.js';
import '../../Scheduler/localization/Lt.js';

const locale = {

    localeName : 'LT',
    localeDesc : 'Lietuvių',
    localeCode : 'lt-LT',
    localeRtl  : false,

    EventEdit : {
        Calendar  : 'Kalendorius',
        'All day' : 'Visą dieną',
        day       : 'Diena',
        week      : 'Savaitė',
        month     : 'Mėnuo',
        year      : 'Metai',
        decade    : 'Dešimtmetis'
    },

    EventMenu : {
        duplicateEvent : 'Dubliuoti įvykį',
        copy           : 'kopijuoti'
    },

    Calendar : {
        toggleSidebar : 'Perjungti šoninės juostos matomumą',
        Today         : 'Šiandien',
        Tomorrow      : 'Rytoj',
        next          : range => `Kitas ${range}`,
        previous      : range => `Ankstesnis ${range}`,
        plusMore      : value => `+${value} daugiau`,
        moreEvents    : value => `${value} daugiau įvykių`,
        allDay        : 'Visą dieną',
        endsOn        : d => `Baigiasi ${d}`,
        weekOfYear    : ([y, w]) => `Savaitė ${w}, ${y}`,
        loadFail      : 'Nepavyko įkelti duomenų. Prašome susisiekti su savo sistemos administratoriumi',
        gotoDate      : 'Eiti į datą',
        timeRange     : 'Laiko intervalas'
    },

    CalendarDrag : {
        holdCtrlForRecurrence : 'Palaikykite CTRL pasikartojančiam įvykiui'
    },

    CalendarMixin : {
        eventCount : count => `${count || 'Ne'} įvykis${count === 1 ? '' : 'sek.'}`
    },

    EventTip : {
        'Edit event' : 'Redaguoti įvykį',
        timeFormat   : 'LST'
    },

    ModeSelector : {
        includeWeekends : 'Įtraukti savaitgalius',
        weekends        : 'Savaitgalis'
    },

    AgendaView : {
        Agenda : 'Darbotvarkė'
    },

    MonthView : {
        Month     : 'Mėnuo',
        monthUnit : 'mėnuo'
    },

    WeekView : {
        weekUnit : 'savaitė'
    },

    YearView : {
        Year     : 'Metai',
        yearUnit : 'metai',
        noEvents : 'NNėra įvykių'
    },

    EventList : {
        List   : 'Sąrašas',
        Start  : 'Pradėti',
        Finish : 'Baigti',
        days   : count => `${count > 1 ? `${count} ` : ''}diena${count === 1 ? '' : 's'}`
    },

    DayView : {
        Day                : 'Diena',
        dayUnit            : 'diena',
        daysUnit           : 'dienos',
        expandAllDayRow    : 'Išplėsti visos dienos skirsnį',
        collapseAllDayRow  : 'Suskleisti dienos skirsnį',
        timeFormat         : 'LST',
        timeAxisTimeFormat : 'LST',
        noMatchingDates    : 'Nėra atitinkančių datų'
    },

    DayAgendaView : {
        dayAgenda : 'Dienotvarkė'
    },

    DayResourceView : {
        dayResourceView : 'Dienos ištekliai'
    },

    Sidebar : {
        selectDate      : 'Pasirinkite kalendoriaus datą',
        matchEvents     : 'Rungtynių įvykiai',
        filterEvents    : 'Filtruoti įvykius pagal įvykio pavadinimą',
        filterResources : 'Filtruoti įvykius pagal pasirinktą šaltinį'
    },

    WeekExpander : {
        expandTip   : 'Spustelėkite, kad išplėstumėte eilutę',
        collapseTip : 'Spsutelėkite, kad sutrauktumėte eilutę'
    }
};

export default LocaleHelper.publishLocale(locale);
