// @bryntum/calendar/source/lib/Calendar/localization/Lv.js
import LocaleHelper from '../../Core/localization/LocaleHelper.js';
import '../../Scheduler/localization/Lv.js';

const locale = {

    localeName : 'Lv',
    localeDesc : 'Latviešu',
    localeCode : 'lv-LV',
    localeRtl  : false,

    EventEdit : {
        Calendar  : 'Kalendārs',
        'All day' : 'Visu dienu',
        day       : 'Diena',
        week      : 'Nedēļa',
        month     : 'Mēnesis',
        year      : 'Gads',
        decade    : 'Desmitgade'
    },

    EventMenu : {
        duplicateEvent : 'Duplicēt notikumu',
        copy           : 'kopēt'
    },

    Calendar : {
        toggleSidebar : 'Pārslēgt sānu joslas redzamību',
        Today         : 'Šodiena',
        Tomorrow      : 'Rīt',
        next          : range => `Nākamais ${range}`,
        previous      : range => `Iepriekšējais ${range}`,
        plusMore      : value => `+${value} vairāk`,
        moreEvents    : value => `${value} daugiau įvykių`,
        allDay        : 'Visu dienu',
        endsOn        : d => `Beidzas ${d}`,
        weekOfYear    : ([y, w]) => `Nedēļa ${w}, ${y}`,
        loadFail      : 'Neizdevās ielādēt kalendāra datus. Sazinieties ar sistēmas administratoru.',
        gotoDate      : 'Doties uz datumu',
        timeRange     : 'Laika diapazons'
    },

    CalendarDrag : {
        holdCtrlForRecurrence : 'Turiet nospiestu CTRL, ja notikums ir atkārtojams'
    },

    CalendarMixin : {
        eventCount : count => `${count || 'Nav'} event${count === 1 ? '' : 'notikumu'}`
    },

    EventTip : {
        'Edit event' : 'Rediģēt notikumu',
        timeFormat   : 'LST'
    },

    ModeSelector : {
        includeWeekends : 'Ietvert nedēļas nogales',
        weekends        : 'Nedēļas nogales'
    },

    AgendaView : {
        Agenda : 'Dienaskārtība'
    },

    MonthView : {
        Month     : 'Mēnesis',
        monthUnit : 'mēn.'
    },

    WeekView : {
        weekUnit : 'ned.'
    },

    YearView : {
        Year     : 'Gads',
        yearUnit : 'g.',
        noEvents : 'Nav notikumu'
    },

    EventList : {
        List   : 'Saraksts',
        Start  : 'Sākums',
        Finish : 'Beigas',
        days   : count => `${count > 1 ? `${count} ` : ''}day${count === 1 ? '' : 's'}`
    },

    DayView : {
        Day                : 'Diena',
        dayUnit            : 'd.',
        daysUnit           : 'd.',
        expandAllDayRow    : 'Izvērst visu dienu sadaļu',
        collapseAllDayRow  : 'Sakļaut visu dienu sadaļu',
        timeFormat         : 'LST',
        timeAxisTimeFormat : 'LST',
        noMatchingDates    : 'Nav atbilstošu datumu'
    },

    DayAgendaView : {
        dayAgenda : 'Dienas plāns'
    },

    DayResourceView : {
        dayResourceView : 'Dienas resursi'
    },

    Sidebar : {
        selectDate      : 'Izvēlieties kalendāra datumu',
        matchEvents     : 'Spēles notikumi',
        filterEvents    : 'Filtrēt notikumus pēc notikuma nosaukuma',
        filterResources : 'Filtrēt notikumus pēc izvēlētā resursa'
    },

    WeekExpander : {
        expandTip   : 'Noklikšķinot izvērsiet rindu',
        collapseTip : 'Noklikšķinot sakļaujiet rindu'
    }
};

export default LocaleHelper.publishLocale(locale);
