// @bryntum/calendar/source/lib/Calendar/localization/Ms.js
import LocaleHelper from '../../Core/localization/LocaleHelper.js';
import '../../Scheduler/localization/Ms.js';

const locale = {

    localeName : 'Ms',
    localeDesc : 'Melayu',
    localeCode : 'ms',
    localeRtl  : false,

    EventEdit : {
        Calendar  : 'Kalendar',
        'All day' : 'Sepanjang hari',
        day       : 'Hari',
        week      : 'Minggu',
        month     : 'Bulan',
        year      : 'Tahun',
        decade    : 'Dekad'
    },

    EventMenu : {
        duplicateEvent : 'Duplikat peristiwa',
        copy           : 'salin'
    },

    Calendar : {
        toggleSidebar : 'Togol keterlihatan sidebar',
        Today         : 'Hari ini',
        Tomorrow      : 'Esok',
        next          : range => `Berikutnya ${range}`,
        previous      : range => `Sebelumnya ${range}`,
        plusMore      : value => `+${value} lagi`,
        moreEvents    : value => `${value} acara lagi`,
        allDay        : 'Sepanjang hari',
        endsOn        : d => `Berakhir ${d}`,
        weekOfYear    : ([y, w]) => `Minggu ${w}, ${y}`,
        loadFail      : 'Pemuatan data kalendar gagal. Sila hubungi pentadbir sistem anda',
        gotoDate      : 'Pergi ke tarikh',
        timeRange     : 'Julat masa'
    },

    CalendarDrag : {
        holdCtrlForRecurrence : 'Tahan CTRL untuk peristiwa berulang'
    },

    CalendarMixin : {
        eventCount : count => `${count || 'Jumlah'} peristiwa`
    },

    EventTip : {
        'Edit event' : 'Ubah peristiwa',
        timeFormat   : 'LST'
    },

    ModeSelector : {
        includeWeekends : 'Termasuk hujung minggu',
        weekends        : 'Hujung minggu'
    },

    AgendaView : {
        Agenda : 'Agenda'
    },

    MonthView : {
        Month     : 'Bulan',
        monthUnit : 'bulan'
    },

    WeekView : {
        weekUnit : 'minggu'
    },

    YearView : {
        Year     : 'Tahun',
        yearUnit : 'tahun',
        noEvents : 'Tiada acara'
    },

    EventList : {
        List   : 'Senarai',
        Start  : 'Mula',
        Finish : 'Selesai',
        days   : count => `${count > 1 ? `${count} ` : ''}hari`
    },

    DayView : {
        Day                : 'Hari',
        dayUnit            : 'hari',
        daysUnit           : 'hari',
        expandAllDayRow    : 'Kembangkan bahagian sepanjang hari',
        collapseAllDayRow  : 'Kecilkan bahagian sepanjang hari',
        timeFormat         : 'LST',
        timeAxisTimeFormat : 'LST',
        noMatchingDates    : 'Tiada tarikh yang sepadan'
    },

    DayAgendaView : {
        dayAgenda : 'Agenda Harian'
    },

    DayResourceView : {
        dayResourceView : 'Sumber harian'
    },

    Sidebar : {
        selectDate      : 'Pilih tarikh kalendar',
        matchEvents     : 'Peristiwa perlawanan',
        filterEvents    : 'Tapis peristiwa mengikut nama peristiwa',
        filterResources : 'Tapis peristiwa mengikut sumber yang dipilih'
    },

    WeekExpander : {
        expandTip   : 'Klik untuk kembangkan baris',
        collapseTip : 'Klik untuk kecilkan baris'
    }
};

export default LocaleHelper.publishLocale(locale);
