// @bryntum/calendar/source/lib/Calendar/localization/Nl.js
import LocaleHelper from '../../Core/localization/LocaleHelper.js';
import '../../Scheduler/localization/Nl.js';

const locale = {

    localeName : 'Nl',
    localeDesc : 'Nederlands',
    localeCode : 'nl',
    localeRtl  : false,

    EventEdit : {
        Calendar  : 'Kalender',
        'All day' : 'Hele dag',
        day       : 'Dag',
        week      : 'Week',
        month     : 'Maand',
        year      : 'Jaar',
        decade    : 'Decennium'
    },

    EventMenu : {
        duplicateEvent : 'Kopiëren boeking',
        copy           : 'kopie'
    },

    Calendar : {
        toggleSidebar : 'Zichtbaarheid van de zijbalk wisselen',
        Today         : 'Vandaag',
        Tomorrow      : 'Morgen',
        next          : range => `Volgende ${range}`,
        previous      : range => `Vorige ${range}`,
        plusMore      : value => `+${value} meer`,
        moreEvents    : value => `${value} meer evenementen`,
        allDay        : 'Hele dag',
        endsOn        : d => `Eindigt op ${d}`,
        weekOfYear    : ([y, w]) => `Week ${w}, ${y}`,
        loadFail      : 'Het laden van agendagegevens is mislukt. Neem contact op met uw systeembeheerder',
        gotoDate      : 'Ga naar datum',
        timeRange     : 'Tijdsbereik'
    },

    CalendarDrag : {
        holdCtrlForRecurrence : 'Toets CTRL voor herhalende gebeurtenis'
    },

    CalendarMixin : {
        eventCount : count => `${count || 'No'} event${count === 1 ? '' : 's'}`
    },

    EventTip : {
        'Edit event' : 'Bewerk boeking',
        timeFormat   : 'LT'
    },

    ModeSelector : {
        includeWeekends : 'Inclusief weekenden',
        weekends        : 'Weekenden'
    },

    AgendaView : {
        Agenda : 'Agenda'
    },

    MonthView : {
        Month     : 'Maand',
        monthUnit : 'maand'
    },

    WeekView : {
        weekUnit : 'week'
    },

    YearView : {
        Year     : 'Jaar',
        yearUnit : 'jaar',
        noEvents : 'Geen evenementen'
    },

    EventList : {
        List   : 'Lijst',
        Start  : 'Begin',
        Finish : 'Af hebben',
        days   : count => `${count > 1 ? `${count} ` : ''}dag${count === 1 ? '' : 'en'}`
    },

    DayView : {
        Day                : 'Dag',
        dayUnit            : 'dag',
        daysUnit           : 'dagen',
        expandAllDayRow    : 'Klap dag overzicht uit',
        collapseAllDayRow  : 'Klap dag overzicht in',
        timeFormat         : 'LT',
        timeAxisTimeFormat : 'LT',
        noMatchingDates    : 'Geen overeenkomende datums'
    },

    DayAgendaView : {
        dayAgenda : 'Dagagenda'
    },

    DayResourceView : {
        dayResourceView : 'Dagelijkse bronnen'
    },

    Sidebar : {
        selectDate      : 'Selecteer kalenderdatum',
        matchEvents     : 'Wedstrijdevenementen',
        filterEvents    : 'Filter evenementen op naam van evenement',
        filterResources : 'Filter evenementen op geselecteerde bron'
    },

    WeekExpander : {
        expandTip   : 'Klik om de rij uit te vouwen',
        collapseTip : 'Klik om de rij samen te vouwen'
    }
};

export default LocaleHelper.publishLocale(locale);
