// @bryntum/calendar/source/lib/Calendar/localization/No.js
import LocaleHelper from '../../Core/localization/LocaleHelper.js';
import '../../Scheduler/localization/No.js';

const locale = {

    localeName : 'No',
    localeDesc : 'Norsk',
    localeCode : 'no',
    localeRtl  : false,

    EventEdit : {
        Calendar  : 'Kalender',
        'All day' : 'Hele dagen',
        day       : 'Dag',
        week      : 'Uke',
        month     : 'Måned',
        year      : 'År',
        decade    : 'Tiår'
    },

    EventMenu : {
        duplicateEvent : 'Duplikat hendelse',
        copy           : 'kopi'
    },

    Calendar : {
        toggleSidebar : 'Bytt sidefeltets synlighet',
        Today         : 'I dag',
        Tomorrow      : 'I morgen',
        next          : range => `Neste ${range}`,
        previous      : range => `Forrige ${range}`,
        plusMore      : value => `+${value} mer`,
        moreEvents    : value => `${value} flere hendelser`,
        allDay        : 'Hele dagen',
        endsOn        : d => `Ender ${d}`,
        weekOfYear    : ([y, w]) => `Uke ${w}, ${y}`,
        loadFail      : 'Innlasting av kalenderdata mislyktes. Ta kontakt med systemadministratoren din',
        gotoDate      : 'Gå til dato',
        timeRange     : 'Tidsområde'
    },

    CalendarDrag : {
        holdCtrlForRecurrence : 'Hold CTRL for gjentakende hendelse'
    },

    CalendarMixin : {
        eventCount : count => `${count || 'Ingen'} hendelse${(count && count > 1) ? '' : 'r' }`
    },

    EventTip : {
        'Edit event' : 'Redigere hendelse',
        timeFormat   : 'LST'
    },

    ModeSelector : {
        includeWeekends : 'Inkluder helger',
        weekends        : 'Helger'
    },

    AgendaView : {
        Agenda : 'Dagsorden'
    },

    MonthView : {
        Month     : 'Måned',
        monthUnit : 'måned'
    },

    WeekView : {
        weekUnit : 'uke'
    },

    YearView : {
        Year     : 'År',
        yearUnit : 'år',
        noEvents : 'Ingen hendelser'
    },

    EventList : {
        List   : 'Liste',
        Start  : 'Start',
        Finish : 'Slutt',
        days   : count => `${count > 1 ? `${count} ` : ''}dag${count === 1 ? '' : 'er'}`
    },

    DayView : {
        Day                : 'Dag',
        dayUnit            : 'dag',
        daysUnit           : 'dager',
        expandAllDayRow    : 'Utvid hele dagens del',
        collapseAllDayRow  : 'Skjul hele dagens del',
        timeFormat         : 'LST',
        timeAxisTimeFormat : 'LST',
        noMatchingDates    : ' Ingen samsvarende datoer'
    },

    DayAgendaView : {
        dayAgenda : 'Dagsorden'
    },

    DayResourceView : {
        dayResourceView : 'Daglige ressurser'
    },

    Sidebar : {
        selectDate      : 'Velg kalenderdato',
        matchEvents     : 'Kamphendelser',
        filterEvents    : 'Filtrer hendelser etter hendelsesnavn',
        filterResources : 'Filtrer hendelser etter valgt ressurs'
    },

    WeekExpander : {
        expandTip   : 'Klikk for å utvide rad',
        collapseTip : 'Klikk for å skjule rad'
    }
};

export default LocaleHelper.publishLocale(locale);
