// @bryntum/calendar/source/lib/Calendar/localization/Pl.js
import LocaleHelper from '../../Core/localization/LocaleHelper.js';
import '../../Scheduler/localization/Pl.js';

const locale = {

    localeName : 'Pl',
    localeDesc : 'Polski',
    localeCode : 'pl',
    localeRtl  : false,

    EventEdit : {
        Calendar  : 'Kalendarz',
        'All day' : 'Cały dzień',
        day       : 'Dzień',
        week      : 'Tydzień',
        month     : 'Miesiąc',
        year      : 'Rok',
        decade    : 'Dekada'
    },

    EventMenu : {
        duplicateEvent : 'Zduplikuj wydarzenie',
        copy           : 'kopiuj'
    },

    Calendar : {
        toggleSidebar : 'Przełącz widoczność paska bocznego',
        Today         : 'Dzisiaj',
        Tomorrow      : 'Jutro',
        next          : range => `Następny ${range}`,
        previous      : range => `Poprzedni ${range}`,
        plusMore      : value => `+${value} więcej`,
        moreEvents    : value => `${value} więcej wydarzeń`,
        allDay        : 'Cały dzień',
        endsOn        : d => `Kończy się ${d}`,
        weekOfYear    : ([y, w]) => `Tydzień ${w}, ${y}`,
        loadFail      : 'Nie załadowano danych kalendarza. Skontaktuj się z administratorem systemu',
        gotoDate      : 'Przejdź do daty',
        timeRange     : 'Zakres czasu'
    },

    CalendarDrag : {
        holdCtrlForRecurrence : 'Przytrzymaj CTRL dla wydarzenia cyklicznego'
    },

    CalendarMixin : {
        eventCount : count => `${count || 'Brak'} wydarzeni${(count && count > 1) ? 'a' : 'e' }`
    },

    EventTip : {
        'Edit event' : 'Zmień wydarzenie',
        timeFormat   : 'LST'
    },

    ModeSelector : {
        includeWeekends : 'Uwzględnij weekendy',
        weekends        : 'Weekendy'
    },

    AgendaView : {
        Agenda : 'Program'
    },

    MonthView : {
        Month     : 'Miesiąc',
        monthUnit : 'miesiąc'
    },

    WeekView : {
        weekUnit : 'tydzień'
    },

    YearView : {
        Year     : 'Rok',
        yearUnit : 'rok',
        noEvents : 'Brak wydarzeń'
    },

    EventList : {
        List   : 'Lista',
        Start  : 'Rozpocznij',
        Finish : 'Zakończ',
        days   : count => `${count > 1 ? `${count} ` : ''}${count === 1 ? 'dzień' : 'dni'}`
    },

    DayView : {
        Day                : 'Dzień',
        dayUnit            : 'dzień',
        daysUnit           : 'dni',
        expandAllDayRow    : 'Rozwiń sekcję całodniową',
        collapseAllDayRow  : 'Zwiń sekcję całodniową',
        timeFormat         : 'LST',
        timeAxisTimeFormat : 'LST',
        noMatchingDates    : 'Brak pasujących dat'
    },

    DayAgendaView : {
        dayAgenda : 'Agenda dnia'
    },

    DayResourceView : {
        dayResourceView : 'Zasoby dziennie'
    },

    Sidebar : {
        selectDate      : 'Wybierz datę z kalendarza',
        matchEvents     : 'Wydarzenia meczowe',
        filterEvents    : 'Filtruj wydarzenia według nazwy wydarzenia',
        filterResources : 'Filtruj wydarzenia według wybranego zasobu'
    },

    WeekExpander : {
        expandTip   : 'Kliknij, aby rozwinąć wiersz',
        collapseTip : 'Kliknij, aby zwinąć wiersz'
    }
};

export default LocaleHelper.publishLocale(locale);
