// @bryntum/calendar/source/lib/Calendar/localization/PtBr.js
import LocaleHelper from '../../Core/localization/LocaleHelper.js';
import '../../Scheduler/localization/PtBr.js';

const locale = {

    localeName : 'PtBr',
    localeDesc : 'Português do Brasil',
    localeCode : 'pt-BR',
    localeRtl  : false,

    EventEdit : {
        Calendar  : 'Calendário',
        'All day' : 'Todo o dia',
        day       : 'Dia',
        week      : 'Semana',
        month     : 'Mês',
        year      : 'Ano',
        decade    : 'Década'
    },

    EventMenu : {
        duplicateEvent : 'Duplicar evento',
        copy           : 'copiar'
    },

    Calendar : {
        toggleSidebar : 'Alternar visibilidade da barra lateral',
        Today         : 'Hoje',
        Tomorrow      : 'Amanhã',
        next          : range => `Seguinte ${range}`,
        previous      : range => `Anterior ${range}`,
        plusMore      : value => `+${value} mais`,
        moreEvents    : value => `${value} mais eventos`,
        allDay        : 'Todo o dia',
        endsOn        : d => `Acaba em ${d}`,
        weekOfYear    : ([y, w]) => `Semana ${w}, ${y}`,
        loadFail      : 'Falha ao carregar dados do calendário. Contate o administrador do sistema',
        gotoDate      : 'Ir para a data',
        timeRange     : 'Intervalo de tempo'
    },

    CalendarDrag : {
        holdCtrlForRecurrence : 'Segure continuamente CTRL para um evento recorrente'
    },

    CalendarMixin : {
        eventCount : count => `${count || 'Sem'} evento${(count && count > 1) ? '' : 's'}`
    },

    EventTip : {
        'Edit event' : 'Editar evento',
        timeFormat   : 'LST'
    },

    ModeSelector : {
        includeWeekends : 'Incluir finais-de-semana',
        weekends        : 'Finais-de-semana'
    },

    AgendaView : {
        Agenda : 'Agenda'
    },

    MonthView : {
        Month     : 'Mês',
        monthUnit : 'mês'
    },

    WeekView : {
        weekUnit : 'semana'
    },

    YearView : {
        Year     : 'Ano',
        yearUnit : 'ano',
        noEvents : 'Sem eventos'
    },

    EventList : {
        List   : 'Lista',
        Start  : 'Início',
        Finish : 'Fim',
        days   : count => `${count > 1 ? `${count} ` : ''}día${count === 1 ? '' : 's'}`
    },

    DayView : {
        Day                : 'Dia',
        dayUnit            : 'dia',
        daysUnit           : 'dias',
        expandAllDayRow    : 'Expandir a seção de todo o dia',
        collapseAllDayRow  : 'Fechar a seção de todo o dia',
        timeFormat         : 'LST',
        timeAxisTimeFormat : 'LST',
        noMatchingDates    : 'Nenhuma data correspondente'
    },

    DayAgendaView : {
        dayAgenda : 'Agenda do dia'
    },

    DayResourceView : {
        dayResourceView : 'Recursos diários'
    },

    Sidebar : {
        selectDate      : 'Selecionar data do calendário',
        matchEvents     : 'Eventos da partida',
        filterEvents    : 'Filtrar eventos pelo nome do evento',
        filterResources : 'Filtrar eventos pelo recurso selecionado'
    },

    WeekExpander : {
        expandTip   : 'Clique para expandir a fila',
        collapseTip : 'Clique para fechar a fila'
    }
};

export default LocaleHelper.publishLocale(locale);
