// @bryntum/calendar/source/lib/Calendar/localization/Ro.js
import LocaleHelper from '../../Core/localization/LocaleHelper.js';
import '../../Scheduler/localization/Ro.js';

const locale = {

    localeName : 'Ro',
    localeDesc : 'Română',
    localeCode : 'ro',
    localeRtl  : false,

    EventEdit : {
        Calendar  : 'Calendar',
        'All day' : 'Toată ziua',
        day       : 'Zi',
        week      : 'Săptămână',
        month     : 'Lună',
        year      : 'An',
        decade    : 'Decadă'
    },

    EventMenu : {
        duplicateEvent : 'Duplicare eveniment',
        copy           : 'copiere'
    },

    Calendar : {
        toggleSidebar : 'Comută vizibilitatea barei laterale',
        Today         : 'Zi',
        Tomorrow      : 'Mâine',
        next          : range => `Următor ${range}`,
        previous      : range => `Anterior ${range}`,
        plusMore      : value => `+${value} mai mult`,
        moreEvents    : value => `${value} mai multe evenimente`,
        allDay        : 'Toată ziua',
        endsOn        : d => `Se încheie ${d}`,
        weekOfYear    : ([y, w]) => `Săptămâna ${w}, ${y}`,
        loadFail      : 'Încărcarea datelor din calendar a eșuat. Vă rugăm să contactați administratorul de sistem',
        gotoDate      : 'Mergi la dată',
        timeRange     : 'Interval de timp'
    },

    CalendarDrag : {
        holdCtrlForRecurrence : 'Țineți apăsat CTRL pentru un eveniment recurent'
    },

    CalendarMixin : {
        eventCount : count => `${count || 'Nu'} eveniment${(count && count > 1) ? 'e' : '' }`
    },

    EventTip : {
        'Edit event' : 'Editare eveniment',
        timeFormat   : 'LST'
    },

    ModeSelector : {
        includeWeekends : 'Includeți weekend-uri',
        weekends        : 'Weekend-uri'
    },

    AgendaView : {
        Agenda : 'Ordine de zi'
    },

    MonthView : {
        Month     : 'Lună',
        monthUnit : 'lună'
    },

    WeekView : {
        weekUnit : 'săptămână'
    },

    YearView : {
        Year     : 'An',
        yearUnit : 'an',
        noEvents : 'Niciun eveniment'
    },

    EventList : {
        List   : 'Listă',
        Start  : 'Start',
        Finish : 'Finalizare',
        days   : count => `${count > 1 ? `${count} ` : ''}zi${count === 1 ? '' : 'le'}`
    },

    DayView : {
        Day                : 'Zi',
        dayUnit            : 'zi',
        daysUnit           : 'zile',
        expandAllDayRow    : 'Extindere secțiune toată ziua',
        collapseAllDayRow  : 'Restrângere secțiune toată ziua',
        timeFormat         : 'LST',
        timeAxisTimeFormat : 'LST',
        noMatchingDates    : 'Nu există date corespunzătoare'
    },

    DayAgendaView : {
        dayAgenda : 'Agenda zilnică'
    },

    DayResourceView : {
        dayResourceView : 'Resurse zilnice'
    },

    Sidebar : {
        selectDate      : 'Selectați data din calendar',
        matchEvents     : 'Evenimente meci',
        filterEvents    : 'Filtrează evenimentele după numele evenimentului',
        filterResources : 'Filtrează evenimentele după resursa selectată'
    },

    WeekExpander : {
        expandTip   : 'Faceți clic pentru a extinde rândul',
        collapseTip : 'Faceți clic pentru a restrânge rândul'
    }
};

export default LocaleHelper.publishLocale(locale);
