// @bryntum/calendar/source/lib/Calendar/localization/Ru.js
import LocaleHelper from '../../Core/localization/LocaleHelper.js';
import '../../Scheduler/localization/Ru.js';

const locale = {

    localeName : 'Ru',
    localeDesc : 'Русский',
    localeCode : 'ru',
    localeRtl  : false,

    EventEdit : {
        Calendar  : 'Календарь',
        'All day' : 'На весь день',
        day       : 'День',
        week      : 'Неделя',
        month     : 'Месяц',
        year      : 'Год',
        decade    : 'Десятилетие'
    },

    EventMenu : {
        duplicateEvent : 'копировать событие',
        copy           : 'копия'
    },

    Calendar : {
        toggleSidebar : 'Переключить видимость боковой панели',
        Today         : 'Сегодня',
        Tomorrow      : 'Завтра',
        next          : range => {
            const next = range.includes('неделя') ? 'следующая' : 'следующий';

            return `${next} ${range}`;
        },
        previous : range => {
            const previous = range.includes('неделя') ?  'предыдущая' : 'предыдущий';

            return `${previous} ${range}`;
        },
        plusMore   : value => `Еще +${value}`,
        moreEvents : value => `${value} больше событий`,
        allDay     : 'На весь день',
        endsOn     : d => `Заканчивается ${d}`,
        weekOfYear : ([y, w]) => `Неделя ${w}, ${y}`,
        loadFail   : 'Не удалось загрузить данные календаря. Пожалуйста, обратитесь к системному администратору',
        gotoDate   : 'Перейти к дате',
        timeRange  : 'Диапазон времени'
    },

    CalendarDrag : {
        holdCtrlForRecurrence : 'Удерживайте CTRL для повторяющегося события'
    },

    CalendarMixin : {
        eventCount : count => `${count || 'Нет'} ${count === 1 ? 'событие' : 'событий'}`
    },

    EventTip : {
        'Edit event' : 'Редактировать событие',
        timeFormat   : 'LT'
    },

    ModeSelector : {
        includeWeekends : 'Включить выходные',
        weekends        : 'выходные'
    },

    AgendaView : {
        Agenda : 'Повестка дня'
    },

    MonthView : {
        Month     : 'Месяц',
        monthUnit : 'месяц'
    },

    WeekView : {
        weekUnit : 'неделя'
    },

    YearView : {
        Year     : 'Год',
        yearUnit : 'год',
        noEvents : 'Нет событий'
    },

    EventList : {
        List   : 'Список',
        Start  : 'Начало',
        Finish : 'Конец',
        days   : count => `${count > 1 ? `${count} ` : ''}${count === 1 ? 'день' : 'дни'}`
    },

    DayView : {
        Day                : 'День',
        dayUnit            : 'день',
        daysUnit           : 'дни',
        expandAllDayRow    : 'Развернуть группу событий на весь день',
        collapseAllDayRow  : 'Свернуть группу событий на весь день',
        timeFormat         : 'LT',
        timeAxisTimeFormat : 'LT',
        noMatchingDates    : 'Нет совпадающих дат'
    },

    DayAgendaView : {
        dayAgenda : 'Повестка дня'
    },

    DayResourceView : {
        dayResourceView : 'Дневные ресурсы'
    },

    Sidebar : {
        selectDate      : 'Выберите дату в календаре',
        matchEvents     : 'События матча',
        filterEvents    : 'Фильтровать события по названию',
        filterResources : 'Фильтровать события по выбранному ресурсу'
    },

    WeekExpander : {
        expandTip   : 'Нажмите, чтобы развернуть строку',
        collapseTip : 'Нажмите, чтобы свернуть строку'
    }
};

export default LocaleHelper.publishLocale(locale);
