// @bryntum/calendar/source/lib/Calendar/localization/Sl.js
import LocaleHelper from '../../Core/localization/LocaleHelper.js';
import '../../Scheduler/localization/Sl.js';

const locale = {

    localeName : 'Sl',
    localeDesc : 'Slovensko',
    localeCode : 'sl',
    localeRtl  : false,

    EventEdit : {
        Calendar  : 'Koledar',
        'All day' : 'Ves dan',
        day       : 'Dan',
        week      : 'Teden',
        month     : 'Mesec',
        year      : 'Leto',
        decade    : 'Desetletje'
    },

    EventMenu : {
        duplicateEvent : 'Podvojeni dogodek',
        copy           : 'kopiraj'
    },

    Calendar : {
        toggleSidebar : 'Preklopi vidnost stranske vrstice',
        Today         : 'Danes',
        Tomorrow      : 'Zajtra',
        next          : range => `Naslednji ${range}`,
        previous      : range => `Prejšnji ${range}`,
        plusMore      : value => `+${value} več`,
        moreEvents    : value => `${value} več dogodkov`,
        allDay        : 'Ves dan',
        endsOn        : d => `Konec ${d}`,
        weekOfYear    : ([y, w]) => `Teden ${w}, ${y}`,
        loadFail      : 'Nalaganje podatkov iz koledarja ni uspelo. Obrnite se na skrbnika sistema',
        gotoDate      : 'Pojdi na datum',
        timeRange     : 'Časovno obdobje'
    },

    CalendarDrag : {
        holdCtrlForRecurrence : 'Drži CTRL za ponavljajoči se dogodek'
    },

    CalendarMixin : {
        eventCount : count => `${count || 'Brez'} dogod${(count && count > 1) ? 'ka' : 'ek' }`
    },

    EventTip : {
        'Edit event' : 'Uredi dogodek',
        timeFormat   : 'LST'
    },

    ModeSelector : {
        includeWeekends : 'Vključno z vikendi',
        weekends        : 'Vikendi'
    },

    AgendaView : {
        Agenda : 'Dnevni red'
    },

    MonthView : {
        Month     : 'Mesec',
        monthUnit : 'mesec'
    },

    WeekView : {
        weekUnit : 'Teden'
    },

    YearView : {
        Year     : 'Leto',
        yearUnit : 'Leto',
        noEvents : 'Ni dogodkov'
    },

    EventList : {
        List   : 'Seznam',
        Start  : 'Začetek',
        Finish : 'Konec',
        days   : count => `${count === 1 ? 'dan' : count === 2 ? '2 dneva' : `${count} dni`}`
    },

    DayView : {
        Day                : 'Dan',
        dayUnit            : 'dan',
        daysUnit           : 'dnevi',
        expandAllDayRow    : 'Razširi celodnevni razdelek',
        collapseAllDayRow  : 'Strni celodnevni razdelek',
        timeFormat         : 'LST',
        timeAxisTimeFormat : 'LST',
        noMatchingDates    : 'Ni ujemajočih se datumov'
    },

    DayAgendaView : {
        dayAgenda : 'Dnevni red'
    },

    DayResourceView : {
        dayResourceView : 'Dnevni viri'
    },

    Sidebar : {
        selectDate      : 'Izberite datum v koledarju',
        matchEvents     : 'Dogodki tekme',
        filterEvents    : 'Filtriraj dogodke po imenu dogodka',
        filterResources : 'Filtriraj dogodke po izbranem viru'
    },

    WeekExpander : {
        expandTip   : 'Klikni za razširitev vrstice',
        collapseTip : 'Klikni za strnitev vrstice'
    }
};

export default LocaleHelper.publishLocale(locale);
