// @bryntum/calendar/source/lib/Calendar/localization/Sr.js
import LocaleHelper from '../../Core/localization/LocaleHelper.js';
import '../../Scheduler/localization/Sr.js';

const locale = {

    localeName : 'Sr',
    localeDesc : 'Srpski',
    localeCode : 'sr',
    localeRtl  : false,

    EventEdit : {
        Calendar  : 'Kalendar',
        'All day' : 'Ceo dan',
        day       : 'Dan',
        week      : 'Nedelja',
        month     : 'Mesec',
        year      : 'Godina',
        decade    : 'Decenija'
    },

    EventMenu : {
        duplicateEvent : 'Dupliraj događaj',
        copy           : 'kopiraj'
    },

    Calendar : {
        toggleSidebar : 'Prebaci vidljivost bočne trake',
        Today         : 'Danas',
        Tomorrow      : 'Сутра',
        next          : range => `Sledeći ${range}`,
        previous      : range => `Prethodni ${range}`,
        plusMore      : value => `+${value} više`,
        moreEvents    : value => `${value} још догађаја`,
        allDay        : 'Ceo dan',
        endsOn        : d => `Završava ${d}`,
        weekOfYear    : ([y, w]) => `Nedelja ${w}, ${y}`,
        loadFail      : 'Učitavanje podataka kalendara nije uspelo. Obratite se svom sistemskom administratoru',
        gotoDate      : 'Idi na datum',
        timeRange     : 'Vremenski opseg'
    },

    CalendarDrag : {
        holdCtrlForRecurrence : 'Držite CTRL za događaje koji se ponavljaju'
    },

    CalendarMixin : {
        eventCount : count => `${count || 'Nema'} događaj${(count && count > 1) ? 'e' : '' }`
    },

    EventTip : {
        'Edit event' : 'Uredi događaj',
        timeFormat   : 'LST'
    },

    ModeSelector : {
        includeWeekends : 'Uključi vikende',
        weekends        : 'Vikendi'
    },

    AgendaView : {
        Agenda : 'Rokovnik'
    },

    MonthView : {
        Month     : 'Mesec',
        monthUnit : 'mesec'
    },

    WeekView : {
        weekUnit : 'nedelja'
    },

    YearView : {
        Year     : 'Godina',
        yearUnit : 'godina',
        noEvents : 'Nema događaja'
    },

    EventList : {
        List   : 'Lista',
        Start  : 'Početak',
        Finish : 'Kraj',
        days   : count => `${count > 1 ? `${count} ` : ''}dan${count === 1 ? '' : 'a'}`
    },

    DayView : {
        Day                : 'Dan',
        dayUnit            : 'dan',
        daysUnit           : 'dana',
        expandAllDayRow    : 'Proširi red ceo dan',
        collapseAllDayRow  : 'Skupi red ceo dan',
        timeFormat         : 'LST',
        timeAxisTimeFormat : 'LST',
        noMatchingDates    : 'Nema odgovarajućih datuma'
    },

    DayAgendaView : {
        dayAgenda : ' Дневни ред'
    },

    DayResourceView : {
        dayResourceView : 'Dnevni resursi'
    },

    Sidebar : {
        selectDate      : 'Izaberite datum iz kalendara',
        matchEvents     : 'Utakmice događaja',
        filterEvents    : 'Filtrirajte događaje prema nazivu događaja',
        filterResources : 'Filtrirajte događaje prema izabranom resursu'
    },

    WeekExpander : {
        expandTip   : 'Klkni da raširiš red',
        collapseTip : 'Klikni da skupiš red'
    }
};

export default LocaleHelper.publishLocale(locale);
