// @bryntum/calendar/source/lib/Calendar/localization/SrRs.js
import LocaleHelper from '../../Core/localization/LocaleHelper.js';
import '../../Scheduler/localization/SrRs.js';

const locale = {

    localeName : 'SrRs',
    localeDesc : 'Српски (ћирилица)',
    localeCode : 'sr-RS',
    localeRtl  : false,

    EventEdit : {
        Calendar  : 'Календар',
        'All day' : 'Цео дан',
        day       : 'Дан',
        week      : 'Седмица',
        month     : 'Mесец',
        year      : 'Година',
        decade    : 'Деценија'
    },

    EventMenu : {
        duplicateEvent : 'Дуплирај догађај',
        copy           : 'копирај'
    },

    Calendar : {
        toggleSidebar : 'Промени видљивост бочне траке',
        Today         : 'Данас',
        Tomorrow      : 'Завтра',
        next          : range => `Следећи ${range}`,
        previous      : range => `Претходни ${range}`,
        plusMore      : value => `+${value} више`,
        moreEvents    : value => `${value} ещё событий`,
        allDay        : 'Цео дан',
        endsOn        : d => `Завршава се ${d}`,
        weekOfYear    : ([y, w]) => `Седмица ${w}, ${y}`,
        loadFail      : 'Учитавање података није успело. Контактирајте свог администратора система',
        gotoDate      : 'Иди на датум',
        timeRange     : 'Временски опсег'
    },

    CalendarDrag : {
        holdCtrlForRecurrence : 'Задржите CTRL за понављање догађаја'
    },

    CalendarMixin : {
        eventCount : count => `${count || 'Не'} event${count === 1 ? '' : 'с'}`
    },

    EventTip : {
        'Edit event' : 'Уредите догађај',
        timeFormat   : 'ЛСТ'
    },

    ModeSelector : {
        includeWeekends : 'Укључи викенде',
        weekends        : 'Викенди'
    },

    AgendaView : {
        Agenda : 'Распоред'
    },

    MonthView : {
        Month     : 'Месец',
        monthUnit : 'месец'
    },

    WeekView : {
        weekUnit : 'недеља'
    },

    YearView : {
        Year     : 'Година',
        yearUnit : 'година',
        noEvents : 'Нема догађаја'
    },

    EventList : {
        List   : 'Листа',
        Start  : 'Почетак',
        Finish : 'Крај',
        days   : count => `${count > 1 ? `${count} ` : ''}day${count === 1 ? '' : 'с'}`
    },

    DayView : {
        Day                : 'Дан',
        dayUnit            : 'дан',
        daysUnit           : 'дана',
        expandAllDayRow    : 'Прошири одељак за цео дан',
        collapseAllDayRow  : 'Скупи догађај за цео дан',
        timeFormat         : 'ЛСТ',
        timeAxisTimeFormat : 'LST',
        noMatchingDates    : 'Нема одговарајућих датума'
    },

    DayAgendaView : {
        dayAgenda : 'Дневной график'
    },

    DayResourceView : {
        dayResourceView : 'Ресурси за дан'
    },

    Sidebar : {
        selectDate      : 'Изаберите дату из календаря',
        matchEvents     : 'События матча',
        filterEvents    : 'Фильтровать события по названию',
        filterResources : 'Фильтровать события по выбранному ресурсу'
    },

    WeekExpander : {
        expandTip   : 'Кликните за проширење реда',
        collapseTip : 'Кликните за скупљање реда'
    }
};

export default LocaleHelper.publishLocale(locale);
