// @bryntum/calendar/source/lib/Calendar/localization/SvSE.js
import LocaleHelper from '../../Core/localization/LocaleHelper.js';
import '../../Scheduler/localization/SvSE.js';

const locale = {

    localeName : 'SvSE',
    localeDesc : 'Svenska',
    localeCode : 'sv-SE',
    localeRtl  : false,

    EventEdit : {
        Calendar  : 'Kalender',
        'All day' : 'Heldag',
        day       : 'Dag',
        week      : 'Vecka',
        month     : 'Månad',
        year      : 'År',
        decade    : 'Årtionde'
    },

    EventMenu : {
        duplicateEvent : 'Kopiera uppgift',
        copy           : 'kopia'
    },

    Calendar : {
        toggleSidebar : 'Växla sidofältets synlighet',
        Today         : 'Idag',
        Tomorrow      : 'I morgon',
        next          : range => `Nästa ${range}`,
        previous      : range => `Föregående ${range}`,
        plusMore      : value => `+${value} till`,
        moreEvents    : value => `${value} fler händelser`,
        allDay        : 'Heldag',
        endsOn        : d => `Slutar ${d}`,
        weekOfYear    : ([y, w]) => `Vecka ${w}, ${y}`,
        loadFail      : 'Det gick inte att ladda kalenderdata. Kontakta din systemadministratör',
        gotoDate      : 'Gå till datum',
        timeRange     : 'Tidsintervall'
    },

    CalendarDrag : {
        holdCtrlForRecurrence : 'Håll ner CTRL för att skapa en återkommande uppgift'
    },

    CalendarMixin : {
        eventCount : count => `${count || 'Inga'} uppgift${count === 1 ? '' : 'er'}`
    },

    EventTip : {
        'Edit event' : 'Redigera uppgift',
        timeFormat   : 'LT'
    },

    ModeSelector : {
        includeWeekends : 'Inkludera helger',
        weekends        : 'Helger'
    },

    AgendaView : {
        Agenda : 'Dagordning'
    },

    MonthView : {
        Month     : 'Månad',
        monthUnit : 'månad'
    },

    WeekView : {
        weekUnit : 'Vecka'
    },

    YearView : {
        Year     : 'År',
        yearUnit : 'år',
        noEvents : 'Inga uppgifter'
    },

    EventList : {
        List   : 'Lista',
        Start  : 'Start',
        Finish : 'Avsluta',
        days   : count => `${count > 1 ? `${count} ` : ''}dag${count === 1 ? '' : 'ar'}`
    },

    DayView : {
        Day                : 'Dag',
        dayUnit            : 'dag',
        daysUnit           : 'dagar',
        expandAllDayRow    : 'Expandera heldagssektionen',
        collapseAllDayRow  : 'Fäll ihop heldagssektionen',
        timeFormat         : 'LT',
        timeAxisTimeFormat : 'LT',
        noMatchingDates    : ' Inga matchande datum'
    },

    DayAgendaView : {
        dayAgenda : 'Dagsagenda'
    },

    DayResourceView : {
        dayResourceView : 'Dagliga resurser'
    },

    Sidebar : {
        selectDate      : 'Välj kalendardatum',
        matchEvents     : 'Matcha händelser',
        filterEvents    : 'Filtrera händelser efter händelsenamn',
        filterResources : 'Filtrera händelser efter vald resurs'
    },

    WeekExpander : {
        expandTip   : 'Klicka för att expandera raden',
        collapseTip : 'Klicka för att dölja raden'
    }
};

export default LocaleHelper.publishLocale(locale);
