// @bryntum/calendar/source/lib/Calendar/localization/Uk.js
import LocaleHelper from '../../Core/localization/LocaleHelper.js';
import '../../Scheduler/localization/Uk.js';

const locale = {

    localeName : 'Uk',
    localeDesc : 'Українська',
    localeCode : 'uk-UA',
    localeRtl  : false,

    EventEdit : {
        Calendar  : 'Календар',
        'All day' : 'Увесь день',
        day       : 'День',
        week      : 'Тиждень',
        month     : 'Місяць',
        year      : 'Рік',
        decade    : 'Десятиліття'
    },

    EventMenu : {
        duplicateEvent : 'Копіювати подію',
        copy           : 'Копіювати'
    },

    Calendar : {
        toggleSidebar : 'Перемкнути видимість бічної панелі',
        Today         : 'Сьогодні',
        Tomorrow      : 'Завтра',
        next          : range => `Далі ${range}`,
        previous      : range => `Назад ${range}`,
        plusMore      : value => `Ще +${value}`,
        moreEvents    : value => `${value} ще подій`,
        allDay        : 'Увесь день',
        endsOn        : d => `Завершується ${d}`,
        weekOfYear    : ([y, w]) => `Тиждень ${w}, ${y}`,
        loadFail      : 'Не вдалося завантажити дані календаря. Зверніться до системного адміністратора',
        gotoDate      : 'Перейти до дати',
        timeRange     : 'Діапазон часу'
    },

    CalendarDrag : {
        holdCtrlForRecurrence : 'Утримуйте клавішу CTRL для повторення події'
    },

    CalendarMixin : {
        eventCount : count => `${count || 'Події відсутні'} event${count === 1 ? '' : 'с'}`
    },

    EventTip : {
        'Edit event' : 'Редагувати подію',
        timeFormat   : 'LST'
    },

    ModeSelector : {
        includeWeekends : 'Включити тижні',
        weekends        : 'Тижні'
    },

    AgendaView : {
        Agenda : 'Порядок денний'
    },

    MonthView : {
        Month     : 'Місяць',
        monthUnit : 'міс.'
    },

    WeekView : {
        weekUnit : 'тиж.'
    },

    YearView : {
        Year     : 'Рік',
        yearUnit : 'р.',
        noEvents : 'Події відсутні'
    },

    EventList : {
        List   : 'Список',
        Start  : 'Початок',
        Finish : 'Завершення',
        days   : count => `${count > 1 ? `${count} ` : ''}day${count === 1 ? '' : 'Дні'}`
    },

    DayView : {
        Day                : 'День',
        dayUnit            : 'дн.',
        daysUnit           : 'дн.',
        expandAllDayRow    : 'Розгорнути розділ з розкладом дня',
        collapseAllDayRow  : 'Згорнути  розділ з розкладом дня',
        timeFormat         : 'LST',
        timeAxisTimeFormat : 'LST',
        noMatchingDates    : 'Немає відповідних дат'
    },

    DayAgendaView : {
        dayAgenda : 'Порядок денний'
    },

    DayResourceView : {
        dayResourceView : 'Ресурси на день'
    },

    Sidebar : {
        selectDate      : 'Виберіть дату в календарі',
        matchEvents     : 'Події матчу',
        filterEvents    : 'Фільтрувати події за назвою',
        filterResources : 'Фільтрувати події за обраним ресурсом'
    },

    WeekExpander : {
        expandTip   : 'Натисніть, щоб розгорнути рядок',
        collapseTip : 'Натисніть, щоб згорнути рядок'
    }
};

export default LocaleHelper.publishLocale(locale);
