// @bryntum/calendar/source/lib/Calendar/localization/Vi.js
import LocaleHelper from '../../Core/localization/LocaleHelper.js';
import '../../Scheduler/localization/Vi.js';

const locale = {

    localeName : 'Vi',
    localeDesc : 'Tiếng Việt',
    localeCode : 'vi',
    localeRtl  : false,

    EventEdit : {
        Calendar  : 'Lịch',
        'All day' : 'Cả ngày',
        day       : 'Ngày',
        week      : 'Tuần',
        month     : 'Tháng',
        year      : 'Năm',
        decade    : 'Thập kỷ'
    },

    EventMenu : {
        duplicateEvent : 'Sự kiện lặp lại',
        copy           : 'sao chép'
    },

    Calendar : {
        toggleSidebar : 'Chuyển đổi tính hiển thị của thanh bên',
        Today         : 'Hôm nay',
        Tomorrow      : 'Ngày mai',
        next          : range => `Sau ${range}`,
        previous      : range => `Trước ${range}`,
        plusMore      : value => `+ thêm ${value}`,
        moreEvents    : value => `${value} sự kiện khác`,
        allDay        : 'Cả ngày',
        endsOn        : d => `Kết thúc vào ngày ${d}`,
        weekOfYear    : ([y, w]) => `Tuần trong năm${w}, ${y}`,
        loadFail      : 'Tải dữ liệu lịch không thành công. Vui lòng liên hệ với quản trị hệ thống của quý vị',
        gotoDate      : 'Đi đến ngày',
        timeRange     : 'Khoảng thời gian'
    },

    CalendarDrag : {
        holdCtrlForRecurrence : 'Giữ nút CTRL cho một sự kiện lặp lại'
    },

    CalendarMixin : {
        eventCount : count => `${count || 'không'} sự kiện`
    },

    EventTip : {
        'Edit event' : 'Chỉnh sửa sự kiện',
        timeFormat   : 'LST'
    },

    ModeSelector : {
        includeWeekends : 'Bao gồm những ngày cuối tuần',
        weekends        : 'Những ngày cuối tuần'
    },

    AgendaView : {
        Agenda : 'Lịch trình'
    },

    MonthView : {
        Month     : 'Tháng',
        monthUnit : 'Đơn vị tháng'
    },

    WeekView : {
        weekUnit : 'Đơn vị tuần'
    },

    YearView : {
        Year     : 'Năm',
        yearUnit : 'Đơn vị năm',
        noEvents : 'Không có sự kiện'
    },

    EventList : {
        List   : 'Danh sách',
        Start  : 'Bắt đầu',
        Finish : 'Kết thúc',
        days   : count => `${count} ngày`
    },

    DayView : {
        Day                : 'Ngày',
        dayUnit            : 'Đơn vị ngày',
        daysUnit           : 'Đơn vị ngày',
        expandAllDayRow    : 'Mở rộng phần cả ngày',
        collapseAllDayRow  : 'Thu gọn phần cả ngày',
        timeFormat         : 'LST',
        timeAxisTimeFormat : 'LST',
        noMatchingDates    : 'Không có ngày phù hợp'
    },

    DayAgendaView : {
        dayAgenda : 'Lịch trình trong ngày'
    },

    DayResourceView : {
        dayResourceView : 'Nguồn tài nguyên hàng ngày'
    },

    Sidebar : {
        selectDate      : 'Chọn ngày trên lịch',
        matchEvents     : 'Sự kiện trận đấu',
        filterEvents    : 'Lọc sự kiện theo tên sự kiện',
        filterResources : 'Lọc sự kiện theo tài nguyên đã chọn'
    },

    WeekExpander : {
        expandTip   : 'Nhấp để mở rộng hàng',
        collapseTip : 'Nhấp để thu gọn hàng'
    }
};

export default LocaleHelper.publishLocale(locale);
