// @bryntum/calendar/source/lib/Calendar/localization/ZhCn.js
import LocaleHelper from '../../Core/localization/LocaleHelper.js';
import '../../Scheduler/localization/ZhCn.js';

const locale = {

    localeName : 'ZhCn',
    localeDesc : '中文（中国）',
    localeCode : 'zh-CN',
    localeRtl  : false,

    EventEdit : {
        Calendar  : '日历',
        'All day' : '全天',
        day       : '日',
        week      : '周',
        month     : '月',
        year      : '年',
        decade    : '十年'
    },

    EventMenu : {
        duplicateEvent : '重复性事件',
        copy           : '复制'
    },

    Calendar : {
        toggleSidebar : '切换侧边栏可见性',
        Today         : '今天',
        Tomorrow      : '明天',
        next          : range => `下一 ${range}`,
        previous      : range => `上一 ${range}`,
        plusMore      : value => `+${value} 更多`,
        moreEvents    : value => `${value} 个更多事件`,
        allDay        : '全天',
        endsOn        : d => `结束 ${d}`,
        weekOfYear    : ([y, w]) => `周 ${w}, ${y}`,
        loadFail      : '日历数据加载失败。请联系系统管理员',
        gotoDate      : '跳转到日期',
        timeRange     : '时间范围'
    },

    CalendarDrag : {
        holdCtrlForRecurrence : '按住CTRL键以重复事件'
    },

    CalendarMixin : {
        eventCount : count => `${count || '无'}事件`
    },

    EventTip : {
        'Edit event' : '编辑事件',
        timeFormat   : 'LST'
    },

    ModeSelector : {
        includeWeekends : '包括周末',
        weekends        : '周末'
    },

    AgendaView : {
        Agenda : '日程'
    },

    MonthView : {
        Month     : '月',
        monthUnit : '月'
    },

    WeekView : {
        weekUnit : '周'
    },

    YearView : {
        Year     : '年',
        yearUnit : '年',
        noEvents : '没有活动'
    },

    EventList : {
        List   : '列表',
        Start  : '开始',
        Finish : '完成',
        days   : count => `${count}天`
    },

    DayView : {
        Day                : '日',
        dayUnit            : '天',
        daysUnit           : '天',
        expandAllDayRow    : '展开全天内容',
        collapseAllDayRow  : '折叠全天内容',
        timeFormat         : 'LST',
        timeAxisTimeFormat : 'LST',
        noMatchingDates    : '没有匹配的日期'
    },

    DayAgendaView : {
        dayAgenda : '日程安排'
    },

    DayResourceView : {
        dayResourceView : '日资源'
    },

    Sidebar : {
        selectDate      : '选择日历日期',
        matchEvents     : '比赛事件',
        filterEvents    : '按事件名称筛选事件',
        filterResources : '按所选资源筛选事件'
    },

    WeekExpander : {
        expandTip   : '展开提示',
        collapseTip : '折叠提示'
    }
};

export default LocaleHelper.publishLocale(locale);
