// @bryntum/calendar/source/lib/Engine/scheduling/DurationConverterMixin.ts
import { AnyConstructor, Mixin } from "../../ChronoGraph/class/BetterMixin.js"
import { CalculationIterator } from "../../ChronoGraph/primitives/Calculation.js"
import { calculate, field } from "../../ChronoGraph/replica/Entity.js"
import DateHelper from "../../Core/helper/DateHelper.js"
import { ChronoModelMixin } from "../chrono/ChronoModelMixin.js"
import { model_field } from "../chrono/ModelFieldAtom.js"
import { Duration, TimeUnit } from "./Types.js"

/**
 * This mixin provides the duration converting functionality - the [[convertDuration]] method. It requires (inherit from) [[ChronoModelMixin]].
 */
export class DurationConverterMixin extends Mixin(
    [ ChronoModelMixin ],
    (base : AnyConstructor<ChronoModelMixin, typeof ChronoModelMixin>) => {


    class DurationConverterMixin extends base {

        @field()
        unitsInMs               : { [ key : string ] : number }

        /**
         * The number of hours per day.
         *
         * **Please note:** the value **does not define** the amount of **working** time per day
         * for that purpose one should use calendars.
         *
         * The value is used when converting the duration from one unit to another.
         * So when user enters a duration of, for example, `5 days` the system understands that it
         * actually means `120 hours` and schedules accordingly.
         */
        @model_field({ type : 'number', defaultValue : 24 })
        hoursPerDay             : number

        /**
         * The number of days per week.
         *
         * **Please note:** the value **does not define** the amount of **working** time per week
         * for that purpose one should use calendars.
         *
         * The value is used when converting the duration from one unit to another.
         * So when user enters a duration of, for example, `2 weeks` the system understands that it
         * actually means `14 days` (which is then converted to [[hoursPerDay|hours]]) and schedules accordingly.
         */
        @model_field({ type : 'number', defaultValue : 7 })
        daysPerWeek             : number

        /**
         * The number of days per month.
         *
         * **Please note:** the value **does not define** the amount of **working** time per month
         * for that purpose one should use calendars.
         *
         * The value is used when converting the duration from one unit to another.
         * So when user enters a duration of, for example, `1 month` the system understands that it
         * actually means `30 days` (which is then converted to [[hoursPerDay|hours]]) and schedules accordingly.
         */
        @model_field({ type : 'number', defaultValue : 30 })
        daysPerMonth            : number


        isElapsedDurationUnit (durationUnit : TimeUnit) : boolean {
            return durationUnit === TimeUnit.ElapsedMinute ||
                durationUnit === TimeUnit.ElapsedHour ||
                durationUnit === TimeUnit.ElapsedDay ||
                durationUnit === TimeUnit.ElapsedWeek ||
                durationUnit === TimeUnit.ElapsedMonth ||
                durationUnit === TimeUnit.ElapsedQuarter ||
                durationUnit === TimeUnit.ElapsedYear
        }


        @calculate('unitsInMs')
        * calculateUnitsInMs () {
            const hoursPerDay   = yield this.$.hoursPerDay
            const daysPerWeek   = yield this.$.daysPerWeek
            const daysPerMonth  = yield this.$.daysPerMonth

            return {
                millisecond : 1,
                second      : 1000,
                minute      : 60 * 1000,
                eminute     : 60 * 1000,
                hour        : 60 * 60 * 1000,
                ehour       : 60 * 60 * 1000,
                day         : hoursPerDay * 60 * 60 * 1000,
                eday        : 24 * 60 * 60 * 1000,
                week        : daysPerWeek * hoursPerDay * 60 * 60 * 1000,
                eweek       : 7 * 24 * 60 * 60 * 1000,
                month       : daysPerMonth * hoursPerDay * 60 * 60 * 1000,
                emonth      : 30 * 24 * 60 * 60 * 1000,
                quarter     : 3 * daysPerMonth * hoursPerDay * 60 * 60 * 1000,
                equarter    : 3 * 30 * 24 * 60 * 60 * 1000,
                year        : 4 * 3 * daysPerMonth * hoursPerDay * 60 * 60 * 1000,
                eyear       : 4 * 3 * 30 * 24 * 60 * 60 * 1000
            }
        }


        /**
         * Converts duration value from one time unit to another
         * @param duration Duration value
         * @param fromUnit Duration value time unit
         * @param toUnit   Target time unit to convert the value to
         */
        convertDuration (duration : Duration, fromUnit : TimeUnit, toUnit : TimeUnit) : Duration {
            return this.run('$convertDuration', duration, fromUnit, toUnit)
        }


        /**
         * Converts duration value from one time unit to another.
         * A generator function able to embed into the Engine calculations.
         * @param duration Duration value
         * @param fromUnit Duration value time unit
         * @param toUnit   Target time unit to convert the value to
         */
        * $convertDuration (duration : Duration, fromUnit : TimeUnit, toUnit : TimeUnit) : CalculationIterator<Duration> {
            toUnit   = DateHelper.normalizeUnit(toUnit) as TimeUnit;
            fromUnit = DateHelper.normalizeUnit(fromUnit) as TimeUnit;

            if (!fromUnit || !toUnit) throw new Error("Conversion unit not provided")

            const unitsInMs : this[ 'unitsInMs' ] = yield this.$.unitsInMs

            let result  = duration

            if (fromUnit !== toUnit) {
                result  = duration * unitsInMs[ fromUnit ] / unitsInMs[ toUnit ]
            }

            return result
        }

    }

    return DurationConverterMixin
}){}
