// @bryntum/calendar/source/lib/Engine/quark/model/gantt/ConstrainedByParentMixin.ts
import { Transaction } from '../../../../ChronoGraph/chrono/Transaction.js'
import { AnyConstructor, Mixin } from '../../../../ChronoGraph/class/Mixin.js'
import { CalculationIterator } from '../../../../ChronoGraph/primitives/Calculation.js'
import { prototypeValue } from '../../../../ChronoGraph/util/Helpers.js'
import DateHelper from '../../../../Core/helper/DateHelper.js'
import Localizable from '../../../../Core/localization/Localizable.js'
import { ConflictResolution, ConstraintInterval, ConstraintIntervalDescription } from '../../../chrono/Conflict.js'
import { DateInterval } from '../../../scheduling/DateInterval.js'
import { ConstraintIntervalSide } from "../../../scheduling/Types.js"
import { format } from '../../../util/Functions.js'
import { BaseEventMixin } from '../scheduler_basic/BaseEventMixin.js'
import { HasChildrenMixin } from '../scheduler_basic/HasChildrenMixin.js'
import { ConstrainedEarlyEventMixin } from '../scheduler_pro/ConstrainedEarlyEventMixin.js'
import "../../../localization/En.js"
import { ConstrainedByParentScheduleMixin } from "./ConstrainedByParentScheduleMixin.js"
import { GanttProjectMixin } from './GanttProjectMixin.js'

//---------------------------------------------------------------------------------------------------------------------
/**
 * This mixin makes the event to "inherit" the constraints from its parent event.
 */
export class ConstrainedByParentMixin extends Mixin(
    [
        BaseEventMixin,
        HasChildrenMixin,
        ConstrainedEarlyEventMixin
    ],
    (base : AnyConstructor<
        BaseEventMixin &
        HasChildrenMixin &
        ConstrainedEarlyEventMixin
        ,
        typeof BaseEventMixin &
        typeof HasChildrenMixin &
        typeof ConstrainedEarlyEventMixin
    >) => {

    const superProto : InstanceType<typeof base> = base.prototype


    class ConstrainedByParentMixin extends base {
        // @ts-ignore
        static scheduleMixinClass  : typeof ConstrainedByParentScheduleMixin   =
            Mixin(
                [ base.scheduleMixinClass, ConstrainedByParentScheduleMixin ],
                base => base
            )


        project : GanttProjectMixin


        * calculateStartDateConstraintIntervals () : CalculationIterator<this[ 'startDateConstraintIntervals' ]> {
            const intervals : DateInterval[] = yield* superProto.calculateStartDateConstraintIntervals.call(this)

            const parentEvent : HasChildrenMixin & ConstrainedByParentMixin = yield this.$.parentEvent

            if (parentEvent?.graph) {
                // Child inherits its parent tasks constraints
                // unless the parent is manually scheduled (in this case the parent breaks the inheritance)
                if (!this.project.manuallyScheduledParentBreaksConstraintInheritance || !(yield* parentEvent.isManuallyScheduled())) {
                    const parentIntervals = yield parentEvent.$.startDateConstraintIntervals

                    intervals.push.apply(intervals, parentIntervals)
                }
            }

            return intervals
        }


        * calculateEndDateConstraintIntervals () : CalculationIterator<this[ 'endDateConstraintIntervals' ]> {
            const intervals : DateInterval[] = yield* superProto.calculateEndDateConstraintIntervals.call(this)

            const parentEvent : HasChildrenMixin & ConstrainedByParentMixin = yield this.$.parentEvent

            if (parentEvent?.graph) {
                // Child inherits its parent tasks constraints
                // unless the parent is manually scheduled (in this case the parent breaks the inheritance)
                if (!this.project.manuallyScheduledParentBreaksConstraintInheritance || !(yield* parentEvent.isManuallyScheduled())) {
                    const parentIntervals = yield parentEvent.$.endDateConstraintIntervals

                    intervals.push.apply(intervals, parentIntervals)
                }
            }

            return intervals
        }
    }

    return ConstrainedByParentMixin
}){}


/**
 * Class implements resolving a scheduling conflict happened due to a parent event
 * [[ConstrainedByParentMixin.manuallyScheduled|manually scheduled state]].
 * It resolves the conflict by setting the event [[ConstrainedByParentMixin.manuallyScheduled|manuallyScheduled]] to `false`.
 */
export class DisableManuallyScheduledConflictResolution extends Localizable(ConflictResolution) {

    static $name = 'RemoveManuallyScheduledParentConstraintConflictResolution'

    event           : ConstrainedByParentMixin

    construct () {
        super.construct(...arguments)
        this.event = this.interval.owner as ConstrainedByParentMixin
    }

    getDescription () : string {
        const { event } = this

        return format(this.L('L{descriptionTpl}'), event.name || event.id)
    }

    /**
     * Resolves the conflict by setting the event [[ConstrainedByParentMixin.manuallyScheduled|manuallyScheduled]] to `false`.
     */
    resolve () {
        this.event.manuallyScheduled = false
    }
}


/**
 * Description builder for an [[ManuallyScheduledParentConstraintInterval|manual parent constraint interval]].
 */
export class ManuallyScheduledParentConstraintIntervalDescription extends ConstraintIntervalDescription {

    static $name = 'ManuallyScheduledParentConstraintIntervalDescription'

    /**
     * Returns description for the provided event constraint interval.
     * @param interval Constraint interval
     */
    static getDescription (interval : ManuallyScheduledParentConstraintInterval) : string {
        interval = interval.reflectionOf as ManuallyScheduledParentConstraintInterval || interval

        let tpl : string

        switch (interval.side) {
            case ConstraintIntervalSide.Start :
                tpl = this.L('L{startDescriptionTpl}')
                break

            case ConstraintIntervalSide.End :
                tpl = this.L('L{endDescriptionTpl}')
                break
        }

        return format(tpl, ...this.getDescriptionParameters(interval))
    }

    static getDescriptionParameters (interval : ManuallyScheduledParentConstraintInterval) : any[] {
        // resolve a reflection to its original interval to show not mapped but real dates
        interval = interval.reflectionOf as ManuallyScheduledParentConstraintInterval || interval

        const event         = interval.owner

        return [
            DateHelper.format(interval.startDate, this.L('L{dateFormat}')),
            DateHelper.format(interval.endDate, this.L('L{dateFormat}')),
            event.name || event.id
        ]
    }

}


/**
 * Class implements an interval applied by a [[ConstrainedByParentMixin.manuallyScheduled|manually scheduled]] parent event.
 * The interval suggests the only resolution option - disabling [[ConstrainedByParentMixin.manuallyScheduled|manually scheduled]] mode.
 */
export class ManuallyScheduledParentConstraintInterval extends ConstraintInterval {

    get isManuallyScheduledParentConstraintInterval () : boolean {
        return true
    }

    /**
     * The event applying the interval.
     */
    owner                                       : ConstrainedByParentMixin

    /**
     * Class implementing "Disable manual scheduling" resolution.
     */
    @prototypeValue(DisableManuallyScheduledConflictResolution)
    resetManuallyScheduledConflictResolutionClass : typeof DisableManuallyScheduledConflictResolution

    /**
     * Class implementing the interval description builder.
     */
    @prototypeValue(ManuallyScheduledParentConstraintIntervalDescription)
    descriptionBuilderClass                     : typeof ManuallyScheduledParentConstraintIntervalDescription

    getDescription () : string {
        return this.descriptionBuilderClass.getDescription(this)
    }

    isAffectedByTransaction (transaction? : Transaction) : boolean {
        const event = this.owner

        transaction = transaction || event.graph.activeTransaction

        const manuallyScheduledQuark = transaction.entries.get(event.$.manuallyScheduled)

        // new constrained event or modified constraint
        return !transaction.baseRevision.hasIdentifier(event.$$) ||
            manuallyScheduledQuark && !manuallyScheduledQuark.isShadow()
    }

    /**
     * Returns possible resolution options for cases when
     * the interval takes part in a conflict.
     *
     * The interval suggests the only resolution option - disabling manual scheduling.
     */
    getResolutions () : ConflictResolution[] {
        return this.resolutions || (this.resolutions = [
            this.resetManuallyScheduledConflictResolutionClass.new({ interval : this })
        ])
    }
}
