// @bryntum/calendar-vue-3/src/components/BryntumTimelineHistogramProps.ts
import type { AjaxStore, AjaxStoreConfig, CellCopyPaste, CellCopyPasteConfig, CellEdit, CellEditConfig, CellMenu, CellMenuConfig, CellTooltip, CellTooltipConfig, ColumnAutoWidth, ColumnAutoWidthConfig, ColumnDragToolbar, ColumnDragToolbarConfig, ColumnLines, ColumnLinesConfig, ColumnPicker, ColumnPickerConfig, ColumnRename, ColumnRenameConfig, ColumnReorder, ColumnReorderConfig, ColumnResize, ColumnResizeConfig, ColumnStore, ColumnStoreConfig, ContainerItemConfig, DomConfig, DurationUnit, EventColor, FileDrop, FileDropConfig, FillHandle, FillHandleConfig, Filter, FilterBar, FilterBarConfig, FilterConfig, FormulaProviderConfig, GridColumnConfig, GridExcelExporter, GridExcelExporterConfig, GridLockRows, GridLockRowsConfig, GridPdfExport, GridPdfExportConfig, GridPrint, GridPrintConfig, GridRowResize, GridRowResizeConfig, GridSelectionMode, Group, GroupConfig, GroupSummary, GroupSummaryConfig, HeaderMenu, HeaderMenuConfig, HeaderZoom, HeaderZoomConfig, Histogram, HistogramConfig, HistogramRenderData, HistogramSeries, KeyMapConfig, Mask, MaskConfig, MergeCells, MergeCellsConfig, Model, ModelConfig, NonWorkingTime, NonWorkingTimeConfig, PagingToolbarConfig, Pan, PanConfig, PanelCollapserConfig, PanelCollapserOverlayConfig, PanelHeader, PreserveScrollOptions, QuickFind, QuickFindConfig, RecurrenceConfirmationPopup, RecurrenceConfirmationPopupConfig, RegionResize, RegionResizeConfig, ResourceMenu, ResourceMenuConfig, ResourceModel, ResourceTimeRangeModel, ResourceTimeRanges, ResourceTimeRangesConfig, RowCopyPaste, RowCopyPasteConfig, RowEdit, RowEditConfig, RowExpander, RowExpanderConfig, RowReorder, RowReorderConfig, ScaleColumnConfig, ScalePoint, SchedulerScheduleTooltip, SchedulerScheduleTooltipConfig, SchedulerTimeRanges, SchedulerTimeRangesConfig, ScrollManager, ScrollManagerConfig, Scroller, ScrollerConfig, Search, SearchConfig, Sort, SortConfig, Split, SplitConfig, StateProvider, StickyCells, StickyCellsConfig, Store, StoreConfig, Stripe, StripeConfig, SubGridConfig, Summary, SummaryConfig, TabConfig, TimeAxis, TimeAxisConfig, TimeAxisHeaderMenu, TimeAxisHeaderMenuConfig, TimelineBase, TimelineHistogramListeners, ToolConfig, ToolbarConfig, ToolbarItems, Tooltip, TooltipConfig, Tree, TreeConfig, TreeGroup, TreeGroupConfig, ViewPresetConfig, VisibleDate, Widget } from '@bryntum/calendar';
import {  } from '@bryntum/calendar';

export declare type BryntumTimelineHistogramProps = {
    // Configs
    /**
     * Element (or element id) to adopt as this Widget's encapsulating element. The widget's
     * content will be placed inside this element.
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Scheduler/view/TimelineHistogram#config-adopt)
     */
    adopt ? : HTMLElement|string
    /**
     * A function used for aggregating child records histogram data entries to their parent entry.
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Scheduler/view/TimelineHistogram#config-aggregateDataEntry)
     * @param {object} aggregateDataEntry.aggregated Target parent data entry to aggregate the entry into.
     * @param {object} aggregateDataEntry.entry Current entry to aggregate into `aggregated`.
     * @param {number} aggregateDataEntry.arrayIndex Index of current array (index of the record among other records being aggregated).
     * @param {object[]} aggregateDataEntry.entryIndex Index of `entry` in the current array.
     * @returns {object} Return value becomes the value of the `aggregated` parameter on the next invocation of this function.
     */
    aggregateDataEntry ? : ((aggregateDataEntry: { aggregated: object, entry: object, arrayIndex: number, entryIndex: object[] }) => object)|string
    /**
     * When `true` the component will automatically calculate data for group records
     * based on the groups members data by calling [getGroupRecordHistogramData](https://bryntum.com/products/calendar/docs/api/Scheduler/view/mixin/TimelineHistogramGrouping#function-getGroupRecordHistogramData) method.
     */
    aggregateHistogramDataForGroups ? : Boolean
    /**
     * When this widget is a child of a [Container](https://bryntum.com/products/calendar/docs/api/Core/widget/Container), it will by default be participating in a
     * flexbox layout. This config allows you to set this widget's
     * [align-self](https://developer.mozilla.org/en-US/docs/Web/CSS/align-self) style.
     */
    alignSelf ? : string
    /**
     * Set to `true` to animate row removals caused by filtering.
     */
    animateFilterRemovals ? : Boolean
    /**
     * Set to `false` to prevent remove row animation and remove the delay related to that.
     */
    animateRemovingRows ? : Boolean
    /**
     * When the [Tree](https://bryntum.com/products/calendar/docs/api/Grid/feature/Tree) feature is in use and the Store is a tree store, this
     * config may be set to `true` to visually animate branch node expand and collapse operations.
     * *This is not supported in Scheduler and Gantt*
     */
    animateTreeNodeToggle ? : Boolean
    /**
     * Element (or the id of an element) to append this widget's element to. Can be configured, or set once at
     * runtime. To access the element of a rendered widget, see [element](https://bryntum.com/products/calendar/docs/api/Core/widget/Widget#property-element).
     */
    appendTo ? : HTMLElement|string
    /**
     * A localizable string (May contain `'L{}'` tokens which resolve in the locale file) to inject
     * into an element which will be linked using the `aria-describedby` attribute.
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Scheduler/view/TimelineHistogram#config-ariaDescription)
     */
    ariaDescription ? : string
    /**
     * A localizable string (May contain `'L{}'` tokens which resolve in the locale file) to inject as
     * the `aria-label` attribute.
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Scheduler/view/TimelineHistogram#config-ariaLabel)
     */
    ariaLabel ? : string
    /**
     * The `aria-live` value for this widget.
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Scheduler/view/TimelineHistogram#config-ariaLive)
     * @deprecated Will be removed in 7.0
     */
    ariaLive ? : string
    /**
     * You can set this option to `false` to make the timeline panel start and end on the exact provided
     * [startDate](https://bryntum.com/products/calendar/docs/api/Scheduler/view/TimelineBase#config-startDate)/[endDate](#Scheduler/view/TimelineBase#config-endDate) w/o adjusting them.
     */
    autoAdjustTimeAxis ? : Boolean
    /**
     * Automatically set grids height to fit all rows (no scrolling in the grid). In general you should avoid
     * using `autoHeight: true`, since it will bypass Grids virtual rendering and render all rows at once, which
     * in a larger grid is really bad for performance.
     */
    autoHeight ? : Boolean
    /**
     * Controls how much space to leave between stacked event bars in px.
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Scheduler/view/TimelineHistogram#config-barMargin)
     */
    barMargin ? : number
    /**
     * A Function which returns the tooltip text to display when hovering a bar.
     * The following parameters are passed:
     * @param {object} context The tooltip context info
     * @param {object} context.datum The histogram bar being hovered info
     * @param {Core.widget.Tooltip} context.tip The tooltip instance
     * @param {HTMLElement} context.element The Element for which the Tooltip is monitoring mouse movement
     * @param {HTMLElement} context.activeTarget The target element that triggered the show
     * @param {Event} context.event The raw DOM event
     * @param {Core.data.Model} data.record The record which value the hovered bar displays.
     * @returns {string} Tooltip HTML content
     */
    barTooltipTemplate ? : (context: { datum: object, tip: Tooltip, element: HTMLElement, activeTarget: HTMLElement, event: Event }, data: { record: Model }) => string
    /**
     * A Config object representing the configuration of a [Toolbar](https://bryntum.com/products/calendar/docs/api/Core/widget/Toolbar),
     * or array of config objects representing the child items of a Toolbar. Another way to add a bbar is to use [strips](https://bryntum.com/products/calendar/docs/api/Core/widget/Panel#config-strips).
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Scheduler/view/TimelineHistogram#config-bbar)
     */
    bbar ? : (ContainerItemConfig|string)[]|ToolbarConfig|PagingToolbarConfig|null
    /**
     * Custom CSS classes to add to the panel's body element.
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Scheduler/view/TimelineHistogram#config-bodyCls)
     */
    bodyCls ? : string|object
    /**
     * An object where property names with a truthy value indicate which events should bubble up the ownership
     * hierarchy when triggered.
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Scheduler/view/TimelineHistogram#config-bubbleEvents)
     */
    bubbleEvents ? : object
    /**
     * This config defines the size of the start and end invisible parts of the timespan when [infiniteScroll](https://bryntum.com/products/calendar/docs/api/Scheduler/view/mixin/TimelineScroll#config-infiniteScroll) set to `true`.
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Scheduler/view/TimelineHistogram#config-bufferCoef)
     */
    bufferCoef ? : number
    /**
     * This config defines the scroll limit, which, when exceeded will cause a timespan shift.
     * The limit is calculated as the `panelWidth * [bufferCoef](https://bryntum.com/products/calendar/docs/api/Scheduler/view/mixin/TimelineScroll#config-bufferCoef) * bufferThreshold`. During scrolling, if the left or right side
     * has less than that of the rendered content - a shift is triggered.
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Scheduler/view/TimelineHistogram#config-bufferThreshold)
     */
    bufferThreshold ? : number
    /**
     * Set to `false` to not call onXXX method names (e.g. `onShow`, `onClick`), as an easy way to listen for events.
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Scheduler/view/TimelineHistogram#config-callOnFunctions)
     */
    callOnFunctions ? : Boolean
    /**
     * Set to `false` to crop text in grid cells without ellipsis (...). When enabled, cells containing pure
     * use `display : block`, instead of `display : flex` to allow ellipsis to work.
     * <strong>NOTE</strong> Only supported in browsers that support `:has()` CSS selector
     */
    cellEllipsis ? : Boolean
    /**
     * Custom CSS classes to add to element.
     * May be specified as a space separated string, or as an object in which property names
     * with truthy values are used as the class names:
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Scheduler/view/TimelineHistogram#config-cls)
     */
    cls ? : string|object
    /**
     * Controls whether the panel is collapsed (the body of the panel is hidden while only the header is
     * visible). Only valid if the panel is [collapsible](https://bryntum.com/products/calendar/docs/api/Core/widget/Panel#config-collapsible).
     */
    collapsed ? : Boolean
    /**
     * This config enables collapsibility for the panel. See [collapsed](https://bryntum.com/products/calendar/docs/api/Core/widget/Panel#config-collapsed).
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Scheduler/view/TimelineHistogram#config-collapsible)
     */
    collapsible ? : Boolean|PanelCollapserConfig|PanelCollapserOverlayConfig
    /**
     * Set to `false` to not show column lines. End result might be overruled by/differ between themes.
     */
    columnLines ? : Boolean
    /**
     * Accepts column definitions for the grid during initialization. They will be used to create
     * [Column](https://bryntum.com/products/calendar/docs/api/Grid/column/Column) instances that are added to a [ColumnStore](#Grid/data/ColumnStore).
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Scheduler/view/TimelineHistogram#config-columns)
     */
    columns ? : ColumnStore|GridColumnConfig[]|ColumnStoreConfig
    config ? : object
    /**
     * Custom CSS classes to add to the [contentElement](https://bryntum.com/products/calendar/docs/api/Core/widget/Widget#property-contentElement).
     * May be specified as a space separated string, or as an object in which property names
     * with truthy values are used as the class names:
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Scheduler/view/TimelineHistogram#config-contentElementCls)
     */
    contentElementCls ? : string|object
    /**
     * Event which is used to show context menus.
     * Available options are: 'contextmenu', 'click', 'dblclick'.
     */
    contextMenuTriggerEvent ? : 'contextmenu'|'click'|'dblclick'
    /**
     * Convenient shortcut to set data in grids store both during initialization and at runtime. Can also be
     * used to retrieve data at runtime, although we do recommend interacting with Grids store instead using
     * the [store](https://bryntum.com/products/calendar/docs/api/Grid/view/GridBase#property-store) property.
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Scheduler/view/TimelineHistogram#config-data)
     */
    data ? : object[]|Model[]|ModelConfig[]
    /**
     * Record field from which the histogram data will be collected.
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Scheduler/view/TimelineHistogram#config-dataModelField)
     */
    dataModelField ? : string
    /**
     * Object to apply to elements dataset (each key will be used as a data-attribute on the element)
     */
    dataset ? : Record<string, string>
    /**
     * Region to which columns are added when they have none specified
     */
    defaultRegion ? : string
    /**
     * Set to `true` to destroy the store when the grid is destroyed.
     */
    destroyStore ? : Boolean
    /**
     * Disable or enable the widget. It is similar to [readOnly](https://bryntum.com/products/calendar/docs/api/Core/widget/Widget#config-readOnly) except a disabled widget
     * cannot be focused, uses a different rendition (usually greyish) and does not allow selecting its value.
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Scheduler/view/TimelineHistogram#config-disabled)
     */
    disabled ? : Boolean|'inert'
    /**
     * Set to `true` to not get a warning when calling [getState](https://bryntum.com/products/calendar/docs/api/Grid/view/GridBase#function-getState) when there is a column
     * configured without an `id`. But the recommended action is to always configure columns with an `id` when
     * using states.
     */
    disableGridColumnIdWarning ? : Boolean
    /**
     * Set to `true` to not get a warning when using another base class than GridRowModel for your grid data. If
     * you do, and would like to use the full feature set of the grid then include the fields from GridRowModel
     * in your model definition.
     */
    disableGridRowModelWarning ? : Boolean
    /**
     * Defines how dates will be formatted in tooltips etc. This config has priority over similar config on the
     * view preset. For allowed values see [format](https://bryntum.com/products/calendar/docs/api/Core/helper/DateHelper#function-format-static).
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Scheduler/view/TimelineHistogram#config-displayDateFormat)
     */
    displayDateFormat ? : string
    /**
     * Controls the placement of this widget when it is added to a [panel's ](https://bryntum.com/products/calendar/docs/api/Core/widget/Panel)
     * [strips collection](https://bryntum.com/products/calendar/docs/api/Core/widget/Panel#config-strips). Typical values for this config are `'top'`,
     * `'bottom'`, `'left'`, or `'right'`, which cause the widget to be placed on that side of the panel's
     * body. Such widgets are called "edge strips".
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Scheduler/view/TimelineHistogram#config-dock)
     */
    dock ? : 'top'|'bottom'|'left'|'right'|'start'|'end'|'header'|'pre-header'|object
    /**
     * Make this Panel a docked drawer which slides out from one side of the browser viewport by default.
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Scheduler/view/TimelineHistogram#config-drawer)
     * @param {'start','left','end','right','top','bottom'} side The side of the viewport to dock the drawer to.  * `'start'` means the `inline-start` side. * `'end'` means the `inline-end` side.
     * @param {string,number} size The size of the drawer in its collapsible axis.
     * @param {boolean} inline If using the [appendTo](https://bryntum.com/products/calendar/docs/api/Core/widget/Panel#config-appendTo) config to place the drawer inside a host widget, set may be set to `true` to make the drawer inline within that host. Note that, if using this, the layout of the host element must have `flex-direction` set appropriately.
     */
    drawer ? : Boolean|{side?: 'start'|'left'|'end'|'right'|'top'|'bottom', size?: string|number, inline?: Boolean}
    /**
     * Decimal precision used when displaying durations, used by tooltips and DurationColumn.
     * Specify `false` to use raw value
     */
    durationDisplayPrecision ? : number|Boolean
    /**
     * An object specifying attributes to assign to the root element of this widget.
     * Set `null` value to attribute to remove it.
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Scheduler/view/TimelineHistogram#config-elementAttributes)
     */
    elementAttributes ? : Record<string, string|null>
    /**
     * Text or HTML, or a [DomConfig](https://bryntum.com/products/calendar/docs/api/Core/helper/DomHelper#typedef-DomConfig) block to display when there is no data to display in the grid
     */
    emptyText ? : string|DomConfig
    /**
     * Set to `false` if you don't want event bar DOM updates to animate.
     */
    enableEventAnimations ? : Boolean
    /**
     * Enables showing occurrences of recurring events across the scheduler's time axis. If you want to disable
     * the recurrence popup, you can choose set the `defaultAction` to `future` to affect all future
     * occurrences, or `single` to just affect the currently selected event.
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Scheduler/view/TimelineHistogram#config-enableRecurringEvents)
     */
    enableRecurringEvents ? : Boolean|{
        defaultAction?: 'single'|'future'
    }
    /**
     * Configure this as `true` to allow elements within cells to be styled as `position: sticky`.
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Scheduler/view/TimelineHistogram#config-enableSticky)
     */
    enableSticky ? : Boolean
    /**
     * Set to `true` to allow text selection in the grid cells. Note, this cannot be used simultaneously with the
     * `RowReorder` feature.
     */
    enableTextSelection ? : Boolean
    /**
     * Set to `true` to listen for CTRL-Z (CMD-Z on Mac OS) keyboard event and trigger undo (redo when SHIFT is
     * pressed). Only applicable when using a [StateTrackingManager](https://bryntum.com/products/calendar/docs/api/Core/data/stm/StateTrackingManager).
     */
    enableUndoRedoKeys ? : Boolean
    /**
     * The end date of the timeline (if not configure with [infiniteScroll](https://bryntum.com/products/calendar/docs/api/Scheduler/view/TimelineBase#config-infiniteScroll)).
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Scheduler/view/TimelineHistogram#config-endDate)
     */
    endDate ? : Date|string
    /**
     * Event color used by default. Events and resources can specify their own color, with priority order being:
     * Event -&gt; Resource -&gt; Scheduler default.
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Scheduler/view/TimelineHistogram#config-eventColor)
     */
    eventColor ? : EventColor
    /**
     * Event style used by default. Events and resources can specify their own style, with priority order being:
     * Event -&gt; Resource -&gt; Scheduler default. Determines the appearance of the event by assigning a CSS class
     * to it. Available styles are:
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Scheduler/view/TimelineHistogram#config-eventStyle)
     */
    eventStyle ? : 'plain'|'border'|'colored'|'hollow'|'line'|'dashed'|'minimal'|'rounded'|'calendar'|'interday'|null
    extraData ? : any
    /**
     * Set to `true` to stretch the last column in a grid with all fixed width columns
     * to fill extra available space if the grid's width is wider than the sum of all
     * configured column widths.
     */
    fillLastColumn ? : Boolean
    /**
     * Specify `true` to force rendered events/tasks to fill entire ticks. This only affects rendering, start
     * and end dates retain their value on the data level.
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Scheduler/view/TimelineHistogram#config-fillTicks)
     */
    fillTicks ? : Boolean
    /**
     * Use fixed row height. Setting this to `true` will configure the underlying RowManager to use fixed row
     * height, which sacrifices the ability to use rows with variable height to gain a fraction better
     * performance.
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Scheduler/view/TimelineHistogram#config-fixedRowHeight)
     */
    fixedRowHeight ? : Boolean
    /**
     * When this widget is a child of a [Container](https://bryntum.com/products/calendar/docs/api/Core/widget/Container), it will by default be participating in a
     * flexbox layout. This config allows you to set this widget's
     * [flex](https://developer.mozilla.org/en-US/docs/Web/CSS/flex) style.
     * This may be configured as a single number or a `&lt;flex-grow&gt; &lt;flex-shrink&gt; &lt;flex-basis&gt;` format string.
     * numeric-only values are interpreted as the `flex-grow` value.
     */
    flex ? : number|string
    /**
     * Config object of a footer. May contain a `dock`, `html` and a `cls` property. A footer is not a widget,
     * but rather plain HTML that follows the last element of the panel's body and [strips](https://bryntum.com/products/calendar/docs/api/Core/widget/Panel#config-strips).
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Scheduler/view/TimelineHistogram#config-footer)
     */
    footer ? : {
        dock?: 'top'|'right'|'bottom'|'left'|'start'|'end'
        html?: string
        cls?: string
    }|string
    /**
     * Set to `true` to force the time columns to fit to the available space (horizontal or vertical depends on mode).
     * Note that setting [suppressFit](https://bryntum.com/products/calendar/docs/api/Scheduler/view/TimelineBase#config-suppressFit) to `true`, will disable `forceFit` functionality. Zooming
     * cannot be used when `forceFit` is set.
     */
    forceFit ? : Boolean
    /**
     * An object which names formula prefixes which will be applied to all columns configured with
     * `formula : true`.
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Scheduler/view/TimelineHistogram#config-formulaProviders)
     */
    formulaProviders ? : Record<string, FormulaProviderConfig>
    /**
     * Refresh entire row when a record changes (`true`) or, if possible, only the cells affected (`false`).
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Scheduler/view/TimelineHistogram#config-fullRowRefresh)
     */
    fullRowRefresh ? : Boolean
    /**
     * A Function which returns a CSS class name to add to a rectangle element.
     * The following parameters are passed:
     * @param {HistogramSeries} series The series being rendered
     * @param {DomConfig} domConfig The rectangle configuration object
     * @param {object} datum The datum being rendered
     * @param {number} index The index of the datum being rendered
     * @param {HistogramRenderData} renderData Current render data giving access to the record, row and cell being rendered.
     * @returns {string} CSS classes of the rectangle element
     */
    getBarClass ? : (series: HistogramSeries, domConfig: DomConfig, datum: object, index: number, renderData: HistogramRenderData) => string
    /**
     * A Function which if provided should return a DOM configuration object for a bar
     * (a `RECT` element representing a single "bar"-type value).
     * The function is passed a default prepared DOM configuration
     * in an argument which then can be processed and returned.
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Scheduler/view/TimelineHistogram#config-getBarDOMConfig)
     * @param {HistogramSeries} series The series being rendered
     * @param {DomConfig} domConfig The rectangle DOM configuration object
     * @param {object} datum The datum being rendered
     * @param {number} index The index of the datum being rendered
     * @param {HistogramRenderData} renderData Current render data giving access to the record, row and cell being rendered.
     * @returns {object} Resulting DOM configuration object. If no value returned the bar is not displayed.
     */
    getBarDOMConfig ? : (series: HistogramSeries, domConfig: DomConfig, datum: object, index: number, renderData: HistogramRenderData) => object
    /**
     * A Function which returns the text to render inside a bar.
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Scheduler/view/TimelineHistogram#config-getBarText)
     * @param {object} datum The datum being rendered
     * @param {number} index The index of the datum being rendered
     * @param {HistogramSeries} series The series (provided if histogram widget [singleTextForAllBars](https://bryntum.com/products/calendar/docs/api/Core/widget/graph/Histogram#config-singleTextForAllBars) is `false`)
     * @param {HistogramRenderData} renderData Current render data giving access to the record, row and cell being rendered.
     * @returns {string} Text to render inside the bar
     */
    getBarText ? : (datum: object, index: number, series: HistogramSeries, renderData: HistogramRenderData) => string
    /**
     * A Function which returns a DOM configuration object for text elements.
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Scheduler/view/TimelineHistogram#config-getBarTextDOMConfig)
     * @param {DomConfig} domConfig The `TEXT` element DOM configuration object
     * @param {object} datum The datum being rendered
     * @param {number} index The index of the datum being rendered
     * @param {HistogramSeries} series The series (provided if histogram widget [singleTextForAllBars](https://bryntum.com/products/calendar/docs/api/Core/widget/graph/Histogram#config-singleTextForAllBars) is `false`)
     * @param {HistogramRenderData} renderData Current render data giving access to the record, row and cell being rendered.
     * @returns {object} The `TEXT` element DOM configuration object. If no value returned the text element is not displayed.
     */
    getBarTextDOMConfig ? : (domConfig: DomConfig, datum: object, index: number, series: HistogramSeries, renderData: HistogramRenderData) => object
    /**
     * Function that extracts a record histogram data entry for aggregating.
     * By default it returns the entry as is. Override the function if you need a more complex way
     * to retrieve the value for aggregating.
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Scheduler/view/TimelineHistogram#config-getDataEntryForAggregating)
     * @param {object} getDataEntryForAggregating.entry Current data entry.
     * @returns {object} Entry to aggregate
     */
    getDataEntryForAggregating ? : ((getDataEntryForAggregating: { entry: object }) => object)|string
    /**
     * A Function which returns a CSS class name to add to a path element
     * built for an `outline` type series.
     * The following parameters are passed:
     * @param {HistogramSeries} series The series being rendered
     * @param {object[]} data The series data
     * @param {HistogramRenderData} renderData Current render data giving access to the record, row and cell being rendered.
     * @returns {string} CSS class name of the path element
     */
    getOutlineClass ? : (series: HistogramSeries, data: object[], renderData: HistogramRenderData) => string
    /**
     * A Function which if provided should return a DOM configuration object for
     * a `path` element built for an `outline` type series.
     * The function accepts a default prepared DOM configuration
     * in an argument which then can be processed and returned.
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Scheduler/view/TimelineHistogram#config-getOutlineDOMConfig)
     * @param {HistogramSeries} series The series being rendered
     * @param {DomConfig} domConfig The `path` element DOM configuration object
     * @param {object[]} data The series data
     * @returns {object} Resulting DOM configuration object. If no value returned the path is not displayed.
     */
    getOutlineDOMConfig ? : (series: HistogramSeries, domConfig: DomConfig, data: object[]) => object
    /**
     * A function, or name of a function which builds histogram data for the provided record.
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Scheduler/view/TimelineHistogram#config-getRecordData)
     * @param {Core.data.Model} getRecordData.record Record to get histogram data for.
     * @param {object} aggregationContext Context object passed in case the data is being retrieved as a part of some parent record data collecting.
     * @returns {object} Histogram data.
     */
    getRecordData ? : ((getRecordData: { record: Model }, aggregationContext?: object) => object)|string
    /**
     * A function called for each row to determine its height. It is passed a [record](https://bryntum.com/products/calendar/docs/api/Core/data/Model) and
     * expected to return the desired height of that records row. If the function returns a falsy value, Grids
     * configured [rowHeight](https://bryntum.com/products/calendar/docs/api/Grid/view/GridBase#config-rowHeight) is used.
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Scheduler/view/TimelineHistogram#config-getRowHeight)
     * @param {Core.data.Model} getRowHeight.record Record to determine row height for
     * @returns {number} Desired row height
     */
    getRowHeight ? : (getRowHeight: { record: Model }) => number
    /**
     * When set to `true` (default) the component reacts on time axis changes
     * (zooming or changing the displayed time span), clears the histogram data cache of the records
     * and then refreshes the view.
     */
    hardRefreshOnTimeAxisReconfigure ? : Boolean
    /**
     * A config [object](https://bryntum.com/products/calendar/docs/api/Core/widget/Panel#typedef-PanelHeader) for the panel's header or a string in place of a `title`.
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Scheduler/view/TimelineHistogram#config-header)
     */
    header ? : string|Boolean|PanelHeader
    /**
     * Widget's height, used to set element style.height. Either specify a valid height string or a number, which
     * will get 'px' appended. We recommend using CSS as the primary way to control height, but in some cases
     * this config is convenient.
     */
    height ? : string|number
    /**
     * Configure with true to make widget initially hidden.
     */
    hidden ? : Boolean
    /**
     * Set to `true` to hide the footer elements
     */
    hideFooters ? : Boolean
    /**
     * Set to `true` to hide the column header elements
     */
    hideHeaders ? : Boolean
    /**
     * Set to `true` to hide the Grid's horizontal scrollbar(s)
     */
    hideHorizontalScrollbar ? : Boolean
    /**
     * By default the row hover effect is not visible in the Scheduler part of the grid.
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Scheduler/view/TimelineHistogram#config-hideRowHover)
     */
    hideRowHover ? : Boolean
    /**
     * An instance or a configuration object of the underlying [Histogram](https://bryntum.com/products/calendar/docs/api/Core/widget/graph/Histogram)
     * component that is used under the hood to render actual charts.
     * In case a configuration object is provided the built class is defined with
     * [histogramWidgetClass](https://bryntum.com/products/calendar/docs/api/Scheduler/view/TimelineHistogramBase#config-histogramWidgetClass) config.
     */
    histogramWidget ? : Histogram|HistogramConfig
    /**
     * The class used for building the [histogram widget](https://bryntum.com/products/calendar/docs/api/Scheduler/view/TimelineHistogramBase#property-histogramWidget)
     */
    histogramWidgetClass ? : Histogram
    /**
     * A CSS class to add to hovered row elements
     */
    hoverCls ? : string
    /**
     * Widget id, if not specified one will be generated. Also used for lookups through Widget.getById
     */
    id ? : string
    /**
     * Set to `true` to ignore reacting to DOM events (mouseover/mouseout etc) while scrolling. Useful if you
     * want to maximize scroll performance.
     */
    ignoreDomEventsWhileScrolling ? : Boolean
    /**
     * Determines if the widgets read-only state should be controlled by its parent.
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Scheduler/view/TimelineHistogram#config-ignoreParentReadOnly)
     */
    ignoreParentReadOnly ? : Boolean
    /**
     * Set to `true` to automatically adjust the panel timespan during scrolling in the time dimension,
     * when the scroller comes close to the start/end edges.
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Scheduler/view/TimelineHistogram#config-infiniteScroll)
     */
    infiniteScroll ? : Boolean
    /**
     * A function that initializes a target group record entry.
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Scheduler/view/TimelineHistogram#config-initAggregatedDataEntry)
     * @returns {object} Target aggregated entry
     */
    initAggregatedDataEntry ? : (() => object)|string
    /**
     * Element (or element id) to insert this widget before. If provided, [appendTo](https://bryntum.com/products/calendar/docs/api/Core/widget/Widget#config-appendTo) config is ignored.
     */
    insertBefore ? : HTMLElement|string
    /**
     * Element (or element id) to append this widget element to, as a first child. If provided, [appendTo](https://bryntum.com/products/calendar/docs/api/Core/widget/Widget#config-appendTo) config is ignored.
     */
    insertFirst ? : HTMLElement|string
    /**
     * See [Keyboard shortcuts](https://bryntum.com/products/calendar/docs/api/Grid/view/Grid#keyboard-shortcuts) for details
     */
    keyMap ? : Record<string, KeyMapConfig>
    /**
     * The listener set for this object.
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Scheduler/view/TimelineHistogram#config-listeners)
     */
    listeners ? : TimelineHistogramListeners
    /**
     * A [Mask](https://bryntum.com/products/calendar/docs/api/Core/widget/Mask) config object, or a message to be shown when a store is performing a remote
     * operation, or Crud Manager is loading data from the sever. Set to `null` to disable default load mask.
     */
    loadMask ? : string|MaskConfig|null
    /**
     * A [Mask](https://bryntum.com/products/calendar/docs/api/Core/widget/Mask) config object to adjust the [maskDefaults](#Core/widget/Widget#config-maskDefaults)
     * when data is loading. The message and optional configuration from the
     * [loadMask](https://bryntum.com/products/calendar/docs/api/Core/mixin/LoadMaskable#config-loadMask) config take priority over these options, just as they do
     * for `maskDefaults`, respectively.
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Scheduler/view/TimelineHistogram#config-loadMaskDefaults)
     */
    loadMaskDefaults ? : MaskConfig
    /**
     * A [Mask](https://bryntum.com/products/calendar/docs/api/Core/widget/Mask) config object to adjust the [maskDefaults](#Core/widget/Widget#config-maskDefaults)
     * when an error occurs loading data.
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Scheduler/view/TimelineHistogram#config-loadMaskError)
     */
    loadMaskError ? : MaskConfig|Mask|Boolean
    /**
     * Set to `false` to disable localization of this object.
     */
    localizable ? : Boolean
    /**
     * Time in ms until a longpress is triggered
     */
    longPressTime ? : number
    /**
     * When `true`, events are sized and positioned based on rowHeight, resourceMargin and barMargin settings.
     * Set this to `false` if you want to control height and vertical position using CSS instead.
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Scheduler/view/TimelineHistogram#config-managedEventSizing)
     */
    managedEventSizing ? : Boolean
    /**
     * Widget's margin. This may be configured as a single number or a `TRBL` format string.
     * numeric-only values are interpreted as pixels.
     */
    margin ? : number|string
    /**
     * Grids change the `maskDefaults` to cover only their `body` element.
     */
    maskDefaults ? : MaskConfig
    /**
     * Set to `true` to apply the default mask to the widget. Alternatively, this can be the mask message or a
     * [Mask](https://bryntum.com/products/calendar/docs/api/Core/widget/Mask) config object.
     */
    masked ? : Boolean|string|MaskConfig
    /**
     * Get/set end date limit of the timeline. Actions such as timeline scrolling, all types of timeline zooms and shifts will respect this limit.
     */
    maxDate ? : Date|string
    /**
     * The element's maxHeight. Can be either a String or a Number (which will have 'px' appended). Note that
     * like [height](https://bryntum.com/products/calendar/docs/api/Core/widget/Widget#config-height), *reading* the value will return the numeric value in pixels.
     */
    maxHeight ? : string|number
    /**
     * The maximum time axis unit to display non-working ranges for ('hour' or 'day' etc).
     * When zooming to a view with a larger unit, no non-working time elements will be rendered.
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Scheduler/view/TimelineHistogram#config-maxTimeAxisUnit)
     */
    maxTimeAxisUnit ? : DurationUnit
    /**
     * The elements maxWidth. Can be either a String or a Number (which will have 'px' appended). Note that
     * like [width](https://bryntum.com/products/calendar/docs/api/Core/widget/Widget#config-width), *reading* the value will return the numeric value in pixels.
     */
    maxWidth ? : string|number
    /**
     * The maximum zoom level to which [zoomIn](https://bryntum.com/products/calendar/docs/api/Scheduler/view/mixin/TimelineZoomable#function-zoomIn) will work. Defaults to the number of
     * [ViewPresets](https://bryntum.com/products/calendar/docs/api/Scheduler/preset/ViewPreset) available, see [presets](#Scheduler/view/mixin/TimelineViewPresets#property-presets)
     * for information. Unless you have modified the collection of available presets, the max zoom level is
     * milliseconds.
     */
    maxZoomLevel ? : number
    /**
     * Get/set start date limit of the timeline. Actions such as timeline scrolling, all types of timeline zooms and shifts will respect this limit.
     */
    minDate ? : Date|string
    /**
     * Grid's `min-height`. Defaults to `10em` to be sure that the Grid always has a height wherever it is
     * inserted.
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Scheduler/view/TimelineHistogram#config-minHeight)
     */
    minHeight ? : string|number
    /**
     * The elements minWidth. Can be either a String or a Number (which will have 'px' appended). Note that
     * like [width](https://bryntum.com/products/calendar/docs/api/Core/widget/Widget#config-width), *reading* the value will return the numeric value in pixels.
     */
    minWidth ? : string|number
    /**
     * The minimum zoom level to which [zoomOut](https://bryntum.com/products/calendar/docs/api/Scheduler/view/mixin/TimelineZoomable#function-zoomOut) will work. Defaults to 0 (year ticks)
     */
    minZoomLevel ? : number
    /**
     * Grid monitors window resize by default.
     */
    monitorResize ? : Boolean
    /**
     * The owning Widget of this Widget. If this Widget is directly contained (that is, it is one of the
     * [items](https://bryntum.com/products/calendar/docs/api/Core/widget/Container#property-items) of a Container), this config will be ignored. In this case
     * the owner is <strong>always</strong> the encapsulating Container.
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Scheduler/view/TimelineHistogram#config-owner)
     */
    owner ? : Widget
    /**
     * Partners this Timeline panel with another Timeline in order to sync their region sizes (sub-grids like locked, normal will get the same width),
     * start and end dates, view preset, zoom level and scrolling position. All these values will be synced with the timeline defined as the `partner`.
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Scheduler/view/TimelineHistogram#config-partner)
     */
    partner ? : TimelineBase
    /**
     * Specify plugins (an array of classes) in config
     */
    plugins ? : Function[]
    /**
     * True to preserve focused cell after loading new data
     */
    preserveFocusOnDatasetChange ? : Boolean
    /**
     * Preserve the grid's vertical scroll position when changesets are applied, as in the case of remote
     * changes, or when stores are configured with [syncDataOnLoad](https://bryntum.com/products/calendar/docs/api/Core/data/Store#config-syncDataOnLoad).
     */
    preserveScroll ? : PreserveScrollOptions|Boolean
    /**
     * Specify `true` to preserve vertical scroll position after store actions that trigger a `refresh` event,
     * such as loading new data and filtering.
     */
    preserveScrollOnDatasetChange ? : Boolean
    /**
     * An array of [ViewPreset](https://bryntum.com/products/calendar/docs/api/Scheduler/preset/ViewPreset) config objects
     * which describes the available timeline layouts for this scheduler.
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Scheduler/view/TimelineHistogram#config-presets)
     */
    presets ? : ViewPresetConfig[]
    /**
     * Prevent tooltip from being displayed on touch devices. Useful for example for buttons that display a
     * menu on click etc, since the tooltip would be displayed at the same time.
     */
    preventTooltipOnTouch ? : Boolean
    /**
     * Set to `true` to make the grid read-only, by disabling any UIs for modifying data.
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Scheduler/view/TimelineHistogram#config-readOnly)
     */
    readOnly ? : Boolean
    /**
     * The confirmation dialog shown when a recurring event is edited.
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Scheduler/view/TimelineHistogram#config-recurrenceConfirmationPopup)
     */
    recurrenceConfirmationPopup ? : RecurrenceConfirmationPopupConfig|RecurrenceConfirmationPopup
    relayStoreEvents ? : Boolean
    /**
     * Configure this property to allow the widget/component to be resized. Pressing <kbd>Shift</kbd> while resizing will
     * constrain the aspect ratio.
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Scheduler/view/TimelineHistogram#config-resizable)
     */
    resizable ? : Boolean|{
        minWidth?: number
        maxWidth?: number
        minHeight?: number
        maxHeight?: number
        handles?: object
    }
    /**
     * Set to `false` to only measure cell contents when double-clicking the edge between column headers.
     */
    resizeToFitIncludesHeader ? : Boolean
    /**
     * An empty function by default, but provided so that you can override it. This function is called each time
     * a resource time range is rendered into the schedule. It's called with `resourceTimeRangeRecord`, `resourceRecord`,
     * and `renderData` params.
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Scheduler/view/TimelineHistogram#config-resourceTimeRangeRenderer)
     * @param {object} detail An object containing the information needed to render a ResourceTimeRangeModel.
     * @param {Scheduler.model.ResourceTimeRangeModel} detail.resourceTimeRangeRecord The resource time range record.
     * @param {Scheduler.model.ResourceModel} detail.resourceRecord The resource record.
     * @param {object} detail.renderData An object containing details about the event rendering.
     * @returns {string,DomConfig,DomConfig[]} A simple string, or a DomConfig object describing the markup
     */
    resourceTimeRangeRenderer ? : (detail: { resourceTimeRangeRecord: ResourceTimeRangeModel, resourceRecord: ResourceModel, renderData: object }) => string|DomConfig|DomConfig[]
    /**
     * "Break points" for which responsive config to use for columns and css.
     */
    responsiveLevels ? : Record<string, number|string>
    /**
     * Configure as `true` to have the component display a translucent ripple when its
     * [focusElement](https://bryntum.com/products/calendar/docs/api/Core/widget/Widget#property-focusElement), or [element](#Core/widget/Widget#property-element) is tapped <em>if the
     * current theme supports ripples</em>. Out of the box, only the Material theme supports ripples.
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Scheduler/view/TimelineHistogram#config-ripple)
     */
    ripple ? : Boolean|{
        delegate?: string
        color?: string
        radius?: number
        clip?: string
    }
    /**
     * If you are rendering this widget to a shadow root inside a web component, set this config to the shadowRoot. If not inside a web component, set it to `document.body`
     */
    rootElement ? : ShadowRoot|HTMLElement
    /**
     * Row height in pixels. This allows the default height for rows to be controlled. Note that it may be
     * overriden by specifying a [rowHeight](https://bryntum.com/products/calendar/docs/api/Grid/data/GridRowModel#field-rowHeight) on a per record basis, or from
     * a column [renderer](https://bryntum.com/products/calendar/docs/api/Grid/column/Column#config-renderer).
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Scheduler/view/TimelineHistogram#config-rowHeight)
     */
    rowHeight ? : number
    /**
     * Set to `false` to not show row lines. End result might be overruled by/differ between themes.
     */
    rowLines ? : Boolean
    /**
     * This may be configured as `true` to make the widget's element use the `direction:rtl` style.
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Scheduler/view/TimelineHistogram#config-rtl)
     */
    rtl ? : Boolean
    /**
     * An object with configuration for the [ScaleColumn](https://bryntum.com/products/calendar/docs/api/Scheduler/column/ScaleColumn).
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Scheduler/view/TimelineHistogram#config-scaleColumn)
     */
    scaleColumn ? : ScaleColumnConfig
    /**
     * Array of objects representing the [scale column](https://bryntum.com/products/calendar/docs/api/Scheduler/view/mixin/TimelineHistogramScaleColumn#config-scaleColumn) scale points.
     * The config basically is a mapping to the column [scalePoints](https://bryntum.com/products/calendar/docs/api/Scheduler/column/ScaleColumn#config-scalePoints)
     * config.
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Scheduler/view/TimelineHistogram#config-scalePoints)
     */
    scalePoints ? : ScalePoint[]
    /**
     * Configures whether the grid is scrollable in the `Y` axis. This is used to configure a [Scroller](https://bryntum.com/products/calendar/docs/api/Core/helper/util/Scroller).
     * See the [scrollerClass](https://bryntum.com/products/calendar/docs/api/Grid/view/GridBase#config-scrollerClass) config option.
     */
    scrollable ? : Boolean|ScrollerConfig|Scroller
    /**
     * The class to instantiate to use as the [scrollable](https://bryntum.com/products/calendar/docs/api/Grid/view/GridBase#config-scrollable). Defaults to [Scroller](#Core/helper/util/Scroller).
     */
    scrollerClass ? : typeof Scroller
    /**
     * Configuration values for the [ScrollManager](https://bryntum.com/products/calendar/docs/api/Core/util/ScrollManager) class on initialization. Returns the
     * [ScrollManager](https://bryntum.com/products/calendar/docs/api/Core/util/ScrollManager) at runtime.
     */
    scrollManager ? : ScrollManagerConfig|ScrollManager
    /**
     * Selection configuration settings, change these properties to control how selection works and what can be
     * selected.
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Scheduler/view/TimelineHistogram#config-selectionMode)
     */
    selectionMode ? : GridSelectionMode
    /**
     * Object enumerating data series for the histogram.
     * The object keys are treated as the series identifiers and values are objects that
     * must contain two properties:
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Scheduler/view/TimelineHistogram#config-series)
     */
    series ? : Record<string, HistogramSeries>
    /**
     * Set to `true` if you want to display a tooltip when hovering an allocation bar. You can also pass a
     * [Tooltip#configs](https://bryntum.com/products/calendar/docs/api/Core/widget/Tooltip#configs) config object.
     * Please use [barTooltipTemplate](https://bryntum.com/products/calendar/docs/api/Scheduler/view/TimelineHistogramBase#config-barTooltipTemplate) function to customize the tooltip contents.
     */
    showBarTip ? : Boolean|TooltipConfig
    /**
     * Configure as `true` to have the grid show a red "changed" tag in cells whose
     * field value has changed and not yet been committed.
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Scheduler/view/TimelineHistogram#config-showDirty)
     */
    showDirty ? : Boolean|{
        duringEdit?: Boolean
    }
    /**
     * Set to `true` to snap to the current time resolution increment while interacting with scheduled events.
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Scheduler/view/TimelineHistogram#config-snap)
     */
    snap ? : Boolean
    /**
     * Affects drag drop and resizing of events when [snap](https://bryntum.com/products/calendar/docs/api/Scheduler/view/mixin/TimelineDateMapper#config-snap)
     * is enabled.
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Scheduler/view/TimelineHistogram#config-snapRelativeToEventStartDate)
     */
    snapRelativeToEventStartDate ? : Boolean
    /**
     * The start date of the timeline (if not configure with [infiniteScroll](https://bryntum.com/products/calendar/docs/api/Scheduler/view/TimelineBase#config-infiniteScroll)).
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Scheduler/view/TimelineHistogram#config-startDate)
     */
    startDate ? : Date|string
    /**
     * This value can be one of the following:
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Scheduler/view/TimelineHistogram#config-stateful)
     */
    stateful ? : Boolean|object|string[]
    /**
     * The events that, when fired by this component, should trigger it to save its state by calling
     * [saveState](https://bryntum.com/products/calendar/docs/api/Core/mixin/State#function-saveState).
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Scheduler/view/TimelineHistogram#config-statefulEvents)
     */
    statefulEvents ? : object|string[]
    /**
     * The key to use when saving this object's state in the [stateProvider](https://bryntum.com/products/calendar/docs/api/Core/mixin/State#config-stateProvider). If this config is
     * not assigned, and [stateful](https://bryntum.com/products/calendar/docs/api/Core/mixin/State#config-stateful) is not set to `false`, the [id](#Core/widget/Widget#config-id)
     * (if explicitly specified) will be used as the `stateId`.
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Scheduler/view/TimelineHistogram#config-stateId)
     */
    stateId ? : string
    /**
     * The `StateProvider` to use to save and restore this object's [state](https://bryntum.com/products/calendar/docs/api/Core/mixin/State#property-state). By default, `state`
     * will be saved using the [default state provider](https://bryntum.com/products/calendar/docs/api/Core/state/StateProvider#property-instance-static).
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Scheduler/view/TimelineHistogram#config-stateProvider)
     */
    stateProvider ? : StateProvider
    /**
     * The properties of this settings object controls how grid is restored from state data.
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Scheduler/view/TimelineHistogram#config-stateSettings)
     */
    stateSettings ? : {
        restoreUnconfiguredColumns?: Boolean
    }
    /**
     * When set, the text in the major time axis header sticks in the scrolling viewport as long as possible.
     */
    stickyHeaders ? : Boolean
    /**
     * Store that holds records to display in the grid, or a store config object. If the configuration contains
     * a `readUrl`, an `AjaxStore` will be created.
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Scheduler/view/TimelineHistogram#config-store)
     */
    store ? : Store|StoreConfig|AjaxStore|AjaxStoreConfig
    /**
     * An object containing widgets keyed by name. By default (when no `type` is given), strips are
     * [toolbars](https://bryntum.com/products/calendar/docs/api/Core/widget/Toolbar). If you want to pass an array, you can use
     * the toolbar's [items](https://bryntum.com/products/calendar/docs/api/Core/widget/Container#config-items).
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Scheduler/view/TimelineHistogram#config-strips)
     */
    strips ? : Record<string, ContainerItemConfig>
    /**
     * An object containing sub grid configuration objects keyed by a `region` property.
     * By default, grid has a 'locked' region (if configured with locked columns) and a 'normal' region.
     * The 'normal' region defaults to use `flex: 1`.
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Scheduler/view/TimelineHistogram#config-subGridConfigs)
     */
    subGridConfigs ? : Record<string, SubGridConfig>
    /**
     * Set to `true` to prevent auto calculating of a minimal [tickSize](https://bryntum.com/products/calendar/docs/api/Scheduler/view/mixin/TimelineEventRendering#property-tickSize)
     * to always fit the content to the screen size. Setting this property on `true` will disable [forceFit](https://bryntum.com/products/calendar/docs/api/Scheduler/view/TimelineBase#config-forceFit) behaviour.
     */
    suppressFit ? : Boolean
    /**
     * A [Mask](https://bryntum.com/products/calendar/docs/api/Core/widget/Mask) config object, or a message to be shown when Crud Manager
     * is persisting changes on the server. Set to `null` to disable default sync mask.
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Scheduler/view/TimelineHistogram#config-syncMask)
     */
    syncMask ? : string|MaskConfig|null
    /**
     * A configuration for the [tab](https://bryntum.com/products/calendar/docs/api/Core/widget/Tab) created for this widget when it is placed in a
     * [TabPanel](https://bryntum.com/products/calendar/docs/api/Core/widget/TabPanel). For example, this config can be used to control the icon of the `tab` for
     * this widget:
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Scheduler/view/TimelineHistogram#config-tab)
     */
    tab ? : Boolean|TabConfig
    /**
     * When this container is used as a tab in a TabPanel, these items are added to the
     * [TabBar](https://bryntum.com/products/calendar/docs/api/Core/widget/TabBar) when this container is the active tab.
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Scheduler/view/TimelineHistogram#config-tabBarItems)
     */
    tabBarItems ? : ToolbarItems[]|Widget[]
    /**
     * A Config object representing the configuration of a [Toolbar](https://bryntum.com/products/calendar/docs/api/Core/widget/Toolbar),
     * or array of config objects representing the child items of a Toolbar.
     * This creates a toolbar docked to the top of the panel immediately below the header.
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Scheduler/view/TimelineHistogram#config-tbar)
     */
    tbar ? : (ContainerItemConfig|string)[]|ToolbarConfig|PagingToolbarConfig|null
    /**
     * The width/height (depending on vertical / horizontal mode) of all the time columns.
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Scheduler/view/TimelineHistogram#config-tickSize)
     */
    tickSize ? : number
    /**
     * A [TimeAxis](https://bryntum.com/products/calendar/docs/api/Scheduler/data/TimeAxis) config object or instance, used to create a backing data store of
     * 'ticks' providing the input date data for the time axis of timeline panel. Created automatically if none
     * supplied.
     */
    timeAxis ? : TimeAxisConfig|TimeAxis
    /**
     * Gets/sets the current time resolution object, which contains a unit identifier and an increment count
     * `{ unit, increment }`. This value means minimal task duration you can create using UI.
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Scheduler/view/TimelineHistogram#config-timeResolution)
     */
    timeResolution ? : object|number
    /**
     * Set to a time zone or a UTC offset. This will set the projects
     * [timeZone](https://bryntum.com/products/calendar/docs/api/Scheduler/model/ProjectModel#config-timeZone) config accordingly. As this config is only a referer,
     * please see project's config [documentation](https://bryntum.com/products/calendar/docs/api/Scheduler/model/ProjectModel#config-timeZone) for more
     * information.
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Scheduler/view/TimelineHistogram#config-timeZone)
     */
    timeZone ? : string|number
    /**
     * A title to display in the header or owning TabPanel. Causes creation and docking of a header
     * to the top if no header is configured.
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Scheduler/view/TimelineHistogram#config-title)
     */
    title ? : string
    /**
     * The [tools](https://bryntum.com/products/calendar/docs/api/Core/widget/Tool) to add either before or after the `title` in the Panel header. Each
     * property name is the reference by which an instantiated tool may be retrieved from the live
     * `[tools](https://bryntum.com/products/calendar/docs/api/Core/widget/mixin/Toolable#property-tools)` property.
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Scheduler/view/TimelineHistogram#config-tools)
     */
    tools ? : Record<string, ToolConfig>|null
    /**
     * Animation transition duration in milliseconds.
     */
    transitionDuration ? : number
    type ? : 'timelinehistogram'
    /**
     * Custom CSS class name suffixes to apply to the elements rendered by this widget. This may be specified
     * as a space separated string, an array of strings, or as an object in which property names with truthy
     * values are used as the class names.
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Scheduler/view/TimelineHistogram#config-ui)
     */
    ui ? : 'plain'|'toolbar'|string|object
    /**
     * A string key used to lookup a predefined [ViewPreset](https://bryntum.com/products/calendar/docs/api/Scheduler/preset/ViewPreset) (e.g. 'weekAndDay', 'hourAndDay'),
     * managed by [PresetManager](https://bryntum.com/products/calendar/docs/api/Scheduler/preset/PresetManager). See [PresetManager](#Scheduler/preset/PresetManager) for more information.
     * Or a config object for a viewPreset.
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Scheduler/view/TimelineHistogram#config-viewPreset)
     */
    viewPreset ? : string|ViewPresetConfig
    /**
     * A date to bring into view initially on the scrollable timeline.
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Scheduler/view/TimelineHistogram#config-visibleDate)
     */
    visibleDate ? : Date|VisibleDate
    /**
     * Integer number indicating the size of timespan during zooming. When zooming, the timespan is adjusted to make
     * the scrolling area `visibleZoomFactor` times wider than the timeline area itself. Used in
     * [zoomToSpan](https://bryntum.com/products/calendar/docs/api/Scheduler/view/mixin/TimelineZoomable#function-zoomToSpan) and [zoomToLevel](#Scheduler/view/mixin/TimelineZoomable#function-zoomToLevel) functions.
     */
    visibleZoomFactor ? : number
    /**
     * A valid JS day index between 0-6 (0: Sunday, 1: Monday etc.) to be considered the start day of the week.
     * When omitted, the week start day is retrieved from the active locale class.
     */
    weekStartDay ? : number
    /**
     * A widgets weight determines its position among siblings when added to a [Container](https://bryntum.com/products/calendar/docs/api/Core/widget/Container).
     * Higher weights go further down.
     */
    weight ? : number
    /**
     * Widget's width, used to set element style.width. Either specify a valid width string or a number, which
     * will get 'px' appended. We recommend using CSS as the primary way to control width, but in some cases
     * this config is convenient.
     */
    width ? : string|number
    /**
     * An object with format `{ fromDay, toDay, fromHour, toHour }` that describes the working days and hours.
     * This object will be used to populate TimeAxis [include](https://bryntum.com/products/calendar/docs/api/Scheduler/data/TimeAxis#config-include) property.
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Scheduler/view/TimelineHistogram#config-workingTime)
     */
    workingTime ? : {
        fromDay?: number
        toDay?: number
        fromHour?: number
        toHour?: number
    }
    /**
     * Whether the originally rendered timespan should be preserved while zooming. By default, it is set to `false`,
     * meaning the timeline panel will adjust the currently rendered timespan to limit the amount of HTML content to
     * render. When setting this option to `true`, be careful not to allow to zoom a big timespan in seconds
     * resolution for example. That will cause <strong>a lot</strong> of HTML content to be rendered and affect performance. You
     * can use [minZoomLevel](https://bryntum.com/products/calendar/docs/api/Scheduler/view/mixin/TimelineZoomable#config-minZoomLevel) and [maxZoomLevel](#Scheduler/view/mixin/TimelineZoomable#config-maxZoomLevel) config options for that.
     */
    zoomKeepsOriginalTimespan ? : Boolean
    /**
     * If true, you can zoom in and out on the time axis using CTRL-key + mouse wheel.
     */
    zoomOnMouseWheel ? : Boolean
    /**
     * True to zoom to time span when double-clicking a time axis cell.
     */
    zoomOnTimeAxisDoubleClick ? : Boolean

    // Features
    /**
     * Allows using `[Ctrl/CMD + C]`, `[Ctrl/CMD + X]` and `[Ctrl/CMD + V]` to cut, copy and paste cell or cell ranges. Also
     * makes cut, copy and paste actions available via the cell context menu.
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Grid/feature/CellCopyPaste)
     */
    cellCopyPasteFeature ? : object|Boolean|string|CellCopyPaste|CellCopyPasteConfig
    /**
     * Adding this feature to the grid and other Bryntum products which are based on the Grid (i.e. Scheduler, SchedulerPro, and Gantt)
     * enables cell editing. Any subclass of [Field](https://bryntum.com/products/calendar/docs/api/Core/widget/Field) can be used
     * as editor for the [Column](https://bryntum.com/products/calendar/docs/api/Grid/column/Column). The most popular are:
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Grid/feature/CellEdit)
     */
    cellEditFeature ? : object|Boolean|string|CellEdit|CellEditConfig
    /**
     * Right click to display context menu for cells.
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Grid/feature/CellMenu)
     */
    cellMenuFeature ? : object|Boolean|string|CellMenu|CellMenuConfig
    /**
     * Displays a tooltip when hovering cells.
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Grid/feature/CellTooltip)
     */
    cellTooltipFeature ? : object|Boolean|string|CellTooltip|CellTooltipConfig
    /**
     * Enables the [autoWidth](https://bryntum.com/products/calendar/docs/api/Grid/column/Column#config-autoWidth) config for a grid's columns.
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Grid/feature/ColumnAutoWidth)
     */
    columnAutoWidthFeature ? : object|Boolean|string|ColumnAutoWidth|ColumnAutoWidthConfig
    /**
     * Displays a toolbar while dragging column headers. Drop on a button in the toolbar to activate a certain function,
     * for example to group by that column. This feature simplifies certain operations on touch devices.
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Grid/feature/ColumnDragToolbar)
     */
    columnDragToolbarFeature ? : object|Boolean|string|ColumnDragToolbar|ColumnDragToolbarConfig
    /**
     * Displays column lines for ticks, with a different styling for major ticks (by default they are darker). If this
     * feature is disabled, no lines are shown. If it's enabled, line are shown for the tick level which is set in current
     * ViewPreset. Please see [columnLinesFor](https://bryntum.com/products/calendar/docs/api/Scheduler/preset/ViewPreset#field-columnLinesFor) config for details.
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Scheduler/feature/ColumnLines)
     */
    columnLinesFeature ? : object|Boolean|string|ColumnLines|ColumnLinesConfig
    /**
     * Displays a column picker (to show/hide columns) in the header context menu. Columns can be displayed in sub menus
     * by region or tag. Grouped headers are displayed as menu hierarchies.
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Grid/feature/ColumnPicker)
     */
    columnPickerFeature ? : object|Boolean|string|ColumnPicker|ColumnPickerConfig
    /**
     * Allows user to rename columns by either right-clicking column header or using keyboard shortcuts when column header
     * is focused.
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Grid/feature/ColumnRename)
     */
    columnRenameFeature ? : object|Boolean|string|ColumnRename|ColumnRenameConfig
    /**
     * Allows user to reorder columns by dragging headers. To get notified about column reorder listen to `change` event
     * on [columns](https://bryntum.com/products/calendar/docs/api/Grid/data/ColumnStore) store.
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Grid/feature/ColumnReorder)
     */
    columnReorderFeature ? : object|Boolean|string|ColumnReorder|ColumnReorderConfig
    /**
     * Enables user to resize columns by dragging a handle on the right hand side of the header. To get notified about column
     * resize listen to `change` event on [columns](https://bryntum.com/products/calendar/docs/api/Grid/data/ColumnStore) store.
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Grid/feature/ColumnResize)
     */
    columnResizeFeature ? : object|Boolean|string|ColumnResize|ColumnResizeConfig
    /**
     * A feature that allows exporting Grid data to Excel or CSV without involving the server. It uses
     * [TableExporter](https://bryntum.com/products/calendar/docs/api/Grid/util/TableExporter) class as data provider, 3rd party provider to generate XLS files, and
     * [Microsoft XML specification](https://msdn.microsoft.com/en-us/library/office/documentformat.openxml.spreadsheet.aspx).
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Grid/feature/experimental/ExcelExporter)
     */
    excelExporterFeature ? : object|Boolean|string|GridExcelExporter|GridExcelExporterConfig
    /**
     * An experimental feature that lets users drop files on a Widget. The widget fires an event when a file is dropped onto it.
     * In the event, you get access to the raw files as strings, that were parsed by calling `readAsBinaryString`.
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Grid/feature/experimental/FileDrop)
     */
    fileDropFeature ? : object|Boolean|string|FileDrop|FileDropConfig
    /**
     * This feature adds a fill handle to a Grid range selection, which when dragged, fills the cells being dragged over
     * with values based on the values in the original selected range. This is similar to functionality normally seen in
     * various spreadsheet applications.
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Grid/feature/FillHandle)
     */
    fillHandleFeature ? : object|Boolean|string|FillHandle|FillHandleConfig
    /**
     * Feature that allows filtering of the grid by settings filters on columns. The actual filtering is done by the store.
     * For info on programmatically handling filters, see [StoreFilter](https://bryntum.com/products/calendar/docs/api/Core/data/mixin/StoreFilter).
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Grid/feature/Filter)
     */
    filterFeature ? : object|Boolean|string|Filter|FilterConfig
    /**
     * Feature that allows filtering of the grid by entering filters on column headers.
     * The actual filtering is done by the store.
     * For info on programmatically handling filters, see [StoreFilter](https://bryntum.com/products/calendar/docs/api/Core/data/mixin/StoreFilter).
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Grid/feature/FilterBar)
     */
    filterBarFeature ? : object|Boolean|string|FilterBar|FilterBarConfig
    /**
     * Enables rendering and handling of row groups. The actual grouping is done in the store, but triggered by `[shift]` +
     * clicking headers or by using two finger tap (one on header, one anywhere on grid). Use `[shift]` + `[alt]` + click to
     * remove a column grouper.
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Grid/feature/Group)
     */
    groupFeature ? : object|Boolean|string|Group|GroupConfig
    /**
     * A special version of the Grid GroupSummary feature that enables summaries within scheduler. To use a single summary
     * it is easiest to configure [renderer](https://bryntum.com/products/calendar/docs/api/Scheduler/feature/GroupSummary#config-renderer), for multiple summaries see [summaries](#Scheduler/feature/GroupSummary#config-summaries).
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Scheduler/feature/GroupSummary)
     */
    groupSummaryFeature ? : object|Boolean|string|GroupSummary|GroupSummaryConfig
    /**
     * Right click column header or focus it and press SPACE key to show the context menu for headers.
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Grid/feature/HeaderMenu)
     */
    headerMenuFeature ? : object|Boolean|string|HeaderMenu|HeaderMenuConfig
    /**
     * Enables users to click and drag to zoom to a date range in Scheduler's header time axis. Only supported in horizontal
     * mode.
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Scheduler/feature/HeaderZoom)
     */
    headerZoomFeature ? : object|Boolean|string|HeaderZoom|HeaderZoomConfig
    /**
     * This feature allows records which satisfy a certain condition to be locked at the top of the grid.
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Grid/feature/LockRows)
     */
    lockRowsFeature ? : object|Boolean|string|GridLockRows|GridLockRowsConfig
    /**
     * This feature merges cells that have the same value in sorted (or [optionally](https://bryntum.com/products/calendar/docs/api/Grid/feature/MergeCells#config-sortedOnly) any) columns
     * configured to [mergeCells](https://bryntum.com/products/calendar/docs/api/Grid/column/Column#config-mergeCells).
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Grid/feature/MergeCells)
     */
    mergeCellsFeature ? : object|Boolean|string|MergeCells|MergeCellsConfig
    /**
     * Feature that allows styling of weekends (and other non-working time) by adding timeRanges for those days.
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Scheduler/feature/NonWorkingTime)
     */
    nonWorkingTimeFeature ? : object|Boolean|string|NonWorkingTime|NonWorkingTimeConfig
    /**
     * Makes the scheduler's timeline pannable by dragging with the mouse. Try it out in the demo below.
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Scheduler/feature/Pan)
     */
    panFeature ? : object|Boolean|string|Pan|PanConfig
    /**
     * Generates PDF/PNG files from the Grid component.
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Grid/feature/export/PdfExport)
     */
    pdfExportFeature ? : object|Boolean|string|GridPdfExport|GridPdfExportConfig
    /**
     * Allows printing Grid contents using browser print dialog.
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Grid/feature/export/GridPrint)
     */
    printFeature ? : object|Boolean|string|GridPrint|GridPrintConfig
    /**
     * Feature that allows the user to search in a column by focusing a cell and typing. Navigate between hits using the
     * keyboard, [f3] or [ctrl]/[cmd] + [g] moves to next, also pressing [shift] moves to previous.
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Grid/feature/QuickFind)
     */
    quickFindFeature ? : object|Boolean|string|QuickFind|QuickFindConfig
    /**
     * Makes the splitter between grid [sections](https://bryntum.com/products/calendar/docs/api/Grid/view/SubGrid) draggable, to let users resize, and
     * collapse/expand the sections.
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Grid/feature/RegionResize)
     */
    regionResizeFeature ? : object|Boolean|string|RegionResize|RegionResizeConfig
    /**
     * Applicable only for Scheduler in `vertical` mode. Right click resource header cells to display a context menu.
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Scheduler/feature/ResourceMenu)
     */
    resourceMenuFeature ? : object|Boolean|string|ResourceMenu|ResourceMenuConfig
    /**
     * Feature that draws resource time ranges, shaded areas displayed behind events. These zones are similar to events in
     * that they have a start and end date but different in that they do not take part in the event layout, and they always
     * occupy full row height.
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Scheduler/feature/ResourceTimeRanges)
     */
    resourceTimeRangesFeature ? : object|Boolean|string|ResourceTimeRanges|ResourceTimeRangesConfig
    /**
     * Allow using [Ctrl/CMD + C/X] and [Ctrl/CMD + V] to copy/cut-and-paste rows. Also makes cut, copy and paste actions
     * available via the cell context menu.
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Grid/feature/RowCopyPaste)
     */
    rowCopyPasteFeature ? : object|Boolean|string|RowCopyPaste|RowCopyPasteConfig
    /**
     * This feature allows editing of entire rows in a grid in a docked panel which by default slides out from the right.
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Grid/feature/RowEdit)
     */
    rowEditFeature ? : object|Boolean|string|RowEdit|RowEditConfig
    /**
     * Enables expanding of Grid rows by either row click or double click, or by adding a separate Grid column which renders
     * a button that expands or collapses the row.
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Grid/feature/RowExpander)
     */
    rowExpanderFeature ? : object|Boolean|string|RowExpander|RowExpanderConfig
    /**
     * This feature implements support for project transactions and used by default in Gantt. For general RowReorder feature
     * documentation see [RowReorder](https://bryntum.com/products/calendar/docs/api/Grid/feature/RowReorder).
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Scheduler/feature/RowReorder)
     */
    rowReorderFeature ? : object|Boolean|string|RowReorder|RowReorderConfig
    /**
     * Enables user to change row height by dragging the bottom row border. After a resize operation, the [rowHeight](https://bryntum.com/products/calendar/docs/api/Grid/data/GridRowModel#field-rowHeight)
     * field of the record is updated (when [applyToAllRows](https://bryntum.com/products/calendar/docs/api/Grid/feature/RowResize#config-applyToAllRows) is `false`).
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Grid/feature/RowResize)
     */
    rowResizeFeature ? : object|Boolean|string|GridRowResize|GridRowResizeConfig
    /**
     * Feature that displays a tooltip containing the time at the mouse position when hovering empty parts of the schedule.
     * To hide the schedule tooltip, just disable this feature:
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Scheduler/feature/ScheduleTooltip)
     */
    scheduleTooltipFeature ? : object|Boolean|string|SchedulerScheduleTooltip|SchedulerScheduleTooltipConfig
    /**
     * <div class="external-example " data-file="Grid/feature/Search.js"></div>
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Grid/feature/Search)
     */
    searchFeature ? : object|Boolean|string|Search|SearchConfig
    /**
     * Allows sorting of grid by clicking (or tapping) headers, also displays which columns grid is sorted by (numbered if
     * using multisort). Use modifier keys for multisorting: [Ctrl/CMD + click] to add sorter, [Ctrl/CMD + Alt + click] to remove sorter.
     * The actual sorting is done by the store, see [Store.sort()](https://bryntum.com/products/calendar/docs/api/Core/data/mixin/StoreSort#function-sort).
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Grid/feature/Sort)
     */
    sortFeature ? : object|Boolean|string|Sort|SortConfig
    /**
     * This feature allows splitting the Scheduler into multiple views, either by using the cell context menu, or
     * programmatically by calling [split()](https://bryntum.com/products/calendar/docs/api/Scheduler/feature/Split#function-split).
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Scheduler/feature/Split)
     */
    splitFeature ? : object|Boolean|string|Split|SplitConfig
    /**
     * A feature which pins configurable content from a grid row to the top of the grid
     * while the row scrolls off the top but is still visible.
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Grid/feature/StickyCells)
     */
    stickyCellsFeature ? : object|Boolean|string|StickyCells|StickyCellsConfig
    /**
     * Stripes rows by adding alternating CSS classes to all row elements (`b-even` and `b-odd`).
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Grid/feature/Stripe)
     */
    stripeFeature ? : object|Boolean|string|Stripe|StripeConfig
    /**
     * A special version of the Grid Summary feature. This feature displays a summary row in the grid footer.
     * For regular columns in the locked section - specify type of summary on columns, available types are:
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Scheduler/feature/Summary)
     */
    summaryFeature ? : object|Boolean|string|Summary|SummaryConfig
    /**
     * Adds scheduler specific menu items to the timeline header context menu.
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Scheduler/feature/TimeAxisHeaderMenu)
     */
    timeAxisHeaderMenuFeature ? : object|Boolean|string|TimeAxisHeaderMenu|TimeAxisHeaderMenuConfig
    /**
     * Feature that renders global ranges of time in the timeline. Use this feature to visualize a `range` like a 1 hr lunch
     * or some important point in time (a `line`, i.e. a range with 0 duration). This feature can also show a current time
     * indicator if you set [showCurrentTimeLine](https://bryntum.com/products/calendar/docs/api/Scheduler/feature/TimeRanges#config-showCurrentTimeLine) to true. To style the rendered elements, use the
     * [cls](https://bryntum.com/products/calendar/docs/api/Scheduler/model/TimeSpan#field-cls) field of the `TimeSpan` class.
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Scheduler/feature/TimeRanges)
     */
    timeRangesFeature ? : object|Boolean|string|SchedulerTimeRanges|SchedulerTimeRangesConfig
    /**
     * Feature that makes the grid work more like a tree. Included by default in [TreeGrid](https://bryntum.com/products/calendar/docs/api/Grid/view/TreeGrid). Requires
     * exactly one [TreeColumn](https://bryntum.com/products/calendar/docs/api/Grid/column/TreeColumn) among grids columns. That column will have its renderer replaced with a
     * tree renderer that adds padding and icon to give the appearance of a tree. The original renderer is preserved and
     * also called.
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Grid/feature/Tree)
     */
    treeFeature ? : object|Boolean|string|Tree|TreeConfig
    /**
     * A feature that allows transforming a flat dataset (or the leaves of a hierarchical) into a tree by specifying a
     * record field per parent level. Parents are generated based on each leaf's value for those fields.
     * ...
     * [View online docs...](https://bryntum.com/products/calendar/docs/api/Grid/feature/TreeGroup)
     */
    treeGroupFeature ? : object|Boolean|string|TreeGroup|TreeGroupConfig

}
