// @bryntum/calendar/source/lib/Scheduler/localization/Ca.js
import LocaleHelper from '../../Core/localization/LocaleHelper.js';
import '../../Grid/localization/Ca.js';

const locale = {

    localeName : 'Ca',
    localeDesc : 'Català',
    localeCode : 'ca-ES',
    localeRtl  : false,

    Object : {
        newEvent    : 'Esdeveniment nou',
        information : 'Informació',
        showDetails : 'Mostra detalls'
    },

    ResourceInfoColumn : {
        eventCountText : data => data + ' esdeveniment' + (data !== 1 ? 's' : '')
    },

    ResourceCollapseColumn : {
        toggleResourceLayout : 'Commuta la columna del disseny del recurs. ESPAI per al menú contextual.'
    },

    Dependencies : {
        from    : 'De',
        to      : 'A',
        valid   : 'Vàlid',
        invalid : 'No vàlid'
    },

    DependencyType : {
        SS           : 'II',
        SF           : 'IF',
        FS           : 'FI',
        FF           : 'FF',
        StartToStart : "D'inici a inici",
        StartToEnd   : "D'inici a final",
        EndToStart   : 'De final a inici',
        EndToEnd     : 'De final a final',
        short        : [
            'II',
            'IF',
            'FI',
            'FF'
        ],
        long : [
            "D'inici a inici",
            "D'inici a final",
            'De final a inici',
            'De final a final'
        ]
    },

    DependencyEdit : {
        From              : 'De',
        To                : 'A',
        Type              : 'Tipus',
        Lag               : 'Retard',
        'Edit dependency' : 'Edita la dependència',
        Save              : 'Desa',
        Delete            : 'Suprimeix',
        Cancel            : 'Cancel·la',
        StartToStart      : "D'inici a inici",
        StartToEnd        : "D'inici a final",
        EndToStart        : 'De final a inici',
        EndToEnd          : 'De final a final'
    },

    EventEdit : {
        Name         : 'Nom',
        Resource     : 'Recurs',
        Start        : 'Inici',
        startTime    : 'Hora d’inici',
        End          : 'Final',
        endTime      : 'Hora de finalització',
        Save         : 'Desa',
        Delete       : 'Suprimeix',
        Cancel       : 'Cancel·la',
        'Edit event' : "Edita l'esdeveniment",
        Repeat       : 'Repeteix'
    },

    EventDrag : {
        eventOverlapsExisting : "L'esdeveniment se superposa a un esdeveniment existent per a aquest recurs",
        noDropOutsideTimeline : "És possible que l'esdeveniment no es pugui treure completament de la cronologia"
    },

    SchedulerBase : {
        'Add event'      : 'Afegeix un esdeveniment',
        'Delete event'   : "Suprimeix l'esdeveniment",
        'Unassign event' : "Desassigna l'esdeveniment",
        color            : 'Color',
        eventAriaLabel   : (eventRecord, startDate) => `${eventRecord.name}, comença a ${startDate} ${eventRecord.duration > 0 ? `durant ${eventRecord.fullDuration.toString(false)}` : ''}`
    },

    TimeAxisHeaderMenu : {
        pickZoomLevel   : 'Zoom',
        activeDateRange : 'Interval de dates',
        startText       : "Data d'inici",
        endText         : 'Data de finalització',
        todayText       : 'Avui'
    },

    TimeAxisColumn : {
        ariaLabel : (startDate, endDate) => `Eix temporal mostrant de ${startDate} a ${endDate}`
    },

    EventCopyPaste : {
        copyEvent  : "Copia l'esdeveniment",
        cutEvent   : "Retalla l'esdeveniment",
        pasteEvent : "Enganxa l'esdeveniment"
    },

    EventFilter : {
        filterEvents : 'Filtra les tasques',
        byName       : 'Per nom'
    },

    TimeRanges : {
        showCurrentTimeLine : 'Mostra la cronologia actual'
    },

    PresetManager : {
        secondAndMinute : {
            displayDateFormat : 'll LTS',
            name              : 'Segons'
        },
        minuteAndHour : {
            topDateFormat     : 'ddd MM/DD, hA',
            displayDateFormat : 'll LST'
        },
        hourAndDay : {
            topDateFormat     : 'ddd MM/DD',
            middleDateFormat  : 'LST',
            displayDateFormat : 'll LST',
            name              : 'Dia'
        },
        day : {
            name : 'Dia/hores'
        },
        week : {
            name : 'Setmana/hores'
        },
        dayAndWeek : {
            displayDateFormat : 'll LST',
            name              : 'Setmana/dies'
        },
        dayAndMonth : {
            name : 'Mes'
        },
        weekAndDay : {
            displayDateFormat : 'll LST',
            name              : 'Setmana'
        },
        weekAndMonth : {
            name : 'Setmanes'
        },
        weekAndDayLetter : {
            name : 'Setmanes/dies feiners'
        },
        weekDateAndMonth : {
            name : 'Mesos/setmanes'
        },
        monthAndYear : {
            name : 'Mesos'
        },
        year : {
            name : 'Anys'
        },
        manyYears : {
            name : 'Diversos anys'
        }
    },

    RecurrenceConfirmationPopup : {
        'delete-title'              : 'Esteu suprimint un esdeveniment',
        'delete-all-message'        : "Voleu suprimir totes les repeticions d'aquest esdeveniment?",
        'delete-further-message'    : "Voleu suprimir aquesta i totes les repeticions futures de l'esdeveniment o només la repetició seleccionada?",
        'delete-only-this-message'  : "Vols suprimir aquesta ocurrència d'aquest esdeveniment?",
        'delete-further-btn-text'   : 'Suprimeix tots els esdeveniments futurs',
        'delete-only-this-btn-text' : 'Suprimeix només aquest esdeveniment',
        'update-title'              : 'Esteu canviant un esdeveniment que es repeteix',
        'update-all-message'        : "Voleu canviar totes les repeticions d'aquest esdeveniment?",
        'update-further-message'    : "Voleu canviar només aquesta repetició de l'esdeveniment o aquesta i totes les repeticions futures?",
        'update-only-this-message'  : "Vols canviar aquesta ocurrència d'aquest esdeveniment?",
        'update-further-btn-text'   : 'Tots els esdeveniments futurs',
        'update-only-this-btn-text' : 'Només aquest esdeveniment',
        Yes                         : 'Sí',
        Cancel                      : 'Cancel·la',
        width                       : 600
    },

    RecurrenceLegend : {
        ' and '                         : ' i ',
        Daily                           : 'Diàriament',
        'Weekly on {1}'                 : ({ days }) => `Setmanalment ${days}`,
        'Monthly on {1}'                : ({ days }) => `Mensualment ${days}`,
        'Yearly on {1} of {2}'          : ({ days, months }) => `Anualment ${days} de ${months}`,
        'Every {0} days'                : ({ interval }) => `Cada ${interval} dies`,
        'Every {0} weeks on {1}'        : ({ interval, days }) => `Cada ${interval} setmanes ${days}`,
        'Every {0} months on {1}'       : ({ interval, days }) => `Cada ${interval} mesos ${days}`,
        'Every {0} years on {1} of {2}' : ({ interval, days, months }) => `Cada ${interval} anys ${days} de ${months}`,
        position1                       : 'el primer',
        position2                       : 'el segon',
        position3                       : 'el tercer',
        position4                       : 'el quart',
        position5                       : 'el cinquè',
        'position-1'                    : "l'últim",
        day                             : 'dia',
        weekday                         : 'dia feiner',
        'weekend day'                   : 'dia del cap de setmana',
        daysFormat                      : ({ position, days }) => `${position} ${days}`
    },

    RecurrenceEditor : {
        'Repeat event'      : "Repeteix l'esdeveniment",
        Cancel              : 'Cancel·la',
        Save                : 'Desa',
        Frequency           : 'Freqüència',
        Every               : 'Cada',
        DAILYintervalUnit   : 'dia/es',
        WEEKLYintervalUnit  : 'setmana/es',
        MONTHLYintervalUnit : 'mes/os',
        YEARLYintervalUnit  : 'any/s',
        Each                : 'Cada',
        on                  : 'El',
        the                 : 'El',
        'On the'            : 'El dia',
        'End repeat'        : 'Finalitza la repetició',
        'time(s)'           : 'vegada/es',
        Days                : 'Dies',
        Months              : 'meses'
    },

    RecurrenceDaysCombo : {
        day           : 'dia',
        weekday       : 'dia feiner',
        'weekend day' : 'dia del cap de setmana'
    },

    RecurrencePositionsCombo : {
        position1    : 'primer',
        position2    : 'segon',
        position3    : 'tercer',
        position4    : 'quart',
        position5    : 'cinquè',
        'position-1' : 'darrer'
    },

    RecurrenceStopConditionCombo : {
        Never     : 'Mai',
        After     : 'Després de',
        'On date' : 'El dia'
    },

    RecurrenceFrequencyCombo : {
        None    : 'No hi ha cap repetició',
        Daily   : 'Diàriament',
        Weekly  : 'Setmanalment {0}',
        Monthly : 'Mensualment',
        Yearly  : 'Anualment'
    },

    RecurrenceCombo : {
        None   : 'Cap',
        Custom : 'Personalitzat...'
    },

    Summary : {
        'Summary for' : date => `Resum del dia ${date}`
    },

    ScheduleRangeCombo : {
        completeview : 'Planificació completa',
        currentview  : 'Planificació visible',
        daterange    : 'Interval de dates',
        completedata : 'Planificació completa (de tots els esdeveniments)'
    },

    SchedulerExportDialog : {
        'Schedule range' : 'Interval de planificació',
        'Export from'    : 'De',
        'Export to'      : 'A'
    },

    ExcelExporter : {
        'No resource assigned' : 'No hi ha cap recurs assignat'
    },

    CrudManagerView : {
        serverResponseLabel : 'Resposta del servidor:'
    },

    DurationColumn : {
        Duration : 'Durada'
    }
};

export default LocaleHelper.publishLocale(locale);
