// @bryntum/calendar/source/lib/Scheduler/localization/Da.js
import LocaleHelper from '../../Core/localization/LocaleHelper.js';
import '../../Grid/localization/Da.js';

const locale = {

    localeName : 'Da',
    localeDesc : 'Dansk',
    localeCode : 'da',
    localeRtl  : false,

    Object : {
        newEvent    : 'ny begivenhed',
        information : 'Information',
        showDetails : 'Vis detaljer'
    },

    ResourceInfoColumn : {
        eventCountText : data => data + ' begivenhed' + (data !== 1 ? 'er' : '')
    },

    ResourceCollapseColumn : {
        toggleResourceLayout : 'Skift ressource layout kolonne. SPACE for kontekstmenu.'
    },

    Dependencies : {
        from    : 'Fra',
        to      : 'Til',
        valid   : 'Gyldig',
        invalid : 'Ugyldig'
    },

    DependencyType : {
        SS           : 'SS',
        SF           : 'SA',
        FS           : 'AS',
        FF           : 'AA',
        StartToStart : 'Start-til-Start',
        StartToEnd   : 'Start-til-Afslutning',
        EndToStart   : 'Afslutning-til-Start',
        EndToEnd     : 'Afslutning-til-Afslutning',
        short        : [
            'SS',
            'SA',
            'AS',
            'AA'
        ],
        long : [
            'Start-til-Start',
            'Start-til-Afslutning',
            'Afslutning-til-Start',
            'Afslutning-til-Afslutning'
        ]
    },

    DependencyEdit : {
        From              : 'Fra',
        To                : 'Til',
        Type              : 'Type',
        Lag               : 'Forsinkelse',
        'Edit dependency' : 'Redigér afhængighed',
        Save              : 'Gem',
        Delete            : 'Slet',
        Cancel            : 'Annullér',
        StartToStart      : 'Start-Start',
        StartToEnd        : 'Start-Afslutning',
        EndToStart        : 'Afslutning-Start',
        EndToEnd          : 'Afslutning-Afslutning'
    },

    EventEdit : {
        Name         : 'Navn',
        Resource     : 'Ressource',
        Start        : 'Start',
        startTime    : 'Starttidspunkt',
        End          : 'Afslutning',
        endTime      : 'Sluttidspunkt',
        Save         : 'Gem',
        Delete       : 'Slet',
        Cancel       : 'Annullér',
        'Edit event' : 'Redigér begivenhed',
        Repeat       : 'Gentag'
    },

    EventDrag : {
        eventOverlapsExisting : 'Begivenhed overlapper med eksisterende begivenhed for denne ressource',
        noDropOutsideTimeline : 'Begivenhed må ikke slippes helt uden for tidslinjen'
    },

    SchedulerBase : {
        'Add event'      : 'Tilføj begivenhed',
        'Delete event'   : 'Slet begivenhed',
        'Unassign event' : 'Fjern tilknytning til begivenhed',
        color            : 'Farve',
        eventAriaLabel   : (eventRecord, startDate) => `${eventRecord.name}, starter kl. ${startDate} ${eventRecord.duration > 0 ? `i ${eventRecord.fullDuration.toString(false)}` : ''}`
    },

    TimeAxisHeaderMenu : {
        pickZoomLevel   : 'Zoom',
        activeDateRange : 'Datointerval',
        startText       : 'Startdato',
        endText         : 'Slutdato',
        todayText       : 'Idag'
    },

    TimeAxisColumn : {
        ariaLabel : (startDate, endDate) => `Tidsakse viser ${startDate} til ${endDate}`
    },

    EventCopyPaste : {
        copyEvent  : 'Kopiér begivenhed',
        cutEvent   : 'Klip begivenhed',
        pasteEvent : 'Indsæt begivenhed'
    },

    EventFilter : {
        filterEvents : 'Filtrér opgaver',
        byName       : 'Efter navn'
    },

    TimeRanges : {
        showCurrentTimeLine : 'Vis aktuel tidslinje'
    },

    PresetManager : {
        secondAndMinute : {
            displayDateFormat : 'll LTS',
            name              : 'Sekunder'
        },
        minuteAndHour : {
            topDateFormat     : 'ddd DD.MM, HH:mm',
            displayDateFormat : 'll LST'
        },
        hourAndDay : {
            topDateFormat     : 'ddd DD.MM',
            middleDateFormat  : 'LST',
            displayDateFormat : 'll LST',
            name              : 'Dag'
        },
        day : {
            name : 'Dag/timer'
        },
        week : {
            name : 'Uge/timer'
        },
        dayAndWeek : {
            displayDateFormat : 'll LST',
            name              : 'Uge/dage'
        },
        dayAndMonth : {
            name : 'Måned'
        },
        weekAndDay : {
            displayDateFormat : 'll LST',
            name              : 'Uge'
        },
        weekAndMonth : {
            name : 'Uger'
        },
        weekAndDayLetter : {
            name : 'Uger/hverdage'
        },
        weekDateAndMonth : {
            name : 'Måneder/uger'
        },
        monthAndYear : {
            name : 'Måneder'
        },
        year : {
            name : 'Flere år'
        },
        manyYears : {
            name : 'Flere år'
        }
    },

    RecurrenceConfirmationPopup : {
        'delete-title'              : 'Du er ved at slette en begivenhed',
        'delete-all-message'        : 'Vil du slette alle forekomster af denne begivenhed?',
        'delete-further-message'    : 'Vil du slette denne og alle fremtidige forekomster af denne begivenhed, eller kun den valgte forekomst?',
        'delete-only-this-message'  : 'Vil du slette denne forekomst af denne begivenhed?',
        'delete-further-btn-text'   : 'Slet alle fremtidige begivenheder',
        'delete-only-this-btn-text' : 'Slet kun denne begivenhed',
        'update-title'              : 'Du er ved at ændre en gentagende begivenhed',
        'update-all-message'        : 'Vil du ændre alle forekomster af denne begivenhed?',
        'update-further-message'    : 'Vil du ændre netop denne forekomst af begivenheden, eller denne og alle fremtidige forekomster?',
        'update-only-this-message'  : 'Vil du ændre denne forekomst af denne begivenhed?',
        'update-further-btn-text'   : 'Alle fremtidige begivenheder',
        'update-only-this-btn-text' : 'Kun denne begivenhed',
        Yes                         : 'Ja',
        Cancel                      : 'Annullér',
        width                       : 600
    },

    RecurrenceLegend : {
        ' and '                         : ' og ',
        Daily                           : 'Daglig',
        'Weekly on {1}'                 : ({ days }) => `Ugentligt ${days}`,
        'Monthly on {1}'                : ({ days }) => `Månedligt ${days}`,
        'Yearly on {1} of {2}'          : ({ days, months }) => `Årligt ${days} af ${months}`,
        'Every {0} days'                : ({ interval }) => `For hver ${interval} dage`,
        'Every {0} weeks on {1}'        : ({ interval, days }) => `For hver ${interval} uger ${days}`,
        'Every {0} months on {1}'       : ({ interval, days }) => `For hver ${interval} måneder ${days}`,
        'Every {0} years on {1} of {2}' : ({ interval, days, months }) => `For hver ${interval} år ${days} i ${months}`,
        position1                       : 'den første',
        position2                       : 'den anden',
        position3                       : 'den tredje',
        position4                       : 'den fjerde',
        position5                       : 'den femte',
        'position-1'                    : 'den sidste',
        day                             : 'dag',
        weekday                         : 'ugedag',
        'weekend day'                   : 'dag i weekenden',
        daysFormat                      : ({ position, days }) => `${position} ${days}`
    },

    RecurrenceEditor : {
        'Repeat event'      : 'Gentag begivenhed',
        Cancel              : 'Annullér',
        Save                : 'Gem',
        Frequency           : 'Frekvens',
        Every               : 'Alle',
        DAILYintervalUnit   : 'dag(e)',
        WEEKLYintervalUnit  : 'uge(r)',
        MONTHLYintervalUnit : 'Måned(er)',
        YEARLYintervalUnit  : 'År',
        Each                : 'For hver',
        on                  : 'Den',
        the                 : 'Den',
        'On the'            : 'på',
        'End repeat'        : 'Stop gentagelse',
        'time(s)'           : 'tid(er)',
        Days                : 'Dage',
        Months              : 'Måneder'
    },

    RecurrenceDaysCombo : {
        day           : 'dag',
        weekday       : 'ugedag',
        'weekend day' : 'dag i weekenden'
    },

    RecurrencePositionsCombo : {
        position1    : 'første',
        position2    : 'anden',
        position3    : 'tredje',
        position4    : 'fjerde',
        position5    : 'femte',
        'position-1' : 'sidste'
    },

    RecurrenceStopConditionCombo : {
        Never     : 'Aldrig',
        After     : 'Efter',
        'On date' : 'På dato'
    },

    RecurrenceFrequencyCombo : {
        None    : 'Ingen gentagelse',
        Daily   : 'Dagligt',
        Weekly  : 'Ugentligt',
        Monthly : 'Månedligt',
        Yearly  : 'Årligt'
    },

    RecurrenceCombo : {
        None   : 'Ingen',
        Custom : 'Tilpasset...'
    },

    Summary : {
        'Summary for' : date => `Resumé for ${date}`
    },

    ScheduleRangeCombo : {
        completeview : 'Komplet tidsplan',
        currentview  : 'Synlig tidsplan',
        daterange    : 'Datointerval',
        completedata : 'Komplet tidsplan (for alle arrangementer)'
    },

    SchedulerExportDialog : {
        'Schedule range' : 'Tidsplan rækkevidde',
        'Export from'    : 'Fra',
        'Export to'      : 'Til'
    },

    ExcelExporter : {
        'No resource assigned' : 'Ingen ressourcer tilknyttet'
    },

    CrudManagerView : {
        serverResponseLabel : 'Svar fra server:'
    },

    DurationColumn : {
        Duration : 'Varighed'
    }
};

export default LocaleHelper.publishLocale(locale);
