// @bryntum/calendar/source/lib/Scheduler/localization/Eu.js
import LocaleHelper from '../../Core/localization/LocaleHelper.js';
import '../../Grid/localization/Eu.js';

const locale = {

    localeName : 'Eu',
    localeDesc : 'Euskara',
    localeCode : 'eu-ES',
    localeRtl  : false,

    Object : {
        newEvent    : 'Gertaera berria',
        information : 'Informazioa',
        showDetails : 'Xehetasunak erakutsi'
    },

    ResourceInfoColumn : {
        eventCountText : data => data + ' gertaera' + (data !== 1 ? 's' : '')
    },

    ResourceCollapseColumn : {
        toggleResourceLayout : 'Errekurtsoen diseinuaren zutabea aldatu. SPACE testuinguru-menua.'
    },

    Dependencies : {
        from    : 'Hasiera',
        to      : 'Amaiera',
        valid   : 'Baliozkoa',
        invalid : 'Baliogabea'
    },

    DependencyType : {
        SS           : 'HH',
        SF           : 'HA',
        FS           : 'AH',
        FF           : 'AA',
        StartToStart : 'Hasieratik-hasierara',
        StartToEnd   : 'Hasieratik-amaierara',
        EndToStart   : 'Amaieratik-hasierara',
        EndToEnd     : 'Amaieratik-amaierara',
        short        : [
            'HH',
            'HA',
            'AH',
            'AA'
        ],
        long : [
            'Hasieratik-hasierara',
            'Hasieratik-amaierara',
            'Amaieratik-hasierara',
            'Amaieratik-amaierara'
        ]
    },

    DependencyEdit : {
        From              : 'Hasiera',
        To                : 'Amaiera',
        Type              : 'Mota',
        Lag               : 'Atzerapena',
        'Edit dependency' : 'Editatu mendekotasuna',
        Save              : 'Gorde',
        Delete            : 'Ezabatu',
        Cancel            : 'Utzi',
        StartToStart      : 'Hasieratik hasierara',
        StartToEnd        : 'Hasieratik amaierara',
        EndToStart        : 'Amaieratik hasierara',
        EndToEnd          : 'Amaieratik amaierara'
    },

    EventEdit : {
        Name         : 'Izena',
        Resource     : 'Baliabidea',
        Start        : 'Hasi',
        startTime    : 'Hasiera-ordua',
        End          : 'Amaiera',
        endTime      : 'Amaiera-ordua',
        Save         : 'Gorde',
        Delete       : 'Ezabatu',
        Cancel       : 'Utzi',
        'Edit event' : 'Editatu gertaera',
        Repeat       : 'Errepikatu'
    },

    EventDrag : {
        eventOverlapsExisting : 'Gertaera baliabide honen gertaera batean gainjartzen da',
        noDropOutsideTimeline : 'Baliteke gertaera guztiz ez kentzea denbora-lerrotik kanpo'
    },

    SchedulerBase : {
        'Add event'      : 'Gehitu gertaera',
        'Delete event'   : 'Ezabatu gertaera',
        'Unassign event' : 'Kendu gertaeraren esleipena',
        color            : 'Kolorea',
        eventAriaLabel   : (eventRecord, startDate) => `${eventRecord.name}, hasten da ${startDate} ${eventRecord.duration > 0 ? `${eventRecord.fullDuration.toString(false)} irauten du` : ''}`
    },

    TimeAxisHeaderMenu : {
        pickZoomLevel   : 'Zooma',
        activeDateRange : 'Data-tartea',
        startText       : 'Hasiera-data',
        endText         : 'Amaiera-data',
        todayText       : 'Gaur'
    },

    TimeAxisColumn : {
        ariaLabel : (startDate, endDate) => `Denbora ardatza erakusten ${startDate} eta ${endDate}`
    },

    EventCopyPaste : {
        copyEvent  : 'Kopiatu gertaera',
        cutEvent   : 'Moztu gertaera',
        pasteEvent : 'Iraganeko gertaera'
    },

    EventFilter : {
        filterEvents : 'Iragazi zereginak',
        byName       : 'Izenaren arabera'
    },

    TimeRanges : {
        showCurrentTimeLine : 'Erakutsi uneko denbora-eskala'
    },

    PresetManager : {
        secondAndMinute : {
            displayDateFormat : 'll LTS',
            name              : 'Segundoak'
        },
        minuteAndHour : {
            topDateFormat     : 'ddd MM/DD, hA',
            displayDateFormat : 'll LST'
        },
        hourAndDay : {
            topDateFormat     : 'ddd MM/DD',
            middleDateFormat  : 'LST',
            displayDateFormat : 'll LST',
            name              : 'Eguna'
        },
        day : {
            name : 'Eguna/orduak'
        },
        week : {
            name : 'Astea/orduak'
        },
        dayAndWeek : {
            displayDateFormat : 'll LST',
            name              : 'Astea/egunak'
        },
        dayAndMonth : {
            name : 'Month'
        },
        weekAndDay : {
            displayDateFormat : 'll LST',
            name              : 'Astea'
        },
        weekAndMonth : {
            name : 'Asteak'
        },
        weekAndDayLetter : {
            name : 'Asteak/Lanegunak'
        },
        weekDateAndMonth : {
            name : 'Hilabeteak/Asteak'
        },
        monthAndYear : {
            name : 'Hilabeteak'
        },
        year : {
            name : 'Urteak'
        },
        manyYears : {
            name : 'Urte bat baino gehiago'
        }
    },

    RecurrenceConfirmationPopup : {
        'delete-title'              : 'Gertaera bat ezabatzera zoaz',
        'delete-all-message'        : 'Gertaera honen agerraldi guztiak ezabatu nahi dituzu?',
        'delete-further-message'    : 'Gertaeraren agerraldi hau bakarrik ezabatu nahi al duzu, edo hautatutakoa bakarrik?',
        'delete-only-this-message'  : 'Gertaera honen agerpen hau ezabatu nahi duzu?',
        'delete-further-btn-text'   : 'Ezabatu etorkizuneko gertaera guztiak',
        'delete-only-this-btn-text' : 'Ezabatu gertaera hau soilik',
        'update-title'              : 'Gertaera errepikari bat aldatzera zoaz',
        'update-all-message'        : 'Gertaera honen agerraldi guztiak aldatu nahi dituzu?',
        'update-further-message'    : 'Gertaeraren agerraldi hau bakarrik aldatu nahi al duzu, edo hau eta etorkizuneko guztiak?',
        'update-only-this-message'  : 'Gertaera honen agerpen hau aldatu nahi duzu?',
        'update-further-btn-text'   : 'Etorkizuneko gertaera guztiak',
        'update-only-this-btn-text' : 'Gertaera hau soilik',
        Yes                         : 'Bai',
        Cancel                      : 'Utzi',
        width                       : 600
    },

    RecurrenceLegend : {
        ' and '                         : ' eta ',
        Daily                           : 'Egunero',
        'Weekly on {1}'                 : ({ days }) => `${days} astean behin`,
        'Monthly on {1}'                : ({ days }) => `H${days} hilabetean behin`,
        'Yearly on {1} of {2}'          : ({ days, months }) => `${days} urtean behin, hilabete honetan: ${months}`,
        'Every {0} days'                : ({ interval }) => `${interval} egunean behin`,
        'Every {0} weeks on {1}'        : ({ interval, days }) => `${days} astean behin, egun honetan: ${interval}`,
        'Every {0} months on {1}'       : ({ interval, days }) => `${interval} hilean behin, egun honetan: ${days}`,
        'Every {0} years on {1} of {2}' : ({ interval, days, months }) => `${interval} urtean behin, ${months}ren ${days} egunean`,
        position1                       : 'lehen',
        position2                       : 'bigarren',
        position3                       : 'hirugarren',
        position4                       : 'laugarren',
        position5                       : 'bosgarren',
        'position-1'                    : 'azken',
        day                             : 'egun',
        weekday                         : 'asteguna',
        'weekend day'                   : 'asteburuko eguna',
        daysFormat                      : ({ position, days }) => `${position} ${days}`
    },

    RecurrenceEditor : {
        'Repeat event'      : 'Errepikatu gertaera',
        Cancel              : 'Utzi',
        Save                : 'Gorde',
        Frequency           : 'Maiztasuna',
        Every               : 'Maiztasuna:',
        DAILYintervalUnit   : 'egun',
        WEEKLYintervalUnit  : 'aste',
        MONTHLYintervalUnit : 'hilabete',
        YEARLYintervalUnit  : 'urte',
        Each                : 'Bakoitza',
        on                  : 'Du',
        the                 : 'Du',
        'On the'            : 'Egun honetan',
        'End repeat'        : 'Amaitu errepikapena',
        'time(s)'           : '(s) bider',
        Days                : 'Egunean',
        Months              : 'Hilean'
    },

    RecurrenceDaysCombo : {
        day           : 'egun',
        weekday       : 'asteguna',
        'weekend day' : 'asteburuko eguna'
    },

    RecurrencePositionsCombo : {
        position1    : 'lehen',
        position2    : 'bigarren',
        position3    : 'hirugarren',
        position4    : 'laugarren',
        position5    : 'bosgarren',
        'position-1' : 'azken'
    },

    RecurrenceStopConditionCombo : {
        Never     : 'Inoiz ez',
        After     : 'geroago',
        'On date' : 'Data honetan'
    },

    RecurrenceFrequencyCombo : {
        None    : 'Ez errepikatu',
        Daily   : 'Egunero',
        Weekly  : 'Astero egun honetan: {0}',
        Monthly : 'Hilero',
        Yearly  : 'Urtero'
    },

    RecurrenceCombo : {
        None   : 'Bat ere ez',
        Custom : 'Pertsonalizatua...'
    },

    Summary : {
        'Summary for' : date => `Laburpena ${date}`
    },

    ScheduleRangeCombo : {
        completeview : 'Antolaketa osoa',
        currentview  : 'Ikusgai dagoen antolaketa',
        daterange    : 'Data-tartea',
        completedata : 'Antolaketa osoa (gertaera guztietarako)'
    },

    SchedulerExportDialog : {
        'Schedule range' : 'Antolaketa-tartea',
        'Export from'    : 'Hasiera',
        'Export to'      : 'Amaiera'
    },

    ExcelExporter : {
        'No resource assigned' : 'Ez da esleitu baliabiderik'
    },

    CrudManagerView : {
        serverResponseLabel : 'Zerbitzariaren erantzuna:'
    },

    DurationColumn : {
        Duration : 'Iraupena'
    }
};

export default LocaleHelper.publishLocale(locale);
