// @bryntum/calendar/source/lib/Scheduler/localization/Gl.js
import LocaleHelper from '../../Core/localization/LocaleHelper.js';
import '../../Grid/localization/Gl.js';

const locale = {

    localeName : 'Gl',
    localeDesc : 'Galego',
    localeCode : 'gl-ES',
    localeRtl  : false,

    Object : {
        newEvent    : 'Novo evento',
        information : 'Information',
        showDetails : 'Amosar detalles'
    },

    ResourceInfoColumn : {
        eventCountText : data => data + ' evento' + (data !== 1 ? 's' : '')
    },

    ResourceCollapseColumn : {
        toggleResourceLayout : 'Alternar columna de deseño de recursos. ESPAZO para o menú contextual.'
    },

    Dependencies : {
        from    : 'Desde',
        to      : 'Até',
        valid   : 'Válido',
        invalid : 'Non válido'
    },

    DependencyType : {
        SS           : 'II',
        SF           : 'IF',
        FS           : 'FI',
        FF           : 'FF',
        StartToStart : 'De inicio a inicio',
        StartToEnd   : 'De inicio a fin',
        EndToStart   : 'De fin a inicio',
        EndToEnd     : 'De fin a fin',
        short        : [
            'II',
            'IF',
            'FI',
            'FF'
        ],
        long : [
            'De inicio a inicio',
            'De inicio a fin',
            'De fin a inicio',
            'De fin a fin'
        ]
    },

    DependencyEdit : {
        From              : 'Desde',
        To                : 'Até',
        Type              : 'Tipo',
        Lag               : 'Atraso',
        'Edit dependency' : 'Editar dependencia',
        Save              : 'Gardar',
        Delete            : 'Eliminar',
        Cancel            : 'Cancelar',
        StartToStart      : 'De inicio a inicio',
        StartToEnd        : 'De inicio a fin',
        EndToStart        : 'De fin a inicio',
        EndToEnd          : 'De fin a fin'
    },

    EventEdit : {
        Name         : 'Nome',
        Resource     : 'Recurso',
        Start        : 'Inicio',
        startTime    : 'Hora de comezo',
        End          : 'Finalización',
        endTime      : 'Hora de finalización',
        Save         : 'Gardar',
        Delete       : 'Eliminar',
        Cancel       : 'Cancelar',
        'Edit event' : 'Editar evento',
        Repeat       : 'Repetir'
    },

    EventDrag : {
        eventOverlapsExisting : 'O evento está superposto co evento existente para este recurso',
        noDropOutsideTimeline : 'É posible que o evento non se poida excluír completamente fóra da liña de tempo'
    },

    SchedulerBase : {
        'Add event'      : 'Engadir evento',
        'Delete event'   : 'Eliminar evento',
        'Unassign event' : 'Evento non atribuído',
        color            : 'Cor',
        eventAriaLabel   : (eventRecord, startDate) => `${eventRecord.name}, comeza ás ${startDate} ${eventRecord.duration > 0 ? `durante ${eventRecord.fullDuration.toString(false)}` : ''}`
    },

    TimeAxisHeaderMenu : {
        pickZoomLevel   : 'Zoom',
        activeDateRange : 'Intervalo de datas',
        startText       : 'Data de inicio',
        endText         : 'Data de finalización',
        todayText       : 'Hoxe'
    },

    TimeAxisColumn : {
        ariaLabel : (startDate, endDate) => `Eixo temporal mostrando de ${startDate} a ${endDate}`
    },

    EventCopyPaste : {
        copyEvent  : 'Copiar evento',
        cutEvent   : 'Cortar evento',
        pasteEvent : 'Pegar evento'
    },

    EventFilter : {
        filterEvents : 'Filtrar tarefas',
        byName       : 'Por nome'
    },

    TimeRanges : {
        showCurrentTimeLine : 'Mostrar liña de tempo actual'
    },

    PresetManager : {
        secondAndMinute : {
            displayDateFormat : 'll LTS',
            name              : 'Segundos'
        },
        minuteAndHour : {
            topDateFormat     : 'ddd MM/DD, hA',
            displayDateFormat : 'll LST'
        },
        hourAndDay : {
            topDateFormat     : 'ddd MM/DD',
            middleDateFormat  : 'LST',
            displayDateFormat : 'll LST',
            name              : 'Día'
        },
        day : {
            name : 'Día/horas'
        },
        week : {
            name : 'Semana/horas'
        },
        dayAndWeek : {
            displayDateFormat : 'll LST',
            name              : 'Semana/días'
        },
        dayAndMonth : {
            name : 'Mes'
        },
        weekAndDay : {
            displayDateFormat : 'll LST',
            name              : 'Semana'
        },
        weekAndMonth : {
            name : 'Semanas'
        },
        weekAndDayLetter : {
            name : 'Semanas/días da semana'
        },
        weekDateAndMonth : {
            name : 'Meses/semanas'
        },
        monthAndYear : {
            name : 'Meses'
        },
        year : {
            name : 'Anos'
        },
        manyYears : {
            name : 'Varios anos'
        }
    },

    RecurrenceConfirmationPopup : {
        'delete-title'              : 'Está a eliminar un evento',
        'delete-all-message'        : 'Quere eliminar todas as ocorrencias deste evento?',
        'delete-further-message'    : 'Quere eliminar esta e todas as ocorrencias futuras deste evento ou só a ocorrencia seleccionada?',
        'delete-only-this-message'  : 'Queres eliminar esta ocorrencia deste evento?',
        'delete-further-btn-text'   : 'Eliminar todos os eventos futuros',
        'delete-only-this-btn-text' : 'Eliminar só este evento',
        'update-title'              : 'Está a cambiar un evento repetitivo',
        'update-all-message'        : 'Quere cambiar todas as ocorrencias deste evento?',
        'update-further-message'    : 'Quere cambiar só esta ocorrencia do evento ou esta e todas as ocorrencias futuras?',
        'update-only-this-message'  : 'Queres cambiar esta ocorrencia deste evento?',
        'update-further-btn-text'   : 'Todos os eventos futuros',
        'update-only-this-btn-text' : 'Só este evento',
        Yes                         : 'Si',
        Cancel                      : 'Cancelar',
        width                       : 600
    },

    RecurrenceLegend : {
        ' and '                         : ' e ',
        Daily                           : 'Diario',
        'Weekly on {1}'                 : ({ days }) => `Semanalmente o ${days}`,
        'Monthly on {1}'                : ({ days }) => `Mensualmente os ${days}`,
        'Yearly on {1} of {2}'          : ({ days, months }) => `Anualmente os ${days} de ${months}`,
        'Every {0} days'                : ({ interval }) => `Cada ${interval} días`,
        'Every {0} weeks on {1}'        : ({ interval, days }) => `Cada ${interval} semanas o ${days}`,
        'Every {0} months on {1}'       : ({ interval, days }) => `Cada ${interval} meses o ${days}`,
        'Every {0} years on {1} of {2}' : ({ interval, days, months }) => `Cada ${interval} anos o ${days} de ${months}`,
        position1                       : 'o primeiro',
        position2                       : 'o segundo',
        position3                       : 'o terceiro',
        position4                       : 'o cuarto',
        position5                       : 'o quinto',
        'position-1'                    : 'o último',
        day                             : 'día',
        weekday                         : 'día da semana',
        'weekend day'                   : 'día da fin de semana',
        daysFormat                      : ({ position, days }) => `${position} ${days}`
    },

    RecurrenceEditor : {
        'Repeat event'      : 'Repetir evento',
        Cancel              : 'Cancelar',
        Save                : 'Gardar',
        Frequency           : 'Frecuencia',
        Every               : 'Cada',
        DAILYintervalUnit   : 'día(s)',
        WEEKLYintervalUnit  : 'semana(s)',
        MONTHLYintervalUnit : 'mes(es)',
        YEARLYintervalUnit  : 'ano(s)',
        Each                : 'Cada',
        on                  : 'No',
        the                 : 'No',
        'On the'            : 'No',
        'End repeat'        : 'Finalizar repetición',
        'time(s)'           : 'vez(veces)',
        Days                : 'Días',
        Months              : 'Meses'
    },

    RecurrenceDaysCombo : {
        day           : 'día',
        weekday       : 'día da semana',
        'weekend day' : 'día da fin de semana'
    },

    RecurrencePositionsCombo : {
        position1    : 'primeiro',
        position2    : 'segundo',
        position3    : 'terceiro',
        position4    : 'cuarto',
        position5    : 'quinto',
        'position-1' : 'último'
    },

    RecurrenceStopConditionCombo : {
        Never     : 'Nunca',
        After     : 'Despois de',
        'On date' : 'Nesa data'
    },

    RecurrenceFrequencyCombo : {
        None    : 'Sen repeticións',
        Daily   : 'Diario',
        Weekly  : 'Semanalmente o {0}',
        Monthly : 'Mensualmente',
        Yearly  : 'Anualmente'
    },

    RecurrenceCombo : {
        None   : 'Ningún',
        Custom : 'Personalizado...'
    },

    Summary : {
        'Summary for' : date => `Resumo para ${date}`
    },

    ScheduleRangeCombo : {
        completeview : 'Completar programación',
        currentview  : 'Programación visible',
        daterange    : 'Intervalo de datas',
        completedata : 'Completar programación (para todos os eventos)'
    },

    SchedulerExportDialog : {
        'Schedule range' : 'Intervalo de programación',
        'Export from'    : 'Desde',
        'Export to'      : 'Até'
    },

    ExcelExporter : {
        'No resource assigned' : 'Non hai recursos atribuídos'
    },

    CrudManagerView : {
        serverResponseLabel : 'Resposta do servidor:'
    },

    DurationColumn : {
        Duration : 'Duración'
    }
};

export default LocaleHelper.publishLocale(locale);
