// @bryntum/calendar/source/lib/Scheduler/localization/Kk.js
import LocaleHelper from '../../Core/localization/LocaleHelper.js';
import '../../Grid/localization/Kk.js';

const locale = {

    localeName : 'Kk',
    localeDesc : 'Қазақ тілі (KZ)',
    localeCode : 'kk-KZ',
    localeRtl  : false,

    Object : {
        newEvent    : 'Жаңа оқиға',
        information : 'Ақпарат',
        showDetails : 'Толығырақ көрсету'
    },

    ResourceInfoColumn : {
        eventCountText : data => data + 'оқиға' + (data !== 1 ? 'с' : '')
    },

    ResourceCollapseColumn : {
        toggleResourceLayout : 'Ресурс орналасу бағанын ауыстыру. SPACE мәзір үшін.'
    },

    Dependencies : {
        from    : 'Басталу күні',
        to      : 'дейін',
        valid   : 'Жарамды',
        invalid : 'Жарамсыз'
    },

    DependencyType : {
        SS           : 'Бастау-бастау',
        SF           : 'Бастау-аяқтау',
        FS           : 'Аяқтау-бастау',
        FF           : 'Аяқтау-аяқтау',
        StartToStart : 'Бастау-бастау',
        StartToEnd   : 'Бастау-аяқтау',
        EndToStart   : 'Аяқтау-бастау',
        EndToEnd     : 'Аяқтау-аяқтау',
        short        : [
            'Бастау-бастау',
            'Бастау-аяқтау',
            'Аяқтау-бастау',
            'Аяқтау-аяқтау'
        ],
        long : [
            'Бастау-бастау',
            'Бастау-аяқтау',
            'Аяқтау-бастау',
            'Аяқтау-аяқтау'
        ]
    },

    DependencyEdit : {
        From              : 'Басталу күні',
        To                : 'дейін',
        Type              : 'Түрі',
        Lag               : 'Кешігу',
        'Edit dependency' : 'Тәуелділікті өңдеу',
        Save              : 'Сақтау',
        Delete            : 'Жою',
        Cancel            : 'Баc тарту',
        StartToStart      : 'Бастау-бастау',
        StartToEnd        : 'Бастау-аяқтау',
        EndToStart        : 'Аяқтау-бастау',
        EndToEnd          : 'Аяқтау-аяқтау'
    },

    EventEdit : {
        Name         : 'Аты',
        Resource     : 'Ресурс',
        Start        : 'Басталу',
        startTime    : 'Бастау уақыты',
        End          : 'Аяқталуы',
        endTime      : 'Аяқталу уақыты',
        Save         : 'Сақтау',
        Delete       : 'Жою',
        Cancel       : 'Баc тарту',
        'Edit event' : 'Оқиғаны өңдеу',
        Repeat       : 'Қайталау'
    },

    EventDrag : {
        eventOverlapsExisting : 'Оқиға осы ресурс үшін бар оқиғаны қабаттастырады',
        noDropOutsideTimeline : 'Уақыт шкаласынан тыс оқиғаны толығымен жою мүмкін емес'
    },

    SchedulerBase : {
        'Add event'      : 'Оқиға қосу',
        'Delete event'   : 'Оқиғаны жою',
        'Unassign event' : 'Оқиғаны тағайындаудан бас тарту',
        color            : 'Түсі',
        eventAriaLabel   : (eventRecord, startDate) => `${eventRecord.name}, басталады ${startDate} ${eventRecord.duration > 0 ? `ұзақтығы ${eventRecord.fullDuration.toString(false)}` : ''}`
    },

    TimeAxisHeaderMenu : {
        pickZoomLevel   : 'Ұлғайту',
        activeDateRange : 'Күн ауқымы',
        startText       : 'Басталу күні',
        endText         : 'Аяқталу күні',
        todayText       : 'Бүгін'
    },

    TimeAxisColumn : {
        ariaLabel : (startDate, endDate) => `Уақыт осі ${startDate} - ${endDate} көрсетеді`
    },

    EventCopyPaste : {
        copyEvent  : 'Оқиғаны көшіру',
        cutEvent   : 'Оқиғаны қию',
        pasteEvent : 'Оқиғаны қою'
    },

    EventFilter : {
        filterEvents : 'Тапсырмаларды сүзгілеу',
        byName       : 'Аты бойынша'
    },

    TimeRanges : {
        showCurrentTimeLine : 'Ағымдағы уақыт шкаласын көрсету'
    },

    PresetManager : {
        secondAndMinute : {
            displayDateFormat : 'll LTS',
            name              : 'Секунд'
        },
        minuteAndHour : {
            topDateFormat     : 'ккк АА/КК, сағТД',
            displayDateFormat : 'll LST'
        },
        hourAndDay : {
            topDateFormat     : 'ккк АА/КК',
            middleDateFormat  : 'LST',
            displayDateFormat : 'll LST',
            name              : 'Күн'
        },
        day : {
            name : 'Күн/сағат'
        },
        week : {
            name : 'Апта/сағат'
        },
        dayAndWeek : {
            displayDateFormat : 'll LST',
            name              : 'Апта/күндер'
        },
        dayAndMonth : {
            name : 'Ай'
        },
        weekAndDay : {
            displayDateFormat : 'll LST',
            name              : 'Апта'
        },
        weekAndMonth : {
            name : 'Апта'
        },
        weekAndDayLetter : {
            name : 'Апталар/жұмыс күндері'
        },
        weekDateAndMonth : {
            name : 'Айлар/апталар'
        },
        monthAndYear : {
            name : 'Ай'
        },
        year : {
            name : 'Жыл'
        },
        manyYears : {
            name : 'Бірнеше жыл'
        }
    },

    RecurrenceConfirmationPopup : {
        'delete-title'              : 'Оқиға жойылуда',
        'delete-all-message'        : 'Осы оқиғаның барлық жағдайларын өзгерту қажет пе?',
        'delete-further-message'    : 'Оқиғаның тек осы жағдайын немесе осы және болашақтағы барлық жағдайларды жою қажет пе?',
        'delete-only-this-message'  : 'Осы оқиғаның осы пайда болуын жойғыңыз келе ме?',
        'delete-further-btn-text'   : 'Барлық болашақ оқиғаларды жою',
        'delete-only-this-btn-text' : 'Тек осы оқиғаны жою',
        'update-title'              : 'Қайталанатын оқиға өзгертілуде',
        'update-all-message'        : 'Осы оқиғаның барлық жағдайларын өзгерту қажет пе?',
        'update-further-message'    : 'Оқиғаның тек осы жағдайын немесе осы және болашақтағы барлық жағдайларды өзгерту қажет пе?',
        'update-only-this-message'  : 'Осы оқиғаның осы пайда болуын өзгерткіңіз келе ме?',
        'update-further-btn-text'   : 'Барлық болашақ оқиғалар',
        'update-only-this-btn-text' : 'Тек осы оқиға',
        Yes                         : 'Иә',
        Cancel                      : 'Баc тарту',
        width                       : 600
    },

    RecurrenceLegend : {
        ' and '                         : ' және ',
        Daily                           : 'Күн сайын',
        'Weekly on {1}'                 : ({ days }) => `Апта сайын ${days} күні`,
        'Monthly on {1}'                : ({ days }) => `Ай сайын ${days} күні`,
        'Yearly on {1} of {2}'          : ({ days, months }) => `Жыл сайын ${months} айының ${days} күні`,
        'Every {0} days'                : ({ interval }) => `Әр ${interval} күн сайын`,
        'Every {0} weeks on {1}'        : ({ interval, days }) => `Әр ${interval} апта сайын ${days}`,
        'Every {0} months on {1}'       : ({ interval, days }) => `Әр ${interval} ай сайын ${days}`,
        'Every {0} years on {1} of {2}' : ({ interval, days, months }) => `${months} ішінен әр ${interval} жыл сайын ${days}`,
        position1                       : 'бірінші',
        position2                       : 'екінші',
        position3                       : 'үшінші',
        position4                       : 'төртінші',
        position5                       : 'бесінші',
        'position-1'                    : 'соңғы',
        day                             : 'күн',
        weekday                         : 'жұмыс күні',
        'weekend day'                   : 'демалыс күні',
        daysFormat                      : ({ position, days }) => `${position} ${days}`
    },

    RecurrenceEditor : {
        'Repeat event'      : 'Оқиғаны қайталау',
        Cancel              : 'Баc тарту',
        Save                : 'Сақтау',
        Frequency           : 'Жиілік',
        Every               : 'Әр',
        DAILYintervalUnit   : 'күн/күндер',
        WEEKLYintervalUnit  : 'апта/апталар',
        MONTHLYintervalUnit : 'ай(лар)',
        YEARLYintervalUnit  : 'жыл/жылдар',
        Each                : 'Әрбір',
        on                  : 'the',
        the                 : 'the',
        'On the'            : 'күні',
        'End repeat'        : 'Қайталауды аяқтау',
        'time(s)'           : 'рет',
        Days                : 'күн',
        Months              : 'ай'
    },

    RecurrenceDaysCombo : {
        day           : 'күн',
        weekday       : 'жұмыс күні',
        'weekend day' : 'демалыс күні'
    },

    RecurrencePositionsCombo : {
        position1    : 'бірінші',
        position2    : 'екінші',
        position3    : 'үшінші',
        position4    : 'төртінші',
        position5    : 'бесінші',
        'position-1' : 'соңғы'
    },

    RecurrenceStopConditionCombo : {
        Never     : 'Ешқашан',
        After     : 'кейін',
        'On date' : 'күні'
    },

    RecurrenceFrequencyCombo : {
        None    : 'Қайталау жоқ',
        Daily   : 'Күн сайын',
        Weekly  : 'Апта сайын {0}',
        Monthly : 'Ай сайын',
        Yearly  : 'Жыл сайын'
    },

    RecurrenceCombo : {
        None   : 'Жоқ',
        Custom : 'Реттелетін...'
    },

    Summary : {
        'Summary for' : date => `үшін түйіндеме ${date}`
    },

    ScheduleRangeCombo : {
        completeview : 'Кестені аяқтау',
        currentview  : 'Көрінетін кесте',
        daterange    : 'Күн ауқымы',
        completedata : 'Кестені аяқтау (барлық оқиғалар үшін)'
    },

    SchedulerExportDialog : {
        'Schedule range' : 'Жоспарлау ауқымы',
        'Export from'    : 'Басталу күні',
        'Export to'      : 'дейін'
    },

    ExcelExporter : {
        'No resource assigned' : 'Тағайындалған ресурс жоқ'
    },

    CrudManagerView : {
        serverResponseLabel : 'Сервер жауабы:'
    },

    DurationColumn : {
        Duration : 'Ұзақтығы'
    }
};

export default LocaleHelper.publishLocale(locale);
