// @bryntum/calendar/source/lib/Scheduler/localization/Lt.js
import LocaleHelper from '../../Core/localization/LocaleHelper.js';
import '../../Grid/localization/Lt.js';

const locale = {

    localeName : 'LT',
    localeDesc : 'Lietuvių',
    localeCode : 'lt-LT',
    localeRtl  : false,

    Object : {
        newEvent    : 'Naujas įvykis',
        information : 'Information',
        showDetails : 'Rodyti išsamią informaciją'
    },

    ResourceInfoColumn : {
        eventCountText : data => data + ' įvykis' + (data !== 1 ? 's' : '')
    },

    ResourceCollapseColumn : {
        toggleResourceLayout : 'Perjungti išteklių išdėstymo stulpelį. SPACE kontekstiniam meniu.'
    },

    Dependencies : {
        from    : 'Nuo',
        to      : 'Iki',
        valid   : 'Galioja',
        invalid : 'Netinkama'
    },

    DependencyType : {
        SS           : 'PrPr',
        SF           : 'PrPa',
        FS           : 'PaPr',
        FF           : 'PaPa',
        StartToStart : 'Nuo pradžios iki pradžios',
        StartToEnd   : 'Nuo pradžios iki pabaigos',
        EndToStart   : 'Nuo pabaigos iki pradžios',
        EndToEnd     : 'Nuo pabaigos iki pabaigos',
        short        : [
            'PrPr',
            'PrPa',
            'PaPr',
            'PaPa'
        ],
        long : [
            'Nuo pradžios iki pradžios',
            'Nuo pradžios iki pabaigos',
            'Nuo pabaigos iki pradžios',
            'Nuo pabaigos iki pabaigos'
        ]
    },

    DependencyEdit : {
        From              : 'Nuo',
        To                : 'Iki',
        Type              : 'Tipas',
        Lag               : 'Lag',
        'Edit dependency' : 'Redaguoti priklausomybę',
        Save              : 'Įrašyti',
        Delete            : 'Naikinti',
        Cancel            : 'Atšaukti',
        StartToStart      : 'Nuo pradžios iki pradžios',
        StartToEnd        : 'Nuo pradžios iki pabaigos',
        EndToStart        : 'Nuo pabaigos iki pradžios',
        EndToEnd          : 'Tiesioginis'
    },

    EventEdit : {
        Name         : 'Pavadinimas',
        Resource     : 'Ištekliai',
        Start        : 'Pradėti',
        startTime    : 'Pradžios laikas',
        End          : 'Baigti',
        endTime      : 'Pabaigos laikas',
        Save         : 'Įrašyti',
        Delete       : 'Naikinti',
        Cancel       : 'Atšaukti',
        'Edit event' : 'Redaguoti įvykį',
        Repeat       : 'Kartoti'
    },

    EventDrag : {
        eventOverlapsExisting : 'Įvykis persidengia su esamu šio ištekliaus įvykiu',
        noDropOutsideTimeline : 'Gali būti, kad įvykis nebus visiškai pašalintas už laiko planavimo juostos ribų'
    },

    SchedulerBase : {
        'Add event'      : 'Įtraukti įvykį',
        'Delete event'   : 'Naikinti įvykį',
        'Unassign event' : 'Atšaukti įvykį',
        color            : 'Spalva',
        eventAriaLabel   : (eventRecord, startDate) => `${eventRecord.name}, prasideda ${startDate} ${eventRecord.duration > 0 ? `trukmė ${eventRecord.fullDuration.toString(false)}` : ''}`
    },

    TimeAxisHeaderMenu : {
        pickZoomLevel   : 'Mastelis',
        activeDateRange : 'Datų diapazonas',
        startText       : 'Pradžios data',
        endText         : 'Pabaigos data',
        todayText       : 'Šiandien'
    },

    TimeAxisColumn : {
        ariaLabel : (startDate, endDate) => `Laiko ašis rodo nuo ${startDate} iki ${endDate}`
    },

    EventCopyPaste : {
        copyEvent  : 'Kopijuoti įvykį',
        cutEvent   : 'Kirpti įvykį',
        pasteEvent : 'Įklijuoti įvykį'
    },

    EventFilter : {
        filterEvents : 'Filtruoti užduotis',
        byName       : 'Pagal pavadinimą'
    },

    TimeRanges : {
        showCurrentTimeLine : 'Rodyti dabartinę laiko planavimo juostą'
    },

    PresetManager : {
        secondAndMinute : {
            displayDateFormat : 'll LTS',
            name              : 'Sekundės'
        },
        minuteAndHour : {
            topDateFormat     : 'MM/DD, ddd, val.',
            displayDateFormat : 'll LST'
        },
        hourAndDay : {
            topDateFormat     : 'MM/DD, ddd',
            middleDateFormat  : 'LST',
            displayDateFormat : 'll LST',
            name              : 'Diena'
        },
        day : {
            name : 'Diena / valandos'
        },
        week : {
            name : 'Savaitė / valandos'
        },
        dayAndWeek : {
            displayDateFormat : 'll LST',
            name              : 'Savaitė / dienos'
        },
        dayAndMonth : {
            name : 'Mėnuo'
        },
        weekAndDay : {
            displayDateFormat : 'll LST',
            name              : 'Savaitė'
        },
        weekAndMonth : {
            name : 'Savaitės'
        },
        weekAndDayLetter : {
            name : 'Savaitės / savaitės dienos'
        },
        weekDateAndMonth : {
            name : 'Mėnesiai / savaitės'
        },
        monthAndYear : {
            name : 'Mėnesiai'
        },
        year : {
            name : 'Metai'
        },
        manyYears : {
            name : 'Keli metai'
        }
    },

    RecurrenceConfirmationPopup : {
        'delete-title'              : 'Naikinate įvykį',
        'delete-all-message'        : 'Ar norite naikinti visus šio įvykio atvejus?',
        'delete-further-message'    : 'Ar norite naikinti šį ir visus būsimus šio įvykio atvejus, ar tik pasirinktus atvejus?',
        'delete-only-this-message'  : 'Ar norite ištrinti šio įvykio šį atsitikimą?',
        'delete-further-btn-text'   : 'Naikinti visus būsimus įvykius',
        'delete-only-this-btn-text' : 'Naikinti tik šį įvykį',
        'update-title'              : 'Keičiate pasikartojantį įvykį',
        'update-all-message'        : 'Ar norite pakeisti visus šio įvykio atvejus?',
        'update-further-message'    : 'Ar norite keisti tik šį įvykio atvejį, ar šį ir visus būsimus atvejus?',
        'update-only-this-message'  : 'Ar norite pakeisti šio įvykio šį atsitikimą?',
        'update-further-btn-text'   : 'Visi būsimi įvykiai',
        'update-only-this-btn-text' : 'Tik šis įvykis',
        Yes                         : 'Taip',
        Cancel                      : 'Atšaukti',
        width                       : 600
    },

    RecurrenceLegend : {
        ' and '                         : ' and ',
        Daily                           : 'Kasdien',
        'Weekly on {1}'                 : ({ days }) => `Savaitinis ${days}`,
        'Monthly on {1}'                : ({ days }) => `Mėnesinis ${days}`,
        'Yearly on {1} of {2}'          : ({ days, months }) => `Metinis ${days} iš ${months}`,
        'Every {0} days'                : ({ interval }) => `Kas ${interval} dienas`,
        'Every {0} weeks on {1}'        : ({ interval, days }) => `Kas ${interval} savaites ${days}`,
        'Every {0} months on {1}'       : ({ interval, days }) => `Kas ${interval} mėnesius ${days}`,
        'Every {0} years on {1} of {2}' : ({ interval, days, months }) => `Kiekvienus ${interval} metus ${months} mėn. ${days}`,
        position1                       : 'pirmas',
        position2                       : 'sekundė',
        position3                       : 'trečias',
        position4                       : 'ketvirtas',
        position5                       : 'penktas',
        'position-1'                    : 'paskutinis',
        day                             : 'diena',
        weekday                         : 'darbo diena',
        'weekend day'                   : 'savaitgalis',
        daysFormat                      : ({ position, days }) => `${position} ${days}`
    },

    RecurrenceEditor : {
        'Repeat event'      : 'Kartoti įvykį',
        Cancel              : 'Atšaukti',
        Save                : 'Įrašyti',
        Frequency           : 'Dažnumas',
        Every               : 'Kas',
        DAILYintervalUnit   : 'diena (-os)',
        WEEKLYintervalUnit  : 'sav.',
        MONTHLYintervalUnit : 'mėnuo (-iai)',
        YEARLYintervalUnit  : 'metai',
        Each                : 'Kiekvienas',
        on                  : 'aukščiau',
        the                 : 'į',
        'On the'            : 'Data',
        'End repeat'        : 'Baigti kartoti',
        'time(s)'           : 'kartais (-ai)',
        Days                : 'Dienas',
        Months              : 'Mėnesius'
    },

    RecurrenceDaysCombo : {
        day           : 'diena',
        weekday       : 'darbo diena',
        'weekend day' : 'savaitgalis'
    },

    RecurrencePositionsCombo : {
        position1    : 'pirmas',
        position2    : 'antras',
        position3    : 'trečias',
        position4    : 'ketvirtas',
        position5    : 'penktas',
        'position-1' : 'paskutinis'
    },

    RecurrenceStopConditionCombo : {
        Never     : 'Niekada',
        After     : 'po',
        'On date' : 'Nustatytai datai'
    },

    RecurrenceFrequencyCombo : {
        None    : 'Nekartoti',
        Daily   : 'Kasdien',
        Weekly  : 'Kas savaitę {0}',
        Monthly : 'Kas mėnesį',
        Yearly  : 'Kasmet'
    },

    RecurrenceCombo : {
        None   : 'Nėra',
        Custom : 'Pasirinktinai...'
    },

    Summary : {
        'Summary for' : date => `Santrauka skirta ${date}`
    },

    ScheduleRangeCombo : {
        completeview : 'Įvykdytas grafikas',
        currentview  : 'Dabartinis grafikas',
        daterange    : 'Datų diapazonas',
        completedata : 'Užbaigti grafiką (visų įvykių)'
    },

    SchedulerExportDialog : {
        'Schedule range' : 'Grafiko intervalas',
        'Export from'    : 'Nuo',
        'Export to'      : 'Iki'
    },

    ExcelExporter : {
        'No resource assigned' : 'Nepriskirti ištekliai'
    },

    CrudManagerView : {
        serverResponseLabel : 'Serverio atsakas:'
    },

    DurationColumn : {
        Duration : 'Trukmė'
    }
};

export default LocaleHelper.publishLocale(locale);
