// @bryntum/calendar/source/lib/Scheduler/localization/Lv.js
import LocaleHelper from '../../Core/localization/LocaleHelper.js';
import '../../Grid/localization/Lv.js';

const locale = {

    localeName : 'Lv',
    localeDesc : 'Latviešu',
    localeCode : 'lv-LV',
    localeRtl  : false,

    Object : {
        newEvent    : 'Jauns notikums',
        information : 'Information',
        showDetails : 'Rādīt detaļas'
    },

    ResourceInfoColumn : {
        eventCountText : data => data + ' notikums' + (data !== 1 ? '(i)' : '')
    },

    ResourceCollapseColumn : {
        toggleResourceLayout : 'Pārslēgt resursu izkārtojuma kolonnu. SPACE konteksta izvēlnei.'
    },

    Dependencies : {
        from    : 'No',
        to      : 'Līdz',
        valid   : 'Derīgs',
        invalid : 'Nederīgs'
    },

    DependencyType : {
        SS           : 'SS',
        SF           : 'SB',
        FS           : 'BS',
        FF           : 'BB',
        StartToStart : 'No sākuma līdz sākumam',
        StartToEnd   : 'No sākuma līdz beigām',
        EndToStart   : 'No beigām līdz sākumam',
        EndToEnd     : 'No beigām līdz beigām',
        short        : [
            'SS',
            'SB',
            'BS',
            'BB'
        ],
        long : [
            'No sākuma līdz sākumam',
            'No sākuma līdz beigām',
            'No beigām līdz sākumam',
            'No beigām līdz beigām'
        ]
    },

    DependencyEdit : {
        From              : 'No',
        To                : 'Līdz',
        Type              : 'Tips',
        Lag               : 'Aizkave',
        'Edit dependency' : 'Rediģēt atkarību',
        Save              : 'Saglabāt',
        Delete            : 'Dzēst',
        Cancel            : 'Atcelt',
        StartToStart      : 'No sākuma līdz sākumam',
        StartToEnd        : 'No sākuma līdz beigām',
        EndToStart        : 'No beigām līdz sākumam',
        EndToEnd          : 'No beigām līdz beigām'
    },

    EventEdit : {
        Name         : 'Nosaukums',
        Resource     : 'Resurss',
        Start        : 'Sākums',
        startTime    : 'Sākuma laiks',
        End          : 'Beigas',
        endTime      : 'Beigu laiks',
        Save         : 'Saglabāt',
        Delete       : 'Dzēst',
        Cancel       : 'Atcelt',
        'Edit event' : 'Rediģēt notikumu',
        Repeat       : 'Atkārtot'
    },

    EventDrag : {
        eventOverlapsExisting : 'Notikums pārklājas ar esošu notikumu šim resursam',
        noDropOutsideTimeline : 'Notikums var nebūt pilnībā atmests ārpus laika skalas'
    },

    SchedulerBase : {
        'Add event'      : 'Pievienot notikumu',
        'Delete event'   : 'Dzēst notikumu',
        'Unassign event' : 'Notikuma piešķires noņemšana',
        color            : 'Krāsa',
        eventAriaLabel   : (eventRecord, startDate) => `${eventRecord.name}, sākas plkst. ${startDate} ${eventRecord.duration > 0 ? `uz ${eventRecord.fullDuration.toString(false)}` : ''}`
    },

    TimeAxisHeaderMenu : {
        pickZoomLevel   : 'Tālummainīt',
        activeDateRange : 'Datumu diapazons',
        startText       : 'Sākuma datums',
        endText         : 'Beigu datums',
        todayText       : 'Today'
    },

    TimeAxisColumn : {
        ariaLabel : (startDate, endDate) => `Laika ass parāda no ${startDate} līdz ${endDate}`
    },

    EventCopyPaste : {
        copyEvent  : 'Kopēt notikumu',
        cutEvent   : 'Izgriezt notikumu',
        pasteEvent : 'Ielīmēt notikumu'
    },

    EventFilter : {
        filterEvents : 'Filtrēt uzdevumus',
        byName       : 'Pēc nosaukuma'
    },

    TimeRanges : {
        showCurrentTimeLine : 'Rādīt pašreizējo laika skalu'
    },

    PresetManager : {
        secondAndMinute : {
            displayDateFormat : 'll LTS',
            name              : 'Sekundes'
        },
        minuteAndHour : {
            topDateFormat     : 'ddd DD/MM, hA',
            displayDateFormat : 'll LST'
        },
        hourAndDay : {
            topDateFormat     : 'ddd DD/MM',
            middleDateFormat  : 'LST',
            displayDateFormat : 'll LST',
            name              : 'Diena'
        },
        day : {
            name : 'Diena/stundas'
        },
        week : {
            name : 'Nedēļa/stundas'
        },
        dayAndWeek : {
            displayDateFormat : 'll LST',
            name              : 'Nedēļa/dienas'
        },
        dayAndMonth : {
            name : 'Mēnesis'
        },
        weekAndDay : {
            displayDateFormat : 'll LST',
            name              : 'Nedēļa'
        },
        weekAndMonth : {
            name : 'Nedēļas'
        },
        weekAndDayLetter : {
            name : 'Nedēļas/darbdienas'
        },
        weekDateAndMonth : {
            name : 'Mēneši/nedēļas'
        },
        monthAndYear : {
            name : 'Mēneši'
        },
        year : {
            name : 'Gadi'
        },
        manyYears : {
            name : 'Vairāki gadi'
        }
    },

    RecurrenceConfirmationPopup : {
        'delete-title'              : 'Jūs dzēšat notikumu',
        'delete-all-message'        : 'Vai vēlaties izdzēst visus šī notikuma gadījumus?',
        'delete-further-message'    : 'Vai vēlaties dzēst šo un visus turpmākos šī notikuma gadījumus vai tikai atlasīto gadījumu?',
        'delete-only-this-message'  : 'Vai vēlaties dzēst šo notikuma gadījumu?',
        'delete-further-btn-text'   : 'Dzēst visus turpmākos notikumus',
        'delete-only-this-btn-text' : 'Dzēst tikai šo notikumu',
        'update-title'              : 'Jūs maināt atkārtotu notikumu',
        'update-all-message'        : 'Vai vēlaties mainīt visus šī notikuma gadījumus?',
        'update-further-message'    : 'Vai vēlaties mainīt tikai šo notikuma gadījumu vai arī šo un visus turpmākos gadījumus?',
        'update-only-this-message'  : 'Vai vēlaties mainīt šo notikuma gadījumu?',
        'update-further-btn-text'   : 'Visi turpmākie notikumi',
        'update-only-this-btn-text' : 'Tikai šis notikums',
        Yes                         : 'Jā',
        Cancel                      : 'Atcelt',
        width                       : 600
    },

    RecurrenceLegend : {
        ' and '                         : ' un ',
        Daily                           : 'Katru dienu',
        'Weekly on {1}'                 : ({ days }) => `Katru nedēļu šādās dienās: ${days}`,
        'Monthly on {1}'                : ({ days }) => `Katru mēnesi šādās dienās: ${days}`,
        'Yearly on {1} of {2}'          : ({ days, months }) => `Katru gadu šādās dienās: ${days}, šādos mēnešos: ${months}`,
        'Every {0} days'                : ({ interval }) => `Ik pēc ${interval} dienas(ām)`,
        'Every {0} weeks on {1}'        : ({ interval, days }) => `Ik pēc ${interval} nedēļas(ām): ${days}`,
        'Every {0} months on {1}'       : ({ interval, days }) => `Ik pēc ${interval} mēneša(iem): ${days}`,
        'Every {0} years on {1} of {2}' : ({ interval, days, months }) => `Ik pēc ${interval} gada(iem): ${days}, ${months}`,
        position1                       : 'pirmā',
        position2                       : 'otrā',
        position3                       : 'trešā',
        position4                       : 'ceturtā',
        position5                       : 'piektā',
        'position-1'                    : 'pēdējā',
        day                             : 'diena',
        weekday                         : 'darbdiena',
        'weekend day'                   : 'nedēļas nogales diena',
        daysFormat                      : ({ position, days }) => `${position} ${days}`
    },

    RecurrenceEditor : {
        'Repeat event'      : 'Atkārtot notikumu',
        Cancel              : 'Atcelt',
        Save                : 'Saglabāt',
        Frequency           : 'Biežums',
        Every               : 'Ik pēc',
        DAILYintervalUnit   : 'dienas(ām)',
        WEEKLYintervalUnit  : 'nedēļas(ām)',
        MONTHLYintervalUnit : 'mēneša(iem)',
        YEARLYintervalUnit  : 'gada(iem)',
        Each                : 'Katru',
        on                  : 'virs',
        the                 : 'uz',
        'On the'            : 'Šādā datumā:',
        'End repeat'        : 'Beigt atkārtošanu',
        'time(s)'           : 'reizi(es)',
        Days                : 'Dienas',
        Months              : 'Mēneša'
    },

    RecurrenceDaysCombo : {
        day           : 'diena',
        weekday       : 'darbdiena',
        'weekend day' : 'nedēļas nogales diena'
    },

    RecurrencePositionsCombo : {
        position1    : 'pirmais',
        position2    : 'otrais',
        position3    : 'trešais',
        position4    : 'ceturtais',
        position5    : 'piektais',
        'position-1' : 'pēdējais'
    },

    RecurrenceStopConditionCombo : {
        Never     : 'Nekad',
        After     : 'Pēc',
        'On date' : 'Datums'
    },

    RecurrenceFrequencyCombo : {
        None    : 'Neatkārtot',
        Daily   : 'Katru dienu',
        Weekly  : 'Ik nedēļu šādā dienā: {0}',
        Monthly : 'Ik mēnesi',
        Yearly  : 'Ik gadu'
    },

    RecurrenceCombo : {
        None   : 'Nav',
        Custom : 'Pielāgošana...'
    },

    Summary : {
        'Summary for' : date => `Kopsavilkums par ${date}`
    },

    ScheduleRangeCombo : {
        completeview : 'Viss grafiks',
        currentview  : 'Redzamais grafiks',
        daterange    : 'Datumu diapazons',
        completedata : 'Viss grafiks (visiem notikumiem)'
    },

    SchedulerExportDialog : {
        'Schedule range' : 'Grafika diapazons',
        'Export from'    : 'No',
        'Export to'      : 'Līdz'
    },

    ExcelExporter : {
        'No resource assigned' : 'Nav piešķirtu resursu'
    },

    CrudManagerView : {
        serverResponseLabel : 'Servera atbilde:'
    },

    DurationColumn : {
        Duration : 'Ilgums'
    }
};

export default LocaleHelper.publishLocale(locale);
