// @bryntum/calendar/source/lib/Scheduler/localization/SrRs.js
import LocaleHelper from '../../Core/localization/LocaleHelper.js';
import '../../Grid/localization/SrRs.js';

const locale = {

    localeName : 'SrRs',
    localeDesc : 'Српски (ћирилица)',
    localeCode : 'sr-RS',
    localeRtl  : false,

    Object : {
        newEvent    : 'Нови догађај',
        information : 'Информација',
        showDetails : 'Прикажи детаље'
    },

    ResourceInfoColumn : {
        eventCountText : data => data + ' догађај' + (data !== 1 ? 'с' : '')
    },

    ResourceCollapseColumn : {
        toggleResourceLayout : 'Пребаци колону распореда ресурса. SPACE за контекст мени.'
    },

    Dependencies : {
        from    : 'Од',
        to      : 'До',
        valid   : 'Важећи',
        invalid : 'Неважеће'
    },

    DependencyType : {
        SS           : 'ПП',
        SF           : 'ПЗ',
        FS           : 'ЗП',
        FF           : 'ЗЗ',
        StartToStart : 'Од почетка ка почетку',
        StartToEnd   : 'Од почетка ка завршетку',
        EndToStart   : 'Од завршетка ка почетку',
        EndToEnd     : 'Од завршетка ка завршетку',
        short        : [
            'ПП',
            'ПЗ',
            'ЗП',
            'ЗЗ'
        ],
        long : [
            'Од почетка ка почетку',
            'Од почетка ка завршетку',
            'Од завршетка ка почетку',
            'Од завршетка ка завршетку'
        ]
    },

    DependencyEdit : {
        From              : 'Од',
        To                : 'До',
        Type              : 'Тип',
        Lag               : 'Кашњење',
        'Edit dependency' : 'Уреди зависност',
        Save              : 'Сачувај',
        Delete            : 'Избриши',
        Cancel            : 'Откажи',
        StartToStart      : 'Од почетка ка почетку',
        StartToEnd        : 'Од почетка до завршетка',
        EndToStart        : 'Од почетка до краја',
        EndToEnd          : 'С краја на крај'
    },

    EventEdit : {
        Name         : 'Име',
        Resource     : 'Ресурс',
        Start        : 'Почетак',
        startTime    : 'Время начала',
        End          : 'Крај',
        endTime      : 'Время окончания',
        Save         : 'Сачувај',
        Delete       : 'Избриши',
        Cancel       : 'Откажи',
        'Edit event' : 'Уреди догађај',
        Repeat       : 'Понови'
    },

    EventDrag : {
        eventOverlapsExisting : 'Догађај се преклапа са постојећим догађајем за овај ресурс',
        noDropOutsideTimeline : 'Догађај можда неће бити одбачен потпуно изван временске осе'
    },

    SchedulerBase : {
        'Add event'      : 'Додај догађај',
        'Delete event'   : 'Избриши догађај',
        'Unassign event' : 'Опозови доделу догађаја',
        color            : 'Боја',
        eventAriaLabel   : (eventRecord, startDate) => `${eventRecord.name}, почиње у ${startDate} ${eventRecord.duration > 0 ? `траје ${eventRecord.fullDuration.toString(false)}` : ''}`
    },

    TimeAxisHeaderMenu : {
        pickZoomLevel   : 'Зумирај',
        activeDateRange : 'Опсег датума',
        startText       : 'Датум почетка',
        endText         : 'Датум завршетка',
        todayText       : 'Данас'
    },

    TimeAxisColumn : {
        ariaLabel : (startDate, endDate) => `Временска ос показује од ${startDate} до ${endDate}`
    },

    EventCopyPaste : {
        copyEvent  : 'Копирај догађај',
        cutEvent   : 'Исеци догађај',
        pasteEvent : 'Налепи догађај'
    },

    EventFilter : {
        filterEvents : 'Филтрирај задатке',
        byName       : 'Према називу'
    },

    TimeRanges : {
        showCurrentTimeLine : 'Прикажи тренутну временску осу'
    },

    PresetManager : {
        secondAndMinute : {
            displayDateFormat : 'll LTS',
            name              : 'Секунди'
        },
        minuteAndHour : {
            topDateFormat     : 'ддд, MM.ДД, hA',
            displayDateFormat : 'll LST'
        },
        hourAndDay : {
            topDateFormat     : 'ддд, MM.ДД.',
            middleDateFormat  : 'LST',
            displayDateFormat : 'll LST',
            name              : 'Дан'
        },
        day : {
            name : 'Дан/сати'
        },
        week : {
            name : 'Седмица/сати'
        },
        dayAndWeek : {
            displayDateFormat : 'll LST',
            name              : 'Седмица/дани'
        },
        dayAndMonth : {
            name : 'Mесец'
        },
        weekAndDay : {
            displayDateFormat : 'll LST',
            name              : 'Седмица'
        },
        weekAndMonth : {
            name : 'Недеља'
        },
        weekAndDayLetter : {
            name : 'Седмице/радни дани'
        },
        weekDateAndMonth : {
            name : 'Месеци/седмице'
        },
        monthAndYear : {
            name : 'Месеци'
        },
        year : {
            name : 'Године'
        },
        manyYears : {
            name : 'Више година'
        }
    },

    RecurrenceConfirmationPopup : {
        'delete-title'              : 'Бришете догађај',
        'delete-all-message'        : 'Желите ли да избришете сва појављивања овог догађаја?',
        'delete-further-message'    : 'Желите ли да избришете ово и сва будућа појављивања овог догађаја или само изабрано појављивање?',
        'delete-only-this-message'  : 'Вы хотите удалить это появление этого события?',
        'delete-further-btn-text'   : 'Избриши све будуће догађаје',
        'delete-only-this-btn-text' : 'Избриши само овај догађај',
        'update-title'              : 'Мењате догађај који се понавља',
        'update-all-message'        : 'Желите ли да промените сва појављивања овог догађаја?',
        'update-further-message'    : 'Желите ли да промените само ово појављивање догађаја или ово и сва будућа појављивања?',
        'update-only-this-message'  : 'Вы хотите изменить это появление этого события?',
        'update-further-btn-text'   : 'Сви будући догађаји',
        'update-only-this-btn-text' : 'Само овај догађај',
        Yes                         : 'Да',
        Cancel                      : 'Откажи',
        width                       : 600
    },

    RecurrenceLegend : {
        ' and '                         : ' и ',
        Daily                           : 'Свакодневно',
        'Weekly on {1}'                 : ({ days }) => `Недељно током ${days}`,
        'Monthly on {1}'                : ({ days }) => `Mесечно током ${days}`,
        'Yearly on {1} of {2}'          : ({ days, months }) => `Годишње током ${days} of ${months}`,
        'Every {0} days'                : ({ interval }) => `На ${interval} дана`,
        'Every {0} weeks on {1}'        : ({ interval, days }) => `На ${interval} седмице/а дана ${days}`,
        'Every {0} months on {1}'       : ({ interval, days }) => `На ${interval} месеца/и дана ${days}`,
        'Every {0} years on {1} of {2}' : ({ interval, days, months }) => `На ${interval} године/а ${days} месеца ${months}`,
        position1                       : 'први',
        position2                       : 'други',
        position3                       : 'трећи',
        position4                       : 'четврти',
        position5                       : 'пети',
        'position-1'                    : 'последњи',
        day                             : 'дан',
        weekday                         : 'радни дан',
        'weekend day'                   : 'дан викенда',
        daysFormat                      : ({ position, days }) => `${position} ${days}`
    },

    RecurrenceEditor : {
        'Repeat event'      : 'Понављај догађај',
        Cancel              : 'Откажи',
        Save                : 'Сачувај',
        Frequency           : 'Учесталост',
        Every               : 'На',
        DAILYintervalUnit   : 'дан(а)',
        WEEKLYintervalUnit  : 'седмица(е)',
        MONTHLYintervalUnit : 'месец(и)',
        YEARLYintervalUnit  : 'година(е)',
        Each                : 'Сваки',
        on                  : 'на',
        the                 : 'на',
        'On the'            : 'На дан',
        'End repeat'        : 'Заврши понављање',
        'time(s)'           : 'пут(а)',
        Days                : 'дана',
        Months              : 'месеци'
    },

    RecurrenceDaysCombo : {
        day           : 'дан',
        weekday       : 'радни дан',
        'weekend day' : 'дан викенда'
    },

    RecurrencePositionsCombo : {
        position1    : 'први',
        position2    : 'second',
        position3    : 'трећи',
        position4    : 'четврти',
        position5    : 'пети',
        'position-1' : 'последње'
    },

    RecurrenceStopConditionCombo : {
        Never     : 'Никад',
        After     : 'након',
        'On date' : 'На датум'
    },

    RecurrenceFrequencyCombo : {
        None    : 'Без понављања',
        Daily   : 'Свакодневно',
        Weekly  : 'Седмично {0}',
        Monthly : 'Месечно',
        Yearly  : 'Годишње'
    },

    RecurrenceCombo : {
        None   : 'Ниједно',
        Custom : 'Прилагођено...'
    },

    Summary : {
        'Summary for' : date => `Резиме за ${date}`
    },

    ScheduleRangeCombo : {
        completeview : 'Целокупан распоред',
        currentview  : 'Видљиви распоред',
        daterange    : 'Опсег датума',
        completedata : 'Целокупан распоред (за све догађаје)'
    },

    SchedulerExportDialog : {
        'Schedule range' : 'Опсег распореда',
        'Export from'    : 'Од',
        'Export to'      : 'До'
    },

    ExcelExporter : {
        'No resource assigned' : 'Нису додељени ресурси'
    },

    CrudManagerView : {
        serverResponseLabel : 'Одговор сервера:'
    },

    DurationColumn : {
        Duration : 'Трајање'
    }
};

export default LocaleHelper.publishLocale(locale);
