// @bryntum/calendar/source/lib/Scheduler/localization/Uk.js
import LocaleHelper from '../../Core/localization/LocaleHelper.js';
import '../../Grid/localization/Uk.js';

const locale = {

    localeName : 'Uk',
    localeDesc : 'Українська',
    localeCode : 'uk-UA',
    localeRtl  : false,

    Object : {
        newEvent    : 'Створити подію',
        information : 'Інформація',
        showDetails : 'Показати деталі'
    },

    ResourceInfoColumn : {
        eventCountText : data => data + ' подія' + (data !== 1 ? 's' : '')
    },

    ResourceCollapseColumn : {
        toggleResourceLayout : 'Перемкнути стовпець макету ресурсу. SPACE для контекстного меню.'
    },

    Dependencies : {
        from    : 'З',
        to      : 'До',
        valid   : 'Припустимий',
        invalid : 'Неприпустимий'
    },

    DependencyType : {
        SS           : 'ПП',
        SF           : 'ПЗ',
        FS           : 'ЗП',
        FF           : 'ЗЗ',
        StartToStart : 'Початок-початок',
        StartToEnd   : 'Початок-завершення',
        EndToStart   : 'Завершення-початок',
        EndToEnd     : 'Завершення-завершення',
        short        : [
            'ПП',
            'ПЗ',
            'ЗП',
            'ЗЗ'
        ],
        long : [
            'Початок-початок',
            'Початок-завершення',
            'Завершення-початок',
            'Завершення-завершення'
        ]
    },

    DependencyEdit : {
        From              : 'З',
        To                : 'До',
        Type              : 'Тип',
        Lag               : 'Відставання від графіку',
        'Edit dependency' : 'Редагувати залежність',
        Save              : 'Зберегти',
        Delete            : 'Видалити',
        Cancel            : 'Скасувати',
        StartToStart      : 'Початок-початок',
        StartToEnd        : 'Початок‑завершення',
        EndToStart        : 'Завершення‑початок',
        EndToEnd          : 'Завершення‑завершення'
    },

    EventEdit : {
        Name         : 'Назва',
        Resource     : 'Ресурс',
        Start        : 'Початок',
        startTime    : 'Час початку',
        End          : 'Завершення',
        endTime      : 'Час завершення',
        Save         : 'Зберегти',
        Delete       : 'Видалити',
        Cancel       : 'Скасувати',
        'Edit event' : 'Редагувати подію',
        Repeat       : 'Повторити'
    },

    EventDrag : {
        eventOverlapsExisting : 'Подія перекриває наявну подію для цього ресурсу',
        noDropOutsideTimeline : 'За межами цієї часової шкали подію може бути видалено не повністю'
    },

    SchedulerBase : {
        'Add event'      : 'Додати подію',
        'Delete event'   : 'Видалити подію',
        'Unassign event' : 'Скасувати призначення події',
        color            : 'Колір',
        eventAriaLabel   : (eventRecord, startDate) => `${eventRecord.name}, починається о ${startDate} ${eventRecord.duration > 0 ? `на ${eventRecord.fullDuration.toString(false)}` : ''}`
    },

    TimeAxisHeaderMenu : {
        pickZoomLevel   : 'Масштаб',
        activeDateRange : 'Діапазон дат',
        startText       : 'Дата початку',
        endText         : 'Дата завершення',
        todayText       : 'Сьогодні'
    },

    TimeAxisColumn : {
        ariaLabel : (startDate, endDate) => `Часова вісь показує з ${startDate} по ${endDate}`
    },

    EventCopyPaste : {
        copyEvent  : 'Копіювати подію',
        cutEvent   : 'Вирізати подію',
        pasteEvent : 'Вставити подію'
    },

    EventFilter : {
        filterEvents : 'Завдання фільтра',
        byName       : 'За іменем'
    },

    TimeRanges : {
        showCurrentTimeLine : 'Показати поточну часову шкалу'
    },

    PresetManager : {
        secondAndMinute : {
            displayDateFormat : 'll LTS',
            name              : 'Секунди'
        },
        minuteAndHour : {
            topDateFormat     : 'ддд ММ/ДД, гA',
            displayDateFormat : 'll LST'
        },
        hourAndDay : {
            topDateFormat     : 'ддд ММ/ДД',
            middleDateFormat  : 'LST',
            displayDateFormat : 'll LST',
            name              : 'День'
        },
        day : {
            name : 'День/години'
        },
        week : {
            name : 'Тиждень/години'
        },
        dayAndWeek : {
            displayDateFormat : 'll LST',
            name              : 'Тиждень/дні'
        },
        dayAndMonth : {
            name : 'Місяць'
        },
        weekAndDay : {
            displayDateFormat : 'll LST',
            name              : 'Тиждень'
        },
        weekAndMonth : {
            name : 'Тижні'
        },
        weekAndDayLetter : {
            name : 'Тижні/дні тижня'
        },
        weekDateAndMonth : {
            name : 'Місяці/тижні'
        },
        monthAndYear : {
            name : 'Місяці'
        },
        year : {
            name : 'Роки'
        },
        manyYears : {
            name : 'Кілька років'
        }
    },

    RecurrenceConfirmationPopup : {
        'delete-title'              : 'Ви видаляєте подію',
        'delete-all-message'        : 'Видалити всі екземпляри цієї події?',
        'delete-further-message'    : 'Видалити цей і всі майбутні екземпляри цієї події або лише вибрані екземпляри?',
        'delete-only-this-message'  : 'Ви хочете видалити цю появу цієї події?',
        'delete-further-btn-text'   : 'Видалити всі майбутні події',
        'delete-only-this-btn-text' : 'Видалити лише цю подію',
        'update-title'              : 'Ви змінюєте повторювану подію',
        'update-all-message'        : 'Змінити всі екземпляри цієї події?',
        'update-further-message'    : 'Змінити лише цей екземпляр події чи й усі майбутні екземпляри?',
        'update-only-this-message'  : 'Ви хочете змінити цю появу цієї події?',
        'update-further-btn-text'   : 'Усі майбутні події',
        'update-only-this-btn-text' : 'Лише ця подія',
        Yes                         : 'Так',
        Cancel                      : 'Скасувати',
        width                       : 600
    },

    RecurrenceLegend : {
        ' and '                         : ' і ',
        Daily                           : 'Щоденно',
        'Weekly on {1}'                 : ({ days }) => `Щотижня по ${days}`,
        'Monthly on {1}'                : ({ days }) => `${days} числа кожного місяця`,
        'Yearly on {1} of {2}'          : ({ days, months }) => `Щороку ${days} ${months}`,
        'Every {0} days'                : ({ interval }) => `Кожен ${interval}-й день`,
        'Every {0} weeks on {1}'        : ({ interval, days }) => `Кожен ${interval}-й тиждень, день тижня: ${days}`,
        'Every {0} months on {1}'       : ({ interval, days }) => `Кожен ${interval}-й місяць, день тижня: ${days}`,
        'Every {0} years on {1} of {2}' : ({ interval, days, months }) => `Кожен ${interval}-й рік, дата: ${days}, місяць: ${months}`,
        position1                       : 'перший',
        position2                       : 'другий',
        position3                       : 'третій',
        position4                       : 'четвертий',
        position5                       : 'п’ятий',
        'position-1'                    : 'останній',
        day                             : 'дн.',
        weekday                         : 'день тижня',
        'weekend day'                   : 'вихідний день',
        daysFormat                      : ({ position, days }) => `${position} ${days}`
    },

    RecurrenceEditor : {
        'Repeat event'      : 'Повторити подію',
        Cancel              : 'Скасувати',
        Save                : 'Зберегти',
        Frequency           : 'Періодичність',
        Every               : 'Кожн.',
        DAILYintervalUnit   : 'дн.',
        WEEKLYintervalUnit  : 'тиж.',
        MONTHLYintervalUnit : 'міс.',
        YEARLYintervalUnit  : 'р.',
        Each                : 'Кожн.',
        on                  : 'В',
        the                 : 'В',
        'On the'            : 'В',
        'End repeat'        : 'Завершити повторення',
        'time(s)'           : 'раз',
        Days                : 'день',
        Months              : 'місяць'
    },

    RecurrenceDaysCombo : {
        day           : 'дн.',
        weekday       : 'день тижня',
        'weekend day' : 'вихідний день'
    },

    RecurrencePositionsCombo : {
        position1    : 'перший',
        position2    : 'другий',
        position3    : 'третій',
        position4    : 'четвертий',
        position5    : 'п’ятий',
        'position-1' : 'останній'
    },

    RecurrenceStopConditionCombo : {
        Never     : 'Ніколи',
        After     : 'Після',
        'On date' : 'Дата завершення повторення'
    },

    RecurrenceFrequencyCombo : {
        None    : 'Без повтору',
        Daily   : 'Щоденно',
        Weekly  : 'Щотижня в {0}',
        Monthly : 'Щомісячно',
        Yearly  : 'Щорічно'
    },

    RecurrenceCombo : {
        None   : 'Немає',
        Custom : 'Настроювані...'
    },

    Summary : {
        'Summary for' : date => `Резюме для ${date}`
    },

    ScheduleRangeCombo : {
        completeview : 'Увесь розклад',
        currentview  : 'Видима область розкладу',
        daterange    : 'Діапазон дат',
        completedata : 'Увесь розклад (усі події)'
    },

    SchedulerExportDialog : {
        'Schedule range' : 'Діапазон розкладу',
        'Export from'    : 'З',
        'Export to'      : 'До'
    },

    ExcelExporter : {
        'No resource assigned' : 'Немає виділених ресурсів'
    },

    CrudManagerView : {
        serverResponseLabel : 'Відповідь сервера:'
    },

    DurationColumn : {
        Duration : 'Тривалість'
    }
};

export default LocaleHelper.publishLocale(locale);
