// @bryntum/calendar/source/lib/Scheduler/localization/ZhTw.js
import LocaleHelper from '../../Core/localization/LocaleHelper.js';
import '../../Grid/localization/ZhTw.js';

const locale = {

    localeName : 'ZhTw',
    localeDesc : '繁體中文 (台灣)',
    localeCode : 'zh-TW',
    localeRtl  : false,

    Object : {
        newEvent    : '新增事件',
        information : '資訊',
        showDetails : '顯示詳細資料'
    },

    ResourceInfoColumn : {
        eventCountText : data => data + ' 事件' + (data !== 1 ? '秒' : '')
    },

    ResourceCollapseColumn : {
        toggleResourceLayout : '切換資源佈局列。按 SPACE 開啟上下文選單。'
    },

    Dependencies : {
        from    : '從',
        to      : '到',
        valid   : '有效',
        invalid : '無效'
    },

    DependencyType : {
        SS           : '開始到開始',
        SF           : '開始到完成',
        FS           : '完成到開始',
        FF           : '完成到完成',
        StartToStart : '開始到開始',
        StartToEnd   : '開始到完成',
        EndToStart   : '完成到開始',
        EndToEnd     : '完成到完成',
        short        : [
            '開始到開始',
            '開始到完成',
            '完成到開始',
            '完成到完成'
        ],
        long : [
            '開始到開始',
            '開始到完成',
            '完成到開始',
            '完成到完成'
        ]
    },

    DependencyEdit : {
        From              : '從',
        To                : '到',
        Type              : '類型',
        Lag               : '滯後',
        'Edit dependency' : '編輯相依性',
        Save              : '儲存',
        Delete            : '刪除',
        Cancel            : '取消',
        StartToStart      : '開始到開始',
        StartToEnd        : '開始到結束',
        EndToStart        : '結束到開始',
        EndToEnd          : '結束到結束'
    },

    EventEdit : {
        Name         : '名稱',
        Resource     : '資源',
        Start        : '開始',
        startTime    : '開始時間',
        End          : '結束',
        endTime      : '結束時間',
        Save         : '儲存',
        Delete       : '刪除',
        Cancel       : '取消',
        'Edit event' : '編輯事件',
        Repeat       : '重複'
    },

    EventDrag : {
        eventOverlapsExisting : '事件與此資源的現有事件重疊',
        noDropOutsideTimeline : '事件可能無法在時間表外完整清除'
    },

    SchedulerBase : {
        'Add event'      : '新增事件',
        'Delete event'   : '刪除事件',
        'Unassign event' : '取消指派事件',
        color            : '色彩',
        eventAriaLabel   : (eventRecord, startDate) => `${eventRecord.name}，開始於 ${startDate} ${eventRecord.duration > 0 ? `持續 ${eventRecord.fullDuration.toString(false)}` : ''}`
    },

    TimeAxisHeaderMenu : {
        pickZoomLevel   : '縮放',
        activeDateRange : '日期範圍',
        startText       : '開始日期',
        endText         : '結束日期',
        todayText       : '今天'
    },

    TimeAxisColumn : {
        ariaLabel : (startDate, endDate) => `時間軸顯示從 ${startDate} 到 ${endDate}`
    },

    EventCopyPaste : {
        copyEvent  : '複製事件',
        cutEvent   : '剪下事件',
        pasteEvent : '貼上事件'
    },

    EventFilter : {
        filterEvents : '篩選工作',
        byName       : '依名稱'
    },

    TimeRanges : {
        showCurrentTimeLine : '顯示目前時間表'
    },

    PresetManager : {
        secondAndMinute : {
            displayDateFormat : 'll LTS',
            name              : '秒'
        },
        minuteAndHour : {
            topDateFormat     : 'MM/DD，ddd，hA',
            displayDateFormat : 'll LST'
        },
        hourAndDay : {
            topDateFormat     : 'MM/DD，ddd',
            middleDateFormat  : 'LST',
            displayDateFormat : 'll LST',
            name              : '天'
        },
        day : {
            name : '天/小時'
        },
        week : {
            name : '週/小時'
        },
        dayAndWeek : {
            displayDateFormat : 'll LST',
            name              : '週/天'
        },
        dayAndMonth : {
            name : '月'
        },
        weekAndDay : {
            displayDateFormat : 'll LST',
            name              : '週'
        },
        weekAndMonth : {
            name : '週'
        },
        weekAndDayLetter : {
            name : '週/工作日'
        },
        weekDateAndMonth : {
            name : '月/週'
        },
        monthAndYear : {
            name : '月'
        },
        year : {
            name : '年'
        },
        manyYears : {
            name : '多年'
        }
    },

    RecurrenceConfirmationPopup : {
        'delete-title'              : '您正在刪除事件',
        'delete-all-message'        : '您想要刪除此事件的所有發生次數嗎?',
        'delete-further-message'    : '您想要刪除此事件的此次及之後的所有未來事件，還是僅刪除選取的事件?',
        'delete-only-this-message'  : '您要删除此事件的这一发生吗？',
        'delete-further-btn-text'   : '刪除所有未來事件',
        'delete-only-this-btn-text' : '僅刪除此事件',
        'update-title'              : '您正在變更重複事件',
        'update-all-message'        : '您想要變更此事件的所有發生次數嗎?',
        'update-further-message'    : '您只要變更事件的這次發生，還是變更這一次及之後的所有未來事件?',
        'update-only-this-message'  : '您要更改此事件的这一发生吗？',
        'update-further-btn-text'   : '所有未來事件',
        'update-only-this-btn-text' : '僅此事件',
        Yes                         : '是',
        Cancel                      : '取消',
        width                       : 600
    },

    RecurrenceLegend : {
        ' and '                         : ' 和 ',
        Daily                           : '每天',
        'Weekly on {1}'                 : ({ days }) => `每周在 ${days}`,
        'Monthly on {1}'                : ({ days }) => `每月在 ${days}`,
        'Yearly on {1} of {2}'          : ({ days, months }) => `每年在 ${months} 的 ${days}`,
        'Every {0} days'                : ({ interval }) => `每隔 ${interval} 天`,
        'Every {0} weeks on {1}'        : ({ interval, days }) => `每隔 ${days} 週的${interval}`,
        'Every {0} months on {1}'       : ({ interval, days }) => `每隔 ${interval} 個月的${days}`,
        'Every {0} years on {1} of {2}' : ({ interval, days, months }) => `每隔 ${interval} 年 ${months} 的${days}`,
        position1                       : '第一個',
        position2                       : '第二個',
        position3                       : '第三個',
        position4                       : '第四個',
        position5                       : '第五個',
        'position-1'                    : '最後一個',
        day                             : '天',
        weekday                         : '工作日',
        'weekend day'                   : '週末',
        daysFormat                      : ({ position, days }) => `${position} ${days}`
    },

    RecurrenceEditor : {
        'Repeat event'      : '重複事件',
        Cancel              : '取消',
        Save                : '儲存',
        Frequency           : '頻率',
        Every               : '每隔',
        DAILYintervalUnit   : '天',
        WEEKLYintervalUnit  : '週',
        MONTHLYintervalUnit : '個月',
        YEARLYintervalUnit  : '年',
        Each                : '每個',
        on                  : '在',
        the                 : '在',
        'On the'            : '在',
        'End repeat'        : '結束重複',
        'time(s)'           : '次',
        Days                : '天',
        Months              : '個月的'
    },

    RecurrenceDaysCombo : {
        day           : '天',
        weekday       : '工作日',
        'weekend day' : '週末'
    },

    RecurrencePositionsCombo : {
        position1    : '第一次',
        position2    : 'second',
        position3    : '第三',
        position4    : '第四',
        position5    : '第五',
        'position-1' : '最後'
    },

    RecurrenceStopConditionCombo : {
        Never     : '永不',
        After     : '之後',
        'On date' : '於日期'
    },

    RecurrenceFrequencyCombo : {
        None    : '沒有重複',
        Daily   : '每天',
        Weekly  : '每週{0}',
        Monthly : '每月',
        Yearly  : '每年'
    },

    RecurrenceCombo : {
        None   : '無',
        Custom : '自訂...'
    },

    Summary : {
        'Summary for' : date => `总结 ${date}`
    },

    ScheduleRangeCombo : {
        completeview : '完整排程',
        currentview  : '可見排程',
        daterange    : '日期範圍',
        completedata : '完整排程 (所有事件)'
    },

    SchedulerExportDialog : {
        'Schedule range' : '排程範圍',
        'Export from'    : '從',
        'Export to'      : '到'
    },

    ExcelExporter : {
        'No resource assigned' : '未指派任何資源'
    },

    CrudManagerView : {
        serverResponseLabel : '伺服器回覆:'
    },

    DurationColumn : {
        Duration : '期間'
    }
};

export default LocaleHelper.publishLocale(locale);
