// @bryntum/calendar/source/lib/Core/localization/Ca.js
import LocaleHelper from '../../Core/localization/LocaleHelper.js';

const locale = {

    localeName : 'Ca',
    localeDesc : 'Català',
    localeCode : 'ca-ES',
    localeRtl  : false,

    Object : {
        Yes      : 'Sí',
        No       : 'No',
        Cancel   : 'Cancel·la',
        Ok       : "D'acord",
        Week     : 'Setmana',
        None     : 'Cap',
        previous : 'Anterior',
        next     : 'Següent',
        to       : 'a',
        at       : 'a',
        on       : 'el',
        editing  : what => `Editant ${what}`,
        settings : 'Configuració',
        close    : 'Tancar',
        go       : 'Vés-hi',
        save     : 'Desa',
        cancel   : 'Cancel·la',
        revert   : 'Reverteix'
    },

    CheckboxGroup : {
        wrongNbrOfOptionsSelected : min => `Seleccioneu exactament ${min}`,
        tooFewOptionsSelected     : min => `Seleccioneu almenys ${min}`,
        tooManyOptionsSelected    : max => `Seleccioneu com a màxim ${max}`
    },

    CodeEditor : {
        apply        : 'Aplicar',
        autoApply    : 'Aplicació automàtica',
        downloadCode : 'Descarregar codi',
        editor       : 'Editor de codi',
        viewer       : 'Visualitzador de codi'
    },

    ColorPicker : {
        noColor : 'Sense color'
    },

    Combo : {
        noResults          : 'No hi ha cap resultat',
        recordNotCommitted : "No s'ha pogut afegir el registre",
        addNewValue        : value => `Afegeix ${value}`
    },

    FilePicker : {
        file : 'Fitxer'
    },

    Field : {
        badInput              : 'Valor del camp no vàlid',
        patternMismatch       : 'El valor ha de coincidir amb un patró específic',
        rangeOverflow         : value => `El valor ha de ser igual o inferior a ${value.max}`,
        rangeUnderflow        : value => `El valor ha de ser igual o superior a ${value.min}`,
        stepMismatch          : "El valor ha d'ajustar-se al pas",
        tooLong               : 'El valor ha de ser més curt',
        tooShort              : 'El valor ha de ser més llarg',
        typeMismatch          : 'Cal que el valor estigui en un format especial',
        valueMissing          : 'Aquest camp és obligatori',
        invalidValue          : 'Valor del camp no vàlid',
        minimumValueViolation : 'Infracció del valor mínim',
        maximumValueViolation : 'Infracció del valor màxim',
        fieldRequired         : 'Aquest camp és obligatori',
        validateFilter        : "El valor s'ha de seleccionar de la llista"
    },

    DateField : {
        invalidDate : format => 'Entrada de data no vàlida, el format esperat és ' + format
    },

    DatePicker : {
        gotoPrevYear  : "Ves a l'any anterior",
        gotoPrevMonth : 'Ves al mes anterior',
        gotoNextMonth : 'Ves al mes següent',
        gotoNextYear  : "Ves a l'any següent"
    },

    Hint : {
        dontShowAgain : 'No tornis a mostrar'
    },

    NumberFormat : {
        locale   : 'ca-ES',
        currency : 'EUR'
    },

    DurationField : {
        invalidUnit : 'Unitat no vàlida'
    },

    TimeField : {
        invalidTime : 'Entrada de temps no vàlida'
    },

    TimePicker : {
        hour   : 'Hora',
        minute : 'Minut',
        second : 'Segon'
    },

    List : {
        loading   : "S'està carregant...",
        selectAll : 'Selecciona-ho tot'
    },

    GridBase : {
        loadMask : "S'està carregant...",
        syncMask : "S'estan desant els canvis, espereu-vos..."
    },

    PagingToolbar : {
        firstPage         : 'Ves a la primera pàgina',
        prevPage          : 'Ves a la pàgina anterior',
        page              : 'Pàgina',
        nextPage          : 'Ves a la pàgina següent',
        lastPage          : "Ves a l'última pàgina",
        reload            : 'Torna a carregar la pàgina actual',
        noRecords         : 'No hi ha cap registre per visualitzar',
        pageCountTemplate : data => `de ${data.lastPage}`,
        summaryTemplate   : data => `S'estan mostrant els registres ${data.start} - ${data.end} de ${data.allCount}`
    },

    PanelCollapser : {
        Collapse : 'Redueix',
        Expand   : 'Expandeix'
    },

    Popup : {
        close : 'Tanca'
    },

    TabPanel : {
        containsInvalidFields : 'La pestanya conté camps no vàlids'
    },

    Toolbar : {
        showOverflowingItems : 'Mostra els elements desbordants',
        scrollForMore        : 'Desplaça’t per veure més'
    },

    UndoRedo : {
        Undo           : 'Desfés',
        Redo           : 'Refés',
        UndoLastAction : "Desfés l'última acció",
        RedoLastAction : "Refés l'última acció desfeta",
        NoActions      : 'No hi ha cap element a la cua de desfer'
    },

    FieldFilterPicker : {
        equals                 : 'és igual a',
        doesNotEqual           : 'no és igual a',
        isEmpty                : 'buit',
        isNotEmpty             : 'no està buit',
        contains               : 'conté',
        doesNotContain         : 'no conté',
        startsWith             : 'comença per',
        endsWith               : 'acaba amb',
        isOneOf                : 'un de',
        isNotOneOf             : 'no és un de',
        isGreaterThan          : 'major que',
        isLessThan             : 'inferior a',
        isGreaterThanOrEqualTo : 'major o igual que',
        isLessThanOrEqualTo    : 'menor o igual que',
        isBetween              : 'entre',
        isNotBetween           : 'no és entre',
        isBefore               : 'abans',
        isAfter                : 'després',
        timeEquals             : 'el temps és igual',
        timeDoesNotEqual       : 'el temps no és igual',
        isToday                : 'avui',
        isTomorrow             : 'demà',
        isYesterday            : 'ahir',
        isThisWeek             : 'aquesta setmana',
        isNextWeek             : 'la setmana que ve',
        isLastWeek             : 'la setmana passada',
        isThisMonth            : 'aquest mes',
        isNextMonth            : 'el mes que ve',
        isLastMonth            : 'el mes passat',
        isThisYear             : 'aquest any',
        isNextYear             : "l'any que ve",
        isLastYear             : "l'any passat",
        isYearToDate           : 'any fins a la data',
        isTrue                 : 'cert',
        isFalse                : 'fals',
        selectAProperty        : 'Seleccioneu una propietat',
        selectAnOperator       : 'Seleccioneu un operador',
        caseSensitive          : 'Distingeix entre majúscules i minúscules',
        and                    : 'i',
        dateFormat             : 'D/M/AA',
        selectValue            : 'Seleccioneu el valor',
        selectOneOrMoreValues  : 'Seleccioneu un o més valors',
        enterAValue            : 'Introduïu un valor',
        enterANumber           : 'Introduïu un número',
        selectADate            : 'Seleccioneu una data',
        selectATime            : "Seleccioneu l'hora"
    },

    FieldFilterPickerGroup : {
        addFilter : 'Afegeix un filtre',
        applyCond : applied => applied ? 'Omet el filtre (actualment aplicat)' : 'Aplica el filtre (actualment omès)'
    },

    DateHelper : {
        locale         : 'ca-ES',
        weekStartDay   : 0,
        nonWorkingDays : {
            0 : true,
            6 : true
        },
        weekends : {
            0 : true,
            6 : true
        },
        unitNames : [
            { single : 'mil·lisegon', plural : 'mil·lisegons', abbrev : 'ms' },
            { single : 'emillisecond', plural : 'ems', abbrev : 'ems' },
            { single : 'segon', plural : 'segons', abbrev : 's' },
            { single : 'esecond', plural : 'eseconds', abbrev : 'es' },
            { single : 'minut', plural : 'minuts', abbrev : 'min' },
            { single : 'eminute', plural : 'eminutes', abbrev : 'emin' },
            { single : 'hora', plural : 'hores', abbrev : 'h' },
            { single : 'ehour', plural : 'ehours', abbrev : 'eh' },
            { single : 'dia', plural : 'dies', abbrev : 'd' },
            { single : 'eday', plural : 'edays', abbrev : 'ed' },
            { single : 'setmana', plural : 'setmanes', abbrev : 'set' },
            { single : 'eweek', plural : 'eweeks', abbrev : 'ew' },
            { single : 'mes', plural : 'mesos', abbrev : 'm' },
            { single : 'emonth', plural : 'emonths', abbrev : 'emon' },
            { single : 'trimestre', plural : 'trimestres', abbrev : 't' },
            { single : 'equarter', plural : 'equarters', abbrev : 'eq' },
            { single : 'any', plural : 'anys', abbrev : 'a' },
            { single : 'eyear', plural : 'eyears', abbrev : 'eyr' },
            { single : 'dècada', plural : 'dècades', abbrev : 'dèc' },
            { single : 'edecade', plural : 'edecades', abbrev : 'edec' }
        ],
        unitAbbreviations : [
            ['ms'],
            [],
            ['s'],
            [],
            ['min'],
            [],
            ['h'],
            [],
            ['d'],
            [],
            ['set', 'setm.'],
            [],
            ['m'],
            [],
            ['t', 'trim.', 'tri'],
            [],
            ['a'],
            [],
            ['dèc'],
            []
        ],
        parsers : {
            L   : 'MM/DD/AAAA',
            LT  : 'HH:mm A',
            LTS : 'HH:mm:ss A'
        },
        ordinalSuffix : number => {
            const hasSpecialCase = ['11', '12', '13'].find((n) => number.endsWith(n));

            let suffix = 'è';

            if (!hasSpecialCase) {
                const lastDigit = number[number.length - 1];
                suffix = { 1 : 'r', 2 : 'n', 3 : 'r', 4 : 't' }[lastDigit] || 'è';
            }

            return number + suffix;
        },
        decade : date => {
            const year = typeof date === 'number' ? date : date.getFullYear();
            return `${year - (year % 10)}s`;
        }
    }
};

export default LocaleHelper.publishLocale(locale);
