// @bryntum/calendar/source/lib/Core/localization/Gl.js
import LocaleHelper from '../../Core/localization/LocaleHelper.js';

const locale = {

    localeName : 'Gl',
    localeDesc : 'Galego',
    localeCode : 'gl-ES',
    localeRtl  : false,

    Object : {
        Yes      : 'Si',
        No       : 'Non',
        Cancel   : 'Cancelar',
        Ok       : 'Aceptar',
        Week     : 'Semana',
        None     : 'Non',
        previous : 'Anterior',
        next     : 'Seguinte',
        to       : 'a',
        at       : 'ás',
        on       : 'en',
        editing  : what => `Editando ${what}`,
        settings : 'Axustes',
        close    : 'Pechar',
        go       : 'Vai',
        save     : 'Gardar',
        cancel   : 'Cancelar',
        revert   : 'Reverter'
    },

    CheckboxGroup : {
        wrongNbrOfOptionsSelected : min => `Seleccione exactamente ${min}`,
        tooFewOptionsSelected     : min => `Seleccione polo menos ${min}`,
        tooManyOptionsSelected    : max => `Seleccione non máis de ${max}`
    },

    CodeEditor : {
        apply        : 'Aplicar',
        autoApply    : 'Aplicación automática',
        downloadCode : 'Descargar código',
        editor       : 'Editor de código',
        viewer       : 'Visor de código'
    },

    ColorPicker : {
        noColor : 'Sen cor'
    },

    Combo : {
        noResults          : 'Non hai resultados',
        recordNotCommitted : 'Non se engadiu o rexistro',
        addNewValue        : value => `Engadir ${value}`
    },

    FilePicker : {
        file : 'Ficheiro'
    },

    Field : {
        badInput              : 'Valor de campo non válido',
        patternMismatch       : 'O valor debe coincidir cun padrón específico',
        rangeOverflow         : value => `O valor debe ser menor ou igual que ${value.max}`,
        rangeUnderflow        : value => `O valor debe ser maior ou igual que ${value.min}`,
        stepMismatch          : 'O valor debe axustarse ao paso',
        tooLong               : 'O valor debe ser máis curto',
        tooShort              : 'O valor debe ser máis longo',
        typeMismatch          : 'É necesario que o valor teña un formato especial',
        valueMissing          : 'Este campo é obrigatorio',
        invalidValue          : 'Valor de campo non válido',
        minimumValueViolation : 'Infracción do valor mínimo',
        maximumValueViolation : 'Infracción do valor máximo',
        fieldRequired         : 'Este campo é obrigatorio',
        validateFilter        : 'O valor debe estar seleccionado na lista'
    },

    DateField : {
        invalidDate : format => 'Entrada de data non válida, o formato esperado é ' + format
    },

    DatePicker : {
        gotoPrevYear  : 'Ir ao ano anterior',
        gotoPrevMonth : 'Ir ao mes anterior',
        gotoNextMonth : 'Ir ao mes seguinte',
        gotoNextYear  : 'Ir ao ano seguinte'
    },

    Hint : {
        dontShowAgain : 'Non mostrar de novo'
    },

    NumberFormat : {
        locale   : 'gl-ES',
        currency : 'EUR'
    },

    DurationField : {
        invalidUnit : 'Unidade non válida'
    },

    TimeField : {
        invalidTime : 'Entrada de hora non válida'
    },

    TimePicker : {
        hour   : 'Hora',
        minute : 'Minuto',
        second : 'Segundo'
    },

    List : {
        loading   : 'Cargando...',
        selectAll : 'Seleccionar todo'
    },

    GridBase : {
        loadMask : 'Cargando...',
        syncMask : 'Gardando modificacións. Agarde...'
    },

    PagingToolbar : {
        firstPage         : 'Ir á primeira páxina',
        prevPage          : 'Ir á páxina anterior',
        page              : 'Páxina',
        nextPage          : 'Ir á páxina seguinte',
        lastPage          : 'Ir á última páxina',
        reload            : 'Recargar a páxina actual',
        noRecords         : 'Non hai rexistros para mostrar',
        pageCountTemplate : data => `de ${data.lastPage}`,
        summaryTemplate   : data => `Mostrando os rexistros ${data.start} - ${data.end} de ${data.allCount}`
    },

    PanelCollapser : {
        Collapse : 'Contraer',
        Expand   : 'Expandir'
    },

    Popup : {
        close : 'Pechar'
    },

    TabPanel : {
        containsInvalidFields : 'A lapela contén campos non válidos'
    },

    Toolbar : {
        showOverflowingItems : 'Mostrar elementos desbordados',
        scrollForMore        : 'Desprazar para máis'
    },

    UndoRedo : {
        Undo           : 'Desfacer',
        Redo           : 'Refacer',
        UndoLastAction : 'Desfacer a última acción',
        RedoLastAction : 'Refacer a última acción de desfacer',
        NoActions      : 'Non hai elementos na fila de desfacer'
    },

    FieldFilterPicker : {
        equals                 : 'igual a',
        doesNotEqual           : 'non é igual a',
        isEmpty                : 'baleiro',
        isNotEmpty             : 'non está baleiro',
        contains               : 'contén',
        doesNotContain         : 'non contén',
        startsWith             : 'comeza por',
        endsWith               : 'finaliza en',
        isOneOf                : 'un de',
        isNotOneOf             : 'non é un de',
        isGreaterThan          : 'maior que',
        isLessThan             : 'menor que',
        isGreaterThanOrEqualTo : 'maior ou igual que',
        isLessThanOrEqualTo    : 'menor ou igual que',
        isBetween              : 'entre',
        isNotBetween           : 'non entre',
        isBefore               : 'antes de',
        isAfter                : 'despois de',
        timeEquals             : 'o tempo é igual',
        timeDoesNotEqual       : 'o tempo non é igual',
        isToday                : 'hoxe',
        isTomorrow             : 'mañá',
        isYesterday            : 'onte',
        isThisWeek             : 'esta semana',
        isNextWeek             : 'a próxima semana',
        isLastWeek             : 'última semana',
        isThisMonth            : 'este mes',
        isNextMonth            : 'o próximo mes',
        isLastMonth            : 'último mes',
        isThisYear             : 'este ano',
        isNextYear             : 'ano seguinte',
        isLastYear             : 'último ano',
        isYearToDate           : 'até o presente',
        isTrue                 : 'verdadeiro',
        isFalse                : 'falso',
        selectAProperty        : 'Seleccionar unha propiedade',
        selectAnOperator       : 'Seleccionar un operador',
        caseSensitive          : 'Diferenciar maiúsculas e minúsculas',
        and                    : 'e',
        dateFormat             : 'D/M/AA',
        selectValue            : 'Seleccione o valor',
        selectOneOrMoreValues  : 'Seleccionar un ou varios valores',
        enterAValue            : 'Introducir un valor',
        enterANumber           : 'Introducir un número',
        selectADate            : 'Seleccionar unha data',
        selectATime            : 'Selecciona hora'
    },

    FieldFilterPickerGroup : {
        addFilter : 'Engadir filtro',
        applyCond : applied => applied ? 'Omitir filtro (aplicado actualmente)' : 'Aplicar filtro (omitido actualmente)'
    },

    DateHelper : {
        locale         : 'gl-ES',
        weekStartDay   : 0,
        nonWorkingDays : {
            0 : true,
            6 : true
        },
        weekends : {
            0 : true,
            6 : true
        },
        unitNames : [
            { single : 'milisegundo', plural : 'milisegundos', abbrev : 'ms' },
            { single : 'emillisecond', plural : 'ems', abbrev : 'ems' },
            { single : 'segundo', plural : 'segundos', abbrev : 's' },
            { single : 'esecond', plural : 'eseconds', abbrev : 'es' },
            { single : 'minuto', plural : 'minutos', abbrev : 'min' },
            { single : 'eminute', plural : 'eminutes', abbrev : 'emin' },
            { single : 'hora', plural : 'horas', abbrev : 'h' },
            { single : 'ehour', plural : 'ehours', abbrev : 'eh' },
            { single : 'día', plural : 'días', abbrev : 'd' },
            { single : 'eday', plural : 'edays', abbrev : 'ed' },
            { single : 'semana', plural : 'semanas', abbrev : 'sem' },
            { single : 'eweek', plural : 'eweeks', abbrev : 'ew' },
            { single : 'mes', plural : 'meses', abbrev : 'mes' },
            { single : 'emonth', plural : 'emonths', abbrev : 'emon' },
            { single : 'trimestre', plural : 'trimestres', abbrev : 't' },
            { single : 'equarter', plural : 'equarters', abbrev : 'eq' },
            { single : 'ano', plural : 'anos', abbrev : 'ano' },
            { single : 'eyear', plural : 'eyears', abbrev : 'eyr' },
            { single : 'década', plural : 'décadas', abbrev : 'déc' },
            { single : 'edecade', plural : 'edecades', abbrev : 'edec' }
        ],
        unitAbbreviations : [
            ['mil'],
            [],
            ['s', 's'],
            [],
            ['min', 'min'],
            [],
            ['h', 'h'],
            [],
            ['d'],
            [],
            ['s', 'sem'],
            [],
            ['mes', 'mes', 'mes'],
            [],
            ['t', 'tri', 'trim'],
            [],
            ['a', 'ano'],
            [],
            ['déc'],
            []
        ],
        parsers : {
            L   : 'MM/DD/AAAA',
            LT  : 'HH:mm A',
            LTS : 'HH:mm:ss A'
        },
        ordinalSuffix : number => {
            const hasSpecialCase = ['11', '12', '13'].find((n) => number.endsWith(n));

            let suffix = 'º';

            if (!hasSpecialCase) {
                const lastDigit = number[number.length - 1];
                suffix = { 1 : 'º', 2 : 'º', 3 : 'º' }[lastDigit] || 'º';
            }

            return number + suffix;
        },
        decade : date => {
            const year = typeof date === 'number' ? date : date.getFullYear();
            return `anos ${year - (year % 10)}`;
        }
    }
};

export default LocaleHelper.publishLocale(locale);
