// @bryntum/calendar/source/lib/Core/localization/Sk.js
import LocaleHelper from '../../Core/localization/LocaleHelper.js';

const locale = {

    localeName : 'Sk',
    localeDesc : 'Slovenský',
    localeCode : 'sk',
    localeRtl  : false,

    Object : {
        Yes      : 'Áno',
        No       : 'Nie',
        Cancel   : 'Zrušiť',
        Ok       : 'OK',
        Week     : 'Týždeň',
        None     : 'Žiadny',
        previous : 'Predchádzajúci',
        next     : 'Nasledujúci',
        to       : 'do',
        at       : 'o',
        on       : 'na',
        editing  : what => `Úprava ${what}`,
        settings : 'Nastavenia',
        close    : 'Zatvoriť',
        go       : 'Choď',
        save     : 'Uložiť',
        cancel   : 'Zrušiť',
        revert   : 'Obnoviť'
    },

    CheckboxGroup : {
        wrongNbrOfOptionsSelected : min => `Vyberte presne ${min}`,
        tooFewOptionsSelected     : min => `Vyberte aspoň ${min}`,
        tooManyOptionsSelected    : max => `Vyberte najviac ${max}`
    },

    CodeEditor : {
        apply        : 'Použiť',
        autoApply    : 'Automatické použitie',
        downloadCode : 'Stiahnuť kód',
        editor       : 'Editor kódu',
        viewer       : 'Prehliadač kódu'
    },

    ColorPicker : {
        noColor : 'Žiadna farba'
    },

    Combo : {
        noResults          : 'Žiadne výsledky',
        recordNotCommitted : 'Záznam sa nepodarilo pridať',
        addNewValue        : value => `Pridať ${value}`
    },

    FilePicker : {
        file : 'Súbor'
    },

    Field : {
        badInput              : 'Neplatná hodnota poľa',
        patternMismatch       : 'Hodnota by sa mala zhodovať so špecifickým vzorom',
        rangeOverflow         : value => `Hodnota mus byť menšia alebo rovná ${value.max}`,
        rangeUnderflow        : value => `Hodnota musí byť väčšia alebo rovná ${value.min}`,
        stepMismatch          : 'Hodnota by sa mala zhodovať s krokom',
        tooLong               : 'Hodnota by mala byť kratšia',
        tooShort              : 'Hodnota by mala byť dlhšia',
        typeMismatch          : 'Požaduje sa, aby hodnota bola v špeciálnom formáte',
        valueMissing          : 'Toto políčko sa požaduje',
        invalidValue          : 'Neplatná hodnota políčka',
        minimumValueViolation : 'Narušenie minimálnej hodnoty',
        maximumValueViolation : 'Narušenie maximálnej hodnoty',
        fieldRequired         : 'Toto políčko sa požaduje',
        validateFilter        : 'Hodnota musí byť zvolená zo zoznamu'
    },

    DateField : {
        invalidDate : format => 'Neplatný dátumový vstup, očakávaný formát je ' + format
    },

    DatePicker : {
        gotoPrevYear  : 'Prejsť na predchádzajúci rok',
        gotoPrevMonth : 'Prejsť na predchádzajúci mesiac',
        gotoNextMonth : 'Prejsť na nasledujúci mesiac',
        gotoNextYear  : 'Prejsť na nalsedujúci rok'
    },

    Hint : {
        dontShowAgain : 'Nezobrazovať znova'
    },

    NumberFormat : {
        locale   : 'sk',
        currency : 'EUR'
    },

    DurationField : {
        invalidUnit : 'Neplatná jednotka'
    },

    TimeField : {
        invalidTime : 'Vloženie neplatného času'
    },

    TimePicker : {
        hour   : 'Hodina',
        minute : 'Minúta',
        second : 'Sekunda'
    },

    List : {
        loading   : 'Načítavanie...',
        selectAll : 'Vybrať všetko'
    },

    GridBase : {
        loadMask : 'Načítavanie...',
        syncMask : 'ukladajú sa zmeny, čakajte...'
    },

    PagingToolbar : {
        firstPage         : 'Prejsť na prvú stranu',
        prevPage          : 'Prejsť na predchádzajúcu stranu',
        page              : 'Strana',
        nextPage          : 'Prejsť na nasledujúcu stranu',
        lastPage          : 'Prejsť na poslednú stranu',
        reload            : 'Znovu načítať súčasnú stranu',
        noRecords         : 'Žiadne záznamy na zobrazenie',
        pageCountTemplate : data => `z ${data.lastPage}`,
        summaryTemplate   : data => `Zobrazujú sa záznamy ${data.start} - ${data.end} z ${data.allCount}`
    },

    PanelCollapser : {
        Collapse : 'Zbaliť',
        Expand   : 'Rozbaliť'
    },

    Popup : {
        close : 'Zatvoriť vyskakovacie okno'
    },

    TabPanel : {
        containsInvalidFields : 'Karta obsahuje neplatné polia'
    },

    Toolbar : {
        showOverflowingItems : 'Zobraziť pretečené položky',
        scrollForMore        : 'Posuňte pre viac'
    },

    UndoRedo : {
        Undo           : 'Vrátiť späť',
        Redo           : 'Znovu vykonať',
        UndoLastAction : 'Vrátiť späť poslednú akciu',
        RedoLastAction : 'Znovu urobiť poslednú nevykonanú akciu',
        NoActions      : 'Žiadne položky v rade na vrátenie späť'
    },

    FieldFilterPicker : {
        equals                 : 'rovná sa',
        doesNotEqual           : 'nerovná sa',
        isEmpty                : 'je prázdne',
        isNotEmpty             : 'nie je prázdne',
        contains               : 'obsahuje',
        doesNotContain         : 'neobsahuje',
        startsWith             : 'začína na',
        endsWith               : 'končí na',
        isOneOf                : 'je jeden z',
        isNotOneOf             : 'nie je jedno z',
        isGreaterThan          : 'je väčšie než',
        isLessThan             : 'je menšie než',
        isGreaterThanOrEqualTo : 'je väčšie alebo sa rovná',
        isLessThanOrEqualTo    : 'je menšie alebo sa rovná',
        isBetween              : 'je medzi',
        isNotBetween           : 'nie je medzi',
        isBefore               : 'je pred',
        isAfter                : 'je po',
        timeEquals             : 'čas sa rovná',
        timeDoesNotEqual       : 'čas sa nerovná',
        isToday                : 'je dnes',
        isTomorrow             : 'je zajtra',
        isYesterday            : 'je včera',
        isThisWeek             : 'je tento týždeň',
        isNextWeek             : 'je budúci týždeň',
        isLastWeek             : 'je minulý týždeň',
        isThisMonth            : 'je tento mesiac',
        isNextMonth            : 'je budúci mesiac',
        isLastMonth            : 'je minulý mesiac',
        isThisYear             : 'je tento rok',
        isNextYear             : 'je budúci rok',
        isLastYear             : 'je minulý rok',
        isYearToDate           : 'je rok do dnešného dňa',
        isTrue                 : 'je správne',
        isFalse                : 'je nesprávne',
        selectAProperty        : 'Vyberte vlastnosť',
        selectAnOperator       : 'Vyberte operátora',
        caseSensitive          : 'Rozlišuje malé a veľké písmená',
        and                    : 'a',
        dateFormat             : 'D/M/YY',
        selectValue            : 'Vyberte hodnotu',
        selectOneOrMoreValues  : 'Vyberte jednu alebo viac hodnôt',
        enterAValue            : 'Zadajte hodnotu',
        enterANumber           : 'Zadajte číslo',
        selectADate            : 'Vyberte dátum',
        selectATime            : 'Vybrať čas'
    },

    FieldFilterPickerGroup : {
        addFilter : 'Pridajte filter',
        applyCond : applied => applied ? 'Preskočiť filter (momentálne použitý)' : 'Použiť filter (momentálne preskočený)'
    },

    DateHelper : {
        locale         : 'sk',
        weekStartDay   : 1,
        nonWorkingDays : {
            0 : true,
            6 : true
        },
        weekends : {
            0 : true,
            6 : true
        },
        unitNames : [
            { single : 'milisekunda', plural : 'ms', abbrev : 'ms' },
            { single : 'emillisecond', plural : 'ems', abbrev : 'ems' },
            { single : 'sekunda', plural : 'sekundy', abbrev : 's' },
            { single : 'esecond', plural : 'eseconds', abbrev : 'es' },
            { single : 'minúta', plural : 'minúty', abbrev : 'min' },
            { single : 'eminute', plural : 'eminutes', abbrev : 'emin' },
            { single : 'hodina', plural : 'hodiny', abbrev : 'h' },
            { single : 'ehour', plural : 'ehours', abbrev : 'eh' },
            { single : 'deň', plural : 'dni', abbrev : 'd' },
            { single : 'eday', plural : 'edays', abbrev : 'ed' },
            { single : 'týždeň', plural : 'týždne', abbrev : 'tžd' },
            { single : 'eweek', plural : 'eweeks', abbrev : 'ew' },
            { single : 'mesiac', plural : 'mesiace', abbrev : 'msc' },
            { single : 'emonth', plural : 'emonths', abbrev : 'emon' },
            { single : 'štvrť', plural : 'štvrtiny', abbrev : '' },
            { single : 'equarter', plural : 'equarters', abbrev : 'eq' },
            { single : 'rok', plural : 'roky', abbrev : 'rk' },
            { single : 'eyear', plural : 'eyears', abbrev : 'eyr' },
            { single : 'dekáda', plural : 'dekády', abbrev : 'dek' },
            { single : 'edecade', plural : 'edecades', abbrev : 'edec' }
        ],
        unitAbbreviations : [
            ['mil'],
            [],
            ['s', 'sec'],
            [],
            ['m', 'min'],
            [],
            ['h', 'h'],
            [],
            ['d'],
            [],
            ['tžd', 'tžd'],
            [],
            ['msc', 'msc', 'msc'],
            [],
            ['štvrť', 'štvrť', 'štvrť'],
            [],
            ['rk', 'rk'],
            [],
            ['dek'],
            []
        ],
        parsers : {
            L   : 'D. M. YYYY.',
            LT  : 'HH:mm',
            LTS : 'HH:mm:ss A'
        },
        ordinalSuffix : number => number + '.',
        decade        : date => {
            const year = typeof date === 'number' ? date : date.getFullYear();
            return `${year - (year % 10)}. roky`;
        }
    }
};

export default LocaleHelper.publishLocale(locale);
