// @bryntum/calendar/source/lib/Core/localization/SrRs.js
import LocaleHelper from '../../Core/localization/LocaleHelper.js';

const locale = {

    localeName : 'SrRs',
    localeDesc : 'Српски (ћирилица)',
    localeCode : 'sr-RS',
    localeRtl  : false,

    Object : {
        Yes      : 'Да',
        No       : 'Не',
        Cancel   : 'Откажи',
        Ok       : 'У реду',
        Week     : 'Седмица',
        None     : 'Ниједан',
        previous : 'Претходни',
        next     : 'Следећи',
        to       : 'до',
        at       : 'в',
        on       : 'на',
        editing  : what => `Редактирование ${what}`,
        settings : 'Настройки',
        close    : 'Закрыть',
        go       : 'Иди',
        save     : 'Сачувај',
        cancel   : 'Откажи',
        revert   : 'Врати'
    },

    CheckboxGroup : {
        wrongNbrOfOptionsSelected : min => `Изаберите тачно ${min}`,
        tooFewOptionsSelected     : min => `Изаберите најмање ${min}`,
        tooManyOptionsSelected    : max => `Изаберите највише ${max}`
    },

    CodeEditor : {
        apply        : 'Примени',
        autoApply    : 'Аутоматска примена',
        downloadCode : 'Преузми код',
        editor       : 'Уређивач кода',
        viewer       : 'Прегледач кода'
    },

    ColorPicker : {
        noColor : 'Без боје'
    },

    Combo : {
        noResults          : 'Нема резултата',
        recordNotCommitted : 'Није могуће додати запис',
        addNewValue        : value => `Додај ${value}`
    },

    FilePicker : {
        file : 'Датотека'
    },

    Field : {
        badInput              : 'Неважећа вредност поља',
        patternMismatch       : 'Вредност треба да се подудара са одређеним обрасцем',
        rangeOverflow         : value => `Вредност мора да буде мања или једнака са ${value.max}`,
        rangeUnderflow        : value => `Вредност мора да буде већа или једнака са ${value.min}`,
        stepMismatch          : 'Вредност треба да се уклапа у корак',
        tooLong               : 'Вредност треба да буде краћа',
        tooShort              : 'Вредност треба да буде дужа',
        typeMismatch          : 'Потребно је да вредност буде у посебном формату',
        valueMissing          : 'Ово поље је обавезно',
        invalidValue          : 'Неважећа вредност поља',
        minimumValueViolation : 'Кршење ограничења минималне вредности',
        maximumValueViolation : 'Кршење ограничења максималне вредности',
        fieldRequired         : 'Ово поље је обавезно',
        validateFilter        : 'Вредност мора да се изабере са листе'
    },

    DateField : {
        invalidDate : format => 'Неважећи унос датума, очекивани формат је ' + format
    },

    DatePicker : {
        gotoPrevYear  : 'Иди на претходну годину',
        gotoPrevMonth : 'Иди на претходни месец',
        gotoNextMonth : 'Иди на следећи месец',
        gotoNextYear  : 'Иди на следећу годину'
    },

    Hint : {
        dontShowAgain : 'Не приказуј поново'
    },

    NumberFormat : {
        locale   : 'sr-RS',
        currency : 'RSD'
    },

    DurationField : {
        invalidUnit : 'Неважећа јединица'
    },

    TimeField : {
        invalidTime : 'Неважећи унос времена'
    },

    TimePicker : {
        hour   : 'Час',
        minute : 'Минут',
        second : 'Секунда'
    },

    List : {
        loading   : 'Учитавање...',
        selectAll : 'Изабери све'
    },

    GridBase : {
        loadMask : 'Учитавање...',
        syncMask : 'Чување промена, сачекајте...'
    },

    PagingToolbar : {
        firstPage         : 'Иди на прву страницу',
        prevPage          : 'Иди на претходну страницу',
        page              : 'Страница',
        nextPage          : 'Иди на следећу страницу',
        lastPage          : 'Иди на последњу страницу',
        reload            : 'Поново учитај тренутну страницу',
        noRecords         : 'Нема записа за приказивање',
        pageCountTemplate : data => `од ${data.lastPage}`,
        summaryTemplate   : data => `Приказ записа ${data.start} - ${data.end} oд ${data.allCount}`
    },

    PanelCollapser : {
        Collapse : 'Скупи',
        Expand   : 'Развиј'
    },

    Popup : {
        close : 'Затвори'
    },

    TabPanel : {
        containsInvalidFields : 'Таб садржи неважећа поља'
    },

    Toolbar : {
        showOverflowingItems : 'Прикажи преливене ставке',
        scrollForMore        : 'Померите за више'
    },

    UndoRedo : {
        Undo           : 'Опозови',
        Redo           : 'Понови',
        UndoLastAction : 'Опозови последњу радњу',
        RedoLastAction : 'Понови последњу поништену радњу',
        NoActions      : 'Нема ставки у реду чекања за опозив'
    },

    FieldFilterPicker : {
        equals                 : 'једнако',
        doesNotEqual           : 'није једнако',
        isEmpty                : 'празно',
        isNotEmpty             : 'није празно',
        contains               : 'садржи',
        doesNotContain         : 'не садржи',
        startsWith             : 'почиње на',
        endsWith               : 'завршава се са',
        isOneOf                : 'један од',
        isNotOneOf             : 'није један од',
        isGreaterThan          : 'веће је од',
        isLessThan             : 'мање је од',
        isGreaterThanOrEqualTo : 'веће је од или једнако',
        isLessThanOrEqualTo    : 'мање је од или је једнако',
        isBetween              : 'између',
        isNotBetween           : 'није између',
        isBefore               : 'пре',
        isAfter                : 'после',
        timeEquals             : 'време је једнако',
        timeDoesNotEqual       : 'време није једнако',
        isToday                : 'данас',
        isTomorrow             : 'сутра',
        isYesterday            : 'јуче',
        isThisWeek             : 'ове седмице',
        isNextWeek             : 'следећа седмица',
        isLastWeek             : 'прошле недеље',
        isThisMonth            : 'овог месеца',
        isNextMonth            : 'следећи месец',
        isLastMonth            : 'прошлог месеца',
        isThisYear             : 'ове године',
        isNextYear             : 'следећа година',
        isLastYear             : 'прошле године',
        isYearToDate           : 'ове године до данас',
        isTrue                 : 'тачно',
        isFalse                : 'нетачно',
        selectAProperty        : 'Изаберите својство',
        selectAnOperator       : 'Изаберите оператора',
        caseSensitive          : 'Разликује велика и мала слова',
        and                    : 'и',
        dateFormat             : 'Д.M.ГГ.',
        selectValue            : 'Изаберите вредност',
        selectOneOrMoreValues  : 'Изаберите једну или више вредности',
        enterAValue            : 'Унесите вредност',
        enterANumber           : 'Унесите број',
        selectADate            : 'Изаберите датум',
        selectATime            : 'Изаберите време'
    },

    FieldFilterPickerGroup : {
        addFilter : 'Додај филтер',
        applyCond : applied => applied ? 'Прескочи филтер (тренутно примењен)' : 'Примени филтер (тренутно прескочен)'
    },

    DateHelper : {
        locale         : 'sr-RS',
        weekStartDay   : 0,
        nonWorkingDays : {
            0 : true,
            6 : true
        },
        weekends : {
            0 : true,
            6 : true
        },
        unitNames : [
            { single : 'милисекунда', plural : 'мс', abbrev : 'мс' },
            { single : 'emillisecond', plural : 'ems', abbrev : 'ems' },
            { single : 'секунда', plural : 'секунде', abbrev : 'с' },
            { single : 'esecond', plural : 'eseconds', abbrev : 'es' },
            { single : 'минут', plural : 'минута', abbrev : 'мин' },
            { single : 'eminute', plural : 'eminutes', abbrev : 'emin' },
            { single : 'час', plural : 'часова', abbrev : 'ч' },
            { single : 'ehour', plural : 'ehours', abbrev : 'eh' },
            { single : 'дан', plural : 'дана', abbrev : 'д' },
            { single : 'eday', plural : 'edays', abbrev : 'ed' },
            { single : 'седмица', plural : 'седмице', abbrev : 'сед' },
            { single : 'eweek', plural : 'eweeks', abbrev : 'ew' },
            { single : 'месец', plural : 'месеци', abbrev : 'мес' },
            { single : 'emonth', plural : 'emonths', abbrev : 'emon' },
            { single : 'квартал', plural : 'квартала', abbrev : 'к' },
            { single : 'equarter', plural : 'equarters', abbrev : 'eq' },
            { single : 'година', plural : 'године', abbrev : 'год.' },
            { single : 'eyear', plural : 'eyears', abbrev : 'eyr' },
            { single : 'деценија', plural : 'деценије', abbrev : 'дец' },
            { single : 'edecade', plural : 'edecades', abbrev : 'edec' }
        ],
        unitAbbreviations : [
            ['мил'],
            [],
            ['с', 'сек'],
            [],
            ['м', 'мин'],
            [],
            ['ч', 'ч.'],
            [],
            ['д'],
            [],
            ['сед', 'сд.'],
            [],
            ['ме', 'мес', 'мсц'],
            [],
            ['к', 'кварт.', 'кврт.'],
            [],
            ['г', 'год.'],
            [],
            ['дец'],
            []
        ],
        parsers : {
            L   : 'ДД.MM.ГГГГ.',
            LT  : 'ЧЧ:mm A',
            LTS : 'ЧЧ:мм:сс A'
        },
        ordinalSuffix : number => {
            const hasSpecialCase = ['11', '12', '13'].find((n) => number.endsWith(n));

            let suffix = '.';

            if (!hasSpecialCase) {
                const lastDigit = number[number.length - 1];
                suffix = { 1 : '.', 2 : '.', 3 : '.' }[lastDigit] || '.';
            }

            return number + suffix;
        },
        decade : date => {
            const year = typeof date === 'number' ? date : date.getFullYear();
            return `${year - (year % 10)}-их`;
        }
    }
};

export default LocaleHelper.publishLocale(locale);
