// @bryntum/calendar/source/lib/Core/localization/Vi.js
import LocaleHelper from '../../Core/localization/LocaleHelper.js';

const locale = {

    localeName : 'Vi',
    localeDesc : 'Tiếng Việt',
    localeCode : 'vi',
    localeRtl  : false,

    Object : {
        Yes      : 'Có',
        No       : 'Không',
        Cancel   : 'Hủy bỏ ',
        Ok       : 'OK',
        Week     : 'Tuần',
        None     : 'Không',
        previous : 'Trước',
        next     : 'Tiếp theo',
        to       : 'đến',
        at       : 'tại',
        on       : 'trên',
        editing  : what => `Chỉnh sửa ${what}`,
        settings : 'Cài đặt',
        close    : 'Đóng',
        go       : 'Đi',
        save     : 'Lưu',
        cancel   : 'Hủy bỏ',
        revert   : 'Phục hồi'
    },

    CheckboxGroup : {
        wrongNbrOfOptionsSelected : min => `Chọn chính xác ${min}`,
        tooFewOptionsSelected     : min => `Chọn ít nhất ${min}`,
        tooManyOptionsSelected    : max => `Chọn không quá ${max}`
    },

    CodeEditor : {
        apply        : 'Áp dụng',
        autoApply    : 'Tự động áp dụng',
        downloadCode : 'Tải mã',
        editor       : 'Trình soạn thảo mã',
        viewer       : 'Trình xem mã'
    },

    ColorPicker : {
        noColor : 'Không màu'
    },

    Combo : {
        noResults          : 'Ko có kết quả',
        recordNotCommitted : 'Không thể thêm hồ sơ',
        addNewValue        : value => `Thêm ${value}`
    },

    FilePicker : {
        file : 'Tập tin'
    },

    Field : {
        badInput              : 'Giá trị trường không hợp lệ',
        patternMismatch       : 'Giá trị phải phù hợp với một mẫu cụ thể',
        rangeOverflow         : value => `Giá trị phải nhỏ hơn hoặc bằng ${value.max}`,
        rangeUnderflow        : value => `Giá trị phải lớn hơn hoặc bằng ${value.min}`,
        stepMismatch          : 'Giá trị phải phù hợp với bước',
        tooLong               : 'Giá trị phải ngắn hơn',
        tooShort              : 'Giá trị phải dài hơn',
        typeMismatch          : 'Giá trị bắt buộc phải ở định dạng đặc biệt',
        valueMissing          : 'Trường này là bắt buộc',
        invalidValue          : 'Giá trị trường không hợp lệ',
        minimumValueViolation : 'Vi phạm giá trị tối thiểu',
        maximumValueViolation : 'Vi phạm giá trị tối đa',
        fieldRequired         : 'Trường này là bắt buộc',
        validateFilter        : 'Giá trị phải được chọn từ danh sách'
    },

    DateField : {
        invalidDate : format => 'Đầu vào ngày không hợp lệ, định dạng mong đợi là ' + format
    },

    DatePicker : {
        gotoPrevYear  : 'Chuyển đến năm trước',
        gotoPrevMonth : 'Chuyển đến tháng trước',
        gotoNextMonth : 'Chuyển đến tháng sau',
        gotoNextYear  : 'Sang năm sau'
    },

    Hint : {
        dontShowAgain : 'Không hiển thị lại'
    },

    NumberFormat : {
        locale   : 'vi',
        currency : 'VND'
    },

    DurationField : {
        invalidUnit : 'Đơn vị không hợp lệ'
    },

    TimeField : {
        invalidTime : 'Đầu vào thời gian không hợp lệ'
    },

    TimePicker : {
        hour   : 'Giờ',
        minute : 'Phút',
        second : 'Thứ hai'
    },

    List : {
        loading   : 'Đang tải...',
        selectAll : 'Chọn Tất cả'
    },

    GridBase : {
        loadMask : 'Đang tải...',
        syncMask : 'Đang lưu các thay đổi, vui lòng đợi ...'
    },

    PagingToolbar : {
        firstPage         : 'Đi đến trang đầu tiên',
        prevPage          : 'Đi đến trang trước',
        page              : 'Trang',
        nextPage          : 'Đi đến trang tiếp theo',
        lastPage          : 'Đi đến trang cuối cùng',
        reload            : 'Tải lại trang hiện tại',
        noRecords         : 'Không có hồ sơ nào để hiển thị',
        pageCountTemplate : data => `/ ${data.lastPage}`,
        summaryTemplate   : data => `Hiển thị hồ sơ ${data.start} - ${data.end} of ${data.allCount}`
    },

    PanelCollapser : {
        Collapse : 'Thu gọn',
        Expand   : 'Mở rộng'
    },

    Popup : {
        close : 'Đóng cửa sổ bật lên'
    },

    TabPanel : {
        containsInvalidFields : 'Tab chứa các trường không hợp lệ'
    },

    Toolbar : {
        showOverflowingItems : 'Hiển thị các mục tràn',
        scrollForMore        : 'Cuộn để xem thêm'
    },

    UndoRedo : {
        Undo           : 'Hoàn tác',
        Redo           : 'Làm lại',
        UndoLastAction : 'Hoàn tác hành động cuối cùng',
        RedoLastAction : 'Làm lại hành động đã hoàn tác cuối cùng',
        NoActions      : 'Không có mục nào trong hàng đợi hoàn tác'
    },

    FieldFilterPicker : {
        equals                 : 'bằng',
        doesNotEqual           : 'không bằng',
        isEmpty                : 'trống rỗng',
        isNotEmpty             : 'không trống rỗng',
        contains               : 'chứa',
        doesNotContain         : 'không chứa',
        startsWith             : 'bắt đầu với',
        endsWith               : 'kết thúc với',
        isOneOf                : 'là một trong những',
        isNotOneOf             : 'không phải là một trong những',
        isGreaterThan          : 'lớn hơn',
        isLessThan             : 'nhỏ hơn',
        isGreaterThanOrEqualTo : 'lớn hơn hoặc bằng',
        isLessThanOrEqualTo    : 'nhỏ hơn hoặc bằng',
        isBetween              : 'ở giữa',
        isNotBetween           : 'không ở giữa',
        isBefore               : 'trước',
        isAfter                : 'sau',
        timeEquals             : 'thời gian bằng',
        timeDoesNotEqual       : 'thời gian không bằng',
        isToday                : 'hôm nay',
        isTomorrow             : 'ngày mai',
        isYesterday            : 'hôm qua',
        isThisWeek             : 'tuần này',
        isNextWeek             : 'tuần sau',
        isLastWeek             : 'tuần trước',
        isThisMonth            : 'tháng này',
        isNextMonth            : 'tháng sau',
        isLastMonth            : 'tháng trước',
        isThisYear             : 'năm nay',
        isNextYear             : 'năm sau',
        isLastYear             : 'năm trước',
        isYearToDate           : 'năm cho đến nay',
        isTrue                 : 'đúng',
        isFalse                : 'sai',
        selectAProperty        : 'Chọn một tài sản',
        selectAnOperator       : 'Chọn một nhà điều hành',
        caseSensitive          : 'Phân biệt chữ hoa chữ thường',
        and                    : 'và',
        dateFormat             : 'D/M/YY',
        selectValue            : 'Chọn một giá trị',
        selectOneOrMoreValues  : 'Chọn một hoặc nhiều giá trị',
        enterAValue            : 'Nhập một giá trị',
        enterANumber           : 'Nhập một số',
        selectADate            : 'Nhập một ngày',
        selectATime            : 'Chọn thời gian'
    },

    FieldFilterPickerGroup : {
        addFilter : 'Thêm bộ lọc',
        applyCond : applied => applied ? 'Bỏ qua bộ lọc (hiện đang áp dụng)' : 'Áp dụng bộ lọc (hiện đang bỏ qua)'
    },

    DateHelper : {
        locale         : 'vi',
        weekStartDay   : 1,
        nonWorkingDays : {
            0 : true,
            6 : true
        },
        weekends : {
            0 : true,
            6 : true
        },
        unitNames : [
            { single : 'milligiây', plural : 'ms', abbrev : 'ms' },
            { single : 'emillisecond', plural : 'ems', abbrev : 'ems' },
            { single : 'giây', plural : 'giây', abbrev : 'sec' },
            { single : 'esecond', plural : 'eseconds', abbrev : 'es' },
            { single : 'phút', plural : 'phút', abbrev : 'min' },
            { single : 'eminute', plural : 'eminutes', abbrev : 'emin' },
            { single : 'giờ', plural : 'giờ', abbrev : 'h' },
            { single : 'ehour', plural : 'ehours', abbrev : 'eh' },
            { single : 'ngày', plural : 'ngày', abbrev : 'd' },
            { single : 'eday', plural : 'edays', abbrev : 'ed' },
            { single : 'tuần', plural : 'tuần', abbrev : 'w' },
            { single : 'eweek', plural : 'eweeks', abbrev : 'ew' },
            { single : 'tháng', plural : 'tháng', abbrev : 'mon' },
            { single : 'emonth', plural : 'emonths', abbrev : 'emon' },
            { single : 'quý', plural : 'quý', abbrev : 'q' },
            { single : 'equarter', plural : 'equarters', abbrev : 'eq' },
            { single : 'năm', plural : 'năm', abbrev : 'yr' },
            { single : 'eyear', plural : 'eyears', abbrev : 'eyr' },
            { single : 'thậpkỷ', plural : 'thậpkỷ', abbrev : 'dec' },
            { single : 'edecade', plural : 'edecades', abbrev : 'edec' }
        ],
        unitAbbreviations : [
            ['mil'],
            [],
            ['s', 'sec'],
            [],
            ['m', 'min'],
            [],
            ['h', 'hr'],
            [],
            ['d'],
            [],
            ['w', 'wk'],
            [],
            ['mo', 'mon', 'mnt'],
            [],
            ['q', 'quar', 'qrt'],
            [],
            ['y', 'yr'],
            [],
            ['dec'],
            []
        ],
        parsers : {
            L   : 'DD/MM/YYYY',
            LT  : 'HH:mm',
            LTS : 'HH:mm:ss A'
        },
        ordinalSuffix : number => number,
        decade        : date => {
            const year = typeof date === 'number' ? date : date.getFullYear();
            return `thập niên ${year - (year % 10)}`;
        }
    }
};

export default LocaleHelper.publishLocale(locale);
