// @bryntum/calendar/source/lib/Core/localization/ZhCn.js
import LocaleHelper from '../../Core/localization/LocaleHelper.js';

const locale = {

    localeName : 'ZhCn',
    localeDesc : '中文（中国）',
    localeCode : 'zh-CN',
    localeRtl  : false,

    Object : {
        Yes      : '是',
        No       : '否',
        Cancel   : '取消',
        Ok       : '好',
        Week     : '周',
        None     : '无',
        previous : '上一个',
        next     : '下一个',
        to       : '到',
        at       : '在',
        on       : '在',
        editing  : what => `编辑 ${what}`,
        settings : '设置',
        close    : '关闭',
        go       : '去',
        save     : '保存',
        cancel   : '取消',
        revert   : '还原'
    },

    CheckboxGroup : {
        wrongNbrOfOptionsSelected : min => `请选择正好 ${min}`,
        tooFewOptionsSelected     : min => `请至少选择 ${min}`,
        tooManyOptionsSelected    : max => `请最多选择 ${max}`
    },

    CodeEditor : {
        apply        : '应用',
        autoApply    : '自动应用',
        downloadCode : '下载代码',
        editor       : '代码编辑器',
        viewer       : '代码查看器'
    },

    ColorPicker : {
        noColor : '无色'
    },

    Combo : {
        noResults          : '无结果',
        recordNotCommitted : '无法添加记录',
        addNewValue        : value => `添加 ${value}`
    },

    FilePicker : {
        file : '文件'
    },

    Field : {
        badInput              : '字段值无效',
        patternMismatch       : '值应与特定模式相匹配',
        rangeOverflow         : value => `值必须小于或等于 ${value.max}`,
        rangeUnderflow        : value => `值必须大于或等于 ${value.min}`,
        stepMismatch          : '值应符合步骤',
        tooLong               : '值应更短',
        tooShort              : '值应更长',
        typeMismatch          : '值要采用特殊格式',
        valueMissing          : '该字段为必填',
        invalidValue          : '字段值无效',
        minimumValueViolation : '不符合最小值限制',
        maximumValueViolation : '不符合最大值限制',
        fieldRequired         : '该字段为必填',
        validateFilter        : '必须从列表中选择值'
    },

    DateField : {
        invalidDate : format => '日期输入无效，预期格式为 ' + format
    },

    DatePicker : {
        gotoPrevYear  : '转至上一年',
        gotoPrevMonth : '转至上一月',
        gotoNextMonth : '转至下一月',
        gotoNextYear  : '转至下一年'
    },

    Hint : {
        dontShowAgain : '不再显示'
    },

    NumberFormat : {
        locale   : 'zh-CN',
        currency : 'CNY'
    },

    DurationField : {
        invalidUnit : '单位无效'
    },

    TimeField : {
        invalidTime : '时间输入无效'
    },

    TimePicker : {
        hour   : '时',
        minute : '分',
        second : '秒'
    },

    List : {
        loading   : '加载中……',
        selectAll : '全选'
    },

    GridBase : {
        loadMask : '加载中……',
        syncMask : '正在保存变更，请稍等……'
    },

    PagingToolbar : {
        firstPage         : '转至第一页',
        prevPage          : '转至上一页',
        page              : '页',
        nextPage          : '转至下一页',
        lastPage          : '转至最后一页',
        reload            : '重新载入当前页面',
        noRecords         : '无记录显示',
        pageCountTemplate : data => `的 ${data.lastPage}`,
        summaryTemplate   : data => `显示记录 ${data.start} - ${data.end} 的 ${data.allCount}`
    },

    PanelCollapser : {
        Collapse : '折叠',
        Expand   : '展开'
    },

    Popup : {
        close : '关闭弹窗'
    },

    TabPanel : {
        containsInvalidFields : '选项卡包含无效字段'
    },

    Toolbar : {
        showOverflowingItems : '显示溢出项',
        scrollForMore        : '滚动查看更多'
    },

    UndoRedo : {
        Undo           : '撤销',
        Redo           : '恢复',
        UndoLastAction : '撤销上个操作',
        RedoLastAction : '恢复上个撤销的操作',
        NoActions      : '撤销队列中没有项目'
    },

    FieldFilterPicker : {
        equals                 : '等于',
        doesNotEqual           : '不等于',
        isEmpty                : '为空',
        isNotEmpty             : '不为空',
        contains               : '包含',
        doesNotContain         : '不包含',
        startsWith             : '开始为',
        endsWith               : '结束为',
        isOneOf                : '是之一',
        isNotOneOf             : '不是之一',
        isGreaterThan          : '大于',
        isLessThan             : '小于',
        isGreaterThanOrEqualTo : '大于或等于',
        isLessThanOrEqualTo    : '小于或等于',
        isBetween              : '在之间',
        isNotBetween           : '不在之间',
        isBefore               : '在之前',
        isAfter                : '在之后',
        timeEquals             : '时间等于',
        timeDoesNotEqual       : '时间不等于',
        isToday                : '在今天',
        isTomorrow             : '在明天',
        isYesterday            : '在昨天',
        isThisWeek             : '在本周',
        isNextWeek             : '在下周',
        isLastWeek             : '在上周',
        isThisMonth            : '在本月',
        isNextMonth            : '在下月',
        isLastMonth            : '在上月',
        isThisYear             : '在今年',
        isNextYear             : '在明年',
        isLastYear             : '在去年',
        isYearToDate           : '是年初至今',
        isTrue                 : '为真',
        isFalse                : '为假',
        selectAProperty        : '选择一项属性',
        selectAnOperator       : '选择操作员',
        caseSensitive          : '区分大小写',
        and                    : '和',
        dateFormat             : '年/月/日',
        selectValue            : '选择一个值',
        selectOneOrMoreValues  : '选择一个或多个值',
        enterAValue            : '输入一个值',
        enterANumber           : '输入一个数字',
        selectADate            : '选择一个日期',
        selectATime            : '选择时间'
    },

    FieldFilterPickerGroup : {
        addFilter : '增加过滤条件',
        applyCond : applied => applied ? '跳过筛选器（当前已应用）' : '应用筛选器（当前已跳过）'
    },

    DateHelper : {
        locale         : 'zh-CN',
        weekStartDay   : 0,
        nonWorkingDays : {
            0 : true,
            6 : true
        },
        weekends : {
            0 : true,
            6 : true
        },
        unitNames : [
            { single : '毫秒', plural : '毫秒', abbrev : 'ms' },
            { single : 'emillisecond', plural : 'ems', abbrev : 'ems' },
            { single : '秒', plural : '秒', abbrev : 's' },
            { single : 'esecond', plural : 'eseconds', abbrev : 'es' },
            { single : '分', plural : '分', abbrev : 'min' },
            { single : 'eminute', plural : 'eminutes', abbrev : 'emin' },
            { single : '时', plural : '时', abbrev : 'h' },
            { single : 'ehour', plural : 'ehours', abbrev : 'eh' },
            { single : '天', plural : '天', abbrev : 'd' },
            { single : 'eday', plural : 'edays', abbrev : 'ed' },
            { single : '周', plural : '周', abbrev : 'w' },
            { single : 'eweek', plural : 'eweeks', abbrev : 'ew' },
            { single : '月', plural : '月', abbrev : 'mon' },
            { single : 'emonth', plural : 'emonths', abbrev : 'emon' },
            { single : '季', plural : '季', abbrev : 'q' },
            { single : 'equarter', plural : 'equarters', abbrev : 'eq' },
            { single : '年', plural : '年', abbrev : 'yr' },
            { single : 'eyear', plural : 'eyears', abbrev : 'eyr' },
            { single : '十年', plural : '十年', abbrev : 'dec' },
            { single : 'edecade', plural : 'edecades', abbrev : 'edec' }
        ],
        unitAbbreviations : [
            ['毫秒'],
            [],
            ['s', '秒'],
            [],
            ['m', '分'],
            [],
            ['h', '时'],
            [],
            ['天'],
            [],
            ['w', '周'],
            [],
            ['mo', '月', '月'],
            [],
            ['q', '季', '季'],
            [],
            ['y', '年'],
            [],
            ['十年'],
            []
        ],
        parsers : {
            L   : 'YYYY-MM-DD',
            LT  : 'HH:mm',
            LTS : 'HH:mm:ss A'
        },
        ordinalSuffix : number => '第' + number,
        decade        : date => {
            const year = typeof date === 'number' ? date : date.getFullYear();
            return `${year - (year % 10)}年代`;
        }
    }
};

export default LocaleHelper.publishLocale(locale);
