// @bryntum/calendar-vue-3/lib/chunks/B2mW7dzm.js
import { defineComponent as m, ref as d, provide as w, onMounted as R, getCurrentInstance as b, toRaw as h, watch as F, onBeforeUnmount as S, createElementBlock as u, openBlock as a, createElementVNode as E, Fragment as D, renderList as M, createBlock as c, Teleport as y, resolveDynamicComponent as x, mergeProps as v, toHandlers as k } from "vue";
import { Grid as B } from "@bryntum/calendar";
import { WrapperHelper as r } from "../helper/WrapperHelper.js";
const H = {
  ref: "teleportsHolder",
  style: { display: "none" }
}, G = /* @__PURE__ */ m({
  __name: "BryntumGrid",
  props: {
    adopt: {},
    alignSelf: {},
    animateFilterRemovals: {},
    animateRemovingRows: {},
    animateTreeNodeToggle: {},
    appendTo: {},
    ariaDescription: {},
    ariaLabel: {},
    ariaLive: {},
    autoHeight: {},
    bbar: {},
    bodyCls: {},
    bubbleEvents: {},
    callOnFunctions: {},
    cellEllipsis: {},
    cls: {},
    collapsed: {},
    collapsible: {},
    columnLines: {},
    columns: {},
    config: {},
    contentElementCls: {},
    contextMenuTriggerEvent: {},
    data: {},
    dataset: {},
    defaultRegion: {},
    destroyStore: {},
    disabled: {},
    disableGridColumnIdWarning: {},
    disableGridRowModelWarning: {},
    dock: {},
    drawer: {},
    elementAttributes: {},
    emptyText: {},
    enableSticky: {},
    enableTextSelection: {},
    enableUndoRedoKeys: {},
    extraData: {},
    fillLastColumn: {},
    fixedRowHeight: {},
    flex: {},
    footer: {},
    formulaProviders: {},
    fullRowRefresh: {},
    getRowHeight: { type: Function },
    header: {},
    height: {},
    hidden: {},
    hideFooters: {},
    hideHeaders: {},
    hideHorizontalScrollbar: {},
    hoverCls: {},
    id: {},
    ignoreParentReadOnly: {},
    insertBefore: {},
    insertFirst: {},
    keyMap: {},
    listeners: {},
    loadMask: {},
    loadMaskDefaults: {},
    loadMaskError: {},
    localizable: {},
    longPressTime: {},
    margin: {},
    maskDefaults: {},
    masked: {},
    maxHeight: {},
    maxWidth: {},
    minHeight: {},
    minWidth: {},
    monitorResize: {},
    owner: {},
    plugins: {},
    preserveFocusOnDatasetChange: {},
    preserveScroll: {},
    preserveScrollOnDatasetChange: {},
    preventTooltipOnTouch: {},
    readOnly: {},
    relayStoreEvents: {},
    resizable: {},
    resizeToFitIncludesHeader: {},
    responsiveLevels: {},
    ripple: {},
    rootElement: {},
    rowHeight: {},
    rowLines: {},
    rtl: {},
    scrollable: {},
    scrollerClass: {},
    scrollManager: {},
    selectionMode: {},
    showDirty: {},
    stateful: {},
    statefulEvents: {},
    stateId: {},
    stateProvider: {},
    stateSettings: {},
    store: {},
    strips: {},
    subGridConfigs: {},
    syncMask: {},
    tab: {},
    tabBarItems: {},
    tbar: {},
    title: {},
    tools: {},
    transitionDuration: {},
    type: {},
    ui: {},
    weight: {},
    width: {},
    cellCopyPasteFeature: {},
    cellEditFeature: {},
    cellMenuFeature: {},
    cellTooltipFeature: {},
    columnAutoWidthFeature: {},
    columnDragToolbarFeature: {},
    columnPickerFeature: {},
    columnRenameFeature: {},
    columnReorderFeature: {},
    columnResizeFeature: {},
    excelExporterFeature: {},
    fileDropFeature: {},
    fillHandleFeature: {},
    filterFeature: {},
    filterBarFeature: {},
    groupFeature: {},
    groupSummaryFeature: {},
    headerMenuFeature: {},
    lockRowsFeature: {},
    mergeCellsFeature: {},
    pdfExportFeature: {},
    printFeature: {},
    quickFindFeature: {},
    regionResizeFeature: {},
    rowCopyPasteFeature: {},
    rowEditFeature: {},
    rowExpanderFeature: {},
    rowReorderFeature: {},
    rowResizeFeature: {},
    searchFeature: {},
    sortFeature: {},
    splitFeature: {},
    stickyCellsFeature: {},
    stripeFeature: {},
    summaryFeature: {},
    treeFeature: {},
    treeGroupFeature: {}
  },
  emits: ["beforeCancelCellEdit", "beforeCancelRowEdit", "beforeCellDelete", "beforeCellEditStart", "beforeColumnDragStart", "beforeColumnDropFinalize", "beforeColumnResize", "beforeCopy", "beforeCSVExport", "beforeDestroy", "beforeExcelExport", "beforeFillHandleDragStart", "beforeFinishCellEdit", "beforeFinishRowEdit", "beforeHide", "beforePaste", "beforePdfExport", "beforeRenderRow", "beforeRenderRows", "beforeRowCollapse", "beforeRowExpand", "beforeSelectionChange", "beforeSetRecord", "beforeShow", "beforeStartRowEdit", "beforeStateApply", "beforeStateSave", "beforeToggleGroup", "beforeToggleNode", "cancelCellEdit", "catchAll", "cellClick", "cellContextMenu", "cellDblClick", "cellMenuBeforeShow", "cellMenuItem", "cellMenuShow", "cellMenuToggleItem", "cellMouseEnter", "cellMouseLeave", "cellMouseOut", "cellMouseOver", "collapse", "collapseNode", "columnDrag", "columnDragStart", "columnDrop", "columnResize", "columnResizeStart", "contextMenuItem", "contextMenuToggleItem", "copy", "dataChange", "destroy", "dirtyStateChange", "dragSelecting", "expand", "expandNode", "fileDrop", "fillHandleBeforeDragFinalize", "fillHandleDrag", "fillHandleDragAbort", "fillHandleDragEnd", "fillHandleDragStart", "finishCellEdit", "finishRowEdit", "focusIn", "focusOut", "gridRowBeforeDragStart", "gridRowBeforeDropFinalize", "gridRowDrag", "gridRowDragAbort", "gridRowDragStart", "gridRowDrop", "headerClick", "headerMenuBeforeShow", "headerMenuItem", "headerMenuShow", "headerMenuToggleItem", "hide", "lockRows", "mouseOut", "mouseOver", "paint", "paste", "pdfExport", "readOnly", "recompose", "renderRow", "renderRows", "resize", "responsive", "rowCollapse", "rowExpand", "rowMouseEnter", "rowMouseLeave", "scroll", "selectionChange", "selectionModeChange", "show", "split", "splitterCollapseClick", "splitterDragEnd", "splitterDragStart", "splitterExpandClick", "startCellEdit", "startRowEdit", "subGridCollapse", "subGridExpand", "toggleGroup", "toggleNode", "toolClick", "unlockRows", "unsplit"],
  setup(T, { expose: p, emit: g }) {
    const f = g, s = {
      instanceClass: B,
      instanceName: "Grid",
      isView: !0,
      dataStores: {
        originalStore: "originals",
        store: "data"
      },
      projectStores: !1,
      featureNames: [
        "cellCopyPasteFeature",
        "cellEditFeature",
        "cellMenuFeature",
        "cellTooltipFeature",
        "columnAutoWidthFeature",
        "columnDragToolbarFeature",
        "columnPickerFeature",
        "columnRenameFeature",
        "columnReorderFeature",
        "columnResizeFeature",
        "excelExporterFeature",
        "fileDropFeature",
        "fillHandleFeature",
        "filterFeature",
        "filterBarFeature",
        "groupFeature",
        "groupSummaryFeature",
        "headerMenuFeature",
        "lockRowsFeature",
        "mergeCellsFeature",
        "pdfExportFeature",
        "printFeature",
        "quickFindFeature",
        "regionResizeFeature",
        "rowCopyPasteFeature",
        "rowEditFeature",
        "rowExpanderFeature",
        "rowReorderFeature",
        "rowResizeFeature",
        "searchFeature",
        "sortFeature",
        "splitFeature",
        "stickyCellsFeature",
        "stripeFeature",
        "summaryFeature",
        "treeFeature",
        "treeGroupFeature"
      ],
      configNames: [
        "adopt",
        "animateFilterRemovals",
        "animateRemovingRows",
        "ariaDescription",
        "ariaLabel",
        "autoHeight",
        "bbar",
        "bodyCls",
        "bubbleEvents",
        "callOnFunctions",
        "collapsible",
        "config",
        "contentElementCls",
        "contextMenuTriggerEvent",
        "defaultRegion",
        "destroyStore",
        "disableGridColumnIdWarning",
        "disableGridRowModelWarning",
        "dock",
        "drawer",
        "elementAttributes",
        "enableSticky",
        "enableTextSelection",
        "fillLastColumn",
        "fixedRowHeight",
        "footer",
        "formulaProviders",
        "fullRowRefresh",
        "getRowHeight",
        "header",
        "hideHorizontalScrollbar",
        "hoverCls",
        "ignoreParentReadOnly",
        "listeners",
        "loadMask",
        "loadMaskDefaults",
        "loadMaskError",
        "localizable",
        "maskDefaults",
        "masked",
        "monitorResize",
        "owner",
        "plugins",
        "preserveFocusOnDatasetChange",
        "preserveScrollOnDatasetChange",
        "preventTooltipOnTouch",
        "relayStoreEvents",
        "resizable",
        "resizeToFitIncludesHeader",
        "responsiveLevels",
        "ripple",
        "rootElement",
        "scrollerClass",
        "scrollManager",
        "showDirty",
        "stateful",
        "statefulEvents",
        "stateId",
        "stateProvider",
        "strips",
        "subGridConfigs",
        "syncMask",
        "tab",
        "tabBarItems",
        "tbar",
        "type",
        "ui",
        "weight"
      ],
      propertyConfigNames: [
        "alignSelf",
        "animateTreeNodeToggle",
        "appendTo",
        "ariaLive",
        "cellEllipsis",
        "cls",
        "collapsed",
        "columnLines",
        "columns",
        "data",
        "dataset",
        "disabled",
        "emptyText",
        "enableUndoRedoKeys",
        "extraData",
        "flex",
        "height",
        "hidden",
        "hideFooters",
        "hideHeaders",
        "id",
        "insertBefore",
        "insertFirst",
        "keyMap",
        "longPressTime",
        "margin",
        "maxHeight",
        "maxWidth",
        "minHeight",
        "minWidth",
        "onBeforeCancelCellEdit",
        "onBeforeCancelRowEdit",
        "onBeforeCellDelete",
        "onBeforeCellEditStart",
        "onBeforeColumnDragStart",
        "onBeforeColumnDropFinalize",
        "onBeforeColumnResize",
        "onBeforeCopy",
        "onBeforeCSVExport",
        "onBeforeDestroy",
        "onBeforeExcelExport",
        "onBeforeFillHandleDragStart",
        "onBeforeFinishCellEdit",
        "onBeforeFinishRowEdit",
        "onBeforeHide",
        "onBeforePaste",
        "onBeforePdfExport",
        "onBeforeRenderRow",
        "onBeforeRenderRows",
        "onBeforeRowCollapse",
        "onBeforeRowExpand",
        "onBeforeSelectionChange",
        "onBeforeSetRecord",
        "onBeforeShow",
        "onBeforeStartRowEdit",
        "onBeforeStateApply",
        "onBeforeStateSave",
        "onBeforeToggleGroup",
        "onBeforeToggleNode",
        "onCancelCellEdit",
        "onCatchAll",
        "onCellClick",
        "onCellContextMenu",
        "onCellDblClick",
        "onCellMenuBeforeShow",
        "onCellMenuItem",
        "onCellMenuShow",
        "onCellMenuToggleItem",
        "onCellMouseEnter",
        "onCellMouseLeave",
        "onCellMouseOut",
        "onCellMouseOver",
        "onCollapse",
        "onCollapseNode",
        "onColumnDrag",
        "onColumnDragStart",
        "onColumnDrop",
        "onColumnResize",
        "onColumnResizeStart",
        "onContextMenuItem",
        "onContextMenuToggleItem",
        "onCopy",
        "onDataChange",
        "onDestroy",
        "onDirtyStateChange",
        "onDragSelecting",
        "onExpand",
        "onExpandNode",
        "onFileDrop",
        "onFillHandleBeforeDragFinalize",
        "onFillHandleDrag",
        "onFillHandleDragAbort",
        "onFillHandleDragEnd",
        "onFillHandleDragStart",
        "onFinishCellEdit",
        "onFinishRowEdit",
        "onFocusIn",
        "onFocusOut",
        "onGridRowBeforeDragStart",
        "onGridRowBeforeDropFinalize",
        "onGridRowDrag",
        "onGridRowDragAbort",
        "onGridRowDragStart",
        "onGridRowDrop",
        "onHeaderClick",
        "onHeaderMenuBeforeShow",
        "onHeaderMenuItem",
        "onHeaderMenuShow",
        "onHeaderMenuToggleItem",
        "onHide",
        "onLockRows",
        "onMouseOut",
        "onMouseOver",
        "onPaint",
        "onPaste",
        "onPdfExport",
        "onReadOnly",
        "onRecompose",
        "onRenderRow",
        "onRenderRows",
        "onResize",
        "onResponsive",
        "onRowCollapse",
        "onRowExpand",
        "onRowMouseEnter",
        "onRowMouseLeave",
        "onScroll",
        "onSelectionChange",
        "onSelectionModeChange",
        "onShow",
        "onSplit",
        "onSplitterCollapseClick",
        "onSplitterDragEnd",
        "onSplitterDragStart",
        "onSplitterExpandClick",
        "onStartCellEdit",
        "onStartRowEdit",
        "onSubGridCollapse",
        "onSubGridExpand",
        "onToggleGroup",
        "onToggleNode",
        "onToolClick",
        "onUnlockRows",
        "onUnsplit",
        "preserveScroll",
        "readOnly",
        "rowHeight",
        "rowLines",
        "rtl",
        "scrollable",
        "selectionMode",
        "stateSettings",
        "store",
        "title",
        "tools",
        "transitionDuration",
        "width"
      ],
      propertyNames: [
        "focusVisible",
        "hasChanges",
        "originalStore",
        "parent",
        "selectedCell",
        "selectedCells",
        "selectedRecord",
        "selectedRecords",
        "selectedRows",
        "state",
        "tooltip"
      ],
      eventNames: [
        "beforeCancelCellEdit",
        "beforeCancelRowEdit",
        "beforeCellDelete",
        "beforeCellEditStart",
        "beforeColumnDragStart",
        "beforeColumnDropFinalize",
        "beforeColumnResize",
        "beforeCopy",
        "beforeCSVExport",
        "beforeDestroy",
        "beforeExcelExport",
        "beforeFillHandleDragStart",
        "beforeFinishCellEdit",
        "beforeFinishRowEdit",
        "beforeHide",
        "beforePaste",
        "beforePdfExport",
        "beforeRenderRow",
        "beforeRenderRows",
        "beforeRowCollapse",
        "beforeRowExpand",
        "beforeSelectionChange",
        "beforeSetRecord",
        "beforeShow",
        "beforeStartRowEdit",
        "beforeStateApply",
        "beforeStateSave",
        "beforeToggleGroup",
        "beforeToggleNode",
        "cancelCellEdit",
        "catchAll",
        "cellClick",
        "cellContextMenu",
        "cellDblClick",
        "cellMenuBeforeShow",
        "cellMenuItem",
        "cellMenuShow",
        "cellMenuToggleItem",
        "cellMouseEnter",
        "cellMouseLeave",
        "cellMouseOut",
        "cellMouseOver",
        "collapse",
        "collapseNode",
        "columnDrag",
        "columnDragStart",
        "columnDrop",
        "columnResize",
        "columnResizeStart",
        "contextMenuItem",
        "contextMenuToggleItem",
        "copy",
        "dataChange",
        "destroy",
        "dirtyStateChange",
        "dragSelecting",
        "expand",
        "expandNode",
        "fileDrop",
        "fillHandleBeforeDragFinalize",
        "fillHandleDrag",
        "fillHandleDragAbort",
        "fillHandleDragEnd",
        "fillHandleDragStart",
        "finishCellEdit",
        "finishRowEdit",
        "focusIn",
        "focusOut",
        "gridRowBeforeDragStart",
        "gridRowBeforeDropFinalize",
        "gridRowDrag",
        "gridRowDragAbort",
        "gridRowDragStart",
        "gridRowDrop",
        "headerClick",
        "headerMenuBeforeShow",
        "headerMenuItem",
        "headerMenuShow",
        "headerMenuToggleItem",
        "hide",
        "lockRows",
        "mouseOut",
        "mouseOver",
        "paint",
        "paste",
        "pdfExport",
        "readOnly",
        "recompose",
        "renderRow",
        "renderRows",
        "resize",
        "responsive",
        "rowCollapse",
        "rowExpand",
        "rowMouseEnter",
        "rowMouseLeave",
        "scroll",
        "selectionChange",
        "selectionModeChange",
        "show",
        "split",
        "splitterCollapseClick",
        "splitterDragEnd",
        "splitterDragStart",
        "splitterExpandClick",
        "startCellEdit",
        "startRowEdit",
        "subGridCollapse",
        "subGridExpand",
        "toggleGroup",
        "toggleNode",
        "toolClick",
        "unlockRows",
        "unsplit"
      ]
    }, t = {}, n = d(), i = d(/* @__PURE__ */ new Map());
    return w("teleports", i), R(() => {
      const o = b();
      t.value = r.createWidget({
        me: o,
        widgetData: s,
        emit: f,
        element: n.value,
        processCellContent: r.processCellContent,
        processEventContent: r.processEventContent,
        processWidgetContent: r.processWidgetContent,
        hasFrameworkRenderer: r.hasFrameworkRenderer,
        toRaw: h
      }), r.relayStores(o.data, t.value);
      const l = (e, C) => F(() => o.props[e], C);
      r.watchProps(t.value, s, l);
    }), S(() => {
      var o, l;
      (l = (o = t.value) == null ? void 0 : o.destroy) == null || l.call(o);
    }), p({
      instance: t,
      refElement: n,
      teleports: i
    }), (o, l) => (a(), u("div", {
      ref_key: "refElement",
      ref: n
    }, [
      E("div", H, [
        (a(!0), u(D, null, M(i.value.values(), (e) => (a(), c(y, {
          key: e.key,
          to: e.to
        }, [
          (a(), c(x(e.is), v({
            ref_for: !0,
            ref: e.key,
            "data-key": e.key,
            "data-generation": e.generation
          }, e.bind, { "data-vue": "" }, k(e.on)), null, 16, ["data-key", "data-generation"]))
        ], 8, ["to"]))), 128))
      ], 512)
    ], 512));
  }
});
export {
  G as _
};
