// @bryntum/calendar-vue-3/lib/chunks/C4nZmdju.js
import { defineComponent as m, ref as i, provide as u, onMounted as f, getCurrentInstance as y, toRaw as g, watch as C, onBeforeUnmount as v, createElementBlock as w, openBlock as b } from "vue";
import { MonthView as S } from "@bryntum/calendar";
import { WrapperHelper as l } from "../helper/WrapperHelper.js";
const P = /* @__PURE__ */ m({
  __name: "BryntumMonthView",
  props: {
    activationKey: {},
    adopt: {},
    align: {},
    alignSelf: {},
    allowDragCreate: {},
    allowDragMove: {},
    allowDragResize: {},
    anchor: {},
    animateTimeShift: {},
    appendTo: {},
    ariaDescription: {},
    ariaLabel: {},
    ariaLive: {},
    autoCreate: {},
    autoRowHeight: {},
    bbar: {},
    bodyCls: {},
    bubbleEvents: {},
    callOnFunctions: {},
    cellRenderer: { type: [Function, String] },
    centered: {},
    cls: {},
    collapsed: {},
    collapsible: {},
    config: {},
    constrainTo: {},
    contentElementCls: {},
    dataset: {},
    date: {},
    dateFormat: {},
    dateSeparator: {},
    dayCellNameFormat: {},
    dayCellRenderer: { type: Function },
    dayNameFormat: {},
    defaultBindProperty: {},
    defaultFocus: { type: [Function, String] },
    defaults: {},
    defaultScrollOptions: {},
    descriptionFormat: {},
    descriptionRenderer: { type: Function },
    disabled: {},
    disabledCls: {},
    disabledDates: { type: [Function, String, Array] },
    disableNonWorkingDays: {},
    disableOtherMonthCells: {},
    disableWeekends: {},
    dock: {},
    draggable: {},
    dragUnit: {},
    drawer: {},
    elementAttributes: {},
    emptyCellRenderer: { type: [Function, Object, String] },
    eventFilter: { type: [Function, String] },
    eventHeaderRenderer: { type: [Function, String] },
    eventHeight: {},
    eventRenderer: { type: [Function, String] },
    eventSorter: { type: Function },
    eventSpacing: {},
    extendAllDayEndDay: {},
    extraData: {},
    filterEventResources: {},
    flex: {},
    floating: {},
    footer: {},
    getPrimaryResource: { type: [Function, String] },
    header: {},
    headerRenderer: { type: [Function, String] },
    height: {},
    hidden: {},
    hideAnimation: {},
    hideNonWorkingDays: {},
    hideOtherMonthCells: {},
    hideWhenEmpty: {},
    id: {},
    ignoreParentReadOnly: {},
    insertBefore: {},
    insertFirst: {},
    itemCls: {},
    items: {},
    keyMap: {},
    layout: {},
    layoutStyle: {},
    lazyItems: {},
    listeners: {},
    localeClass: {},
    localizable: {},
    localizableProperties: {},
    margin: {},
    maskDefaults: {},
    masked: {},
    maxDate: {},
    maxEventsPerCell: {},
    maxHeight: {},
    maximizeOnMobile: {},
    maxWidth: {},
    minColumnWidth: {},
    minDate: {},
    minHeight: {},
    minRowHeight: {},
    minWidth: {},
    monitorResize: {},
    month: {},
    namedItems: {},
    nonWorkingDayCls: {},
    nonWorkingDays: {},
    otherMonthCls: {},
    overflowButtonRenderer: { type: Function },
    overflowClickAction: {},
    overflowPopup: {},
    overflowPopupTrigger: {},
    owner: {},
    pastEventCls: {},
    persistShrinkWrappedRows: {},
    positioned: {},
    preventTooltipOnTouch: {},
    readOnly: {},
    relayStoreEvents: {},
    resourceImagePath: {},
    responsive: {},
    responsiveRoot: {},
    responsiveState: {},
    responsiveTarget: {},
    ripple: {},
    rootElement: {},
    rtl: {},
    scrollable: {},
    scrollAction: {},
    selectorButton: {},
    selectorMenuItem: {},
    shortDateFormat: {},
    shortDateTimeFormat: {},
    shortEventCls: {},
    shortEventDuration: {},
    showAnimation: {},
    showBullet: {},
    showResourceAvatars: {},
    showTooltipWhenDisabled: {},
    showWeekColumn: {},
    showWeekNumber: {},
    sixWeeks: {},
    stackMultiDayEvents: {},
    stateful: {},
    statefulEvents: {},
    stateId: {},
    stateProvider: {},
    strictRecordMapping: {},
    strips: {},
    syncCalendarDate: {},
    syncViewDate: {},
    tab: {},
    tabBarItems: {},
    tag: {},
    tbar: {},
    textAlign: {},
    timeFormat: {},
    tip: {},
    title: {},
    todayCls: {},
    tools: {},
    tooltip: {},
    trapFocus: {},
    type: {},
    ui: {},
    weekendCls: {},
    weekRenderer: { type: [Function, String] },
    weekStartDay: {},
    weight: {},
    width: {},
    x: {},
    y: {}
  },
  emits: ["beforeAutoCreate", "beforeChangeDate", "beforeDestroy", "beforeHide", "beforeRefresh", "beforeResponsiveStateChange", "beforeSetRecord", "beforeShow", "beforeShowOverflowPopup", "beforeStateApply", "beforeStateSave", "catchAll", "cellMapPopulated", "collapse", "dateChange", "dayCellPopulated", "destroy", "dirtyStateChange", "emptyCellClick", "eventAutoCreated", "eventPropagate", "expand", "focusIn", "focusOut", "hide", "paint", "readOnly", "recompose", "refresh", "resize", "responsiveStateChange", "show", "showOverflowPopup", "toolClick", "weekFlex", "weekShrinkwrap"],
  setup(D, { expose: s, emit: d }) {
    const p = d, r = {
      instanceClass: S,
      instanceName: "MonthView",
      configNames: [
        "activationKey",
        "adopt",
        "align",
        "anchor",
        "ariaDescription",
        "ariaLabel",
        "bbar",
        "bodyCls",
        "bubbleEvents",
        "callOnFunctions",
        "cellRenderer",
        "centered",
        "collapsible",
        "config",
        "constrainTo",
        "contentElementCls",
        "dayCellNameFormat",
        "dayCellRenderer",
        "dayNameFormat",
        "defaultBindProperty",
        "defaultFocus",
        "defaults",
        "descriptionRenderer",
        "disabledCls",
        "disabledDates",
        "disableNonWorkingDays",
        "disableOtherMonthCells",
        "disableWeekends",
        "dock",
        "draggable",
        "drawer",
        "elementAttributes",
        "emptyCellRenderer",
        "eventFilter",
        "eventHeaderRenderer",
        "eventRenderer",
        "eventSorter",
        "floating",
        "footer",
        "getPrimaryResource",
        "header",
        "headerRenderer",
        "hideAnimation",
        "hideOtherMonthCells",
        "hideWhenEmpty",
        "ignoreParentReadOnly",
        "itemCls",
        "lazyItems",
        "listeners",
        "localeClass",
        "localizable",
        "localizableProperties",
        "maskDefaults",
        "masked",
        "maxEventsPerCell",
        "minColumnWidth",
        "minRowHeight",
        "monitorResize",
        "month",
        "namedItems",
        "nonWorkingDayCls",
        "nonWorkingDays",
        "otherMonthCls",
        "overflowButtonRenderer",
        "overflowClickAction",
        "overflowPopup",
        "overflowPopupTrigger",
        "owner",
        "positioned",
        "preventTooltipOnTouch",
        "relayStoreEvents",
        "resourceImagePath",
        "responsive",
        "responsiveRoot",
        "responsiveState",
        "responsiveTarget",
        "ripple",
        "rootElement",
        "scrollAction",
        "selectorButton",
        "selectorMenuItem",
        "shortEventCls",
        "shortEventDuration",
        "showAnimation",
        "showBullet",
        "showResourceAvatars",
        "showTooltipWhenDisabled",
        "showWeekColumn",
        "sixWeeks",
        "stateful",
        "statefulEvents",
        "stateId",
        "stateProvider",
        "strips",
        "tab",
        "tabBarItems",
        "tag",
        "tbar",
        "textAlign",
        "timeFormat",
        "tip",
        "todayCls",
        "trapFocus",
        "type",
        "ui",
        "weekendCls",
        "weekRenderer",
        "weekStartDay",
        "weight"
      ],
      propertyConfigNames: [
        "alignSelf",
        "allowDragCreate",
        "allowDragMove",
        "allowDragResize",
        "animateTimeShift",
        "appendTo",
        "ariaLive",
        "autoCreate",
        "autoRowHeight",
        "cls",
        "collapsed",
        "dataset",
        "date",
        "dateFormat",
        "dateSeparator",
        "defaultScrollOptions",
        "descriptionFormat",
        "disabled",
        "dragUnit",
        "eventHeight",
        "eventSpacing",
        "extendAllDayEndDay",
        "extraData",
        "filterEventResources",
        "flex",
        "height",
        "hidden",
        "hideNonWorkingDays",
        "id",
        "insertBefore",
        "insertFirst",
        "items",
        "keyMap",
        "layout",
        "layoutStyle",
        "margin",
        "maxDate",
        "maxHeight",
        "maximizeOnMobile",
        "maxWidth",
        "minDate",
        "minHeight",
        "minWidth",
        "onBeforeAutoCreate",
        "onBeforeChangeDate",
        "onBeforeDestroy",
        "onBeforeHide",
        "onBeforeRefresh",
        "onBeforeResponsiveStateChange",
        "onBeforeSetRecord",
        "onBeforeShow",
        "onBeforeShowOverflowPopup",
        "onBeforeStateApply",
        "onBeforeStateSave",
        "onCatchAll",
        "onCellMapPopulated",
        "onCollapse",
        "onDateChange",
        "onDayCellPopulated",
        "onDestroy",
        "onDirtyStateChange",
        "onEmptyCellClick",
        "onEventAutoCreated",
        "onEventPropagate",
        "onExpand",
        "onFocusIn",
        "onFocusOut",
        "onHide",
        "onPaint",
        "onReadOnly",
        "onRecompose",
        "onRefresh",
        "onResize",
        "onResponsiveStateChange",
        "onShow",
        "onShowOverflowPopup",
        "onToolClick",
        "onWeekFlex",
        "onWeekShrinkwrap",
        "pastEventCls",
        "persistShrinkWrappedRows",
        "readOnly",
        "rtl",
        "scrollable",
        "shortDateFormat",
        "shortDateTimeFormat",
        "showWeekNumber",
        "stackMultiDayEvents",
        "strictRecordMapping",
        "syncCalendarDate",
        "syncViewDate",
        "title",
        "tools",
        "tooltip",
        "width",
        "x",
        "y"
      ],
      propertyNames: [
        "anchorSize",
        "dayCellCls",
        "firstVisibleCell",
        "firstVisibleDate",
        "focusVisible",
        "hasChanges",
        "isSettingValues",
        "isValid",
        "lastVisibleCell",
        "lastVisibleDate",
        "parent",
        "record",
        "state",
        "values"
      ],
      eventNames: [
        "beforeAutoCreate",
        "beforeChangeDate",
        "beforeDestroy",
        "beforeHide",
        "beforeRefresh",
        "beforeResponsiveStateChange",
        "beforeSetRecord",
        "beforeShow",
        "beforeShowOverflowPopup",
        "beforeStateApply",
        "beforeStateSave",
        "catchAll",
        "cellMapPopulated",
        "collapse",
        "dateChange",
        "dayCellPopulated",
        "destroy",
        "dirtyStateChange",
        "emptyCellClick",
        "eventAutoCreated",
        "eventPropagate",
        "expand",
        "focusIn",
        "focusOut",
        "hide",
        "paint",
        "readOnly",
        "recompose",
        "refresh",
        "resize",
        "responsiveStateChange",
        "show",
        "showOverflowPopup",
        "toolClick",
        "weekFlex",
        "weekShrinkwrap"
      ]
    }, o = {}, a = i(), n = i(/* @__PURE__ */ new Map());
    return u("teleports", n), f(() => {
      const e = y();
      o.value = l.createWidget({
        me: e,
        widgetData: r,
        emit: p,
        element: a.value,
        toRaw: g
      });
      const t = (c, h) => C(() => e.props[c], h);
      l.watchProps(o.value, r, t);
    }), v(() => {
      var e, t;
      (t = (e = o.value) == null ? void 0 : e.destroy) == null || t.call(e);
    }), s({
      instance: o,
      refElement: a,
      teleports: n
    }), (e, t) => (b(), w("div", {
      ref_key: "refElement",
      ref: a
    }, null, 512));
  }
});
export {
  P as _
};
