// @bryntum/calendar-vue-3/lib/chunks/CcSS6EBU.js
import { defineComponent as u, ref as i, provide as h, onMounted as f, getCurrentInstance as g, toRaw as y, watch as v, onBeforeUnmount as b, createElementBlock as C, openBlock as w } from "vue";
import { ResourceView as S } from "@bryntum/calendar";
import { WrapperHelper as s } from "../helper/WrapperHelper.js";
const F = /* @__PURE__ */ u({
  __name: "BryntumResourceView",
  props: {
    activationKey: {},
    adopt: {},
    align: {},
    alignSelf: {},
    allowDragCreate: {},
    allowDragMove: {},
    allowDragResize: {},
    anchor: {},
    animateTimeShift: {},
    appendTo: {},
    ariaDescription: {},
    ariaLabel: {},
    ariaLive: {},
    autoCreate: {},
    bbar: {},
    bodyCls: {},
    bubbleEvents: {},
    callOnFunctions: {},
    centered: {},
    cls: {},
    collapsed: {},
    collapsible: {},
    config: {},
    constrainTo: {},
    contentElementCls: {},
    dataset: {},
    dateFormat: {},
    dateSeparator: {},
    defaultBindProperty: {},
    defaultFocus: { type: [Function, String] },
    defaults: {},
    defaultScrollOptions: {},
    descriptionFormat: {},
    descriptionRenderer: { type: Function },
    disabled: {},
    dock: {},
    draggable: {},
    dragUnit: {},
    drawer: {},
    elementAttributes: {},
    eventHeaderRenderer: { type: [Function, String] },
    eventHeight: {},
    eventRenderer: { type: [Function, String] },
    eventSorter: { type: Function },
    extendAllDayEndDay: {},
    extraData: {},
    filterEventResources: {},
    flex: {},
    floating: {},
    footer: {},
    getPrimaryResource: { type: [Function, String] },
    header: {},
    height: {},
    hidden: {},
    hideAnimation: {},
    hideNonWorkingDays: {},
    hideWhenEmpty: {},
    id: {},
    ignoreParentReadOnly: {},
    includeTimeRanges: {},
    insertBefore: {},
    insertFirst: {},
    itemCls: {},
    items: {},
    keyMap: {},
    layout: {},
    layoutStyle: {},
    lazyItems: {},
    listeners: {},
    localeClass: {},
    localizable: {},
    localizableProperties: {},
    margin: {},
    maskDefaults: {},
    masked: {},
    maxDate: {},
    maxHeight: {},
    maximizeOnMobile: {},
    maxWidth: {},
    meta: { type: [String, Function] },
    minDate: {},
    minHeight: {},
    minWidth: {},
    monitorResize: {},
    namedItems: {},
    nonWorkingDays: {},
    owner: {},
    pastEventCls: {},
    positioned: {},
    preventTooltipOnTouch: {},
    range: {},
    readOnly: {},
    relayStoreEvents: {},
    resourceImagePath: {},
    resourceWidth: {},
    responsive: {},
    responsiveRoot: {},
    responsiveState: {},
    responsiveTarget: {},
    ripple: {},
    rootElement: {},
    rtl: {},
    scrollable: {},
    scrollAction: {},
    selectorButton: {},
    selectorMenuItem: {},
    shortDateFormat: {},
    shortDateTimeFormat: {},
    shortEventCls: {},
    shortEventDuration: {},
    showAnimation: {},
    showAvatars: {},
    showBullet: {},
    showResourceAvatars: {},
    showTooltipWhenDisabled: {},
    stableResourceOrder: {},
    stateful: {},
    statefulEvents: {},
    stateId: {},
    stateProvider: {},
    strictRecordMapping: {},
    strips: {},
    syncCalendarDate: {},
    syncViewDate: {},
    tab: {},
    tabBarItems: {},
    tag: {},
    tbar: {},
    textAlign: {},
    timeFormat: {},
    title: {},
    tools: {},
    tooltip: {},
    trapFocus: {},
    type: {},
    ui: {},
    view: {},
    viewGap: {},
    weekStartDay: {},
    weight: {},
    width: {},
    x: {},
    y: {}
  },
  emits: ["beforeAutoCreate", "beforeChangeDate", "beforeDestroy", "beforeHide", "beforeResponsiveStateChange", "beforeSetRecord", "beforeShow", "beforeStateApply", "beforeStateSave", "catchAll", "collapse", "destroy", "dirtyStateChange", "eventAutoCreated", "expand", "focusIn", "focusOut", "hide", "paint", "readOnly", "recompose", "refresh", "resize", "responsiveStateChange", "show", "toolClick", "viewCreate"],
  setup(D, { expose: l, emit: c }) {
    const d = c, r = {
      instanceClass: S,
      instanceName: "ResourceView",
      configNames: [
        "activationKey",
        "adopt",
        "align",
        "anchor",
        "ariaDescription",
        "ariaLabel",
        "bbar",
        "bodyCls",
        "bubbleEvents",
        "callOnFunctions",
        "centered",
        "collapsible",
        "config",
        "constrainTo",
        "contentElementCls",
        "defaultBindProperty",
        "defaultFocus",
        "defaults",
        "descriptionRenderer",
        "dock",
        "draggable",
        "drawer",
        "elementAttributes",
        "eventHeaderRenderer",
        "eventHeight",
        "eventRenderer",
        "eventSorter",
        "floating",
        "footer",
        "getPrimaryResource",
        "header",
        "hideAnimation",
        "hideWhenEmpty",
        "ignoreParentReadOnly",
        "includeTimeRanges",
        "itemCls",
        "lazyItems",
        "listeners",
        "localeClass",
        "localizable",
        "localizableProperties",
        "maskDefaults",
        "masked",
        "meta",
        "monitorResize",
        "namedItems",
        "nonWorkingDays",
        "owner",
        "positioned",
        "preventTooltipOnTouch",
        "relayStoreEvents",
        "resourceImagePath",
        "responsive",
        "responsiveRoot",
        "responsiveState",
        "responsiveTarget",
        "ripple",
        "rootElement",
        "scrollAction",
        "selectorButton",
        "selectorMenuItem",
        "shortEventCls",
        "shortEventDuration",
        "showAnimation",
        "showAvatars",
        "showBullet",
        "showResourceAvatars",
        "showTooltipWhenDisabled",
        "stableResourceOrder",
        "stateful",
        "statefulEvents",
        "stateId",
        "stateProvider",
        "strips",
        "tab",
        "tabBarItems",
        "tag",
        "tbar",
        "textAlign",
        "timeFormat",
        "trapFocus",
        "type",
        "ui",
        "weekStartDay",
        "weight"
      ],
      propertyConfigNames: [
        "alignSelf",
        "allowDragCreate",
        "allowDragMove",
        "allowDragResize",
        "animateTimeShift",
        "appendTo",
        "ariaLive",
        "autoCreate",
        "cls",
        "collapsed",
        "dataset",
        "dateFormat",
        "dateSeparator",
        "defaultScrollOptions",
        "descriptionFormat",
        "disabled",
        "dragUnit",
        "extendAllDayEndDay",
        "extraData",
        "filterEventResources",
        "flex",
        "height",
        "hidden",
        "hideNonWorkingDays",
        "id",
        "insertBefore",
        "insertFirst",
        "items",
        "keyMap",
        "layout",
        "layoutStyle",
        "margin",
        "maxDate",
        "maxHeight",
        "maximizeOnMobile",
        "maxWidth",
        "minDate",
        "minHeight",
        "minWidth",
        "onBeforeAutoCreate",
        "onBeforeChangeDate",
        "onBeforeDestroy",
        "onBeforeHide",
        "onBeforeResponsiveStateChange",
        "onBeforeSetRecord",
        "onBeforeShow",
        "onBeforeStateApply",
        "onBeforeStateSave",
        "onCatchAll",
        "onCollapse",
        "onDestroy",
        "onDirtyStateChange",
        "onEventAutoCreated",
        "onExpand",
        "onFocusIn",
        "onFocusOut",
        "onHide",
        "onPaint",
        "onReadOnly",
        "onRecompose",
        "onRefresh",
        "onResize",
        "onResponsiveStateChange",
        "onShow",
        "onToolClick",
        "onViewCreate",
        "pastEventCls",
        "range",
        "readOnly",
        "resourceWidth",
        "rtl",
        "scrollable",
        "shortDateFormat",
        "shortDateTimeFormat",
        "strictRecordMapping",
        "syncCalendarDate",
        "syncViewDate",
        "title",
        "tools",
        "tooltip",
        "view",
        "viewGap",
        "width",
        "x",
        "y"
      ],
      propertyNames: [
        "anchorSize",
        "dayCellCls",
        "firstVisibleCell",
        "firstVisibleDate",
        "focusVisible",
        "hasChanges",
        "isSettingValues",
        "isValid",
        "lastVisibleCell",
        "lastVisibleDate",
        "parent",
        "record",
        "state",
        "values"
      ],
      eventNames: [
        "beforeAutoCreate",
        "beforeChangeDate",
        "beforeDestroy",
        "beforeHide",
        "beforeResponsiveStateChange",
        "beforeSetRecord",
        "beforeShow",
        "beforeStateApply",
        "beforeStateSave",
        "catchAll",
        "collapse",
        "destroy",
        "dirtyStateChange",
        "eventAutoCreated",
        "expand",
        "focusIn",
        "focusOut",
        "hide",
        "paint",
        "readOnly",
        "recompose",
        "refresh",
        "resize",
        "responsiveStateChange",
        "show",
        "toolClick",
        "viewCreate"
      ]
    }, o = {}, a = i(), n = i(/* @__PURE__ */ new Map());
    return h("teleports", n), f(() => {
      const e = g();
      o.value = s.createWidget({
        me: e,
        widgetData: r,
        emit: d,
        element: a.value,
        toRaw: y
      });
      const t = (p, m) => v(() => e.props[p], m);
      s.watchProps(o.value, r, t);
    }), b(() => {
      var e, t;
      (t = (e = o.value) == null ? void 0 : e.destroy) == null || t.call(e);
    }), l({
      instance: o,
      refElement: a,
      teleports: n
    }), (e, t) => (w(), C("div", {
      ref_key: "refElement",
      ref: a
    }, null, 512));
  }
});
export {
  F as _
};
