// @bryntum/calendar-vue-3/lib/chunks/CujPRVnk.js
import { defineComponent as h, ref as l, provide as u, onMounted as b, getCurrentInstance as g, toRaw as f, watch as y, onBeforeUnmount as w, createElementBlock as x, openBlock as A } from "vue";
import { ViewPresetCombo as O } from "@bryntum/calendar";
import { WrapperHelper as r } from "../helper/WrapperHelper.js";
const T = /* @__PURE__ */ h({
  __name: "BryntumViewPresetCombo",
  props: {
    align: {},
    alignSelf: {},
    anchor: {},
    appendTo: {},
    ariaDescription: {},
    ariaLabel: {},
    ariaLive: {},
    badge: {},
    bubbleEvents: {},
    callOnFunctions: {},
    centered: {},
    clearTextOnSelection: {},
    clearWhenInputEmpty: {},
    client: {},
    cls: {},
    config: {},
    constrainTo: {},
    contentElementCls: {},
    dataField: {},
    defaultBindProperty: {},
    disabled: {},
    dock: {},
    draggable: {},
    editable: {},
    elementAttributes: {},
    extraData: {},
    flex: {},
    floating: {},
    height: {},
    hidden: {},
    hideAnimation: {},
    hidePickerOnSelect: {},
    hint: { type: [String, Function] },
    hintHtml: { type: [String, Function] },
    id: {},
    ignoreParentReadOnly: {},
    inputAlign: {},
    inputAttributes: {},
    inputTag: {},
    inputWidth: {},
    insertBefore: {},
    insertFirst: {},
    keyMap: {},
    label: {},
    labelCls: {},
    labelPosition: {},
    labels: {},
    labelWidth: {},
    listCls: {},
    listeners: {},
    listItemTpl: { type: Function },
    localeClass: {},
    localizable: {},
    localizableProperties: {},
    margin: {},
    maskDefaults: {},
    masked: {},
    maxHeight: {},
    maximizeOnMobile: {},
    maxWidth: {},
    minHeight: {},
    minWidth: {},
    monitorResize: {},
    name: {},
    overlayAnchor: {},
    owner: {},
    pickerWidth: {},
    positioned: {},
    presets: {},
    preventTooltipOnTouch: {},
    readOnly: {},
    relayStoreEvents: {},
    ripple: {},
    rootElement: {},
    rtl: {},
    scrollAction: {},
    showAnimation: {},
    showRequiredIndicator: {},
    showTooltipWhenDisabled: {},
    skipValidation: {},
    tab: {},
    tabIndex: {},
    textAlign: {},
    type: {},
    ui: {},
    useFixedDuration: {},
    weight: {},
    width: {},
    x: {},
    y: {}
  },
  emits: ["action", "beforeDestroy", "beforeHide", "beforeShow", "catchAll", "change", "clear", "destroy", "focusIn", "focusOut", "hide", "paint", "readOnly", "recompose", "resize", "select", "show", "trigger"],
  setup(k, { expose: s, emit: c }) {
    const d = c, o = {
      instanceClass: O,
      instanceName: "ViewPresetCombo",
      configNames: [
        "align",
        "anchor",
        "ariaDescription",
        "ariaLabel",
        "bubbleEvents",
        "callOnFunctions",
        "centered",
        "clearTextOnSelection",
        "clearWhenInputEmpty",
        "client",
        "config",
        "constrainTo",
        "contentElementCls",
        "dataField",
        "defaultBindProperty",
        "dock",
        "draggable",
        "elementAttributes",
        "floating",
        "hideAnimation",
        "hidePickerOnSelect",
        "hint",
        "hintHtml",
        "ignoreParentReadOnly",
        "inputAlign",
        "inputAttributes",
        "inputTag",
        "inputWidth",
        "labelCls",
        "labelPosition",
        "labels",
        "labelWidth",
        "listCls",
        "listeners",
        "listItemTpl",
        "localeClass",
        "localizable",
        "localizableProperties",
        "maskDefaults",
        "masked",
        "monitorResize",
        "name",
        "overlayAnchor",
        "owner",
        "pickerWidth",
        "positioned",
        "presets",
        "preventTooltipOnTouch",
        "relayStoreEvents",
        "ripple",
        "rootElement",
        "scrollAction",
        "showAnimation",
        "showTooltipWhenDisabled",
        "skipValidation",
        "tab",
        "tabIndex",
        "textAlign",
        "type",
        "ui",
        "useFixedDuration",
        "weight"
      ],
      propertyConfigNames: [
        "alignSelf",
        "appendTo",
        "ariaLive",
        "badge",
        "cls",
        "disabled",
        "editable",
        "extraData",
        "flex",
        "height",
        "hidden",
        "id",
        "insertBefore",
        "insertFirst",
        "keyMap",
        "label",
        "margin",
        "maxHeight",
        "maximizeOnMobile",
        "maxWidth",
        "minHeight",
        "minWidth",
        "onAction",
        "onBeforeDestroy",
        "onBeforeHide",
        "onBeforeShow",
        "onCatchAll",
        "onChange",
        "onClear",
        "onDestroy",
        "onFocusIn",
        "onFocusOut",
        "onHide",
        "onPaint",
        "onReadOnly",
        "onRecompose",
        "onResize",
        "onSelect",
        "onShow",
        "onTrigger",
        "readOnly",
        "rtl",
        "showRequiredIndicator",
        "width",
        "x",
        "y"
      ],
      propertyNames: [
        "anchorSize",
        "focusVisible",
        "formula",
        "input",
        "multiSelect",
        "parent",
        "picker",
        "placeholder",
        "required",
        "tooltip",
        "value"
      ],
      eventNames: [
        "action",
        "beforeDestroy",
        "beforeHide",
        "beforeShow",
        "catchAll",
        "change",
        "clear",
        "destroy",
        "focusIn",
        "focusOut",
        "hide",
        "paint",
        "readOnly",
        "recompose",
        "resize",
        "select",
        "show",
        "trigger"
      ]
    }, i = {}, n = l(), a = l(/* @__PURE__ */ new Map());
    return u("teleports", a), b(() => {
      const e = g();
      i.value = r.createWidget({
        me: e,
        widgetData: o,
        emit: d,
        element: n.value,
        toRaw: f
      });
      const t = (p, m) => y(() => e.props[p], m);
      r.watchProps(i.value, o, t);
    }), w(() => {
      var e, t;
      (t = (e = i.value) == null ? void 0 : e.destroy) == null || t.call(e);
    }), s({
      instance: i,
      refElement: n,
      teleports: a
    }), (e, t) => (A(), x("div", {
      ref_key: "refElement",
      ref: n
    }, null, 512));
  }
});
export {
  T as _
};
