// @bryntum/calendar-vue-3/lib/chunks/D0lMC7WA.js
import { defineComponent as p, ref as i, provide as u, onMounted as g, getCurrentInstance as f, toRaw as y, watch as b, onBeforeUnmount as C, createElementBlock as D, openBlock as v } from "vue";
import { MonthAgendaView as S } from "@bryntum/calendar";
import { WrapperHelper as s } from "../helper/WrapperHelper.js";
const A = /* @__PURE__ */ p({
  __name: "BryntumMonthAgendaView",
  props: {
    activationKey: {},
    activeDate: {},
    adopt: {},
    agenda: {},
    agendaDateFormat: {},
    agendHeaderRenderer: { type: [Function, String] },
    align: {},
    alignSelf: {},
    allowDragCreate: {},
    allowDragMove: {},
    allowDragResize: {},
    anchor: {},
    animateTimeShift: {},
    appendTo: {},
    ariaDescription: {},
    ariaLabel: {},
    ariaLive: {},
    autoCreate: {},
    bbar: {},
    bodyCls: {},
    bubbleEvents: {},
    callOnFunctions: {},
    cellRenderer: { type: [Function, String] },
    centered: {},
    cls: {},
    collapsed: {},
    collapsible: {},
    config: {},
    constrainTo: {},
    contentElementCls: {},
    dataset: {},
    date: {},
    dateFormat: {},
    dateSeparator: {},
    dayNameFormat: {},
    defaultBindProperty: {},
    defaultFocus: { type: [Function, String] },
    defaults: {},
    defaultScrollOptions: {},
    descriptionFormat: {},
    descriptionRenderer: { type: Function },
    disabled: {},
    disabledCls: {},
    disabledDates: { type: [Function, String, Array] },
    disableNonWorkingDays: {},
    disableOtherMonthCells: {},
    disableWeekends: {},
    dock: {},
    draggable: {},
    dragUnit: {},
    drawer: {},
    editMonth: {},
    elementAttributes: {},
    eventCountTip: {},
    eventDots: {},
    eventFilter: { type: [Function, String] },
    eventHeaderRenderer: { type: [Function, String] },
    eventRenderer: { type: [Function, String] },
    eventSorter: { type: Function },
    eventStore: {},
    extendAllDayEndDay: {},
    extraData: {},
    filterEventResources: {},
    flex: {},
    floating: {},
    focusDisabledDates: {},
    footer: {},
    getPrimaryResource: { type: [Function, String] },
    header: {},
    headerRenderer: { type: [Function, String] },
    height: {},
    hidden: {},
    hideAnimation: {},
    hideNonWorkingDays: {},
    hideOtherMonthCells: {},
    hideWhenEmpty: {},
    highlightSelectedWeek: {},
    id: {},
    ignoreParentReadOnly: {},
    includeYear: {},
    insertBefore: {},
    insertFirst: {},
    itemCls: {},
    items: {},
    keyMap: {},
    layout: {},
    layoutStyle: {},
    lazyItems: {},
    listeners: {},
    localeClass: {},
    localizable: {},
    localizableProperties: {},
    margin: {},
    maskDefaults: {},
    masked: {},
    maxDate: {},
    maxDots: {},
    maxHeight: {},
    maximizeOnMobile: {},
    maxWidth: {},
    minColumnWidth: {},
    minDate: {},
    minHeight: {},
    minRowHeight: {},
    minWidth: {},
    monitorResize: {},
    month: {},
    multiSelect: {},
    namedItems: {},
    nonWorkingDayCls: {},
    nonWorkingDays: {},
    otherMonthCls: {},
    owner: {},
    pastEventCls: {},
    positioned: {},
    preventTooltipOnTouch: {},
    readOnly: {},
    relayStoreEvents: {},
    resourceImagePath: {},
    ripple: {},
    rootElement: {},
    rtl: {},
    scrollable: {},
    scrollAction: {},
    selection: {},
    selectorButton: {},
    selectorMenuItem: {},
    shiftIncrement: {},
    shortDateFormat: {},
    shortDateTimeFormat: {},
    shortEventCls: {},
    showAnimation: {},
    showBullet: {},
    showEvents: {},
    showResourceAvatars: {},
    showTooltipWhenDisabled: {},
    showWeekColumn: {},
    sixWeeks: {},
    stateful: {},
    statefulEvents: {},
    stateId: {},
    stateProvider: {},
    strictRecordMapping: {},
    strips: {},
    syncCalendarDate: {},
    syncViewDate: {},
    tab: {},
    tabBarItems: {},
    tag: {},
    tbar: {},
    textAlign: {},
    timeFormat: {},
    tip: {},
    title: {},
    todayCls: {},
    tools: {},
    tooltip: {},
    trapFocus: {},
    type: {},
    ui: {},
    weekendCls: {},
    weekRenderer: { type: [Function, String] },
    weekStartDay: {},
    weight: {},
    width: {},
    x: {},
    y: {}
  },
  emits: ["beforeAutoCreate", "beforeChangeDate", "beforeDestroy", "beforeHide", "beforeRefresh", "beforeSetRecord", "beforeShow", "beforeStateApply", "beforeStateSave", "catchAll", "collapse", "dateChange", "destroy", "dirtyStateChange", "eventAutoCreated", "expand", "focusIn", "focusOut", "hide", "paint", "readOnly", "recompose", "refresh", "resize", "selectionChange", "show", "toolClick"],
  setup(w, { expose: l, emit: d }) {
    const c = d, n = {
      instanceClass: S,
      instanceName: "MonthAgendaView",
      configNames: [
        "activationKey",
        "activeDate",
        "adopt",
        "agendaDateFormat",
        "agendHeaderRenderer",
        "align",
        "anchor",
        "ariaDescription",
        "ariaLabel",
        "bbar",
        "bodyCls",
        "bubbleEvents",
        "callOnFunctions",
        "cellRenderer",
        "centered",
        "collapsible",
        "config",
        "constrainTo",
        "contentElementCls",
        "dayNameFormat",
        "defaultBindProperty",
        "defaultFocus",
        "defaults",
        "descriptionRenderer",
        "disabledCls",
        "disabledDates",
        "disableNonWorkingDays",
        "disableOtherMonthCells",
        "disableWeekends",
        "dock",
        "draggable",
        "drawer",
        "editMonth",
        "elementAttributes",
        "eventFilter",
        "eventHeaderRenderer",
        "eventRenderer",
        "eventSorter",
        "eventStore",
        "floating",
        "focusDisabledDates",
        "footer",
        "getPrimaryResource",
        "header",
        "headerRenderer",
        "hideAnimation",
        "hideOtherMonthCells",
        "hideWhenEmpty",
        "ignoreParentReadOnly",
        "includeYear",
        "itemCls",
        "lazyItems",
        "listeners",
        "localeClass",
        "localizable",
        "localizableProperties",
        "maskDefaults",
        "masked",
        "minColumnWidth",
        "minRowHeight",
        "monitorResize",
        "month",
        "multiSelect",
        "namedItems",
        "nonWorkingDayCls",
        "nonWorkingDays",
        "otherMonthCls",
        "owner",
        "positioned",
        "preventTooltipOnTouch",
        "relayStoreEvents",
        "resourceImagePath",
        "ripple",
        "rootElement",
        "scrollAction",
        "selectorButton",
        "selectorMenuItem",
        "shortEventCls",
        "showAnimation",
        "showBullet",
        "showResourceAvatars",
        "showTooltipWhenDisabled",
        "showWeekColumn",
        "sixWeeks",
        "stateful",
        "statefulEvents",
        "stateId",
        "stateProvider",
        "strips",
        "tab",
        "tabBarItems",
        "tag",
        "tbar",
        "textAlign",
        "timeFormat",
        "tip",
        "todayCls",
        "trapFocus",
        "type",
        "ui",
        "weekendCls",
        "weekRenderer",
        "weekStartDay",
        "weight"
      ],
      propertyConfigNames: [
        "agenda",
        "alignSelf",
        "allowDragCreate",
        "allowDragMove",
        "allowDragResize",
        "animateTimeShift",
        "appendTo",
        "ariaLive",
        "autoCreate",
        "cls",
        "collapsed",
        "dataset",
        "date",
        "dateFormat",
        "dateSeparator",
        "defaultScrollOptions",
        "descriptionFormat",
        "disabled",
        "dragUnit",
        "eventCountTip",
        "eventDots",
        "extendAllDayEndDay",
        "extraData",
        "filterEventResources",
        "flex",
        "height",
        "hidden",
        "hideNonWorkingDays",
        "highlightSelectedWeek",
        "id",
        "insertBefore",
        "insertFirst",
        "items",
        "keyMap",
        "layout",
        "layoutStyle",
        "margin",
        "maxDate",
        "maxDots",
        "maxHeight",
        "maximizeOnMobile",
        "maxWidth",
        "minDate",
        "minHeight",
        "minWidth",
        "onBeforeAutoCreate",
        "onBeforeChangeDate",
        "onBeforeDestroy",
        "onBeforeHide",
        "onBeforeRefresh",
        "onBeforeSetRecord",
        "onBeforeShow",
        "onBeforeStateApply",
        "onBeforeStateSave",
        "onCatchAll",
        "onCollapse",
        "onDateChange",
        "onDestroy",
        "onDirtyStateChange",
        "onEventAutoCreated",
        "onExpand",
        "onFocusIn",
        "onFocusOut",
        "onHide",
        "onPaint",
        "onReadOnly",
        "onRecompose",
        "onRefresh",
        "onResize",
        "onSelectionChange",
        "onShow",
        "onToolClick",
        "pastEventCls",
        "readOnly",
        "rtl",
        "scrollable",
        "selection",
        "shiftIncrement",
        "shortDateFormat",
        "shortDateTimeFormat",
        "showEvents",
        "strictRecordMapping",
        "syncCalendarDate",
        "syncViewDate",
        "title",
        "tools",
        "tooltip",
        "width",
        "x",
        "y"
      ],
      propertyNames: [
        "anchorSize",
        "dayCellCls",
        "firstVisibleCell",
        "firstVisibleDate",
        "focusVisible",
        "hasChanges",
        "isSettingValues",
        "isValid",
        "lastVisibleCell",
        "lastVisibleDate",
        "parent",
        "record",
        "state",
        "values"
      ],
      eventNames: [
        "beforeAutoCreate",
        "beforeChangeDate",
        "beforeDestroy",
        "beforeHide",
        "beforeRefresh",
        "beforeSetRecord",
        "beforeShow",
        "beforeStateApply",
        "beforeStateSave",
        "catchAll",
        "collapse",
        "dateChange",
        "destroy",
        "dirtyStateChange",
        "eventAutoCreated",
        "expand",
        "focusIn",
        "focusOut",
        "hide",
        "paint",
        "readOnly",
        "recompose",
        "refresh",
        "resize",
        "selectionChange",
        "show",
        "toolClick"
      ]
    }, a = {}, o = i(), r = i(/* @__PURE__ */ new Map());
    return u("teleports", r), g(() => {
      const e = f();
      a.value = s.createWidget({
        me: e,
        widgetData: n,
        emit: c,
        element: o.value,
        toRaw: y
      });
      const t = (h, m) => b(() => e.props[h], m);
      s.watchProps(a.value, n, t);
    }), C(() => {
      var e, t;
      (t = (e = a.value) == null ? void 0 : e.destroy) == null || t.call(e);
    }), l({
      instance: a,
      refElement: o,
      teleports: r
    }), (e, t) => (v(), D("div", {
      ref_key: "refElement",
      ref: o
    }, null, 512));
  }
});
export {
  A as _
};
