// @bryntum/calendar-vue-3/lib/chunks/D8stjqVb.js
import { defineComponent as u, ref as l, provide as m, onMounted as g, getCurrentInstance as b, toRaw as f, watch as y, onBeforeUnmount as x, createElementBlock as k, openBlock as w } from "vue";
import { TextAreaPickerField as A } from "@bryntum/calendar";
import { WrapperHelper as r } from "../helper/WrapperHelper.js";
const T = /* @__PURE__ */ u({
  __name: "BryntumTextAreaPickerField",
  props: {
    adopt: {},
    align: {},
    alignSelf: {},
    anchor: {},
    appendTo: {},
    ariaDescription: {},
    ariaLabel: {},
    ariaLive: {},
    autoComplete: {},
    autoExpand: {},
    autoSelect: {},
    badge: {},
    bubbleEvents: {},
    callOnFunctions: {},
    centered: {},
    clearable: {},
    cls: {},
    config: {},
    constrainTo: {},
    container: {},
    containValues: { type: [String, Function] },
    contentElementCls: {},
    dataField: {},
    dataset: {},
    defaultBindProperty: {},
    disabled: {},
    dock: {},
    draggable: {},
    elementAttributes: {},
    extraData: {},
    flex: {},
    floating: {},
    height: {},
    hidden: {},
    hideAnimation: {},
    highlightExternalChange: {},
    hint: { type: [String, Function] },
    hintHtml: { type: [String, Function] },
    id: {},
    ignoreParentReadOnly: {},
    inline: {},
    inputAlign: {},
    inputAttributes: {},
    inputTag: {},
    inputType: {},
    inputWidth: {},
    insertBefore: {},
    insertFirst: {},
    keyMap: {},
    keyStrokeChangeDelay: {},
    label: {},
    labelCls: {},
    labelPosition: {},
    labels: {},
    labelWidth: {},
    listeners: {},
    localeClass: {},
    localizable: {},
    localizableProperties: {},
    margin: {},
    maskDefaults: {},
    masked: {},
    maxHeight: {},
    maximizeOnMobile: {},
    maxLength: {},
    maxWidth: {},
    minHeight: {},
    minLength: {},
    minWidth: {},
    monitorResize: {},
    name: {},
    owner: {},
    picker: {},
    pickerAlignElement: {},
    placeholder: {},
    positioned: {},
    preventTooltipOnTouch: {},
    readOnly: {},
    relayStoreEvents: {},
    required: {},
    resize: {},
    revertOnEscape: {},
    ripple: {},
    rootElement: {},
    rtl: {},
    scrollAction: {},
    showAnimation: {},
    showRequiredIndicator: {},
    showTooltipWhenDisabled: {},
    skipValidation: {},
    spellCheck: {},
    tab: {},
    tabIndex: {},
    textAlign: {},
    title: {},
    tooltip: {},
    triggers: {},
    type: {},
    ui: {},
    validateOnInput: {},
    value: {},
    weight: {},
    width: {},
    x: {},
    y: {}
  },
  emits: ["action", "beforeDestroy", "beforeHide", "beforeShow", "catchAll", "change", "clear", "destroy", "focusIn", "focusOut", "hide", "input", "paint", "readOnly", "recompose", "resize", "show", "trigger"],
  setup(v, { expose: s, emit: c }) {
    const p = c, a = {
      instanceClass: A,
      instanceName: "TextAreaPickerField",
      configNames: [
        "adopt",
        "align",
        "anchor",
        "ariaDescription",
        "ariaLabel",
        "autoComplete",
        "autoExpand",
        "autoSelect",
        "bubbleEvents",
        "callOnFunctions",
        "centered",
        "clearable",
        "config",
        "constrainTo",
        "container",
        "containValues",
        "contentElementCls",
        "dataField",
        "defaultBindProperty",
        "dock",
        "draggable",
        "elementAttributes",
        "floating",
        "hideAnimation",
        "highlightExternalChange",
        "hint",
        "hintHtml",
        "ignoreParentReadOnly",
        "inline",
        "inputAlign",
        "inputAttributes",
        "inputTag",
        "inputType",
        "inputWidth",
        "keyStrokeChangeDelay",
        "labelCls",
        "labelPosition",
        "labels",
        "labelWidth",
        "listeners",
        "localeClass",
        "localizable",
        "localizableProperties",
        "maskDefaults",
        "masked",
        "maxLength",
        "minLength",
        "monitorResize",
        "name",
        "owner",
        "picker",
        "pickerAlignElement",
        "positioned",
        "preventTooltipOnTouch",
        "relayStoreEvents",
        "resize",
        "revertOnEscape",
        "ripple",
        "rootElement",
        "scrollAction",
        "showAnimation",
        "showTooltipWhenDisabled",
        "skipValidation",
        "spellCheck",
        "tab",
        "tabIndex",
        "textAlign",
        "title",
        "type",
        "ui",
        "validateOnInput",
        "weight"
      ],
      propertyConfigNames: [
        "alignSelf",
        "appendTo",
        "ariaLive",
        "badge",
        "cls",
        "dataset",
        "disabled",
        "extraData",
        "flex",
        "height",
        "hidden",
        "id",
        "insertBefore",
        "insertFirst",
        "keyMap",
        "label",
        "margin",
        "maxHeight",
        "maximizeOnMobile",
        "maxWidth",
        "minHeight",
        "minWidth",
        "onAction",
        "onBeforeDestroy",
        "onBeforeHide",
        "onBeforeShow",
        "onCatchAll",
        "onChange",
        "onClear",
        "onDestroy",
        "onFocusIn",
        "onFocusOut",
        "onHide",
        "onInput",
        "onPaint",
        "onReadOnly",
        "onRecompose",
        "onResize",
        "onShow",
        "onTrigger",
        "placeholder",
        "readOnly",
        "required",
        "rtl",
        "showRequiredIndicator",
        "tooltip",
        "triggers",
        "value",
        "width",
        "x",
        "y"
      ],
      propertyNames: [
        "anchorSize",
        "content",
        "editable",
        "focusVisible",
        "formula",
        "html",
        "input",
        "parent",
        "scrollable"
      ],
      eventNames: [
        "action",
        "beforeDestroy",
        "beforeHide",
        "beforeShow",
        "catchAll",
        "change",
        "clear",
        "destroy",
        "focusIn",
        "focusOut",
        "hide",
        "input",
        "paint",
        "readOnly",
        "recompose",
        "resize",
        "show",
        "trigger"
      ]
    }, n = {}, i = l(), o = l(/* @__PURE__ */ new Map());
    return m("teleports", o), g(() => {
      const e = b();
      n.value = r.createWidget({
        me: e,
        widgetData: a,
        emit: p,
        element: i.value,
        toRaw: f
      });
      const t = (d, h) => y(() => e.props[d], h);
      r.watchProps(n.value, a, t);
    }), x(() => {
      var e, t;
      (t = (e = n.value) == null ? void 0 : e.destroy) == null || t.call(e);
    }), s({
      instance: n,
      refElement: i,
      teleports: o
    }), (e, t) => (w(), k("div", {
      ref_key: "refElement",
      ref: i
    }, null, 512));
  }
});
export {
  T as _
};
