// @bryntum/calendar-vue-3/lib/chunks/DSNZLJ-_.js
import { defineComponent as h, ref as r, provide as u, onMounted as g, getCurrentInstance as f, toRaw as b, watch as y, onBeforeUnmount as w, createElementBlock as x, openBlock as O } from "vue";
import { Slider as v } from "@bryntum/calendar";
import { WrapperHelper as l } from "../helper/WrapperHelper.js";
const z = /* @__PURE__ */ h({
  __name: "BryntumSlider",
  props: {
    adopt: {},
    align: {},
    alignSelf: {},
    anchor: {},
    appendTo: {},
    ariaDescription: {},
    ariaLabel: {},
    ariaLive: {},
    bubbleEvents: {},
    callOnFunctions: {},
    centered: {},
    cls: {},
    config: {},
    constrainTo: {},
    content: {},
    contentElementCls: {},
    dataset: {},
    defaultBindProperty: {},
    disabled: {},
    dock: {},
    draggable: {},
    elementAttributes: {},
    extraData: {},
    flex: {},
    floating: {},
    height: {},
    hidden: {},
    hideAnimation: {},
    html: { type: [String, Function, Array] },
    htmlCls: {},
    id: {},
    ignoreParentReadOnly: {},
    insertBefore: {},
    insertFirst: {},
    keyMap: {},
    label: {},
    listeners: {},
    localeClass: {},
    localizable: {},
    localizableProperties: {},
    margin: {},
    maskDefaults: {},
    masked: {},
    max: {},
    maxHeight: {},
    maximizeOnMobile: {},
    maxWidth: {},
    min: {},
    minHeight: {},
    minWidth: {},
    monitorResize: {},
    owner: {},
    positioned: {},
    preventTooltipOnTouch: {},
    readOnly: {},
    relayStoreEvents: {},
    ripple: {},
    rootElement: {},
    rtl: {},
    scrollable: {},
    scrollAction: {},
    showAnimation: {},
    showTooltip: {},
    showTooltipWhenDisabled: {},
    showValue: {},
    step: {},
    tab: {},
    tag: {},
    text: {},
    textAlign: {},
    title: {},
    tooltip: {},
    triggerChangeOnInput: {},
    type: {},
    ui: {},
    unit: {},
    value: {},
    weight: {},
    width: {},
    x: {},
    y: {}
  },
  emits: ["beforeDestroy", "beforeHide", "beforeShow", "catchAll", "change", "destroy", "focusIn", "focusOut", "hide", "input", "paint", "readOnly", "recompose", "resize", "show"],
  setup(A, { expose: s, emit: c }) {
    const p = c, i = {
      instanceClass: v,
      instanceName: "Slider",
      configNames: [
        "adopt",
        "align",
        "anchor",
        "ariaDescription",
        "ariaLabel",
        "bubbleEvents",
        "callOnFunctions",
        "centered",
        "config",
        "constrainTo",
        "contentElementCls",
        "defaultBindProperty",
        "dock",
        "draggable",
        "elementAttributes",
        "floating",
        "hideAnimation",
        "htmlCls",
        "ignoreParentReadOnly",
        "label",
        "listeners",
        "localeClass",
        "localizable",
        "localizableProperties",
        "maskDefaults",
        "masked",
        "monitorResize",
        "owner",
        "positioned",
        "preventTooltipOnTouch",
        "relayStoreEvents",
        "ripple",
        "rootElement",
        "scrollAction",
        "showAnimation",
        "showTooltip",
        "showTooltipWhenDisabled",
        "showValue",
        "tab",
        "tag",
        "textAlign",
        "title",
        "type",
        "ui",
        "unit",
        "weight"
      ],
      propertyConfigNames: [
        "alignSelf",
        "appendTo",
        "ariaLive",
        "cls",
        "content",
        "dataset",
        "disabled",
        "extraData",
        "flex",
        "height",
        "hidden",
        "html",
        "id",
        "insertBefore",
        "insertFirst",
        "keyMap",
        "margin",
        "max",
        "maxHeight",
        "maximizeOnMobile",
        "maxWidth",
        "min",
        "minHeight",
        "minWidth",
        "onBeforeDestroy",
        "onBeforeHide",
        "onBeforeShow",
        "onCatchAll",
        "onChange",
        "onDestroy",
        "onFocusIn",
        "onFocusOut",
        "onHide",
        "onInput",
        "onPaint",
        "onReadOnly",
        "onRecompose",
        "onResize",
        "onShow",
        "readOnly",
        "rtl",
        "scrollable",
        "step",
        "text",
        "tooltip",
        "triggerChangeOnInput",
        "value",
        "width",
        "x",
        "y"
      ],
      propertyNames: [
        "anchorSize",
        "focusVisible",
        "parent"
      ],
      eventNames: [
        "beforeDestroy",
        "beforeHide",
        "beforeShow",
        "catchAll",
        "change",
        "destroy",
        "focusIn",
        "focusOut",
        "hide",
        "input",
        "paint",
        "readOnly",
        "recompose",
        "resize",
        "show"
      ]
    }, o = {}, n = r(), a = r(/* @__PURE__ */ new Map());
    return u("teleports", a), g(() => {
      const e = f();
      o.value = l.createWidget({
        me: e,
        widgetData: i,
        emit: p,
        element: n.value,
        toRaw: b
      });
      const t = (d, m) => y(() => e.props[d], m);
      l.watchProps(o.value, i, t);
    }), w(() => {
      var e, t;
      (t = (e = o.value) == null ? void 0 : e.destroy) == null || t.call(e);
    }), s({
      instance: o,
      refElement: n,
      teleports: a
    }), (e, t) => (O(), x("div", {
      ref_key: "refElement",
      ref: n
    }, null, 512));
  }
});
export {
  z as _
};
