// @bryntum/calendar-vue-3/lib/chunks/DzgX-pfP.js
import { defineComponent as g, ref as a, provide as h, onMounted as u, getCurrentInstance as f, toRaw as b, watch as y, onBeforeUnmount as w, createElementBlock as x, openBlock as v } from "vue";
import { Button as A } from "@bryntum/calendar";
import { WrapperHelper as l } from "../helper/WrapperHelper.js";
const T = /* @__PURE__ */ g({
  __name: "BryntumButton",
  props: {
    adopt: {},
    align: {},
    alignSelf: {},
    anchor: {},
    appendTo: {},
    ariaDescription: {},
    ariaLabel: {},
    ariaLive: {},
    badge: {},
    behaviorType: {},
    bubbleEvents: {},
    callOnFunctions: {},
    centered: {},
    cls: {},
    color: {},
    config: {},
    constrainTo: {},
    content: {},
    contentElementCls: {},
    dataset: {},
    defaultBindProperty: {},
    disabled: {},
    dock: {},
    draggable: {},
    elementAttributes: {},
    extraData: {},
    flex: {},
    floating: {},
    height: {},
    hidden: {},
    hideAnimation: {},
    href: {},
    html: { type: [String, Function, Array] },
    htmlCls: {},
    icon: {},
    iconAlign: {},
    id: {},
    ignoreParentReadOnly: {},
    insertBefore: {},
    insertFirst: {},
    keyMap: {},
    listeners: {},
    localeClass: {},
    localizable: {},
    localizableProperties: {},
    margin: {},
    maskDefaults: {},
    masked: {},
    maxHeight: {},
    maximizeOnMobile: {},
    maxWidth: {},
    menu: {},
    menuIcon: {},
    minHeight: {},
    minWidth: {},
    monitorResize: {},
    owner: {},
    positioned: {},
    pressed: {},
    pressedCls: {},
    pressedIcon: {},
    preventTooltipOnTouch: {},
    readOnly: {},
    relayStoreEvents: {},
    ripple: {},
    rootElement: {},
    rtl: {},
    scrollable: {},
    scrollAction: {},
    showAnimation: {},
    showTooltipWhenDisabled: {},
    split: {},
    supportsPressedClick: {},
    tab: {},
    tabIndex: {},
    tag: {},
    target: {},
    text: {},
    textAlign: {},
    title: {},
    toggleable: {},
    toggleGroup: {},
    tooltip: {},
    type: {},
    ui: {},
    weight: {},
    width: {},
    x: {},
    y: {}
  },
  emits: ["action", "beforeDestroy", "beforeHide", "beforeShow", "beforeShowMenu", "beforeToggle", "catchAll", "click", "destroy", "focusIn", "focusOut", "hide", "paint", "readOnly", "recompose", "resize", "show", "toggle"],
  setup(k, { expose: s, emit: c }) {
    const p = c, i = {
      instanceClass: A,
      instanceName: "Button",
      configNames: [
        "adopt",
        "align",
        "anchor",
        "ariaDescription",
        "ariaLabel",
        "bubbleEvents",
        "callOnFunctions",
        "centered",
        "config",
        "constrainTo",
        "contentElementCls",
        "defaultBindProperty",
        "dock",
        "draggable",
        "elementAttributes",
        "floating",
        "hideAnimation",
        "htmlCls",
        "ignoreParentReadOnly",
        "listeners",
        "localeClass",
        "localizable",
        "localizableProperties",
        "maskDefaults",
        "masked",
        "monitorResize",
        "owner",
        "positioned",
        "preventTooltipOnTouch",
        "relayStoreEvents",
        "ripple",
        "rootElement",
        "scrollAction",
        "showAnimation",
        "showTooltipWhenDisabled",
        "supportsPressedClick",
        "tab",
        "tag",
        "textAlign",
        "title",
        "type",
        "ui",
        "weight"
      ],
      propertyConfigNames: [
        "alignSelf",
        "appendTo",
        "ariaLive",
        "badge",
        "behaviorType",
        "cls",
        "color",
        "content",
        "dataset",
        "disabled",
        "extraData",
        "flex",
        "height",
        "hidden",
        "href",
        "html",
        "icon",
        "iconAlign",
        "id",
        "insertBefore",
        "insertFirst",
        "keyMap",
        "margin",
        "maxHeight",
        "maximizeOnMobile",
        "maxWidth",
        "menu",
        "menuIcon",
        "minHeight",
        "minWidth",
        "onAction",
        "onBeforeDestroy",
        "onBeforeHide",
        "onBeforeShow",
        "onBeforeShowMenu",
        "onBeforeToggle",
        "onCatchAll",
        "onClick",
        "onDestroy",
        "onFocusIn",
        "onFocusOut",
        "onHide",
        "onPaint",
        "onReadOnly",
        "onRecompose",
        "onResize",
        "onShow",
        "onToggle",
        "pressed",
        "pressedCls",
        "pressedIcon",
        "readOnly",
        "rtl",
        "scrollable",
        "split",
        "tabIndex",
        "target",
        "text",
        "toggleable",
        "toggleGroup",
        "tooltip",
        "width",
        "x",
        "y"
      ],
      propertyNames: [
        "anchorSize",
        "focusVisible",
        "parent"
      ],
      eventNames: [
        "action",
        "beforeDestroy",
        "beforeHide",
        "beforeShow",
        "beforeShowMenu",
        "beforeToggle",
        "catchAll",
        "click",
        "destroy",
        "focusIn",
        "focusOut",
        "hide",
        "paint",
        "readOnly",
        "recompose",
        "resize",
        "show",
        "toggle"
      ]
    }, t = {}, n = a(), r = a(/* @__PURE__ */ new Map());
    return h("teleports", r), u(() => {
      const e = f();
      t.value = l.createWidget({
        me: e,
        widgetData: i,
        emit: p,
        element: n.value,
        toRaw: b
      });
      const o = (d, m) => y(() => e.props[d], m);
      l.watchProps(t.value, i, o);
    }), w(() => {
      var e, o;
      (o = (e = t.value) == null ? void 0 : e.destroy) == null || o.call(e);
    }), s({
      instance: t,
      refElement: n,
      teleports: r
    }), (e, o) => (v(), x("div", {
      ref_key: "refElement",
      ref: n
    }, null, 512));
  }
});
export {
  T as _
};
