// @bryntum/calendar-vue-3/lib/chunks/YreYuYOq.js
import { defineComponent as p, ref as l, provide as f, onMounted as C, getCurrentInstance as m, toRaw as h, watch as w, onBeforeUnmount as R, createElementBlock as D, openBlock as S } from "vue";
import { AgendaView as b } from "@bryntum/calendar";
import { WrapperHelper as i } from "../helper/WrapperHelper.js";
const M = /* @__PURE__ */ p({
  __name: "BryntumAgendaView",
  props: {
    activationKey: {},
    adopt: {},
    alignSelf: {},
    allowDragCreate: {},
    allowDragMove: {},
    allowDragResize: {},
    alwaysShowCurrentDate: {},
    animateFilterRemovals: {},
    animateRemovingRows: {},
    animateTimeShift: {},
    animateTreeNodeToggle: {},
    appendTo: {},
    ariaDescription: {},
    ariaLabel: {},
    ariaLive: {},
    autoCreate: {},
    autoHeight: {},
    bbar: {},
    bodyCls: {},
    bubbleEvents: {},
    callOnFunctions: {},
    cellEllipsis: {},
    cls: {},
    collapsed: {},
    collapsible: {},
    columnLines: {},
    columns: {},
    config: {},
    contentElementCls: {},
    contextMenuTriggerEvent: {},
    data: {},
    dataset: {},
    date: {},
    dateFilter: { type: [Function, String] },
    dateFormat: {},
    dateSeparator: {},
    defaultRegion: {},
    defaultScrollOptions: {},
    descriptionFormat: {},
    descriptionRenderer: { type: Function },
    destroyStore: {},
    disabled: {},
    disableGridColumnIdWarning: {},
    disableGridRowModelWarning: {},
    dock: {},
    dragUnit: {},
    drawer: {},
    elementAttributes: {},
    emptyText: {},
    enableSticky: {},
    enableTextSelection: {},
    enableUndoRedoKeys: {},
    endDate: {},
    eventFilter: { type: [Function, String] },
    eventHeaderRenderer: { type: [Function, String] },
    eventHeight: {},
    eventListTimeFormat: {},
    eventRenderer: { type: [Function, String] },
    eventRowSpacing: {},
    eventSorter: { type: Function },
    eventTimeRenderer: { type: [Function, String] },
    extendAllDayEndDay: {},
    extraData: {},
    fillLastColumn: {},
    filterEventResources: {},
    fixedRowHeight: {},
    flex: {},
    footer: {},
    formulaProviders: {},
    fullRowRefresh: {},
    getPrimaryResource: { type: [Function, String] },
    getRowHeight: { type: Function },
    header: {},
    height: {},
    hidden: {},
    hideEventOverflow: {},
    hideFooters: {},
    hideHeaders: {},
    hideHorizontalScrollbar: {},
    hideNonWorkingDays: {},
    hoverCls: {},
    id: {},
    ignoreParentReadOnly: {},
    insertBefore: {},
    insertFirst: {},
    keyMap: {},
    listeners: {},
    listRangeMenu: {},
    loadMask: {},
    loadMaskDefaults: {},
    loadMaskError: {},
    localizable: {},
    longPressTime: {},
    margin: {},
    maskDefaults: {},
    masked: {},
    maxDate: {},
    maxHeight: {},
    maxWidth: {},
    minDate: {},
    minHeight: {},
    minWidth: {},
    monitorResize: {},
    nonWorkingDays: {},
    offsetStartsBeforeEvents: {},
    owner: {},
    pastEventCls: {},
    plugins: {},
    preserveFocusOnDatasetChange: {},
    preserveScroll: {},
    preserveScrollOnDatasetChange: {},
    preventTooltipOnTouch: {},
    range: {},
    readOnly: {},
    relayStoreEvents: {},
    resizable: {},
    resizeToFitIncludesHeader: {},
    resourceImagePath: {},
    responsive: {},
    responsiveLevels: {},
    responsiveRoot: {},
    responsiveState: {},
    responsiveTarget: {},
    ripple: {},
    rootElement: {},
    rowHeight: {},
    rowLines: {},
    rtl: {},
    scrollable: {},
    scrollerClass: {},
    scrollManager: {},
    selectionMode: {},
    selectorButton: {},
    selectorMenuItem: {},
    settingsButton: {},
    shiftIncrement: {},
    shortDateFormat: {},
    shortDateTimeFormat: {},
    shortEventCls: {},
    shortEventDuration: {},
    showBullet: {},
    showDirty: {},
    showEmptyDates: {},
    showResourceAvatars: {},
    stackMultiDayEvents: {},
    startDate: {},
    stateful: {},
    statefulEvents: {},
    stateId: {},
    stateProvider: {},
    stateSettings: {},
    store: {},
    strips: {},
    subGridConfigs: {},
    syncCalendarDate: {},
    syncMask: {},
    syncViewDate: {},
    tab: {},
    tabBarItems: {},
    tbar: {},
    timeFormat: {},
    title: {},
    tools: {},
    transitionDuration: {},
    type: {},
    ui: {},
    weekStartDay: {},
    weight: {},
    width: {},
    cellCopyPasteFeature: {},
    cellEditFeature: {},
    cellMenuFeature: {},
    cellTooltipFeature: {},
    columnAutoWidthFeature: {},
    columnDragToolbarFeature: {},
    columnPickerFeature: {},
    columnRenameFeature: {},
    columnReorderFeature: {},
    columnResizeFeature: {},
    excelExporterFeature: {},
    fileDropFeature: {},
    fillHandleFeature: {},
    filterFeature: {},
    filterBarFeature: {},
    groupFeature: {},
    groupSummaryFeature: {},
    headerMenuFeature: {},
    lockRowsFeature: {},
    mergeCellsFeature: {},
    pdfExportFeature: {},
    printFeature: {},
    quickFindFeature: {},
    regionResizeFeature: {},
    rowCopyPasteFeature: {},
    rowEditFeature: {},
    rowExpanderFeature: {},
    rowReorderFeature: {},
    rowResizeFeature: {},
    searchFeature: {},
    sortFeature: {},
    splitFeature: {},
    stickyCellsFeature: {},
    stripeFeature: {},
    summaryFeature: {},
    treeFeature: {},
    treeGroupFeature: {}
  },
  emits: ["beforeAutoCreate", "beforeCancelCellEdit", "beforeCancelRowEdit", "beforeCellDelete", "beforeCellEditStart", "beforeChangeDate", "beforeColumnDragStart", "beforeColumnDropFinalize", "beforeColumnResize", "beforeCopy", "beforeCSVExport", "beforeDestroy", "beforeExcelExport", "beforeFillHandleDragStart", "beforeFinishCellEdit", "beforeFinishRowEdit", "beforeHide", "beforePaste", "beforePdfExport", "beforeRenderRow", "beforeRenderRows", "beforeResponsiveStateChange", "beforeRowCollapse", "beforeRowExpand", "beforeSelectionChange", "beforeSetRecord", "beforeShow", "beforeStartRowEdit", "beforeStateApply", "beforeStateSave", "beforeToggleGroup", "beforeToggleNode", "cancelCellEdit", "catchAll", "cellClick", "cellContextMenu", "cellDblClick", "cellMapPopulated", "cellMenuBeforeShow", "cellMenuItem", "cellMenuShow", "cellMenuToggleItem", "cellMouseEnter", "cellMouseLeave", "cellMouseOut", "cellMouseOver", "collapse", "collapseNode", "columnDrag", "columnDragStart", "columnDrop", "columnResize", "columnResizeStart", "contextMenuItem", "contextMenuToggleItem", "copy", "dataChange", "dayCellPopulated", "destroy", "dirtyStateChange", "dragSelecting", "eventAutoCreated", "eventPropagate", "expand", "expandNode", "fileDrop", "fillHandleBeforeDragFinalize", "fillHandleDrag", "fillHandleDragAbort", "fillHandleDragEnd", "fillHandleDragStart", "finishCellEdit", "finishRowEdit", "focusIn", "focusOut", "gridRowBeforeDragStart", "gridRowBeforeDropFinalize", "gridRowDrag", "gridRowDragAbort", "gridRowDragStart", "gridRowDrop", "headerClick", "headerMenuBeforeShow", "headerMenuItem", "headerMenuShow", "headerMenuToggleItem", "hide", "lockRows", "mouseOut", "mouseOver", "paint", "paste", "pdfExport", "rangeChange", "readOnly", "recompose", "refresh", "renderRow", "renderRows", "resize", "responsive", "responsiveStateChange", "rowCollapse", "rowExpand", "rowMouseEnter", "rowMouseLeave", "scroll", "selectionChange", "selectionModeChange", "shiftIncrementChange", "show", "split", "splitterCollapseClick", "splitterDragEnd", "splitterDragStart", "splitterExpandClick", "startCellEdit", "startRowEdit", "subGridCollapse", "subGridExpand", "toggleGroup", "toggleNode", "toolClick", "topDateChange", "unlockRows", "unsplit"],
  setup(E, { expose: s, emit: d }) {
    const u = d, n = {
      instanceClass: b,
      instanceName: "AgendaView",
      configNames: [
        "activationKey",
        "adopt",
        "animateFilterRemovals",
        "animateRemovingRows",
        "ariaDescription",
        "ariaLabel",
        "autoHeight",
        "bbar",
        "bodyCls",
        "bubbleEvents",
        "callOnFunctions",
        "collapsible",
        "config",
        "contentElementCls",
        "contextMenuTriggerEvent",
        "defaultRegion",
        "descriptionRenderer",
        "destroyStore",
        "disableGridColumnIdWarning",
        "disableGridRowModelWarning",
        "dock",
        "drawer",
        "elementAttributes",
        "enableSticky",
        "enableTextSelection",
        "endDate",
        "eventFilter",
        "eventHeaderRenderer",
        "eventHeight",
        "eventListTimeFormat",
        "eventRenderer",
        "eventRowSpacing",
        "eventSorter",
        "fillLastColumn",
        "fixedRowHeight",
        "footer",
        "formulaProviders",
        "fullRowRefresh",
        "getPrimaryResource",
        "getRowHeight",
        "header",
        "hideEventOverflow",
        "hideHorizontalScrollbar",
        "hoverCls",
        "ignoreParentReadOnly",
        "listeners",
        "listRangeMenu",
        "loadMask",
        "loadMaskDefaults",
        "loadMaskError",
        "localizable",
        "maskDefaults",
        "masked",
        "monitorResize",
        "nonWorkingDays",
        "owner",
        "plugins",
        "preserveFocusOnDatasetChange",
        "preserveScrollOnDatasetChange",
        "preventTooltipOnTouch",
        "relayStoreEvents",
        "resizable",
        "resizeToFitIncludesHeader",
        "resourceImagePath",
        "responsive",
        "responsiveLevels",
        "responsiveRoot",
        "responsiveState",
        "responsiveTarget",
        "ripple",
        "rootElement",
        "scrollerClass",
        "scrollManager",
        "selectorButton",
        "selectorMenuItem",
        "shortEventCls",
        "shortEventDuration",
        "showBullet",
        "showDirty",
        "showResourceAvatars",
        "startDate",
        "stateful",
        "statefulEvents",
        "stateId",
        "stateProvider",
        "strips",
        "subGridConfigs",
        "syncMask",
        "tab",
        "tabBarItems",
        "tbar",
        "timeFormat",
        "type",
        "ui",
        "weekStartDay",
        "weight"
      ],
      propertyConfigNames: [
        "alignSelf",
        "allowDragCreate",
        "allowDragMove",
        "allowDragResize",
        "alwaysShowCurrentDate",
        "animateTimeShift",
        "animateTreeNodeToggle",
        "appendTo",
        "ariaLive",
        "autoCreate",
        "cellEllipsis",
        "cls",
        "collapsed",
        "columnLines",
        "columns",
        "data",
        "dataset",
        "date",
        "dateFilter",
        "dateFormat",
        "dateSeparator",
        "defaultScrollOptions",
        "descriptionFormat",
        "disabled",
        "dragUnit",
        "emptyText",
        "enableUndoRedoKeys",
        "eventTimeRenderer",
        "extendAllDayEndDay",
        "extraData",
        "filterEventResources",
        "flex",
        "height",
        "hidden",
        "hideFooters",
        "hideHeaders",
        "hideNonWorkingDays",
        "id",
        "insertBefore",
        "insertFirst",
        "keyMap",
        "longPressTime",
        "margin",
        "maxDate",
        "maxHeight",
        "maxWidth",
        "minDate",
        "minHeight",
        "minWidth",
        "offsetStartsBeforeEvents",
        "onBeforeAutoCreate",
        "onBeforeCancelCellEdit",
        "onBeforeCancelRowEdit",
        "onBeforeCellDelete",
        "onBeforeCellEditStart",
        "onBeforeChangeDate",
        "onBeforeColumnDragStart",
        "onBeforeColumnDropFinalize",
        "onBeforeColumnResize",
        "onBeforeCopy",
        "onBeforeCSVExport",
        "onBeforeDestroy",
        "onBeforeExcelExport",
        "onBeforeFillHandleDragStart",
        "onBeforeFinishCellEdit",
        "onBeforeFinishRowEdit",
        "onBeforeHide",
        "onBeforePaste",
        "onBeforePdfExport",
        "onBeforeRenderRow",
        "onBeforeRenderRows",
        "onBeforeResponsiveStateChange",
        "onBeforeRowCollapse",
        "onBeforeRowExpand",
        "onBeforeSelectionChange",
        "onBeforeSetRecord",
        "onBeforeShow",
        "onBeforeStartRowEdit",
        "onBeforeStateApply",
        "onBeforeStateSave",
        "onBeforeToggleGroup",
        "onBeforeToggleNode",
        "onCancelCellEdit",
        "onCatchAll",
        "onCellClick",
        "onCellContextMenu",
        "onCellDblClick",
        "onCellMapPopulated",
        "onCellMenuBeforeShow",
        "onCellMenuItem",
        "onCellMenuShow",
        "onCellMenuToggleItem",
        "onCellMouseEnter",
        "onCellMouseLeave",
        "onCellMouseOut",
        "onCellMouseOver",
        "onCollapse",
        "onCollapseNode",
        "onColumnDrag",
        "onColumnDragStart",
        "onColumnDrop",
        "onColumnResize",
        "onColumnResizeStart",
        "onContextMenuItem",
        "onContextMenuToggleItem",
        "onCopy",
        "onDataChange",
        "onDayCellPopulated",
        "onDestroy",
        "onDirtyStateChange",
        "onDragSelecting",
        "onEventAutoCreated",
        "onEventPropagate",
        "onExpand",
        "onExpandNode",
        "onFileDrop",
        "onFillHandleBeforeDragFinalize",
        "onFillHandleDrag",
        "onFillHandleDragAbort",
        "onFillHandleDragEnd",
        "onFillHandleDragStart",
        "onFinishCellEdit",
        "onFinishRowEdit",
        "onFocusIn",
        "onFocusOut",
        "onGridRowBeforeDragStart",
        "onGridRowBeforeDropFinalize",
        "onGridRowDrag",
        "onGridRowDragAbort",
        "onGridRowDragStart",
        "onGridRowDrop",
        "onHeaderClick",
        "onHeaderMenuBeforeShow",
        "onHeaderMenuItem",
        "onHeaderMenuShow",
        "onHeaderMenuToggleItem",
        "onHide",
        "onLockRows",
        "onMouseOut",
        "onMouseOver",
        "onPaint",
        "onPaste",
        "onPdfExport",
        "onRangeChange",
        "onReadOnly",
        "onRecompose",
        "onRefresh",
        "onRenderRow",
        "onRenderRows",
        "onResize",
        "onResponsive",
        "onResponsiveStateChange",
        "onRowCollapse",
        "onRowExpand",
        "onRowMouseEnter",
        "onRowMouseLeave",
        "onScroll",
        "onSelectionChange",
        "onSelectionModeChange",
        "onShiftIncrementChange",
        "onShow",
        "onSplit",
        "onSplitterCollapseClick",
        "onSplitterDragEnd",
        "onSplitterDragStart",
        "onSplitterExpandClick",
        "onStartCellEdit",
        "onStartRowEdit",
        "onSubGridCollapse",
        "onSubGridExpand",
        "onToggleGroup",
        "onToggleNode",
        "onToolClick",
        "onTopDateChange",
        "onUnlockRows",
        "onUnsplit",
        "pastEventCls",
        "preserveScroll",
        "range",
        "readOnly",
        "rowHeight",
        "rowLines",
        "rtl",
        "scrollable",
        "selectionMode",
        "settingsButton",
        "shiftIncrement",
        "shortDateFormat",
        "shortDateTimeFormat",
        "showEmptyDates",
        "stackMultiDayEvents",
        "stateSettings",
        "store",
        "syncCalendarDate",
        "syncViewDate",
        "title",
        "tools",
        "transitionDuration",
        "width"
      ],
      propertyNames: [
        "dayCellCls",
        "firstVisibleCell",
        "firstVisibleDate",
        "focusVisible",
        "hasChanges",
        "lastVisibleCell",
        "lastVisibleDate",
        "originalStore",
        "parent",
        "selectedCell",
        "selectedCells",
        "selectedRecord",
        "selectedRecords",
        "selectedRows",
        "state",
        "tooltip"
      ],
      eventNames: [
        "beforeAutoCreate",
        "beforeCancelCellEdit",
        "beforeCancelRowEdit",
        "beforeCellDelete",
        "beforeCellEditStart",
        "beforeChangeDate",
        "beforeColumnDragStart",
        "beforeColumnDropFinalize",
        "beforeColumnResize",
        "beforeCopy",
        "beforeCSVExport",
        "beforeDestroy",
        "beforeExcelExport",
        "beforeFillHandleDragStart",
        "beforeFinishCellEdit",
        "beforeFinishRowEdit",
        "beforeHide",
        "beforePaste",
        "beforePdfExport",
        "beforeRenderRow",
        "beforeRenderRows",
        "beforeResponsiveStateChange",
        "beforeRowCollapse",
        "beforeRowExpand",
        "beforeSelectionChange",
        "beforeSetRecord",
        "beforeShow",
        "beforeStartRowEdit",
        "beforeStateApply",
        "beforeStateSave",
        "beforeToggleGroup",
        "beforeToggleNode",
        "cancelCellEdit",
        "catchAll",
        "cellClick",
        "cellContextMenu",
        "cellDblClick",
        "cellMapPopulated",
        "cellMenuBeforeShow",
        "cellMenuItem",
        "cellMenuShow",
        "cellMenuToggleItem",
        "cellMouseEnter",
        "cellMouseLeave",
        "cellMouseOut",
        "cellMouseOver",
        "collapse",
        "collapseNode",
        "columnDrag",
        "columnDragStart",
        "columnDrop",
        "columnResize",
        "columnResizeStart",
        "contextMenuItem",
        "contextMenuToggleItem",
        "copy",
        "dataChange",
        "dayCellPopulated",
        "destroy",
        "dirtyStateChange",
        "dragSelecting",
        "eventAutoCreated",
        "eventPropagate",
        "expand",
        "expandNode",
        "fileDrop",
        "fillHandleBeforeDragFinalize",
        "fillHandleDrag",
        "fillHandleDragAbort",
        "fillHandleDragEnd",
        "fillHandleDragStart",
        "finishCellEdit",
        "finishRowEdit",
        "focusIn",
        "focusOut",
        "gridRowBeforeDragStart",
        "gridRowBeforeDropFinalize",
        "gridRowDrag",
        "gridRowDragAbort",
        "gridRowDragStart",
        "gridRowDrop",
        "headerClick",
        "headerMenuBeforeShow",
        "headerMenuItem",
        "headerMenuShow",
        "headerMenuToggleItem",
        "hide",
        "lockRows",
        "mouseOut",
        "mouseOver",
        "paint",
        "paste",
        "pdfExport",
        "rangeChange",
        "readOnly",
        "recompose",
        "refresh",
        "renderRow",
        "renderRows",
        "resize",
        "responsive",
        "responsiveStateChange",
        "rowCollapse",
        "rowExpand",
        "rowMouseEnter",
        "rowMouseLeave",
        "scroll",
        "selectionChange",
        "selectionModeChange",
        "shiftIncrementChange",
        "show",
        "split",
        "splitterCollapseClick",
        "splitterDragEnd",
        "splitterDragStart",
        "splitterExpandClick",
        "startCellEdit",
        "startRowEdit",
        "subGridCollapse",
        "subGridExpand",
        "toggleGroup",
        "toggleNode",
        "toolClick",
        "topDateChange",
        "unlockRows",
        "unsplit"
      ]
    }, t = {}, r = l(), a = l(/* @__PURE__ */ new Map());
    return f("teleports", a), C(() => {
      const e = m();
      t.value = i.createWidget({
        me: e,
        widgetData: n,
        emit: u,
        element: r.value,
        toRaw: h
      });
      const o = (c, g) => w(() => e.props[c], g);
      i.watchProps(t.value, n, o);
    }), R(() => {
      var e, o;
      (o = (e = t.value) == null ? void 0 : e.destroy) == null || o.call(e);
    }), s({
      instance: t,
      refElement: r,
      teleports: a
    }), (e, o) => (S(), D("div", {
      ref_key: "refElement",
      ref: r
    }, null, 512));
  }
});
export {
  M as _
};
