// @bryntum/calendar/source/lib/Engine/localization/SvSE.ts
import LocaleHelper from '../../Core/localization/LocaleHelper.js'

const locale = {

    localeName : 'SvSE',
    localeDesc : 'Svenska',
    localeCode : 'sv-SE',
    localeRtl  : false,

    RemoveDependencyCycleEffectResolution : {
        descriptionTpl : 'Ta bort beroende'
    },

    DeactivateDependencyCycleEffectResolution : {
        descriptionTpl : 'Avaktivera beroende'
    },

    CycleEffectDescription : {
        descriptionTpl : 'En cyklisk sekvens upptäcktes av följande uppgifter: {0}'
    },

    EmptyCalendarEffectDescription : {
        descriptionTpl               : '"{2}" - uppgift kan inte utföras. "{0}" kalender innehåller inga arbetstidsintervall.',
        forwardDescriptionTpl        : '"{2}" - uppgift kan inte utföras. "{0}" kalender har ingen arbetstid efter {1}.',
        backwardDescriptionTpl       : '"{2}" - uppgift kan inte utföras. "{0}" kalender har ingen arbetstid före {1}.',
        noIntersectionDescriptionTpl : '"{2}" - uppgift kan inte utföras. Dess "{0}" -kalender och dess resurskalendrar har ingen gemensam arbetstid.'
    },

    Use24hrsEmptyCalendarEffectResolution : {
        descriptionTpl : 'Använd 24-timmarskalender (måndag-fredag).'
    },

    Use8hrsEmptyCalendarEffectResolution : {
        descriptionTpl : 'Använd 8-timmarskalender  (måndag-fredag 08:00-12:00, 13:00-17:00)'
    },

    IgnoreProjectConstraintResolution : {
        descriptionTpl : 'Ignorera projektgränsen och fortsätt med ändringen.'
    },

    ProjectConstraintConflictEffectDescription : {
        startDescriptionTpl : 'Du flyttade uppgiften "{0}" så att den startar {1}. Detta är före projektets startdatum {2}.',
        endDescriptionTpl   : 'Du flyttade uppgiften "{0}" så att den avslutar {1}. Detta är efter projektets slutdatum {2}.'
    },

    HonorProjectConstraintResolution : {
        descriptionTpl : 'Justera uppgiften för att respektera projektgränsen.'
    },

    IgnoreResourceCalendarEmptyCalendarEffectResolution : {
        descriptionTpl : 'Ignorera resurskalendrar och schemalägga uppgiften med sin egen kalender.'
    },

    ConflictEffectDescription : {
        descriptionTpl : 'En schemaläggningskonflikt har uppstått, {0} är oförenlig med {1}'
    },

    ConstraintIntervalDescription : {
        dateFormat : 'LLL'
    },

    ProjectConstraintIntervalDescription : {
        startDateDescriptionTpl : 'Projektets startdatum {0}',
        endDateDescriptionTpl   : 'Projektets slutdatum {0}'
    },

    DependencyType : {
        long : [
            'Start-Till-Start',
            'Start-Till-Avslut',
            'Avslut-Till-Start',
            'Avslut-Till-Avslut'
        ]
    },

    ManuallyScheduledParentConstraintIntervalDescription : {
        startDescriptionTpl : 'Manuellt planerade "{2}" tvingar dess underaktiviteter att inte starta tidigare än den {0}',
        endDescriptionTpl   : 'Manuellt planerade "{2}" tvingar dess underaktiviteter att inte sluta senare än den {1}'
    },

    DisableManuallyScheduledConflictResolution : {
        descriptionTpl : 'Inaktivera manuell planering för "{0}"'
    },

    DependencyConstraintIntervalDescription : {
        descriptionTpl : 'Beroende ({2}) från "{3}" till "{4}"'
    },

    RemoveDependencyResolution : {
        descriptionTpl : 'Ta bort beroende från "{1}" till "{2}"'
    },

    DeactivateDependencyResolution : {
        descriptionTpl : 'Avaktivera beroende från "{1}" till "{2}"'
    },

    DateConstraintIntervalDescription : {
        startDateDescriptionTpl : 'Aktivitet "{2}" {3} {0} villkor',
        endDateDescriptionTpl   : 'Aktivitet "{2}" {3} {1} villkor',
        constraintTypeTpl       : {
            startnoearlierthan  : 'Starta-tidigast',
            finishnoearlierthan : 'Avsluta-tidigast',
            muststarton         : 'Måste-starta',
            mustfinishon        : 'Måste-avslutas',
            startnolaterthan    : 'Starta-senast',
            finishnolaterthan   : 'Avsluta-senast'
        }
    },

    RemoveDateConstraintConflictResolution : {
        descriptionTpl : 'Ta bort "{1}" villkoret för aktiviteten "{0}"'
    },

    PostponeDateConstraintConflictResolution : {
        descriptionTpl : 'Skjut upp lösningen och flagga uppgift "{0}" med konfliktmarkör'
    }
}

export default LocaleHelper.publishLocale(locale)
